// jibo-interaction-memory/src/im/InteractionMemory.ts
/**
 * @fileOverview
 *
 * Created on 4/28/16.
 * @author Andrew Stout <andrew.stout@jibo.com>
 */

// These are for internal typings only and should NOT be exposed publicly or dependencies will break.
import JiboType = require('../../../jibo');
import Node = JiboType.kb.Node;
import Model = JiboType.kb.Model;

import {Log} from 'jibo-log';
import * as _ from 'lodash/lodash.min';

export interface StorageOptions {
  storageFolder: string;
}

export interface InteractionDescription {
  personIDs?: string[];
  skillName?: string;
}

export class InteractionEvent {
  //timestamp: Date; // ms since the epoch

  constructor(public description: InteractionDescription, public timestamp: Date) { /* noop */ }
}

export interface DateProvider {
  getDate: () => Date;
}

export class DefaultDateProvider implements DateProvider {
  getDate() {
    return new Date();
  }
}

enum Edge {
  history = <any>'history',
  stats = <any>'stats'
}

export const DEFAULT_STORAGE_OPTIONS = {
  storageFolder: '/jibo/interactionmemory'
};

export type EventFilter = (event: InteractionEvent) => boolean;

export interface UnivariateMeanAndVariance {
  mu: number;
  sigmasq: number;
}

/* tslint:disable:member-ordering */

export class InteractionMemory {

  private _kbModel: Model;
  private _dateProvider: DateProvider = new DefaultDateProvider();
  private log = new Log('Jibo.InteractionMemory');
  /**
   * Mark JiboType as private to keep it out of the d.ts
   * @private
   */
  private jibo: typeof JiboType;
  static HISTORY_SIZE_LIMIT = 3000; // limit to the size of the history, just as a paranoid safeguard.
  // just because you're paranoid doesn't mean they're not out to get you
  static HISTORY_SIZE_DECREMENT = 100; // how many old events to remove at once

  constructor(jibo: any, options?: StorageOptions) {
    this.log.info("Initializing interaction memory.");

    this.jibo = jibo;
    let storageOptions = options || DEFAULT_STORAGE_OPTIONS;
    this._kbModel = this.jibo.kb.createModel(storageOptions.storageFolder);
  }

  /**
   * Set a date provider for Interaction Memory to use to create dates
   * @param {Object} dateProvider
   */
  setDateProvider(dateProvider: DateProvider) {
    this._dateProvider = dateProvider;
  }

  /**
   * create an InteractionEvent with a timestamp aligned with InteractionMemory's dateProvider.
   * @param description
   * @returns {InteractionEvent}
   */
  createEvent(description:InteractionDescription): InteractionEvent{
    return new InteractionEvent(description, this._dateProvider.getDate());
  }

  private _getOrInitializeHistory() {
    return this._getOrCreateNode(Edge.history)
      .then( (node) => {
        // check existence of array, initialize if nonexistant
        if (!node.data.history) {
          node.data.history = [];
        }
        return node;
      }).catch((error) => {
        if (error.message) {
          error.message = 'Could not get or initialize history: ' + error.message;
        }
        throw error;
      });
  }

  private _getOrInitializeStats() {
    return this._getOrCreateNode(Edge.stats)
      .then( (node) => {
        // check existence of data, initialize if nonexistant
        if (!node.data.userSkillLaunchCounts) {
          node.data.userSkillLaunchCounts = {};
        }
        return node;
      }).catch((error) => {
        if (error.message) {
          error.message = 'Could not get or initialize stats: ' + error.message;
        }
        throw error;
      });
  }

  /**
   * Store the event
   * @param {InteractionDescription} description
   * @returns {Promise<InteractionEvent>}
   */
  noteEvent(description: InteractionDescription): Promise<InteractionEvent>{
    let event = new InteractionEvent(description, this._dateProvider.getDate());

    // store the event in the event history
    let phist = this._getOrInitializeHistory()
      .then( (node) => {
        // push event
        // convert event timestamp to ms before saving
        event.timestamp = <any>Number(event.timestamp);
        node.data.history.push(event);
        // as a paranoid safeguard, limit the size of the history.
        // if it's over size, dump the oldest - in batches, since that'll be a bit more performant
        if (node.data.history.length > InteractionMemory.HISTORY_SIZE_LIMIT){
          node.data.history = node.data.history.slice(InteractionMemory.HISTORY_SIZE_DECREMENT);
        }
        return this._saveNode(node);
      }).catch((error) => {
        this.log.error("Could not note event: ", error);
      });
    // increment user skill launch counts (consider moving this to its own function)
    let pstats = this._getOrInitializeStats()
      .then( (statsNode) => {
        if (description.personIDs && description.skillName) {
          for (let userID of description.personIDs) {
            let u = statsNode.data.userSkillLaunchCounts[userID];
            if (!u) { // userID has no entry in userSkillLaunchCounts
              statsNode.data.userSkillLaunchCounts[userID] = {};
            }
            let us = statsNode.data.userSkillLaunchCounts[userID][description.skillName];
            if (!us) { // skillName has no entry in userSkillLaunchCounts[userID]
              statsNode.data.userSkillLaunchCounts[userID][description.skillName] = 0;
            }
            statsNode.data.userSkillLaunchCounts[userID][description.skillName]++;
          }
        }
        return this._saveNode(statsNode);
      }).catch((error) => {
        this.log.error("Could not note event: ", error);
      });
    return Promise.all([phist, pstats])
	    .then(() => {
	    	return event;
      });
  }

  /**
   * default EventDescription comparison function
   * if present, every person mentioned in the query should be included in the candidate
   * @param {InteractionDescription} query
   * @param {InteractionDescription} candidate
   * @returns {boolean}
   */
  static compareUser(query: InteractionDescription, candidate: InteractionDescription) {
    // if present, every person mentioned in the query should be included in the candidate
    if (query.personIDs && candidate.personIDs) {
      for (let person of query.personIDs) {
        if (candidate.personIDs.indexOf(person) === -1) { // if this person isn't present in the candidate's list
          return false;
        }
      }
      return true;
    }
    return false;
  }

  /**
   * every person mentioned in the query should be included in the candidate
   * skillName should match
   * both must be present
   * @param {InteractionDescription} query
   * @param {InteractionDescription} candidate
   * @returns {boolean}
   */
  static compareUserAndSkill(query: InteractionDescription, candidate: InteractionDescription) {
    // every person mentioned in the query should be included in the candidate
    // skillName should match
    // both must be present
    if (query.personIDs && candidate.personIDs && query.skillName && candidate.skillName) {
      for (let person of query.personIDs) {
        if (candidate.personIDs.indexOf(person) === -1) { // if this person isn't present in the candidate's list
          return false;
        }
      }
      return query.skillName === candidate.skillName;
    } else { // query.personIDs, candidate.personIDs, query.skillName or candidate.skillName not present
      return false;
    }
  }

  /**
   * Every property specified in the query must match; additional properties in the candidate are okay.
   * @param query
   * @param candidate
   * @returns {boolean}
   */
  static compareQueryFields(query: InteractionDescription, candidate: InteractionDescription) {
    let ret = true;
    Object.keys(query).forEach(function(key,index) {
      if (!(candidate[key] && _.isEqual(candidate[key], query[key])) ) {
        ret = false;
      }
    });
    return ret;
  }


  /**
   *
   * @param {InteractionDescription} queryDesc - an InteractionDescription to be matched
   * @param compare - a function returning true when two InteractionDescriptions are considered a match, false otherwise
   * @param {Date|null} startTime - optional start time of search window; i.e., when we stop searching backwards in time
   * @param {Date} [refDate] - optional, a reference date to compute against. If not provided the current time (as per InteractionMemory's dateProvider) will be used.
   * @returns number of milliseconds since a matching event, or Number.POSITIVE_INFINITY to indicate no match
   */
  getTimeSinceLast(queryDesc: InteractionDescription, compare = InteractionMemory.compareUser, startTime?: Date, refDate:Date = this._dateProvider.getDate()): Promise<number> {
    // stupid version: search (chronologically backwards) for matching event
    // eventually, a smarter version does some kind of database thing.
    const query = new InteractionEvent(queryDesc, refDate);
    return this._getOrInitializeHistory()
      .then( (node) => {
        // search through history for matching event
        for (let ei = node.data.history.length - 1; ei >= 0; ei--) {
          let e = node.data.history[ei];  // timestamp comes out of storage as a number (ms)
          // we're progressing in reverse chronological order.
          if (startTime && e.timestamp < startTime.getTime()) {
            // we have (progressing in reverse chronological order), reached events
            // before the start time. Assuming order, we can stop iterating now.
            break;
          }
          // check whether e matches the query
          if (compare(query.description, e.description)) {
            const queryTimestamp = query.timestamp.getTime();
            const eventTimestamp = e.timestamp;
            this.log.info(`Matched. Query timestamp: ${queryTimestamp} Event timestamp: ${eventTimestamp}`);
            // compute time since
            return queryTimestamp - eventTimestamp;
          }
        }
        // if no match, return some indication
        return Number.POSITIVE_INFINITY;
      }).catch((error) => {
        // There's no valid number we can return here, throw an error.
        if (error.message) {
          error.message = 'Could not get time since last: ' + error.message;
        }
        throw error;
      });
  }

  /**
   * returns promise of number of milliseconds since last observed interaction event, or Number.POSITIVE_INFINITY to indicate no events in history
   * @param {Date} [refTime] Reference time (defaults to current time as set by dateProvider)
   * @returns {Promise<number>}
   */
  getTimeSinceLastInteraction(refTime?: Date): Promise<number> {
    const _refTime = refTime || this._dateProvider.getDate();

    return this._getOrInitializeHistory()
      .then( (node) => {
        if (node.data.history.length > 0) {
          let lasttime = node.data.history[node.data.history.length -1].timestamp;
          return _refTime.getTime() - lasttime;  // timestamp comes out of storage as a number (ms)
        } else {
          return Number.POSITIVE_INFINITY;
        }
      }).catch((error) => {
        // There's no valid number we can return here, throw an error.
        if (error.message) {
          error.message = 'Could not get time since last interaction: ' + error.message;
        }
        throw error;
      });
  }

  /**
   * returns promise of number of times user userID has launched skill skillName
   * @param {string} userID
   * @param {string} skillName
   * @returns {Promise<number>}
   */
  getUserSkillCount(userID: string, skillName: string): Promise<Number>{
    return this._getOrInitializeStats()
      .then((statsNode) => {
        let u = statsNode.data.userSkillLaunchCounts[userID];
        if (!u) { // userID has no entry in userSkillLaunchCounts
          return 0;
        } else {
          let us = u[skillName];
          if (!us) { // skillName has no entry in userSkillLaunchCounts[userID]
            return 0;
          } else {
            return us;
          }
        }
      }).catch((error) => {
        if (error.message) {
          error.message = 'Could not get user skill count: ' + error.message;
        }
        throw error;
      });
  }

  /**
   * get all events matching a query
   * @intdocs
   * @param {EventFilter} filter Functor determining desired events
   * @param {Date} [startTime] start of chronological window of interest
   * @param {Date} [endTime] end of chronological window of interest
   */
  async getEventsMatchingFilter(filter: EventFilter, startTime?: Date, endTime?: Date): Promise<Array<InteractionEvent> > {
    // NOTE: like getTimeSinceLast, this implementation relies on the chronological ordering of events
    let filteredEvents = new Array<InteractionEvent>();
    let node;
    try {
      node = await this._getOrInitializeHistory();
    } catch (error) {
      if (error.message) {
        error.message = "Could not get events matching filter: " + error.message;
      }
      throw error;
    }
    for (let ei = node.data.history.length - 1; ei >= 0; ei--) {
      let e = node.data.history[ei];
      // we're progressing in reverse chronological order.
      if (startTime && e.timestamp < startTime.getTime()) {
        // we have (progressing in reverse chronological order), reached events
        // before the start time. Assuming order, we can stop iterating now.
        break;
      }
      if (endTime && e.timestamp > endTime.getTime()) {
        // we haven't yet gotten to the time window of the query; keep iterating
        continue;
      }
      if (filter(e)){
        filteredEvents.push(e); // This reverses the order relative to storage--this will be earliest first.
      }
    }
    return filteredEvents;
  }


  /**
   * Compute the univariate mean and variance of univariate data set.
   * @param {number[]} data
   * @returns {UnivariateMeanAndVariance}
   * @intdocs
   * @static
   */
  static univariateMeanAndVariance(data: number[]): UnivariateMeanAndVariance {
    // assumes uniformly-weighted samples
    // mu = sum(data)/data.length
    // sigma^2 = sum(data^2)/data.length - mu^2
    let sum_x = 0;
    let sum_xsq = 0;
    for (let x of data) {
      sum_x += x;
      sum_xsq += x*x;
    }
    let mu = sum_x/data.length;
    let sigmasq = sum_xsq/data.length - mu*mu;
    return {mu: mu, sigmasq: sigmasq};
  }

  /**
   * compute mean and variance for the time of day (as minutes after 0:00 midnight) of a set of events
   * @static
   * @param {InteractionEvent[]} events
   * @returns {UnivariateMeanAndVariance}
   * @intdocs
   * @memberOf InteractionMemory
   */
  static getTimeOfDaySummaryStatistics(events: InteractionEvent[]): UnivariateMeanAndVariance {
    // apply accessor to get time data
    let accessor = (ie) => {
      let date = new Date(ie.timestamp);
      return date.getHours()*60 + date.getMinutes();
    };
    let data = [];
    for (let ie of events) {
      data.push(accessor(ie));
    }
    // apply univariateMeanAndVariance
    return InteractionMemory.univariateMeanAndVariance(data);
  }


  /**
   * Clear the history.
   * @returns {Promise<void>}
   * @intdocs
   */
  public _clear(): Promise<void> {
    let phist = this._getOrCreateNode(Edge.history)
      .then( (node) => {
        node.data.history = [];
        return this._saveNode(node);
      }).catch((error) => {
        this.log.error("Could not clear: ", error);
      });
    let pstats = this._getOrCreateNode(Edge.stats)
      .then( (statsNode) => {
        statsNode.data.userSkillLaunchCounts = {};
        return this._saveNode(statsNode);
      }).catch((error) => {
        this.log.error("Could not clear: ", error);
      });
    return Promise.all([phist, pstats]).then(() => null);
  }

  public _getHistory(): Promise<InteractionEvent[]> {
    return this._getOrInitializeHistory()
      .then((node) => {
        // convert the history nodes back to InteractionEvents
        // timestamps come out of storage as a number (ms)
        // InteractionEvents have Date objects as their timestamp
        let history = [];
        for (let ei = node.data.history.length - 1; ei >= 0; ei--) {
          let e = node.data.history[ei];
          let event = new InteractionEvent(e.description, new Date(e.timestamp));
          history.push(event);
        }
        return history;
      }).catch((error) => {
        if (error.message) {
          error.message = 'Could not get history: ' + error.message;
        }
        throw error;
      });
  }

  /**
   * Load a history or stats node from the root node, or create one
   * and attach it to the root node if it doesn't already exist.
   * @param {string} edge Name of edge to follow or create. Also
   * serves as node type.
   * @returns {Promise<Node>}
   */
  private _getOrCreateNode(edge:Edge|string):Promise<Node> {
    let edgeName = Edge[edge];
    if (!edgeName) {
      throw new Error(`Invalid 'edge' name: ${edge}`);
    }
    return this._loadRoot()
      .then( (rootNode) => {
        let edges = rootNode.getEdges(edgeName);
        if (edges.length > 0) {
          let id = edges[0];
          return this._loadNode(id);
        } else {
          // no history node yet, so create one
          let node = this._kbModel.createNode(edgeName);  // node type = edge name
          return this._saveNode(node)
            .then( () => {
              rootNode.addEdges(node, edgeName);
              return this._saveNode(rootNode)
                .then( () => {
                  return node;
                });
            });
        }
      }).catch((error) => {
        if (error.message) {
          error.message = 'Could not get or create node: ' + error.message;
        }
        throw error;
      });
  }

  /**
   * A promise wrapper around the jibo.kb.Model#_loadRoot call. Loads the
   * root node for the interaction memory KB model.
   * @returns {Promise<Node>} The root node for the interaction memory model.
   */
  private _loadRoot():Promise<Node> {
    return new Promise( (res, rej) => {
      this._kbModel.loadRoot( (err, rootNode) => {
        if (err) {
          rej(err);
        } else {
          res(rootNode);
        }
      });
    });
  }

  /**
   * A promise wrapper around the jibo.kb.Node#_save call.
   * @param {Node} The node to be saved.
   * @returns {Promise}
   */
  private _saveNode(node:Node) {
    return new Promise( (res, rej) => {
      node.save( (err) => {
        if (err) {
          rej(err);
        } else {
          res();
        }
      });
    });
  }

  /**
   * A promise wrapper around the jibo.kb.Model#_load call.
   * @param {string} id ID of node to load.
   * @returns {Promise<Node>}
   */
  private _loadNode(id:string): Promise<Node> {
    return new Promise<Node>( (res, rej) => {
      this._kbModel.load(id, (err, node) => {
        if (err) {
          rej(err);
        } else {
          res(node);
        }
      });
    });
  }
}

// For testing
if ((<any>global)._imTest) {
  (<any>global)._imTest.InteractionMemory = module.exports;
}
