// @be/main-menu/index.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.bemainMenu = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';
module.exports = function (blackboard, notepad, result, emitter) {
    return {
        'meta': {
            'version': 1,
            'name': 'main',
            'uri': '/Users/jon/Workspace/jibo/buildsdk/skills/main-menu/src/flows/main.flow'
        },
        '011d9d9a-703a-4924-bb70-108ec1726352': function () {
            return {
                'id': '011d9d9a-703a-4924-bb70-108ec1726352',
                'transitions': [{
                        'frm': '011d9d9a-703a-4924-bb70-108ec1726352',
                        'to': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                        'value': ''
                    }],
                'exceptions': [],
                'class': 'Flow.Begin',
                'options': {
                    'inputParameters': () => {
                        return {};
                    }
                }
            };
        },
        'b32b43a7-5ff2-44c8-a7b3-f7691433934b': function () {
            return {
                'id': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                'name': 'Main Menu',
                'transitions': [
                    {
                        'frm': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                        'to': '9b202569-67ce-421f-b9ab-1c601f045f55',
                        'value': 'fun'
                    },
                    {
                        'frm': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                        'to': 'a8161622-b641-4ae0-8a14-b01abca93661',
                        'value': ''
                    },
                    {
                        'frm': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                        'to': '3958d3bf-67ef-4fd3-9e55-f73a5bf68304',
                        'value': 'personal-report'
                    }
                ],
                'exceptions': [
                    {
                        'frm': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                        'to': '4922ff2c-16e8-4923-9ede-d6b51afa93af',
                        'value': '~MenuFailed'
                    },
                    {
                        'frm': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                        'to': '4922ff2c-16e8-4923-9ede-d6b51afa93af',
                        'value': '~InteractionError'
                    }
                ],
                'class': 'Menu',
                'options': {
                    'getConfig': loadCallback => {
                        jibo.loader.load('resources/views/main-menu-verbal.json', (err, config) => {
                            loadCallback(config);
                        });
                    },
                    'onMenuClosed': (wasTimeout, menu, overrideMenuTransition, exception) => {
                        overrideMenuTransition('remain');
                        blackboard.log.debug('Menu was closed with exception:', exception);
                        return exception;
                    },
                    'onItemChosen': (result, menu, overrideMenuTransition) => {
                        overrideMenuTransition('remain');
                        if (menu) {
                            let list = menu.getComponentByIndex(0);
                            let button = list.getComponentById(result.entities.destination);
                            list.actionHandler({
                                type: 'forceIndex',
                                data: {}
                            }, button);
                            blackboard.log.debug('Menu item was chosen with destination: ', result.entities.destination);
                            return result.entities.destination;
                        } else {
                            blackboard.log.warn('onItemChosen supplied an undefined menu parameter');
                            return '~MenuFailed';
                        }
                    },
                    'onPositionalSelect': (commandAsr, intendedIndex, menu) => {
                    }
                }
            };
        },
        '4922ff2c-16e8-4923-9ede-d6b51afa93af': function () {
            return {
                'id': '4922ff2c-16e8-4923-9ede-d6b51afa93af',
                'transitions': [],
                'exceptions': [],
                'class': 'Flow.End',
                'options': {
                    'getTransition': () => {
                        return;
                    }
                }
            };
        },
        '9b202569-67ce-421f-b9ab-1c601f045f55': function () {
            return {
                'id': '9b202569-67ce-421f-b9ab-1c601f045f55',
                'name': 'Fun Stuff Menu',
                'transitions': [{
                        'frm': '9b202569-67ce-421f-b9ab-1c601f045f55',
                        'to': 'a8161622-b641-4ae0-8a14-b01abca93661',
                        'value': ''
                    }],
                'exceptions': [
                    {
                        'frm': '9b202569-67ce-421f-b9ab-1c601f045f55',
                        'to': '4922ff2c-16e8-4923-9ede-d6b51afa93af',
                        'value': '~MenuFailed'
                    },
                    {
                        'frm': '9b202569-67ce-421f-b9ab-1c601f045f55',
                        'to': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                        'value': '~InteractionError.MenuClosed'
                    },
                    {
                        'frm': '9b202569-67ce-421f-b9ab-1c601f045f55',
                        'to': '4922ff2c-16e8-4923-9ede-d6b51afa93af',
                        'value': '~InteractionError.MenuTimeout'
                    }
                ],
                'class': 'Menu',
                'options': {
                    'getConfig': loadCallback => {
                        jibo.loader.load('resources/views/fun-stuff-verbal.json', (err, config) => {
                            loadCallback(config);
                        });
                    },
                    'onMenuClosed': (wasTimeout, menu, overrideMenuTransition, exception) => {
                        overrideMenuTransition('remain');
                        blackboard.log.debug('Menu was closed with exception:', exception);
                        return exception;
                    },
                    'onItemChosen': (result, menu, overrideMenuTransition) => {
                        overrideMenuTransition('remain');
                        if (menu) {
                            let list = menu.getComponentByIndex(0);
                            let button = list.getComponentById(result.entities.destination);
                            list.actionHandler({
                                type: 'forceIndex',
                                data: {}
                            }, button);
                            blackboard.log.debug('Menu item was chosen with destination: ', result.entities.destination);
                            return result.entities.destination;
                        } else {
                            blackboard.log.warn('onItemChosen supplied an undefined menu parameter');
                            return '~MenuFailed';
                        }
                    },
                    'onPositionalSelect': (commandAsr, intendedIndex, menu) => {
                    }
                }
            };
        },
        'a8161622-b641-4ae0-8a14-b01abca93661': function () {
            return {
                'id': 'a8161622-b641-4ae0-8a14-b01abca93661',
                'name': 'Skill Redirect',
                'asset-pack': 'core',
                'transitions': [{
                        'frm': 'a8161622-b641-4ae0-8a14-b01abca93661',
                        'to': '4922ff2c-16e8-4923-9ede-d6b51afa93af',
                        'value': ''
                    }],
                'exceptions': [],
                'class': 'Flow.Eval-Async',
                'options': {
                    'exec': done => {
                        blackboard.log.info('Attempting redirect to skill: ' + this.inTransition);
                        blackboard.skill.redirectToSkill(this.inTransition);
                    },
                    'onStop': () => {
                    }
                }
            };
        },
        '3958d3bf-67ef-4fd3-9e55-f73a5bf68304': function () {
            return {
                'id': '3958d3bf-67ef-4fd3-9e55-f73a5bf68304',
                'name': 'Personal Report Menu',
                'asset-pack': 'core',
                'transitions': [{
                        'frm': '3958d3bf-67ef-4fd3-9e55-f73a5bf68304',
                        'to': 'a8161622-b641-4ae0-8a14-b01abca93661',
                        'value': ''
                    }],
                'exceptions': [
                    {
                        'frm': '3958d3bf-67ef-4fd3-9e55-f73a5bf68304',
                        'to': '4922ff2c-16e8-4923-9ede-d6b51afa93af',
                        'value': '~'
                    },
                    {
                        'frm': '3958d3bf-67ef-4fd3-9e55-f73a5bf68304',
                        'to': 'b32b43a7-5ff2-44c8-a7b3-f7691433934b',
                        'value': '~InteractionError.MenuClosed'
                    }
                ],
                'class': 'Menu',
                'options': {
                    'getConfig': loadCallback => {
                        jibo.loader.load('resources/views/personal-report-verbal.json', (err, config) => {
                            loadCallback(config);
                        });
                    },
                    'onMenuClosed': (wasTimeout, menu, overrideMenuTransition, exception) => {
                        overrideMenuTransition('remain');
                        blackboard.log.debug('Menu was closed with exception:', exception);
                        return exception;
                    },
                    'onItemChosen': (result, menu, overrideMenuTransition) => {
                        overrideMenuTransition('remain');
                        if (menu) {
                            let list = menu.getComponentByIndex(0);
                            let button = list.getComponentById(result.entities.destination);
                            list.actionHandler({
                                type: 'forceIndex',
                                data: {}
                            }, button);
                            blackboard.log.debug('Menu item was chosen with destination: ', result.entities.destination);
                            return result.entities.destination;
                        } else {
                            blackboard.log.warn('onItemChosen supplied an undefined menu parameter');
                            return '~MenuFailed';
                        }
                    },
                    'onPositionalSelect': (commandAsr, intendedIndex, menu) => {
                    }
                }
            };
        }
    };
};
},{}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const be_framework_1 = require("@be/be-framework");
const jibo = require("jibo");
const mainFlow = require('./flows/main');
class MainMenu extends be_framework_1.BeSkill {
    constructor(assetPack) {
        super(assetPack);
        this.flow = null;
        this.skillChosen = false;
        this.flowOverrides = null;
    }
    open(result, refresh, previousSkillName, previousSkillOptions) {
        if (!refresh) {
            if (result && result.intent === 'tap') {
                jibo.mim.silentMenus = true;
                this.log.info('open() from tap');
                this.track('Menu Opened', { modality: 'touch' });
            }
            else if (result && result.asr.text) {
                this.log.info('open() from HJ/global');
                this.track('Menu Opened', { modality: 'voice' });
            }
            else {
                this.log.info('open() from other skill');
                this.track('Menu Opened', { modality: 'return' });
            }
            this.skillChosen = false;
            const options = Object.assign({
                assetPack: this.assetPack,
                enableLogging: true,
                blackboard: {
                    log: this.log,
                    skill: this
                }
            }, this.flowOverrides);
            this.flow = jibo.flow.run(mainFlow, options, (err, status) => {
                this.log.debug('flow ended with status', status);
                if (status !== jibo.bt.Status.INTERRUPTED) {
                    this.log.debug('flow ended without interruption');
                    this.flow.destroy();
                    this.flow = null;
                    this.exit();
                }
                else {
                    this.log.debug('flow ended with interruption, skill chosen', this.skillChosen);
                }
            });
        }
        else {
            this.log.info('open() refresh');
            jibo.globalEvents.shared.nonInterruptingGlobal.emit();
            let alreadyOnMenuMimConfig = {
                mimPath: 'mims/en-us/AlreadyOnMenu.mim',
                assetPack: this.assetPack
            };
            let alreadyOnMenuMim = new jibo.bt.behaviors.Mim(alreadyOnMenuMimConfig);
            alreadyOnMenuMim.start();
        }
    }
    redirectToSkill(skill) {
        this.log.debug('redirectToSkill() given skill: ' + skill);
        if (this.skillChosen || !skill) {
            this.log.warn('redirectToSkill() attempt to skill: ' + skill + ' but a skill has already been chosen.');
            return;
        }
        if (skill === 'fun' || skill === 'personal-report') {
            return false;
        }
        let redirectOptions;
        let isCloud = false;
        if (skill === 'joke') {
            isCloud = true;
            skill = 'chitchat';
            redirectOptions = {
                clientASR: 'tell me a joke'
            };
        }
        else if (skill === 'dance') {
            isCloud = true;
            skill = 'chitchat';
            redirectOptions = {
                clientASR: 'do a dance'
            };
        }
        else if (skill === 'surprise') {
            isCloud = true;
            skill = 'chitchat';
            redirectOptions = {
                clientASR: 'surprise me'
            };
        }
        else if (skill === 'full-report') {
            isCloud = true;
            skill = 'personal-report';
            redirectOptions = {
                clientASR: 'personal report'
            };
        }
        else if (skill === 'weather') {
            isCloud = true;
            skill = 'personal-report';
            redirectOptions = {
                clientASR: 'tell me about the weather'
            };
        }
        else if (skill === 'news') {
            isCloud = true;
            skill = 'personal-report';
            redirectOptions = {
                clientASR: 'tell me about the news'
            };
        }
        else if (skill === 'calendar') {
            isCloud = true;
            skill = 'personal-report';
            redirectOptions = {
                clientASR: 'tell me about my calendar'
            };
        }
        else if (skill === 'commute') {
            isCloud = true;
            skill = 'personal-report';
            redirectOptions = {
                clientASR: 'what does my commute look like'
            };
        }
        else {
            let intent;
            if (skill === 'photobooth') {
                intent = 'createSomePhotos';
                skill = 'create';
            }
            else if (skill === 'snapshot') {
                intent = 'createOnePhoto';
                skill = 'create';
            }
            else {
                intent = 'menu';
            }
            redirectOptions = {
                nlu: {
                    entities: {
                        domain: skill,
                    },
                    intent: intent
                }
            };
        }
        this.skillChosen = true;
        jibo.face.views.changeView({
            remove: true,
            leaveEmpty: true,
            transitionClose: jibo.face.views.TRANSITION.UP
        }, () => {
            if (!isCloud) {
                this.redirect(`@be/${skill}`, redirectOptions);
            }
        });
        if (isCloud) {
            jibo.embodied.listen.disableOnce();
            jibo.jetstream.mimicGlobalTurn(redirectOptions).catch((err) => {
                this.log.error(`Unable to mimic global turn to launch cloud skill ${skill} with data`, redirectOptions, err);
                this.exit();
            });
        }
    }
    close(done) {
        this.log.info('close()');
        if (this.flow) {
            this.log.info('close() flow requires stop and destroy followed by view clean up');
            Promise.all([
                this.flow.stop()
                    .catch((err) => {
                    this.log.error('Error when stopping flow: ', err);
                })
                    .then(() => {
                    this.flow.destroy();
                    this.flow = null;
                }),
                this.cleanupViews()
                    .catch((err) => {
                    this.log.debug('Eror when closing views: ', err);
                })
            ]).then(() => {
                done();
            });
        }
        else {
            this.cleanupViews()
                .then(done);
        }
    }
    cleanupViews() {
        return new Promise((resolve) => {
            jibo.face.views.viewStackCleanup()
                .then(() => {
                resolve();
            })
                .catch(() => {
                this.log.debug('cleanupViews() failed to remove all views, calling done anyway');
                resolve();
            });
        });
    }
}
module.exports = MainMenu;

},{"./flows/main":1,"@be/be-framework":undefined,"jibo":undefined}]},{},[2])(2)
});
