// rule-generator/lib/rule-generator.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.ruleGenerator = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
var otherutils = require('../utils/other-utils');
var permutation = require('./permutation');
var constants = require('./constants');
var priorities = require('./priorities');
//function get_values_in_set(crew_member, result) {
//    if (crew_member.hasOwnProperty(constants.FIRSTNAME_FIELD))
//        result[crew_member[constants.FIRSTNAME_FIELD]] = true;
//    if (crew_member.hasOwnProperty(constants.MIDDLENAME_FIELD))
//        result[crew_member[constants.MIDDLENAME_FIELD]] = true;
//    if (crew_member.hasOwnProperty(constants.LASTNAME_FIELD))
//        result[crew_member[constants.LASTNAME_FIELD]] = true;
//    if (crew_member.hasOwnProperty(constants.NICKNAME_FIELD))
//        result[crew_member[constants.NICKNAME_FIELD]] = true;
//}
function fields2code(arr) {
    var result = '';
    for (var i = 0; i < arr.length; ++i) {
        result += constants.FIELDCODES[arr[i]];
    }
    return result;
}
function convert_to_specific_hash_format(result, crew) {
    var literals = otherutils.get_keys(result);
    for (var i = 0; i < literals.length; ++i) {
        var literal = literals[i];
        var users = '';
        var fields = '';
        var uids_in_priority_order = priorities.get_uids_in_priority_order(result[literal], crew);
        for (var j = 0; j < uids_in_priority_order.length; ++j) {
            var uid = uids_in_priority_order[j];
            users += uid + ',';
            fields += fields2code(result[literal][uid]) + ',';
        }
        users = users.substring(0, users.length - 1);
        fields = fields.substring(0, fields.length - 1);
        result[literal] = { '_users': users, '_fields': fields };
    }
}
function add_dummy_entry(result) {
    var literal = "XXXXXXXXXX";
    var users = "null";
    var fields = "null";
    result[literal] = { '_users': users, '_fields': fields };
}
// The keys are the literal ways of saying a crew
// member.
// The values are a hash containing the key/value
// pairs of slots and values.
// The only two slots are: _users and _fields
// e.g.: result['rose'] = {'_users':'uid001-uid002', '_fields':'f-f'}
//       result['rose smith'] = {_users='uid001', '_fields':'fl'}
//       result['rose garcia'] = {_users='uid002', '_fields':'fl'}
// Fields:f, m, l, n mean first, middle, last and nick name.
// _users are the users matched by that literal, divided by a dash
// _fields are the format of the literal, e.g.: first last would be 'fl'
//         also divided by a dash, in the same order as the users
// E.g.: In the example above, the literal "rose" is matched by both
//       uid001 and uid002, with the first one being matched by the first
//       name and the second being matched by the first name as well.
// E.g.: 'rose garcia' is matched only by uid002, and the format is
//       first last.
function build_userids_per_literal(crew) {
    var result = {};
    for (var i = 0; i < crew.length; ++i) {
        add_all_literals_for_one_user(crew[i], result);
    }
    convert_to_specific_hash_format(result, crew);
    // This is important in case result turns out empty
    add_dummy_entry(result);
    return result;
}
// Returns all possible literal combinations of a crew
// member, and adds to the result hash
// e.g.: for rose linda garcia (the cool girl) it would
// add to result, the following literals:
// rose
// linda
// rose linda
// rose linda garcia
// rose the cool girl garcia
// ... and so on
function add_all_literals_for_one_user(crew_member, result) {
    var only_names = {};
    for (var i = 0; i < constants.FIELDS.length; ++i)
        if (crew_member.hasOwnProperty(constants.FIELDS[i]))
            only_names[constants.FIELDS[i]] = crew_member[constants.FIELDS[i]];
    delete only_names[constants.USERID_FIELD];
    var id = crew_member[constants.USERID_FIELD];
    var string_so_far = '';
    var fields_so_far = [];
    permutation.run_permutation_algorithm(only_names, id, string_so_far, fields_so_far, result);
}
exports.build_userids_per_literal = build_userids_per_literal;

},{"../utils/other-utils":8,"./constants":3,"./permutation":5,"./priorities":6}],2:[function(require,module,exports){
var stringutils = require('../utils/string-utils');
var constants = require('./constants');
// cleaning and validate crew object
function clean_and_validate(crew_members) {
    if (crew_members == null || !Array.isArray(crew_members))
        throw 'input to crew2rule is not an array';
    for (var i = 0; i < crew_members.length; ++i)
        clean_and_validate_single_user(crew_members[i]);
}
function clean_and_validate_single_user(crew_member) {
    if (crew_member == null || typeof (crew_member) != 'object')
        throw ('An element of the crew member list is either null or is not an object');
    if (!crew_member.hasOwnProperty(constants.USERID_FIELD)) {
        throw ('An element of the crew member list does not have a ' + constants.USERID_FIELD + ' field');
    }
    // For each field, clean up from side spaces
    // and internal double spaces.
    // If no content, then delete entry
    for (var i = 0; i < constants.FIELDS.length; ++i)
        if (crew_member.hasOwnProperty(constants.FIELDS[i])) {
            crew_member[constants.FIELDS[i]] = stringutils.process_special_characters(crew_member[constants.FIELDS[i]]);
            crew_member[constants.FIELDS[i]] = stringutils.remove_extra_spaces(crew_member[constants.FIELDS[i]]);
            if (crew_member[constants.FIELDS[i]] == '')
                delete crew_member[constants.FIELDS[i]];
        }
}
exports.clean_and_validate = clean_and_validate;

},{"../utils/string-utils":9,"./constants":3}],3:[function(require,module,exports){
var USERID_FIELD = 'userId';
var FIRSTNAME_FIELD = 'first_name';
var MIDDLENAME_FIELD = 'middle_name';
var LASTNAME_FIELD = 'last_name';
var NICKNAME_FIELD = 'nick_name';
var FIELDS = [USERID_FIELD, FIRSTNAME_FIELD, MIDDLENAME_FIELD,
    LASTNAME_FIELD, NICKNAME_FIELD];
var FIELDCODES = {};
FIELDCODES[USERID_FIELD] = "u";
FIELDCODES[FIRSTNAME_FIELD] = "f";
FIELDCODES[MIDDLENAME_FIELD] = "m";
FIELDCODES[LASTNAME_FIELD] = "l";
FIELDCODES[NICKNAME_FIELD] = "n";
exports.USERID_FIELD = USERID_FIELD;
exports.FIRSTNAME_FIELD = FIRSTNAME_FIELD;
exports.MIDDLENAME_FIELD = MIDDLENAME_FIELD;
exports.LASTNAME_FIELD = LASTNAME_FIELD;
exports.NICKNAME_FIELD = NICKNAME_FIELD;
exports.FIELDS = FIELDS;
exports.FIELDCODES = FIELDCODES;

},{}],4:[function(require,module,exports){
var cleaning = require('./cleaning');
var buildinghash = require('./building-hash');
var hash2rule = require('../rule-writers/hash2rule');
// crew_members is an object expected
// to be in the following format
// [
//   {"userId": "uid001", 
//    "frist_name": "elroy", 
//     "middle_name": "john",
//     "last_name":"jetson",
//     "nick_name":"cool boy"},
//   {"userId": "uid001", 
//    "frist_name": "judy", 
//     "middle_name": "july",
//     "last_name":"jetson",
//     "nick_name":"cool lady"},
// ]
// Defining crew2rule function
function crew2rule(crew_members) {
    if (crew_members == null || !Array.isArray(crew_members))
        throw ('input to crew2rule is null or is not an array');
    var crew = crew_members.slice();
    cleaning.clean_and_validate(crew);
    var crew_hash = buildinghash.build_userids_per_literal(crew);
    var rule = hash2rule.hash2rule(crew_hash);
    return rule;
}
// Exporting crew2rule
exports.crew2rule = crew2rule;

},{"../rule-writers/hash2rule":7,"./building-hash":1,"./cleaning":2}],5:[function(require,module,exports){
var otherutils = require('../utils/other-utils');
var stringutils = require('../utils/string-utils');
// Adds to result all literal combinations of a specific length
function run_permutation_algorithm(all_names, id, string_so_far, fields_so_far, result) {
    // converting all_names to array
    var fields_array = [];
    var names_array = [];
    otherutils.get_keys_and_values(all_names, fields_array, names_array);
    // For each name:
    //     add to string_so_far, add to result
    //     remove the name from all_names
    //     recursively run_permuation_algorithm
    //     return name to all_names
    for (var i = 0; i < names_array.length; ++i) {
        var name = names_array[i];
        var field = fields_array[i];
        var tmp_string_so_far = stringutils.concat_strings_with_space(string_so_far, name);
        fields_so_far.push(field);
        if (!result.hasOwnProperty(tmp_string_so_far))
            result[tmp_string_so_far] = {};
        result[tmp_string_so_far][id] = fields_so_far.slice();
        delete all_names[field];
        run_permutation_algorithm(all_names, id, tmp_string_so_far, fields_so_far, result);
        all_names[field] = name;
        fields_so_far.pop();
    }
}
exports.run_permutation_algorithm = run_permutation_algorithm;

},{"../utils/other-utils":8,"../utils/string-utils":9}],6:[function(require,module,exports){
var constants = require('./constants');
// priority sequences, from top priority
// to low priority
var priorities = [
    [constants.NICKNAME_FIELD, constants.MIDDLENAME_FIELD, constants.LASTNAME_FIELD],
    [constants.FIRSTNAME_FIELD, constants.MIDDLENAME_FIELD, constants.LASTNAME_FIELD],
    [constants.NICKNAME_FIELD, constants.LASTNAME_FIELD],
    [constants.FIRSTNAME_FIELD, constants.LASTNAME_FIELD],
    [constants.NICKNAME_FIELD, constants.MIDDLENAME_FIELD],
    [constants.FIRSTNAME_FIELD, constants.MIDDLENAME_FIELD],
    [constants.NICKNAME_FIELD],
    [constants.FIRSTNAME_FIELD],
    [constants.LASTNAME_FIELD],
    [constants.MIDDLENAME_FIELD],
    [constants.LASTNAME_FIELD, constants.NICKNAME_FIELD],
    [constants.LASTNAME_FIELD, constants.FIRSTNAME_FIELD],
    [constants.NICKNAME_FIELD, constants.LASTNAME_FIELD, constants.MIDDLENAME_FIELD],
    [constants.FIRSTNAME_FIELD, constants.LASTNAME_FIELD, constants.MIDDLENAME_FIELD],
    [constants.MIDDLENAME_FIELD, constants.LASTNAME_FIELD, constants.NICKNAME_FIELD],
    [constants.MIDDLENAME_FIELD, constants.LASTNAME_FIELD, constants.FIRSTNAME_FIELD],
    [constants.MIDDLENAME_FIELD, constants.NICKNAME_FIELD, constants.LASTNAME_FIELD],
    [constants.MIDDLENAME_FIELD, constants.FIRSTNAME_FIELD, constants.LASTNAME_FIELD],
    [constants.LASTNAME_FIELD, constants.NICKNAME_FIELD, constants.MIDDLENAME_FIELD],
    [constants.LASTNAME_FIELD, constants.FIRSTNAME_FIELD, constants.MIDDLENAME_FIELD],
    [constants.LASTNAME_FIELD, constants.MIDDLENAME_FIELD, constants.NICKNAME_FIELD],
    [constants.LASTNAME_FIELD, constants.MIDDLENAME_FIELD, constants.FIRSTNAME_FIELD],
    [constants.MIDDLENAME_FIELD, constants.NICKNAME_FIELD],
    [constants.MIDDLENAME_FIELD, constants.FIRSTNAME_FIELD]
];
function array_equals(arr1, arr2) {
    if (arr1.length != arr2.length)
        return false;
    for (var i = 0; i < arr1.length; ++i) {
        if (arr1[i] != arr2[i])
            return false;
    }
    return true;
}
function has_nick_name(uid, crew) {
    for (var i = 0; i < crew.length; ++i) {
        if (crew[i][constants.USERID_FIELD] == uid && crew[i].hasOwnProperty(constants.NICKNAME_FIELD))
            return true;
    }
    return false;
}
function get_priority(uid, name_sequence, crew) {
    var priority = 0.0;
    for (var i = 0; i < priorities.length; ++i) {
        if (array_equals(name_sequence, priorities[i])) {
            if (has_nick_name(uid, crew)) {
                return priority + 0.5;
            }
            return priority;
        }
        ++priority;
    }
    return priority;
}
function get_uids_with_priorities(literal_hash, crew) {
    var result = [];
    for (var uid in literal_hash) {
        result.push({ key: uid, value: get_priority(uid, literal_hash[uid], crew) });
    }
    return result;
}
function sort_by_value(obj1, obj2) {
    if (obj1.value < obj2.value)
        return 0;
    if (obj1.value == obj2.value && obj1.key < obj2.key)
        return 0;
    return 1;
}
function get_uids_in_priority_order(literal_hash, crew) {
    var uids_with_priorities = get_uids_with_priorities(literal_hash, crew);
    uids_with_priorities = uids_with_priorities.sort(sort_by_value);
    var arr = [];
    for (var i = 0; i < uids_with_priorities.length; ++i)
        arr.push(uids_with_priorities[i].key);
    return arr;
}
exports.get_uids_in_priority_order = get_uids_in_priority_order;

},{"./constants":3}],7:[function(require,module,exports){
function hash2rule(hash) {
    var result = '';
    result += '!use_equivalent_words = true;\n';
    result += 'TopRule = \n';
    var firstone = true;
    for (var key in hash) {
        if (!firstone)
            result += '    |\n';
        else
            firstone = false;
        var split_key = key.split(" ");
        split_key[split_key.length - 1] = '[(' + split_key[split_key.length - 1];
        result += '    (' + split_key.join(" ") + ')?(?(\\\')s)])\n';
        result += add_nl_returns(hash[key], result);
    }
    result += ';\n';
    return result;
}
function add_nl_returns(nl_hash) {
    var result = '';
    for (var key in nl_hash)
        result += "        {" + key + "='" + nl_hash[key] + "'}\n";
    return result;
}
exports.hash2rule = hash2rule;

},{}],8:[function(require,module,exports){
function get_keys(hash) {
    var result = [];
    for (var key in hash)
        if (hash.hasOwnProperty(key))
            result.push(key);
    return result;
}
function get_keys_and_values(hash, keys, values) {
    keys.length = 0;
    values.length = 0;
    for (var key in hash)
        if (hash.hasOwnProperty(key)) {
            keys.push(key);
            values.push(hash[key]);
        }
}
function object_shallow_copy(a) {
    var result = {};
    for (var key in a)
        if (a.hasOwnProperty(key))
            result[key] = a[key];
    return result;
}
exports.get_keys = get_keys;
exports.get_keys_and_values = get_keys_and_values;
exports.object_shallow_copy = object_shallow_copy;

},{}],9:[function(require,module,exports){
function concat_strings_with_space(st1, st2) {
    if (st1 == '')
        return st2;
    else if (st2 == '')
        return st1;
    else
        return st1 + ' ' + st2;
}
function remove_extra_spaces(st) {
    if (typeof (st) != 'string')
        throw ('A specific field value of a specific crew_member is not a string');
    // pos: position in the st as we parse it
    // len: lenght of st
    var len = st.length;
    var pos = 0;
    // If we reached the end of st
    // return
    if (pos == len)
        return st;
    // initializing result to return
    var res = '';
    // Ignore initial spaces
    while (pos < len && is_blank(st[pos]))
        ++pos;
    var parsingSpaces;
    while (pos < len) {
        // At this point we are confident that we
        // have a non-blank character
        res = res + st[pos];
        ++pos;
        // If we find spaces we ignore them
        while (pos < len && is_blank(st[pos])) {
            parsingSpaces = true;
            ++pos;
        }
        // If there are characters after the 
        // spaces, then we add a space to the result
        if (parsingSpaces && pos < len) {
            res += ' ';
            parsingSpaces = false;
        }
    }
    // returning result
    return res;
}
function is_blank(c) {
    if (c == ' ' || c == '\t' || c == '\n' || c == '\r')
        return true;
    return false;
}
function is_ignored_character(c) {
    if (c == ',' ||
        c == '.' ||
        c == '"' ||
        c == '\\' ||
        c == '*' ||
        c == '=' ||
        c == ';' ||
        c == '+' ||
        c == '}' ||
        c == '!' ||
        c == '{' ||
        c == '[' ||
        c == ']' ||
        c == '<' ||
        c == '>' ||
        c == '(' ||
        c == '@' ||
        c == ')' ||
        c == '|' ||
        c == '?' ||
        c == '~' ||
        c == '$' ||
        c == '#' ||
        c == '^')
        return true;
    return false;
}
function is_escaped_character(c) {
    if (c == "'")
        return true;
    return false;
}
function process_special_characters(st) {
    var result = '';
    var tmp = '';
    // Ingoring some special characters
    for (var i = 0; i < st.length; ++i)
        if (!is_ignored_character(st[i]))
            tmp += st[i];
    // escaping some special characters
    for (i = 0; i < tmp.length; ++i)
        if (is_escaped_character(tmp[i]))
            result += '\\' + tmp[i];
        else
            result += tmp[i];
    return result;
}
// Given the crew members, it returns a hash
exports.remove_extra_spaces = remove_extra_spaces;
exports.is_blank = is_blank;
exports.concat_strings_with_space = concat_strings_with_space;
exports.process_special_characters = process_special_characters;

},{}],10:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var crew2rule_js_1 = require("./crew2rule/crew2rule.js");
exports.crew2rule = crew2rule_js_1.crew2rule;

},{"./crew2rule/crew2rule.js":4}]},{},[10])(10)
});
