// jibo-typed-events/lib/jibo-typed-events.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboTypedEvents = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EmitterMode;
(function (EmitterMode) {
    EmitterMode[EmitterMode["ALL_HANDLERS"] = 'ALL_HANDLERS'] = "ALL_HANDLERS";
    EmitterMode[EmitterMode["FIRST_HANDLER"] = 'FIRST_HANDLER'] = "FIRST_HANDLER";
    EmitterMode[EmitterMode["LAST_HANDLER"] = 'LAST_HANDLER'] = "LAST_HANDLER";
})(EmitterMode = exports.EmitterMode || (exports.EmitterMode = {}));
var Event = (function () {
    /**
     * Creates an Event object
     * @param {String} name Name of event
     */
    function Event(name) {
        this.name = name;
        this._onceEvents = new Map();
        this._events = [];
        this._maxListeners = 10;
        /**
         * Mode of order to emit events.
         * In Stack mode, only emits to last event subscribed
         * @type {enum}
         */
        this._emitterMode = EmitterMode.ALL_HANDLERS;
    }
    /**
     * Returns a list of currently subscribed handlers
     * @returns {Handler<Type>[]}
     */
    Event.prototype.listeners = function () {
        return this._events.slice();
    };
    /**
     * Sets the maximum number of listeners
     * @param {number} n
     * @returns {Event} this
     */
    Event.prototype.setMaxListeners = function (n) {
        this._maxListeners = n;
        return this;
    };
    /**
     * Gets the maximum number of listeners
     * @returns {number}
     */
    Event.prototype.getMaxListeners = function () {
        return this._maxListeners;
    };
    /**
     * Subscribes a handler function to only the next event
     * @return {Event} this
     */
    Event.prototype.once = function (handler) {
        var _this = this;
        var handlerWrapper = this._onceEvents.get(handler);
        if (!handlerWrapper) {
            handlerWrapper = function (data) {
                _this._onceEvents.delete(handler);
                _this.removeListener(handlerWrapper);
                handler(data);
            };
            this._onceEvents.set(handler, handlerWrapper);
        }
        return this.on(handlerWrapper);
    };
    /**
     * Subscribes a handler function to this event
     * @return {Event} this
     */
    Event.prototype.on = function (handler) {
        if (this._events.length >= this._maxListeners) {
            var error = new Error();
            console.error("Maximum listeners '" + this._maxListeners + "' reached");
            console.error(error.stack);
        }
        else {
            this._events.push(handler);
            return this;
        }
    };
    /**
     * Unsubscribes most recently added subscriber
     * @return {Event} this
     */
    Event.prototype.removeLastListener = function () {
        this._events.splice(this._events.length - 1, 1);
    };
    /**
     * Unsubscribes oldest added subscriber
     * @return {Event} this
     */
    Event.prototype.removeFirstListener = function () {
        // if events is empty, no error; still gives empty array.
        this._events.splice(0, 1);
    };
    /**
     * Subscribes a handler function to this event
     * @return {Event} this
     */
    Event.prototype.addListener = function (handler) {
        return this.on(handler);
    };
    /**
     * Unsubscribes a handler function from this event
     * @param {Handler} handler
     * @return {Event} this
     */
    Event.prototype.removeListener = function (handler) {
        // We remove the handler if it has been provided through `on`
        var index = this._events.indexOf(handler);
        while (index !== -1) {
            this._events.splice(index, 1);
            index = this._events.indexOf(handler, index);
        }
        // We remove the handler if it has been provided through `once`
        var onceHandler = this._onceEvents.get(handler);
        if (onceHandler) {
            var index_1 = this._events.indexOf(onceHandler);
            while (index_1 !== -1) {
                this._events.splice(index_1, 1);
                index_1 = this._events.indexOf(onceHandler, index_1);
            }
        }
        return this;
    };
    /**
     * Unsubscribes a handler function from this event (alias for removeListener)
     * @param {Handler} handler
     * @return {Event} this
     */
    Event.prototype.off = function (handler) {
        return this.removeListener(handler);
    };
    /**
     * Unsubscribes all handler functions from this event
     * @return {Event} this
     */
    Event.prototype.removeAllListeners = function () {
        this._events = [];
        this._onceEvents.clear();
        return this;
    };
    /**
     * Fires off an instance of this event with optional data
     * @param {Type} [data] The data of the event
     * @return {boolean} true if the event has any listeners, false otherwise
     */
    Event.prototype.emit = function (data) {
        if (this._emitterMode === EmitterMode.ALL_HANDLERS) {
            // We start by creating a copy of handlers
            // this is required to make sure that no handler modifies the
            // underlying collection which could result in one or more
            // handlers not being called
            var copy = this.listeners();
            for (var i = 0; i < copy.length; i++) {
                var handler = copy[i];
                handler(data);
            }
        }
        else if (this._emitterMode === EmitterMode.LAST_HANDLER) {
            var handler = this._events[this._events.length - 1];
            if (handler) {
                handler(data);
            }
            return !!handler;
        }
        else if (this._emitterMode === EmitterMode.FIRST_HANDLER) {
            var handler = this._events[0];
            if (handler) {
                handler(data);
            }
            return !!handler;
        }
        return this._events.length > 0;
    };
    /**
     * Sets mode to stack; event only emits to last subscriber
     * @param {enum} mode of event to set
     */
    Event.prototype.setMode = function (mode) {
        if (!EmitterMode[mode]) {
            throw new Error("Emitter mode of typed events must be ALL_HANDLERS, FIRST_HANDLER, or LAST_HANDLER");
        }
        this._emitterMode = mode;
    };
    /**
     * Waits for event to occur.
     * @param {number} [timeoutMs=-1] Timeout while waiting for event
     * @return {Promise<any>} Resolved with event payload or rejected with 'TIMEOUT' if times out
     */
    Event.prototype.waitFor = function (timeoutMs) {
        var _this = this;
        if (timeoutMs === void 0) { timeoutMs = -1; }
        return new Promise(function (resolve, reject) {
            var timeoutHandle;
            var eventHandler = function (data) {
                if (timeoutHandle) {
                    clearTimeout(timeoutHandle);
                }
                resolve(data);
            };
            // Connect timeout handler
            if (timeoutMs > 0) {
                timeoutHandle = setTimeout(function () {
                    _this.removeListener(eventHandler);
                    reject('TIMEOUT');
                }, timeoutMs);
            }
            _this.once(eventHandler);
        });
    };
    /**
     * Returns the number of listeners
     * @returns {number}
     */
    Event.prototype.listenerCount = function () {
        return this._events.length;
    };
    return Event;
}());
exports.default = Event;

},{}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var events = require("events");
var EventContainer = (function () {
    function EventContainer() {
        /**
         * A regular NodeJS event emitter for untyped
         * events
         * @type {EventEmitter}
         */
        this.untyped = new events.EventEmitter();
    }
    /**
     * Removes all listeners of all events that are top level
     * members of the EventContainer
     */
    EventContainer.prototype.removeAllListeners = function () {
        this.untyped.removeAllListeners();
        this.forEachTypedEvent(function (event) {
            event.removeAllListeners();
        });
    };
    /**
     * Iterates through all event properties of this event container
     * @param {Function} handler
     */
    EventContainer.prototype.forEachTypedEvent = function (handler) {
        var _this = this;
        Object.keys(this).forEach(function (key) {
            var value = _this[key];
            if (value.constructor && value.constructor.name === 'Event') {
                handler(value);
            }
        });
    };
    return EventContainer;
}());
exports.default = EventContainer;

},{"events":undefined}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Event_1 = require("./Event");
exports.Event = Event_1.default;
var EventContainer_1 = require("./EventContainer");
exports.EventContainer = EventContainer_1.default;
var Event_2 = require("./Event");
exports.EmitterMode = Event_2.EmitterMode;

},{"./Event":1,"./EventContainer":2}]},{},[3])(3)
});
