// jibo-cai-utils/lib/jibo-cai-utils.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboCaiUtils = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
/**
 * Created on 7/20/16.
 * @author Tom Donahue <tom.donahue@jibo.com>
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ActionTimelineState;
(function (ActionTimelineState) {
    ActionTimelineState[ActionTimelineState["DISPATCHING"] = 0] = "DISPATCHING";
    ActionTimelineState[ActionTimelineState["NOT_DISPATCHING"] = 1] = "NOT_DISPATCHING";
    ActionTimelineState[ActionTimelineState["CANCELLED"] = 2] = "CANCELLED";
    ActionTimelineState[ActionTimelineState["FAST_FORWARD"] = 3] = "FAST_FORWARD";
})(ActionTimelineState = exports.ActionTimelineState || (exports.ActionTimelineState = {}));
var ActionTimeline = (function () {
    function ActionTimeline() {
        this._state = ActionTimelineState.NOT_DISPATCHING;
    }
    /**
     * Dispatch a timeline of actions
     *
     * @param {TimelineElement[]} timeline
     * @returns {Promise<void>}
     */
    ActionTimeline.prototype.dispatchTimeline = function (timeline) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var startTime;
            return __generator(this, function (_a) {
                this._state = ActionTimelineState.DISPATCHING;
                startTime = process.hrtime();
                this._promiseChain = Promise.resolve();
                timeline.forEach(function (el) {
                    _this._promiseChain = _this._promiseChain
                        .then(function () { return _this._dispatchTimeline(el, startTime); });
                });
                this._promiseChain = this._promiseChain.then(function () {
                    _this._state = ActionTimelineState.NOT_DISPATCHING;
                    _this._timeoutID = null;
                    _this._next = null;
                });
                return [2 /*return*/, this._promiseChain];
            });
        });
    };
    ActionTimeline.prototype._dispatchTimeline = function (el, startTime) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var timeDiff, passedTimeMs, timeToWait;
            return __generator(this, function (_a) {
                timeDiff = process.hrtime(startTime);
                passedTimeMs = timeDiff[0] * 1e3 + timeDiff[1] * 1e-6;
                timeToWait = Math.max(0, el.timeMs - passedTimeMs);
                return [2 /*return*/, this._wait(timeToWait).then(function () {
                        if (_this._state !== ActionTimelineState.CANCELLED) {
                            el.action();
                        }
                    })];
            });
        });
    };
    /**
     * Waits for timeMs and then resolves the promise
     *
     * @param {number} timeMs Time to wait in ms
     * @returns {Promise<void>}
     */
    ActionTimeline.prototype._wait = function (timeMs) {
        var _this = this;
        return new Promise(function (res) {
            _this._next = res;
            if (_this._state === ActionTimelineState.FAST_FORWARD || _this._state === ActionTimelineState.CANCELLED) {
                _this._next();
            }
            else {
                _this._timeoutID = setTimeout(res, timeMs);
            }
        });
    };
    /**
     * Cancel a timeline of actions (remaining actions will not be dispatched)
     *
     * @returns {void}
     */
    ActionTimeline.prototype.cancel = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                if (this._state === ActionTimelineState.DISPATCHING) {
                    this._state = ActionTimelineState.CANCELLED;
                    if (this._timeoutID) {
                        clearTimeout(this._timeoutID);
                        this._timeoutID = null;
                        this._next();
                    }
                    return [2 /*return*/, this._promiseChain.then(function () {
                            _this._state = ActionTimelineState.NOT_DISPATCHING;
                            _this._timeoutID = null;
                            _this._next = null;
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Fast-forward a timeline (i.e. trigger all remaining actions in sequence immediately)
     *
     * @returns {void}
     */
    ActionTimeline.prototype.fastForward = function () {
        if (this._state === ActionTimelineState.DISPATCHING) {
            this._state = ActionTimelineState.FAST_FORWARD;
            if (this._timeoutID) {
                clearTimeout(this._timeoutID);
                this._timeoutID = null;
                this._next();
            }
        }
    };
    /**
     * Get state of ActionTimeline
     *
     * @returns {ActionTimelineState}
     */
    ActionTimeline.prototype._getState = function () {
        return this._state;
    };
    return ActionTimeline;
}());
exports.ActionTimeline = ActionTimeline;

},{}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CacheUtils = (function () {
    function CacheUtils() {
    }
    /**
     * Ensures that the global cache exists for jibo.loader.
     * @param {Jibo} jibo The Jibo Runtime instance on which to ensure global cache is initted.
     */
    CacheUtils.initGlobalCache = function (jibo) {
        // Set up global cache if it doesn't exist
        if (!jibo.loader.getCache(CacheUtils.GlobalCacheName)) {
            var active = jibo.loader.activeCache;
            jibo.loader.addCache(CacheUtils.GlobalCacheName);
            jibo.loader.activeCache = active;
        }
    };
    return CacheUtils;
}());
/**
 * Name of global cache
 * @type {string}
 */
CacheUtils.GlobalCacheName = 'global';
exports.CacheUtils = CacheUtils;

},{}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var events = require("events");
exports.CancelTokenEvents = {
    UNPAUSED: 'UNPAUSED',
    ALL_DONE: 'ALL_DONE',
    CANCELED: 'CANCELED'
};
/**
 * @description Represests
 * @class CancelToken
 */
var CancelToken = (function () {
    /**
     * Creates an instance of CancelToken.
     * @param {Promise<T>} promise - The promise to wrap
     * @param {CancelHandler} [onCancel] - Handler function for cancel events
     * @memberOf CancelToken
     */
    function CancelToken(promise, onCancel) {
        var _this = this;
        this._paused = false;
        this._canceled = false;
        this._emitter = new events.EventEmitter();
        this._onCancel = onCancel;
        this.promise = new Promise(function (resolve, reject) {
            var done = function (error, data) {
                _this._emitter.removeAllListeners();
                if (error) {
                    reject(error);
                }
                else {
                    resolve(data);
                }
            };
            promise.then(function (data) {
                if (_this._paused) {
                    _this._emitter.once(exports.CancelTokenEvents.UNPAUSED, function () { return done(null, data); });
                }
                else if (!_this._canceled) {
                    done(null, data);
                }
            }).catch(function (err) {
                if (_this._paused) {
                    _this._emitter.once(exports.CancelTokenEvents.UNPAUSED, function () { return done(err); });
                }
                else if (!_this._canceled) {
                    done(err);
                }
            });
        });
    }
    /**
     * Sets the paused status of this promise session
     * @param {boolean} paused
     * @memberOf PromiseSession
     */
    CancelToken.prototype.setPaused = function (paused) {
        var oldPaused = this._paused;
        this._paused = paused;
        if (oldPaused && !paused) {
            this._emitter.emit(exports.CancelTokenEvents.UNPAUSED);
        }
    };
    /**
     * Returns the paused status of this promise session
     * @returns {boolean}
     * @memberOf PromiseSession
     */
    CancelToken.prototype.isPaused = function () {
        return this._paused;
    };
    /**
     * Is this token canceled
     * @returns {boolean}
     * @memberOf CancelToken
     */
    CancelToken.prototype.isCanceled = function () {
        return this._canceled;
    };
    /**
     * Cancels the current session. Session should not be reused after calling cancel
     * @memberOf PromiseSession
     */
    CancelToken.prototype.cancel = function () {
        this._canceled = true;
        if (this._onCancel) {
            this._onCancel();
        }
    };
    return CancelToken;
}());
exports.CancelToken = CancelToken;

},{"events":undefined}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var events = require("events");
var CancelToken_1 = require("./CancelToken");
/**
 * @description A class that represents a session for a set of promise tokens
 * @class PromiseSession
 */
var CancelTokenSession = (function () {
    function CancelTokenSession() {
        this._outstanding = new Set();
        this._emitter = new events.EventEmitter();
    }
    /**
     * Pauses all
     * @param {boolean} paused
     * @memberOf CancelTokenSession
     */
    CancelTokenSession.prototype.setPaused = function (paused) {
        this._outstanding.forEach(function (token) { return token.setPaused(paused); });
    };
    /**
     * Cancels the current session. Session should not be reused after calling cancel
     * @memberOf CancelTokenSession
     */
    CancelTokenSession.prototype.cancel = function () {
        this._outstanding.forEach(function (token) {
            token.cancel();
        });
        this._outstanding.clear();
        this._emitter.emit(CancelToken_1.CancelTokenEvents.CANCELED);
    };
    /**
     * Number of outstanding promises
     * @returns {number}
     * @memberOf CancelTokenSession
     */
    CancelTokenSession.prototype.size = function () {
        return this._outstanding.size;
    };
    /**
     * Waits for all outstanding promises to complete
     * This refers to the internal promises, not the wrapped promises that were returned
     * @returns {Promise<void>}
     * @memberOf CancelTokenSession
     */
    CancelTokenSession.prototype.waitForAll = function () {
        var _this = this;
        if (this._outstanding.size === 0) {
            return Promise.resolve();
        }
        else {
            return new Promise(function (resolve, reject) {
                var allDoneHandler = function () { return done(); };
                var canceledHandler = function () { return done(new Error("Promise session canceled")); };
                // When done, either with error or no error, we clean up listeners and resolve/reject
                var done = function (error) {
                    _this._emitter.removeListener(CancelToken_1.CancelTokenEvents.ALL_DONE, allDoneHandler);
                    _this._emitter.removeListener(CancelToken_1.CancelTokenEvents.CANCELED, canceledHandler);
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve();
                    }
                };
                // We wait for either of these events
                _this._emitter.once(CancelToken_1.CancelTokenEvents.ALL_DONE, allDoneHandler);
                _this._emitter.once(CancelToken_1.CancelTokenEvents.CANCELED, canceledHandler);
            });
        }
    };
    /**
     * Wraps a promise with a token and adds to this session
     * @param {Promise<T>} promise
     * @returns {Promise<T>}
     * @memberOf CancelTokenSession
     */
    CancelTokenSession.prototype.wrap = function (promise) {
        var token = new CancelToken_1.CancelToken(promise);
        return this.addToken(token);
    };
    /**
     * Adds a cancel token to this session
     * @template T
     * @param {CancelToken} token
     * @memberOf CancelTokenSession
     */
    CancelTokenSession.prototype.addToken = function (token) {
        var _this = this;
        this._outstanding.add(token);
        // We remove the token from the list out outstanding promises regardless of outcome
        return token.promise
            .then(function (data) {
            _this._removeOutstanding(token);
            return data; // Pass data forward to the caller
        })
            .catch(function (error) {
            _this._removeOutstanding(token);
            throw error; // Allows caller code to handle error
        });
    };
    CancelTokenSession.prototype._removeOutstanding = function (token) {
        this._outstanding.delete(token);
        if (this._outstanding.size === 0) {
            this._emitter.emit(CancelToken_1.CancelTokenEvents.ALL_DONE);
        }
    };
    return CancelTokenSession;
}());
exports.CancelTokenSession = CancelTokenSession;

},{"./CancelToken":3,"events":undefined}],5:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 5/12/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var path = require("path");
var PromiseUtils_1 = require("./PromiseUtils");
var TestUtils_1 = require("./TestUtils");
var PathUtils_1 = require("./PathUtils");
var prify = PromiseUtils_1.PromiseUtils.promisify;
var onRobot = fs.existsSync('/var/jibo/identity.json');
var FileUtils = (function () {
    function FileUtils() {
    }
    /**
     * Returns the root path of project
     * @param {NodeModule|string} pathOrModule Path in project or NodeModule
     * @returns {string}
     * @deprecated
     */
    FileUtils.getProjectRoot = function (pathOrModule) {
        console.warn("'FileUtils.getProjectRoot' is deprecated, please use 'PathUtils.getProjectRoot'");
        return PathUtils_1.PathUtils.getProjectRoot(pathOrModule);
    };
    /**
     * Returns the main export of a project, same as you would get if you
     * did `let exp = require('project-name');`
     * @param {NodeModule|string} pathOrModule Path in project or NodeModule
     * @returns {any}
     * @deprecated
     */
    FileUtils.getProjectMainExport = function (pathOrModule) {
        console.warn("'FileUtils.getProjectMainExport' is deprecated, please use 'TestUtils.getProjectMainExport'");
        return TestUtils_1.TestUtils.getProjectMainExport(pathOrModule);
    };
    /**
     * Read file from disk and return promise
     * fs.readFile doesn't ensure closing file descriptors and can run
     * out of allowed maximum number of open files
     * @param {string} filePath Path to file
     * @param {string} [encoding='utf8'] Encoding of file
     * @param {number} [chunkSize=512] How many bytes to read from file at a time
     * @returns {Promise<string>}
     */
    FileUtils.readFile = function (filePath, encoding, chunkSize) {
        if (encoding === void 0) { encoding = 'utf8'; }
        if (chunkSize === void 0) { chunkSize = 512; }
        return __awaiter(this, void 0, void 0, function () {
            var fd, buffer, bufferSize, stats, bytesRead_1, _loop_1, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, prify(function (h) { return fs.open(filePath, 'r', h); })];
                    case 1:
                        fd = _a.sent();
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 7, 8, 10]);
                        return [4 /*yield*/, prify(function (h) { return fs.fstat(fd, h); })];
                    case 3:
                        stats = _a.sent();
                        bufferSize = stats.size;
                        buffer = new Buffer(bufferSize);
                        bytesRead_1 = 0;
                        _loop_1 = function () {
                            var chunkSizeToUse;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        chunkSizeToUse = chunkSize;
                                        if ((bytesRead_1 + chunkSize) > bufferSize) {
                                            chunkSizeToUse = (bufferSize - bytesRead_1);
                                        }
                                        return [4 /*yield*/, prify(function (h) { return fs.read(fd, buffer, bytesRead_1, chunkSizeToUse, bytesRead_1, h); })];
                                    case 1:
                                        _a.sent();
                                        bytesRead_1 += chunkSizeToUse;
                                        return [2 /*return*/];
                                }
                            });
                        };
                        _a.label = 4;
                    case 4:
                        if (!(bytesRead_1 < bufferSize)) return [3 /*break*/, 6];
                        return [5 /*yield**/, _loop_1()];
                    case 5:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 6: return [3 /*break*/, 10];
                    case 7:
                        error_1 = _a.sent();
                        throw error_1;
                    case 8: return [4 /*yield*/, prify(function (h) { return fs.close(fd, h); })];
                    case 9:
                        _a.sent();
                        return [7 /*endfinally*/];
                    case 10:
                        if (buffer) {
                            return [2 /*return*/, buffer.toString(encoding, 0, bufferSize)];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Finds all files of a certain extension in directories rooted at a certain
     * base directory
     * @param {String} baseDir The base directory path of the search
     * @param {String} extension The extension of files to find
     */
    FileUtils.findAllFilesWithExt = function (baseDir, extension) {
        var selector = function (fileName) {
            var _extension = extension.toLowerCase();
            var ext = path.extname(fileName.toLowerCase());
            return (ext[0] === '.' && ext.substr(1) === _extension);
        };
        return FileUtils.findAllFiles(baseDir, selector);
    };
    /**
     * Finds all files of a certain extension in directories rooted at a certain
     * base directory
     * @param {String} baseDir The base directory path of the search
     * @param {Function} selector A function which takes a filename and returns boolean
     * @return {Promise<string[]>}
     */
    FileUtils.findAllFiles = function (baseDir, selector) {
        return __awaiter(this, void 0, void 0, function () {
            var files;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        files = [];
                        return [4 /*yield*/, FileUtils._findAllFiles(baseDir, selector, files)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, files];
                }
            });
        });
    };
    FileUtils._findAllFiles = function (baseDir, selector, out) {
        return __awaiter(this, void 0, void 0, function () {
            var directories, _fileNames, filePromiseGenerators, pr;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        directories = [];
                        return [4 /*yield*/, prify(function (h) { return fs.readdir(baseDir, h); })];
                    case 1:
                        _fileNames = _a.sent();
                        filePromiseGenerators = _fileNames.map(function (f) {
                            var fullPath = path.resolve(baseDir, f);
                            return function () {
                                var fd;
                                return PromiseUtils_1.PromiseUtils.promisify(function (h) { return fs.open(fullPath, 'r', h); })
                                    .then(function (_fd) {
                                    fd = _fd;
                                    // Here we gather stats on the file descriptor
                                    return prify(function (h) { return fs.fstat(fd, h); });
                                })
                                    .then(function (stat) {
                                    // We determine which files are directories and which
                                    // ones comply with our selector
                                    if (stat.isDirectory()) {
                                        directories.push(fullPath);
                                    }
                                    else if (selector(fullPath)) {
                                        out.push(fullPath);
                                    }
                                    // Now we close the file descriptor
                                    return prify(function (h) { return fs.close(fd, h); });
                                });
                            };
                        });
                        pr = Promise.resolve();
                        filePromiseGenerators.forEach(function (promiseGen) {
                            pr = pr.then(function () { return promiseGen(); });
                        });
                        // Here we wait for processing the files in order to find all the directories
                        return [4 /*yield*/, pr];
                    case 2:
                        // Here we wait for processing the files in order to find all the directories
                        _a.sent();
                        // We also recurse linearly through all the directories
                        directories.forEach(function (dir) {
                            pr = pr.then(function () { return FileUtils._findAllFiles(dir, selector, out); });
                        });
                        // Now we wait for all the recursive calls to take place
                        return [4 /*yield*/, pr];
                    case 3:
                        // Now we wait for all the recursive calls to take place
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Checks to see if we are on the robot or in simulator
     */
    FileUtils.onRobot = function () {
        return onRobot;
    };
    /**
     * Search backward in path hierarchy.
     *
     * Example1: inputPath: '/root/foo/bar/bloop', name: 'glop.js'
     * could output '/root/foo/glop.js' if it existed there, null otherwise
     *
     * Example2: inputPath: '/root/foo/bar/bloop', name: 'blip.js', nestedDirectories: ['test']
     * could output '/root/test/blip.js' if it existed there, null otherwise
     *
     * @param {string} inputPath Starting point for search
     * @param {string} name Folder or file name to find
     * @param {string[]} [nestedDirectories] optional nested directories to search into at each level
     * @returns {Promise<string|void>} Resolves with path or null if nothing found
     */
    FileUtils.searchBackwardInPath = function (inputPath, name, nestedDirectories) {
        return __awaiter(this, void 0, void 0, function () {
            var pathToUse, paths, ind, nested;
            return __generator(this, function (_a) {
                pathToUse = path.dirname(inputPath);
                paths = [];
                FileUtils._gatherBackwardPaths(pathToUse, paths);
                ind = 0;
                nested = nestedDirectories || [];
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        var find = function () {
                            FileUtils._searchBackward(paths[ind++], name, nested)
                                .then(function (result) {
                                if (result) {
                                    resolve(result);
                                }
                                else if (ind < paths.length) {
                                    find();
                                }
                                else {
                                    resolve(null);
                                }
                            }).catch(reject);
                        };
                        find();
                    })];
            });
        });
    };
    FileUtils._gatherBackwardPaths = function (inputPath, collection) {
        collection.push(path.resolve(inputPath));
        var next = path.resolve(path.join(inputPath, '..'));
        if (next === inputPath) {
            return;
        }
        else {
            FileUtils._gatherBackwardPaths(next, collection);
        }
    };
    FileUtils._searchBackward = function (inputPath, name, nestedDirectories) {
        return __awaiter(this, void 0, void 0, function () {
            var paths, ind;
            return __generator(this, function (_a) {
                paths = [path.join(inputPath, name)];
                nestedDirectories.forEach(function (dir) { return paths.push(path.join(inputPath, dir, name)); });
                ind = 0;
                return [2 /*return*/, new Promise(function (resolve) {
                        var find = function () {
                            var pathToUse = paths[ind++];
                            fs.exists(pathToUse, function (exists) {
                                if (exists) {
                                    resolve(pathToUse);
                                }
                                else if (ind < paths.length) {
                                    find();
                                }
                                else {
                                    resolve(null);
                                }
                            });
                        };
                        find();
                    })];
            });
        });
    };
    return FileUtils;
}());
exports.FileUtils = FileUtils;

},{"./PathUtils":6,"./PromiseUtils":9,"./TestUtils":14,"fs":undefined,"path":undefined}],6:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 9/25/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var findRoot = require('find-root');
var PathUtils = (function () {
    function PathUtils() {
    }
    /**
     * Returns the root path of project
     * @param {NodeModule|string} pathOrModule Path in project or NodeModule
     * @returns {string}
     */
    PathUtils.getProjectRoot = function (pathOrModule) {
        if (typeof pathOrModule === 'string') {
            return findRoot(pathOrModule);
        }
        else if (pathOrModule.filename) {
            return findRoot(pathOrModule.filename);
        }
        else {
            throw new Error("Can't find project root, invalid input: '" + pathOrModule + "'");
        }
    };
    return PathUtils;
}());
exports.PathUtils = PathUtils;

},{"find-root":undefined}],7:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 7/6/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var events = require("events");
var path = require("path");
var MockDOFSet = (function () {
    function MockDOFSet(dofs) {
        this._dofs = new Set(dofs);
    }
    MockDOFSet.prototype.hasDOF = function (name) {
        return this._dofs.has(name);
    };
    return MockDOFSet;
}());
var KeysPlayback = (function (_super) {
    __extends(KeysPlayback, _super);
    function KeysPlayback(_animation, _cache, _alias, _options) {
        var _this = _super.call(this) || this;
        _this._animation = _animation;
        _this._cache = _cache;
        _this._alias = _alias;
        _this._options = _options;
        _this._animation.on('EVENT', function (event, instance, payload) {
            _this.emit(event, instance, payload);
        });
        return _this;
    }
    KeysPlayback.prototype.play = function (arg1, arg2, arg3) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var jibo;
            return __generator(this, function (_a) {
                this._options.configBuilder = arg1 || this._options.configBuilder;
                this._options.handleAnimation = arg2 || this._options.handleAnimation;
                this._options.screenCenterOverride = (typeof arg3 === 'boolean') ? arg3 : this._options.screenCenterOverride;
                jibo = this._options.jibo;
                return [2 /*return*/, new Promise(function (resolve) {
                        _this._resolver = resolve;
                        var hasBeenStopped = false;
                        var stopHandler = function () {
                            hasBeenStopped = true;
                            _this._cleanup();
                            _this._resolver();
                        };
                        _this._animation.once('STOPPED', stopHandler);
                        _this._animation.once('CANCELLED', stopHandler);
                        // Check if this animation has any screen content
                        var dofsInScreen = jibo.animate.dofs.SCREEN;
                        var dofsInAnim = new MockDOFSet(_this._animation.builder.dofNames);
                        var dofsNotInAnim = dofsInScreen.minus(dofsInAnim);
                        _this._hasScreenDOFs = (dofsNotInAnim.getDOFs().length !== dofsInScreen.getDOFs().length);
                        // TODO: Switch to system proposed in https://jira.jibo.com/browse/SB-2957 when available
                        if (_this._hasScreenDOFs && !_this._options.disableSetFaceAnim) {
                            jibo.face.eye.animation = _this._animation;
                        }
                        if (_this._options.configBuilder) {
                            _this._options.configBuilder(_this._animation.builder);
                        }
                        // Here, if this animation has screen content and a centering of the screen DOFs was explicitly requested, we center all of the screen DOFs
                        if ((_this._options.screenCenterOverride === false) && _this._hasScreenDOFs) {
                            //console.log("centerRobot()! PlaybackUtils:screenDOFs ("+this._animation.id, this._animation.builder.motion.getMotion().getName()+") on ("+"BODY:"+dofsInScreen.hasDOF("topSection_r")+" EYE:"+dofsInScreen.hasDOF("eyeSubRootBn_t")+" DEFORM:"+dofsInScreen.hasDOF("vertexJoint1_t")+")");
                            jibo.dofArbiter.centerRobot(_this._options.ownerInformation, dofsInScreen, false, function () {
                                if (!hasBeenStopped) {
                                    _this._playAndHandleInstance();
                                }
                            });
                        }
                        else {
                            _this._playAndHandleInstance();
                        }
                    })];
            });
        });
    };
    KeysPlayback.prototype._playAndHandleInstance = function () {
        var dofArbiter = this._options.jibo.dofArbiter;
        this._instance = dofArbiter.playAnimation(this._animation.builder, this._options.ownerInformation, this._options.dofOptions);
        if (this._instance && this._options.handleAnimation) {
            this._options.handleAnimation(this._instance);
        }
    };
    /**
     * Stops playing animation
     */
    KeysPlayback.prototype.stop = function () {
        if (this._instance) {
            this._instance.stop();
        }
    };
    KeysPlayback.prototype._cleanup = function () {
        if (this._animation) {
            if (this._hasScreenDOFs && !this._options.disableSetFaceAnim) {
                this._options.jibo.face.eye.animation = null;
            }
            // Always destroy clean copies
            if (this._animation.isCopy) {
                this._animation.destroy();
                this._animation = null;
            }
            // If this animation was cached and we are not told to override the default unloading, then unload
            if (this._cache) {
                if (!this._options.overrideUnload) {
                    this._options.jibo.loader.unload(this._alias);
                }
            }
            else if (!this._options.overrideDestroy && this._animation) {
                this._animation.destroy();
            }
        }
        this._animation = null;
        this._instance = null;
    };
    return KeysPlayback;
}(events.EventEmitter));
exports.KeysPlayback = KeysPlayback;
var PlaybackUtils = (function () {
    function PlaybackUtils() {
    }
    /**
     * Creates a KeysPlayback object
     * @param {KeysPlaybackOptions} options
     * @returns {Promise<KeysPlayback>}
     */
    PlaybackUtils.createKeysPlayback = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var jibo, PathUtils, assetPack, animPath, alias, cache, animation;
            return __generator(this, function (_a) {
                jibo = options.jibo;
                PathUtils = jibo.utils.PathUtils;
                assetPack = options.assetPack || '';
                if (!options.animPath && !options.animData) {
                    return [2 /*return*/, Promise.reject("Need to provide either 'animPath' and 'animData'")];
                }
                animPath = options.animPath || '';
                alias = PathUtils.setDefaultPath('animations', animPath);
                cache = options.animData ? null : options.cache;
                // If the animation exists in cache then reset it and return it
                if (cache) {
                    animation = jibo.loader.cached(alias, cache);
                    if (animation) {
                        animation = animation.getCleanCopy();
                        return [2 /*return*/, Promise.resolve(new KeysPlayback(animation, cache, alias, options))];
                    }
                }
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        var src = animPath;
                        if (!path.isAbsolute(src)) {
                            src = PathUtils.getAssetUri(alias, assetPack);
                        }
                        var root = options.rootPath || PathUtils.findRoot(src);
                        jibo.loader.load({
                            id: alias,
                            cache: cache,
                            data: options.animData,
                            src: src,
                            upload: true,
                            root: root,
                            type: 'keys',
                            complete: function (error, animation) {
                                if (error) {
                                    return reject(error);
                                }
                                resolve(animation);
                            }
                        });
                    }).then(function (animation) {
                        return new KeysPlayback(animation, cache, alias, options);
                    })];
            });
        });
    };
    return PlaybackUtils;
}());
exports.PlaybackUtils = PlaybackUtils;

},{"events":undefined,"path":undefined}],8:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 4/4/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var PromiseQueueState;
(function (PromiseQueueState) {
    PromiseQueueState[PromiseQueueState["PLAYING"] = 0] = "PLAYING";
    PromiseQueueState[PromiseQueueState["NOT_PLAYING"] = 1] = "NOT_PLAYING";
})(PromiseQueueState = exports.PromiseQueueState || (exports.PromiseQueueState = {}));
var PromiseQueue = (function () {
    function PromiseQueue() {
        this._queueEmptyPr = Promise.resolve();
        this._state = PromiseQueueState.NOT_PLAYING;
        this._queue = [];
    }
    /**
     * Get the state of the PromiseQueue
     */
    PromiseQueue.prototype.getState = function () {
        return this._state;
    };
    /**
     * Add a promise generator
     * @param {PromiseProvider} promiseProvider Promise generator
     */
    PromiseQueue.prototype.add = function (promiseProvider) {
        var _this = this;
        this._queue.push(promiseProvider);
        if (this._state !== PromiseQueueState.PLAYING) {
            this._state = PromiseQueueState.PLAYING;
            this._queueEmptyPr = new Promise(function (resolve, reject) {
                _this._resolver = resolve;
                _this._rejector = reject;
            });
            this._playNext();
        }
    };
    /**
     * Returns a promise that gets resolved either immediately if queue is
     * empty or as soon as the queue empties and/or rejects at first rejection of any
     * promise in the queue
     *
     * @returns {Promise}
     */
    PromiseQueue.prototype.waitUntilEmpty = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this._queueEmptyPr];
            });
        });
    };
    /**
     * Clears current queue.
     * If there is currently an outstanding promise then it will be unaffected
     * by this method since it is not in the queue.
     * Any promise from `waitUntilFinish` will be resolved when currently outstanding promise resolves;
     */
    PromiseQueue.prototype.clear = function () {
        this._queue = [];
    };
    /**
     * Picks out next promise in queue and continues until all are resolved or one was rejected
     * @private
     */
    PromiseQueue.prototype._playNext = function () {
        var _this = this;
        if (this._queue.length > 0) {
            var next = this._queue.shift();
            next()
                .then(function () { return _this._playNext(); })
                .catch(function (e) {
                _this._state = PromiseQueueState.NOT_PLAYING;
                _this._queue = [];
                _this._rejector(e);
            });
        }
        else {
            this._state = PromiseQueueState.NOT_PLAYING;
            this._resolver();
        }
    };
    return PromiseQueue;
}());
exports.PromiseQueue = PromiseQueue;

},{}],9:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 5/12/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var PromiseUtils = (function () {
    function PromiseUtils() {
    }
    /**
     * Installs a global handler for unhandled rejected promises
     * Makes sure to only install one instance.
     */
    PromiseUtils.catchUnhandledRejection = function () {
        var gl = global;
        if (!gl._hasInstalledUncaughtHandler) {
            process.on('unhandledRejection', function (reason, p) {
                console.error("Unhandled Rejection at: Promise '" + JSON.stringify(p) + "'");
                if (reason instanceof Error) {
                    console.error("  message: " + reason.message);
                    console.error("  stack: " + reason.stack);
                }
                else {
                    console.error(reason);
                }
            });
            gl._hasInstalledUncaughtHandler = true;
        }
    };
    /**
     * Returns promise that succeeds when the first input promise succeeds
     * If all input promises fail, then the returned promise fails.
     * @param {Promise<T>[]} promises
     * @returns {Promise<T>}
     */
    PromiseUtils.firstToSucceed = function (promises) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (res, rej) {
                        var errors = [];
                        var errorCount = 0;
                        var _loop_1 = function (i) {
                            promises[i].then(res).catch(function (e) {
                                errors[i] = e;
                                errorCount++;
                                if (errorCount === promises.length) {
                                    rej(errors);
                                }
                            });
                        };
                        for (var i = 0; i < promises.length; i++) {
                            _loop_1(i);
                        }
                    })];
            });
        });
    };
    PromiseUtils.promisify = function (func, firstParamError) {
        if (firstParamError === void 0) { firstParamError = true; }
        return new Promise(function (resolve, reject) {
            func(function (error, result) {
                if (firstParamError) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(result);
                    }
                }
                else {
                    // In this case it is assumed that the data is in the first argument
                    if (result) {
                        console.warn("Data in second argument ignored");
                    }
                    resolve(error);
                }
            });
        });
    };
    return PromiseUtils;
}());
exports.PromiseUtils = PromiseUtils;

},{}],10:[function(require,module,exports){
/**
 * @fileOverview
 * Contains Utilities related to random selection of items
 *
 * Created on 7/17/16.
 * @author Sam Spaulding <sam.spaulding@jibo.com>
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RandomUtils = (function () {
    function RandomUtils() {
    }
    /**
     * Returns a 'data' element from an array of objects with { data: Object, weight: number} fields
     * selected at random with weight proportional to the weight value of the corresponding element
     * normalised by the weights of the other elements
     *
     * @param {WeightedRandomData[]} inputs
     * @returns {Object}
     */
    RandomUtils.weightedRandomSample = function (inputs) {
        var result = {};
        var totalWeight = 0;
        for (var _i = 0, inputs_1 = inputs; _i < inputs_1.length; _i++) {
            var element = inputs_1[_i];
            totalWeight += element.weight;
        }
        var rand = Math.random() * totalWeight;
        var ongoingWeight = 0;
        for (var i = 0; i < inputs.length; ++i) {
            // do not factor in 0 or negative weight elements
            if (inputs[i].weight > 0) {
                ongoingWeight += inputs[i].weight;
                if (rand < ongoingWeight) {
                    result = inputs[i].data;
                    break;
                }
            }
        }
        return result;
    };
    return RandomUtils;
}());
exports.RandomUtils = RandomUtils;

},{}],11:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var events = require("events");
var uuid = require('uuid');
var Session = (function () {
    function Session(options) {
        this.options = options;
        this.id = uuid.v4();
        this.active = false;
        this.hasBeenStopped = false;
    }
    /**
     * Stops the session activity if possible
     * @returns {(any | Promise<any>)}
     * @memberOf Session
     */
    Session.prototype.stop = function () {
        if (!this.hasBeenStopped && this.activity) {
            this.hasBeenStopped = true;
            return this.activity.stop();
        }
    };
    /**
     * Updates the session activity if possible
     * @memberOf Session
     */
    Session.prototype.update = function () {
        if (this.activity && this.activity.update) {
            this.activity.update();
        }
    };
    return Session;
}());
exports.Session = Session;
var SessionManager = (function () {
    /**
     * Creates an instance of SessionManager.
     * @param {OpenMethod} openMethod - Is called to create a new session
     * @param {() => void} done - Gets called when all sessions have completed
     * @param {JiboLog} [logger] optional logger implementation
     * @memberOf SessionManager
     */
    function SessionManager(openMethod, done, logger) {
        this.openMethod = openMethod;
        this.done = done;
        this.logger = logger;
        this.isClosing = false;
        this.emitter = new events.EventEmitter();
        // noop
    }
    /**
     * Should be called when skill is opened
     * @memberOf SessionManager
     */
    SessionManager.prototype.open = function (options) {
        if (this.current) {
            this.current.active = false;
        }
        this.current = null;
        this.pending = null;
        this.isClosing = false;
        this.emitter.removeAllListeners();
        this.replaceSession(options);
    };
    /**
     * Switches skill session
     * @param {*} options
     * @memberOf SessionManager
     */
    SessionManager.prototype.replaceSession = function (options) {
        var session = new Session(options);
        if (this.current) {
            // If we have a current session, we set this as pending and tell current session to stop
            this.pending = session;
            this.current.stop();
        }
        else {
            // We immediately open this session
            this._startSession(session);
        }
    };
    /**
     * Stops current session and waits for it to finish.
     * After any call to this method you have to call `open` again
     * @returns {Promise<any>}
     * @memberOf SessionManager
     */
    SessionManager.prototype.close = function () {
        var _this = this;
        this.isClosing = true;
        this.pending = null;
        var pr = Promise.resolve();
        if (this.current) {
            var ret = this.current.stop();
            if (ret instanceof Promise) {
                pr = ret;
            }
        }
        return pr.then(function () { return _this.waitForCurrent(); });
    };
    /**
     * Waits for the current session to complete, if no session is current then we resolve immediately
     * @returns {Promise<any>}
     * @memberOf SessionManager
     */
    SessionManager.prototype.waitForCurrent = function () {
        if (this.current) {
            return this.current.activity.promise;
        }
        else {
            return Promise.resolve();
        }
    };
    /**
     * Waits for all sessions to complete, if no session is current then we resolve immediately
     * @returns {Promise<any>}
     * @memberOf SessionManager
     */
    SessionManager.prototype.waitForAll = function () {
        var _this = this;
        if (this.current) {
            return new Promise(function (resolve) {
                _this.emitter.once('done', resolve);
            });
        }
        else {
            return Promise.resolve();
        }
    };
    SessionManager.prototype._startSession = function (session) {
        var _this = this;
        this.current = session;
        session.active = true;
        session.activity = this.openMethod(session.options);
        session.activity.promise = session.activity.promise
            .catch(function (err) { return _this.errorLog(err); })
            .then(function () {
            // If this session is the current one, then we set it to null
            if (_this.current === session) {
                _this.current = null;
            }
            session.active = false;
            // If there is a pending session, we open that one
            if (_this.pending) {
                var next = _this.pending;
                _this.pending = null;
                _this._startSession(next);
            }
            else {
                // If we are not in the process of closing then we call the done callback
                if (!_this.isClosing) {
                    _this.done();
                }
                _this.emitter.emit('done');
            }
        });
    };
    /**
     * Prints error to log
     * @private
     * @param {*} msg
     * @param {...any[]} args
     * @memberOf SessionManager
     */
    SessionManager.prototype.errorLog = function (msg) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (this.logger) {
            (_a = this.logger).error.apply(_a, [msg].concat(args));
        }
        else {
            console.error.apply(console, [msg].concat(args));
        }
        var _a;
    };
    return SessionManager;
}());
exports.SessionManager = SessionManager;

},{"events":undefined,"uuid":undefined}],12:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 7/11/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var child_process_1 = require("child_process");
var ProcessResult = (function () {
    function ProcessResult() {
    }
    return ProcessResult;
}());
exports.ProcessResult = ProcessResult;
var TerminalUtils = (function () {
    function TerminalUtils() {
    }
    /**
     * Runs a terminal command
     * @param {string} cmd Command to run
     * @param {string[]} [args] Arguments to command
     * @param {string} [cwd] Working directory to execute command
     * @returns {Promise<T>}
     */
    TerminalUtils.run = function (cmd, args, cwd) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        var inst = cwd ? child_process_1.spawn(cmd, args, { cwd: cwd }) : child_process_1.spawn(cmd, args);
                        var res = new ProcessResult();
                        inst.stdout.on('data', function (data) {
                            res.stdout = new Buffer(data).toString('utf8', 0);
                        });
                        inst.stderr.on('data', function (data) {
                            res.stderr = new Buffer(data).toString('utf8', 0);
                        });
                        inst.on('error', function (err) {
                            res.error = err;
                            res.code = err.code;
                            reject(res);
                        });
                        inst.on('close', function (code) {
                            res.code = code;
                            if (code === 0) {
                                resolve(res);
                            }
                            else {
                                reject(res);
                            }
                        });
                    })];
            });
        });
    };
    return TerminalUtils;
}());
exports.TerminalUtils = TerminalUtils;

},{"child_process":undefined}],13:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 9/24/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var FileUtils_1 = require("./FileUtils");
var events_1 = require("events");
var Type;
(function (Type) {
    Type[Type["NUMBER"] = 'NUMBER'] = "NUMBER";
    Type[Type["BOOLEAN"] = 'BOOLEAN'] = "BOOLEAN";
    Type[Type["STRING"] = 'STRING'] = "STRING";
})(Type = exports.Type || (exports.Type = {}));
/**
 * Parses a boolean value
 * @param {string|boolean} input
 * @return {boolean}
 */
function getBoolean(input) {
    if (typeof input === 'boolean') {
        return input;
    }
    else if (typeof input === 'string') {
        return !!JSON.parse(input.toLowerCase());
    }
    else {
        throw new Error("Can't parse boolean '" + input + "'");
    }
}
/**
 * @class TestConfiguration
 */
var TestConfiguration = (function (_super) {
    __extends(TestConfiguration, _super);
    function TestConfiguration(namespace) {
        var _this = _super.call(this) || this;
        _this.namespace = namespace;
        _this._hasTestData = false;
        return _this;
    }
    /**
     * Initializes test configuration from test file (if it exists)
     * Should be called after all value providers have been provided
     * @param {string} [pathToTestConfig] Path to test config file
     * @return {Promise<boolean>} True if test configuration loaded, false otherwise
     */
    TestConfiguration.prototype.init = function (pathToTestConfig) {
        return __awaiter(this, void 0, void 0, function () {
            var data, error_1, enabled;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, FileUtils_1.FileUtils.readFile(pathToTestConfig)];
                    case 1:
                        data = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        error_1 = _a.sent();
                        return [3 /*break*/, 3];
                    case 3:
                        if (data) {
                            try {
                                data = JSON.parse(data);
                            }
                            catch (error) {
                                throw new Error("Error parsing test config '" + pathToTestConfig + "': " + error);
                            }
                            enabled = getBoolean(data.enabled);
                            if (enabled && data[this.namespace]) {
                                this._data = data[this.namespace];
                                // Verify test values
                                // TOFIX check if test names are expected
                                this._hasTestData = true;
                                return [2 /*return*/, true];
                            }
                        }
                        if (!this._data) {
                            this._data = {};
                            return [2 /*return*/, false];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Returns true if this class has loaded a valid test configuration
     * and that it is enabled
     * @return {boolean}
     */
    TestConfiguration.prototype.hasEnabledTestConfig = function () {
        return this._hasTestData;
    };
    /**
     * Returns true if this class has an overriding test configuration
     * @return {boolean}
     */
    TestConfiguration.prototype.hasOverridingTestConfig = function () {
        return !!this._testConfig;
    };
    /**
     * Clears the overriding test config
     */
    TestConfiguration.prototype.clearOverridingTestConfig = function () {
        this._testConfig = null;
    };
    /**
     * Sets the overriding test config. If null then test config is disabled
     * @param {Object} testConfig Null to disable
     */
    TestConfiguration.prototype.setOverridingTestConfig = function (testConfig) {
        this._testConfig = testConfig;
        if (testConfig) {
            // Verify values
            // TOFIX check if test names are expected
        }
    };
    /**
     * Waits for event emitted on this object
     * @param {string} name
     * @return {Promise<void>}
     */
    TestConfiguration.prototype.waitForEvent = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve) {
                        _this.once(name, function () { return resolve(); });
                    })];
            });
        });
    };
    /**
     * A number value getter
     * @param {string} name
     * @param {function} [provider] A provider of the real value in case it doesn't exist in test config
     * @return {Promise<number|null>}
     */
    TestConfiguration.prototype.getNumberFromTestConfig = function (name, provider) {
        return __awaiter(this, void 0, void 0, function () {
            var val;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._getFromTestConfig(name, Type.NUMBER)];
                    case 1:
                        val = _a.sent();
                        if (val !== null) {
                            return [2 /*return*/, parseFloat(val)];
                        }
                        else if (provider) {
                            return [2 /*return*/, Promise.resolve(provider())];
                        }
                        else {
                            return [2 /*return*/, null];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * A string value getter
     * @param {string} name
     * @param {function} [provider] A provider of the real value in case it doesn't exist in test config
     * @return {Promise<string|null>}
     */
    TestConfiguration.prototype.getStringFromTestConfig = function (name, provider) {
        return __awaiter(this, void 0, void 0, function () {
            var val;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._getFromTestConfig(name, Type.STRING)];
                    case 1:
                        val = _a.sent();
                        if (val !== null) {
                            return [2 /*return*/, (typeof val === 'string') ? val : ('' + val)];
                        }
                        else if (provider) {
                            return [2 /*return*/, Promise.resolve(provider())];
                        }
                        else {
                            return [2 /*return*/, null];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * A boolean value getter
     * @param {string} name
     * @param {function} [provider] A provider of the real value in case it doesn't exist in test config
     * @return {Promise<boolean|null>}
     */
    TestConfiguration.prototype.getBooleanFromTestConfig = function (name, provider) {
        return __awaiter(this, void 0, void 0, function () {
            var val;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._getFromTestConfig(name, Type.BOOLEAN)];
                    case 1:
                        val = _a.sent();
                        if (val !== null) {
                            return [2 /*return*/, getBoolean(val)];
                        }
                        else if (provider) {
                            return [2 /*return*/, Promise.resolve(provider())];
                        }
                        else {
                            return [2 /*return*/, null];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * A generic value getter
     * @param {string} name
     * @param {Type} type
     * @return {Promise<any>}
     * @private
     */
    TestConfiguration.prototype._getFromTestConfig = function (name, type) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!this._data) {
                    throw new Error("Test configuration '" + this.namespace + "' is not initialized");
                }
                // TOFIX check if value name is expected and type thereof
                // If we find the data in the overriding test file
                if (this._testConfig && this._testConfig[name] !== undefined) {
                    return [2 /*return*/, this._testConfig[name]];
                }
                else if (this._data[name] !== undefined) {
                    return [2 /*return*/, this._data[name]];
                }
                return [2 /*return*/, null];
            });
        });
    };
    return TestConfiguration;
}(events_1.EventEmitter));
exports.TestConfiguration = TestConfiguration;

},{"./FileUtils":5,"events":undefined}],14:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 7/17/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PathUtils_1 = require("./PathUtils");
var path = require("path");
var NameSpace = (function () {
    function NameSpace(name) {
        this.name = name;
        /**
         * A container for test dates
         * @type {Map<string, Date>}
         */
        this.dateById = new Map();
        /**
         * A container for test random numbers
         * @type {Map<string, Date>}
         */
        this.randomById = new Map();
        // lint noop
    }
    /**
     * Clear all installed tests
     */
    NameSpace.prototype.clear = function () {
        this.dateById.clear();
        this.randomById.clear();
    };
    return NameSpace;
}());
exports.NameSpace = NameSpace;
var defaultNamespace = new NameSpace('');
var nameSpaces = new Map();
nameSpaces.set('', defaultNamespace);
nameSpaces.set('default', defaultNamespace);
/**
 * Whether testing is enabled or not
 * @type {boolean}
 */
var enableTest = false;
/**
 * A function that returns a date
 * @type {Function}
 * @returns {Date}
 */
var realDateProvider = function () {
    return new Date();
};
/**
 * A function that returns a random number [0,1)
 * @type {Function}
 * @returns {number}
 */
var realRandomProvider = function () {
    return Math.random();
};
var TestUtils = (function () {
    function TestUtils() {
    }
    /**
     * Returns the main export of a project, same as you would get if you
     * did `let exp = require('project-name');`
     * @param {NodeModule|string} pathOrModule Path in project or NodeModule
     * @returns {any}
     */
    TestUtils.getProjectMainExport = function (pathOrModule) {
        var root = PathUtils_1.PathUtils.getProjectRoot(pathOrModule);
        var pack = require(path.join(root, 'package.json'));
        return require(path.join(root, pack.main));
    };
    /**
     * Clear all tests in all namespaces
     */
    TestUtils.clearAllTests = function () {
        nameSpaces.forEach(function (ns) { return ns.clear(); });
    };
    /**
     * Clear all tests in namespace
     * @param {string} [namespace=''] name of namespace
     */
    TestUtils.clearTests = function (namespace) {
        var nsName = namespace || '';
        var ns = nameSpaces.get(nsName);
        if (ns) {
            ns.clear();
        }
    };
    /**
     * Turn on/off testing
     * @param {boolean} enable
     */
    TestUtils.enableTesting = function (enable) {
        enableTest = enable;
    };
    /**
     * Returns namespace, if undefined then returns default namespace
     * @param namespace
     * @returns {any}
     */
    TestUtils.getNamespace = function (namespace) {
        if (!namespace) {
            return defaultNamespace;
        }
        else {
            var ns = nameSpaces.get(namespace);
            if (!ns) {
                ns = new NameSpace(namespace);
                nameSpaces.set(namespace, ns);
            }
            return ns;
        }
    };
    /**
     * Provides a test ID for a particular date
     * @param {Date} date
     * @param {string} id
     * @param {string} [namespace='']
     */
    TestUtils.setTestDate = function (date, id, namespace) {
        var ns = TestUtils.getNamespace(namespace);
        ns.dateById.set(id, date);
    };
    /**
     * Returns a Date object. If testing is enabled and an id is
     * provided then the provided date will be returned
     * @param {string} [id] Test id
     * @param {string} [namespace='']
     * @returns {Date}
     */
    TestUtils.getDate = function (id, namespace) {
        if (enableTest && id) {
            var ns = TestUtils.getNamespace(namespace);
            var testDate = ns.dateById.get(id);
            if (testDate) {
                return testDate;
            }
        }
        return realDateProvider();
    };
    /**
     * Provides a test ID for a particular random number
     * @param {number} randomNum
     * @param {string} id
     * @param {string} [namespace='']
     */
    TestUtils.setTestRandom = function (randomNum, id, namespace) {
        var ns = TestUtils.getNamespace(namespace);
        ns.randomById.set(id, randomNum);
    };
    /**
     * Returns a random number. If testing is enabled and an id is
     * provided then the provided random number will be returned
     * @param {string} [id] Test id
     * @param {string} [namespace]
     * @returns {number}
     */
    TestUtils.getRandom = function (id, namespace) {
        if (enableTest && id) {
            var ns = TestUtils.getNamespace(namespace);
            var random = ns.randomById.get(id);
            if (random) {
                return random;
            }
        }
        return realRandomProvider();
    };
    return TestUtils;
}());
exports.TestUtils = TestUtils;

},{"./PathUtils":6,"path":undefined}],15:[function(require,module,exports){
/**
 * @fileOverview
 *
 * Created on 5/12/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t;
    return { next: verb(0), "throw": verb(1), "return": verb(2) };
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var _timeOutFunc = setTimeout;
var DefaultDateProvider = (function () {
    function DefaultDateProvider() {
    }
    DefaultDateProvider.prototype.getDate = function () {
        return new Date();
    };
    return DefaultDateProvider;
}());
exports.DefaultDateProvider = DefaultDateProvider;
// wanted to use TS enums, but can't get them to work as strings in JS
exports.PartOfDay = {
    MORNING: 'MORNING',
    AFTERNOON: 'AFTERNOON',
    EVENING: 'EVENING',
    NIGHT: 'NIGHT'
};
var TimeUtils = (function () {
    function TimeUtils() {
    }
    /**
     * Sets the date provider that should be used by TimeUtils
     * @param {DateProvider} dateProvider
     */
    TimeUtils.setDateProvider = function (dateProvider) {
        TimeUtils._dateProvider = dateProvider;
    };
    /**
     * Provide the timeout function to use in all time utilities
     * @param {Function.<Function, number>} timeoutFunc
     */
    TimeUtils.setTimeoutFunction = function (timeoutFunc) {
        _timeOutFunc = timeoutFunc;
    };
    /**
     * Waits for timeMs and then resolves the promise
     * @param {number} timeMs Time to wait in ms
     * @returns {Promise<void>}
     */
    TimeUtils.wait = function (timeMs) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (res) {
                        _timeOutFunc(res, timeMs);
                    })];
            });
        });
    };
    /**
     * Dispatch a timeline of actions
     * @param {TimelineElement[]} timeline
     * @returns {Promise<void>}
     */
    TimeUtils.dispatchTimeline = function (timeline) {
        return __awaiter(this, void 0, void 0, function () {
            var startTime, promiseChain;
            return __generator(this, function (_a) {
                startTime = process.hrtime();
                promiseChain = Promise.resolve();
                timeline.forEach(function (el) {
                    promiseChain = promiseChain
                        .then(function () { return TimeUtils._dispatchTimeline(el, startTime); });
                });
                return [2 /*return*/, promiseChain];
            });
        });
    };
    TimeUtils._dispatchTimeline = function (el, startTime) {
        return __awaiter(this, void 0, void 0, function () {
            var timeDiff, passedTimeMs, timeToWait;
            return __generator(this, function (_a) {
                timeDiff = process.hrtime(startTime);
                passedTimeMs = timeDiff[0] * 1e3 + timeDiff[1] * 1e-6;
                timeToWait = Math.max(0, el.timeMs - passedTimeMs);
                return [2 /*return*/, TimeUtils.wait(timeToWait).then(function () { return el.action(); })];
            });
        });
    };
    /**
     * returns the current part of day (morning, afternoon, evening, night)
     * @param {Date} [date] Optional date, defaults to using date provider
     * @returns {PartOfDay}
     */
    TimeUtils.getPartOfDay = function (date) {
        date = date || TimeUtils._dateProvider.getDate();
        var hours = date.getHours();
        if (hours > 5 && hours < 12) {
            return exports.PartOfDay.MORNING;
        }
        else if (hours < 18) {
            return exports.PartOfDay.AFTERNOON;
        }
        else if (hours < 23) {
            return exports.PartOfDay.EVENING;
        }
        else {
            return exports.PartOfDay.NIGHT;
        }
    };
    /**
     * Returns whether it is early in the morning (between 5 and 7 am)
     * @param {Date} [date] Optional date, defaults to using date provider
     * @returns {boolean}
     */
    TimeUtils.isEarlyMorning = function (date) {
        date = date || TimeUtils._dateProvider.getDate();
        return (TimeUtils.getPartOfDay(date) === exports.PartOfDay.MORNING) && (date.getHours() < 7);
    };
    /**
     * Converts milliseconds to seconds
     * @param ms {number}
     * @returns {number}
     */
    TimeUtils.msToSeconds = function (ms) {
        return ms / (1000);
    };
    /**
     * Converts milliseconds to minutes
     * @param ms {number}
     * @returns {number}
     */
    TimeUtils.msToMinutes = function (ms) {
        return ms / (60 * 1000);
    };
    /**
     * Converts milliseconds to hours
     * @param ms {number}
     * @returns {number}
     */
    TimeUtils.msToHours = function (ms) {
        return ms / (60 * 60 * 1000);
    };
    /**
     * Converts seconds to milliseconds
     * @param seconds {number}
     * @returns {number}
     */
    TimeUtils.secondsToMs = function (seconds) {
        return seconds * 1000;
    };
    /**
     * Converts minutes to milliseconds
     * @param minutes {number}
     * @returns {number}
     */
    TimeUtils.minutesToMs = function (minutes) {
        return minutes * 60 * 1000;
    };
    /**
     * Converts hours to milliseconds
     * @param hours {number}
     * @returns {number}
     */
    TimeUtils.hoursToMs = function (hours) {
        return hours * 60 * 60 * 1000;
    };
    return TimeUtils;
}());
/**
 * A class that can provide dates (defaults to `new Date()`)
 * @type {DateProvider}
 * @private
 */
TimeUtils._dateProvider = new DefaultDateProvider();
exports.TimeUtils = TimeUtils;

},{}],16:[function(require,module,exports){
/**
 * Created by Andrew Stout on 6/28/16, based on prior work by Siggi Orn.
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Updatable = (function () {
    function Updatable() {
        this.updatables = new Set();
        // lint noop
    }
    /**
     * Should get called from the main update loop.
     * Iterates through all registered updatables and updates them
     */
    Updatable.prototype.update = function (elapsed) {
        //for(let callback of this.updatables) {
        this.updatables.forEach(function (callback) {
            callback(elapsed); // this should probably be made asynchronous at some point
        });
    };
    /**
     * Add a function to be called on every update of loop.
     * If this function has already been added then this has no effect.
     * @param {function} callback Function to be called at every update.
     * The callback function takes a parameter of the form [seconds, milliseconds]
     * @returns {boolean} false if callback existed in collection and
     * was therefore not added.
     */
    Updatable.prototype.add = function (callback) {
        var exists = this.updatables.has(callback);
        this.updatables.add(callback);
        return !exists;
    };
    /**
     * Stop updating a function on every update of loop
     * @param {function} callback Function to be removed from update loop
     * @returns {boolean}
     */
    Updatable.prototype.remove = function (callback) {
        return this.updatables.delete(callback);
    };
    return Updatable;
}());
exports.Updatable = Updatable;

},{}],17:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./ActionTimeline"));
__export(require("./CacheUtils"));
__export(require("./CancelToken"));
__export(require("./CancelTokenSession"));
__export(require("./FileUtils"));
__export(require("./PathUtils"));
__export(require("./PlaybackUtils"));
__export(require("./PromiseQueue"));
__export(require("./PromiseUtils"));
__export(require("./RandomUtils"));
__export(require("./SessionManager"));
__export(require("./TerminalUtils"));
__export(require("./TestConfiguration"));
__export(require("./TestUtils"));
__export(require("./TimeUtils"));
__export(require("./Updatable"));

},{"./ActionTimeline":1,"./CacheUtils":2,"./CancelToken":3,"./CancelTokenSession":4,"./FileUtils":5,"./PathUtils":6,"./PlaybackUtils":7,"./PromiseQueue":8,"./PromiseUtils":9,"./RandomUtils":10,"./SessionManager":11,"./TerminalUtils":12,"./TestConfiguration":13,"./TestUtils":14,"./TimeUtils":15,"./Updatable":16}],18:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./main"));

},{"./main":17}]},{},[18])(18)
});
