// jibo-common-types/lib/jibo-common-types.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboCommonTypes = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EmotionName;
(function (EmotionName) {
    EmotionName["PLEASED"] = "PLEASED";
    EmotionName["SAD"] = "SAD";
    EmotionName["NEUTRAL"] = "NEUTRAL";
    EmotionName["CONFIDENT"] = "CONFIDENT";
    EmotionName["UNCERTAIN"] = "UNCERTAIN";
    EmotionName["INSECURE"] = "INSECURE";
    EmotionName["PESSIMISTIC"] = "PESSIMISTIC";
    EmotionName["FRUSTRATED"] = "FRUSTRATED";
    EmotionName["HOPEFUL"] = "HOPEFUL";
    EmotionName["DETERMINED"] = "DETERMINED";
    EmotionName["JOYFUL"] = "JOYFUL";
    EmotionName["BASELINE"] = "BASELINE";
})(EmotionName = exports.EmotionName || (exports.EmotionName = {}));

},{}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CircadianState;
(function (CircadianState) {
    CircadianState.ALERT = 'ALERT';
    CircadianState.RELAXED = 'RELAXED';
    CircadianState.DAYTIME_NAP = 'DAYTIME_NAP';
    CircadianState.ASLEEP = 'ASLEEP';
    CircadianState.WAKING_UP = 'WAKING_UP';
    CircadianState.FALLING_ASLEEP = 'FALLING_ASLEEP';
    CircadianState.TIME_OF_DAY = 'TIME_OF_DAY';
    CircadianState.HJ_NO_MATCH = 'HJ_NO_MATCH';
    CircadianState.SELECT_INTENT = 'SELECT_INTENT';
    CircadianState.TURN_AWAY = 'TURN_AWAY';
})(CircadianState = exports.CircadianState || (exports.CircadianState = {}));

},{}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ResultStatus;
(function (ResultStatus) {
    ResultStatus["SUCCEEDED"] = "SUCCEEDED";
    ResultStatus["TIMEOUT"] = "TIMEOUT";
    ResultStatus["INTERRUPTED"] = "INTERRUPTED";
    ResultStatus["CANCELLED"] = "CANCELLED";
})(ResultStatus = exports.ResultStatus || (exports.ResultStatus = {}));
var AttentionMode;
(function (AttentionMode) {
    AttentionMode["OFF"] = "OFF";
    AttentionMode["IDLE"] = "IDLE";
    AttentionMode["RELAXED"] = "RELAXED";
    AttentionMode["NAP"] = "NAP";
    AttentionMode["DISENGAGE"] = "DISENGAGE";
    AttentionMode["ENGAGED"] = "ENGAGED";
    AttentionMode["SPEAKING"] = "SPEAKING";
    AttentionMode["FIXATED"] = "FIXATED";
    AttentionMode["ATTRACTABLE"] = "ATTRACTABLE";
    AttentionMode["MENU"] = "MENU";
    AttentionMode["COMMAND"] = "COMMAND";
})(AttentionMode = exports.AttentionMode || (exports.AttentionMode = {}));
var LEDMode;
(function (LEDMode) {
    LEDMode["OFF"] = "OFF";
    LEDMode["LISTEN"] = "LISTEN";
    LEDMode["ERROR"] = "ERROR";
})(LEDMode = exports.LEDMode || (exports.LEDMode = {}));
var IndexRobotResult;
(function (IndexRobotResult) {
    IndexRobotResult["SUCCEEDED"] = "SUCCEEDED";
    IndexRobotResult["TIMEOUT"] = "TIMEOUT";
    IndexRobotResult["FAULT"] = "FAULT";
})(IndexRobotResult = exports.IndexRobotResult || (exports.IndexRobotResult = {}));
var PlayStatus;
(function (PlayStatus) {
    PlayStatus["OK"] = "OK";
    PlayStatus["REJECTED"] = "REJECTED";
})(PlayStatus = exports.PlayStatus || (exports.PlayStatus = {}));
var ExpressionIds;
(function (ExpressionIds) {
    ExpressionIds[ExpressionIds["EXPRESSION"] = 0] = "EXPRESSION";
})(ExpressionIds = exports.ExpressionIds || (exports.ExpressionIds = {}));

},{}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GlobalCommand;
(function (GlobalCommand) {
    GlobalCommand["SLEEP"] = "SLEEP";
    GlobalCommand["HELP"] = "HELP";
    GlobalCommand["STOP"] = "STOP";
    GlobalCommand["PAUSE"] = "PAUSE";
    GlobalCommand["VOLUME"] = "VOLUME";
    GlobalCommand["WHATCANIDO"] = "WHATCANIDO";
    GlobalCommand["HOLDON"] = "HOLDON";
    GlobalCommand["OVERHERE"] = "OVERHERE";
    GlobalCommand["TURNAWAY"] = "TURNAWAY";
    GlobalCommand["TURNAROUND"] = "TURNAROUND";
})(GlobalCommand = exports.GlobalCommand || (exports.GlobalCommand = {}));
exports.GlobalCommandMap = new Map();
exports.GlobalCommandMap.set("sleep", GlobalCommand.SLEEP);
exports.GlobalCommandMap.set("help", GlobalCommand.HELP);
exports.GlobalCommandMap.set("stop", GlobalCommand.STOP);
exports.GlobalCommandMap.set("pause", GlobalCommand.PAUSE);
exports.GlobalCommandMap.set("volume", GlobalCommand.VOLUME);
exports.GlobalCommandMap.set("whatCanIDo", GlobalCommand.WHATCANIDO);
exports.GlobalCommandMap.set("holdOn", GlobalCommand.HOLDON);
exports.GlobalCommandMap.set("overHere", GlobalCommand.OVERHERE);
exports.GlobalCommandMap.set("turnAway", GlobalCommand.TURNAWAY);
exports.GlobalCommandMap.set("turnAround", GlobalCommand.TURNAROUND);

},{}],5:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const animation_utilities_1 = require("animation-utilities");
const jibo_typed_events_1 = require("jibo-typed-events");
class PositionInSpace {
    constructor(x = 0, y = 0, z = 0) {
        this.x = x;
        this.y = y;
        this.z = z;
    }
}
exports.PositionInSpace = PositionInSpace;
var AudioLPSEntityType;
(function (AudioLPSEntityType) {
    AudioLPSEntityType[AudioLPSEntityType["ENERGY"] = 0] = "ENERGY";
    AudioLPSEntityType[AudioLPSEntityType["HOT_WORD"] = 1] = "HOT_WORD";
    AudioLPSEntityType[AudioLPSEntityType["SPEAKER_ID"] = 2] = "SPEAKER_ID";
})(AudioLPSEntityType = exports.AudioLPSEntityType || (exports.AudioLPSEntityType = {}));
var JiboBodyAxis;
(function (JiboBodyAxis) {
    JiboBodyAxis["PELVIS"] = "PELVIS";
    JiboBodyAxis["TORSO"] = "TORSO";
    JiboBodyAxis["NECK"] = "NECK";
})(JiboBodyAxis = exports.JiboBodyAxis || (exports.JiboBodyAxis = {}));
var IDLabels;
(function (IDLabels) {
    IDLabels["NOT_ENROLLED"] = "NOT_ENROLLED";
    IDLabels["ANYBODY"] = "ANYBODY";
    IDLabels["IDENTIFIED"] = "IDENTIFIED";
    IDLabels["UNKNOWN"] = "UNKNOWN";
    IDLabels["NOT_TRAINED"] = "NOT_TRAINED";
})(IDLabels = exports.IDLabels || (exports.IDLabels = {}));
var SensoryType;
(function (SensoryType) {
    SensoryType.UNKNOWN = 'UNKNOWN';
    SensoryType.PERSON = 'PERSON';
    SensoryType.VOICE = 'VOICE';
    function fromString(name) {
        return SensoryType[name] || SensoryType[name.toUpperCase()];
    }
    SensoryType.fromString = fromString;
})(SensoryType = exports.SensoryType || (exports.SensoryType = {}));
var IdentificationService;
(function (IdentificationService) {
    IdentificationService.UNKNOWN = 'UNKNOWN';
    IdentificationService.ASR = 'ASR';
    IdentificationService.LPS = 'LPS';
    function fromString(name) {
        return IdentificationService[name] || IdentificationService[name.toUpperCase()];
    }
    IdentificationService.fromString = fromString;
})(IdentificationService = exports.IdentificationService || (exports.IdentificationService = {}));
var VisualEntityDescriptionType;
(function (VisualEntityDescriptionType) {
    VisualEntityDescriptionType.PERSON = 'PERSON';
    VisualEntityDescriptionType.MOTION = 'MOTION';
    VisualEntityDescriptionType.UNKNOWN = 'UNKNOWN';
    function fromString(name) {
        return VisualEntityDescriptionType[name] || VisualEntityDescriptionType[name.toUpperCase()];
    }
    VisualEntityDescriptionType.fromString = fromString;
})(VisualEntityDescriptionType = exports.VisualEntityDescriptionType || (exports.VisualEntityDescriptionType = {}));
var IdentificationSource;
(function (IdentificationSource) {
    IdentificationSource.VOICE = 'VOICE';
    IdentificationSource.FACE = 'FACE';
    IdentificationSource.UNKNOWN = 'UNKNOWN';
    function fromString(name) {
        return IdentificationSource[name] || IdentificationSource[name.toUpperCase()];
    }
    IdentificationSource.fromString = fromString;
})(IdentificationSource = exports.IdentificationSource || (exports.IdentificationSource = {}));
class SensoryRecord {
    constructor(entity, position, type, reportedTS) {
        this.position = null;
        Object.assign(this, entity);
        this.type = type;
        this.raw = entity;
        if (position !== null && position !== undefined) {
            this.position = position;
        }
        this.id = entity.id;
        this.confidence = entity.confidence;
        this.receivedAt = animation_utilities_1.Clock.currentTime();
        this.reportedTS = reportedTS;
    }
    update(entity) {
        Object.assign(this, entity);
        this.position = new animation_utilities_1.THREE.Vector3(entity.position.x, entity.position.y, entity.position.z);
        this.lastUpdate = animation_utilities_1.Clock.currentTime();
    }
}
exports.SensoryRecord = SensoryRecord;
class PresenceEvents extends jibo_typed_events_1.EventContainer {
    constructor() {
        super();
        this.presenceStarted = new jibo_typed_events_1.Event("Presence Started");
        this.presenceEnded = new jibo_typed_events_1.Event("Presence Ended");
    }
}
exports.PresenceEvents = PresenceEvents;
class PresenceRecord {
    constructor(presence) {
        this.identificationService = null;
        this.identificationSource = IdentificationSource.UNKNOWN;
        this.events = new PresenceEvents();
        Object.assign(this, presence);
        this.lastUpdate = animation_utilities_1.Clock.currentTime();
        this.ASRIdentified = (this.identificationService === IdentificationService.ASR);
        this.LPSIdentified = (this.identificationService === IdentificationService.LPS);
    }
    update(presence) {
        const presenceState = this.present;
        Object.assign(this, presence);
        if (presenceState !== this.present) {
            if (this.present === true) {
                this.events.presenceStarted.emit();
            }
            else {
                this.events.presenceEnded.emit();
            }
        }
        this.lastUpdate = animation_utilities_1.Clock.currentTime();
    }
}
exports.PresenceRecord = PresenceRecord;
class SpeakerIdInfo {
    constructor(info) {
        this.id = info ? info.id : "";
        this.accepted = info ? info.accepted : false;
        this.highConfidence = info ? info.highConfidence : true;
    }
}
exports.SpeakerIdInfo = SpeakerIdInfo;
class SpeakerRecord {
    constructor(speaker) {
        this.idInfo = speaker ? speaker.idInfo : new SpeakerIdInfo();
        this.entity = speaker ? speaker.entity : -1;
        this.location = speaker ? speaker.location : null;
    }
}
exports.SpeakerRecord = SpeakerRecord;
class RawSpeakerData {
}
exports.RawSpeakerData = RawSpeakerData;

},{"animation-utilities":undefined,"jibo-typed-events":undefined}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class BodyState {
    constructor() {
        this.miscState = new MiscState();
        this.touchState = new TouchState();
        this.powerState = new PowerState();
        this.imuState = new IMUState();
    }
}
exports.BodyState = BodyState;
class MiscState {
    constructor() {
        this.ts = [];
        this.main_board_temp = 0;
        this.cpu_temp = 0;
        this.intake_temp = 0;
        this.exhaust_temp = 0;
        this.neck_motor_temp = 0;
        this.torso_motor_temp = 0;
        this.pelvis_motor_temp = 0;
    }
}
exports.MiscState = MiscState;
class TouchState {
    constructor() {
        this.ts = [];
        this.touched = false;
        this.pad_state = [false, false, false, false, false, false];
        this.changed = [0, 0, 0, 0, 0, 0];
    }
}
exports.TouchState = TouchState;
class PowerState {
    constructor() {
        this.ts = [];
        this.source = "";
        this.vsys = 0;
        this.battery = new BatteryState();
    }
}
exports.PowerState = PowerState;
class BatteryState {
    constructor() {
        this.id = 0;
        this.cycles = 0;
        this.capacity = 0;
        this.max_capacity = 0;
        this.charge_rate = 0;
        this.temp = 0;
    }
}
exports.BatteryState = BatteryState;
class IMUState {
    constructor() {
        this.ts = [];
        this.acc = [];
        this.gyro = [];
        this.up = [];
        this.moving = false;
        this.raw_acc = [];
        this.raw_gyro = [];
    }
}
exports.IMUState = IMUState;
class InputEnergy {
    constructor() {
        this.ts = [null, null];
        this.db_rms = null;
        this.db_high = null;
        this.db_mid = null;
        this.db_low = null;
    }
}
exports.InputEnergy = InputEnergy;
var StatusFlags;
(function (StatusFlags) {
    StatusFlags[StatusFlags["INDEXED"] = 0] = "INDEXED";
    StatusFlags[StatusFlags["ENABLED"] = 1] = "ENABLED";
    StatusFlags[StatusFlags["BRAKED"] = 2] = "BRAKED";
    StatusFlags[StatusFlags["MOVING"] = 3] = "MOVING";
    StatusFlags[StatusFlags["STALLED"] = 4] = "STALLED";
    StatusFlags[StatusFlags["TIMEOUT"] = 5] = "TIMEOUT";
    StatusFlags[StatusFlags["FAULT"] = 6] = "FAULT";
})(StatusFlags = exports.StatusFlags || (exports.StatusFlags = {}));
var FaultFlags;
(function (FaultFlags) {
    FaultFlags[FaultFlags["INDEXED"] = 0] = "INDEXED";
    FaultFlags[FaultFlags["POWER"] = 1] = "POWER";
    FaultFlags[FaultFlags["STALL"] = 2] = "STALL";
    FaultFlags[FaultFlags["ENCODER"] = 3] = "ENCODER";
    FaultFlags[FaultFlags["ENCODER_BACKWARDS"] = 4] = "ENCODER_BACKWARDS";
    FaultFlags[FaultFlags["THERMISTOR"] = 5] = "THERMISTOR";
})(FaultFlags = exports.FaultFlags || (exports.FaultFlags = {}));
var LockoutFlags;
(function (LockoutFlags) {
    LockoutFlags[LockoutFlags["NO_BATTERY"] = 0] = "NO_BATTERY";
    LockoutFlags[LockoutFlags["NOT_UPRIGHT"] = 1] = "NOT_UPRIGHT";
    LockoutFlags[LockoutFlags["FALLING"] = 2] = "FALLING";
})(LockoutFlags = exports.LockoutFlags || (exports.LockoutFlags = {}));
var PowerSource;
(function (PowerSource) {
    PowerSource["BATTERY"] = "BATTERY";
    PowerSource["EXTERNAL"] = "EXTERNAL";
})(PowerSource = exports.PowerSource || (exports.PowerSource = {}));

},{}],7:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./GlobalListenTypes"));
__export(require("./LPSTypes"));
__export(require("./SystemTypes"));
__export(require("./CharacterTypes"));
__export(require("./Circadian"));
__export(require("./Expression"));

},{"./CharacterTypes":1,"./Circadian":2,"./Expression":3,"./GlobalListenTypes":4,"./LPSTypes":5,"./SystemTypes":6}]},{},[7])(7)
});
