// jibo/resources/buttons/bigActionButton.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Container = PIXI.Container;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    lib.bigAction_icon = Container.extend(function () {
        Container.call(this);

    });

    lib.bigAction_border = Container.extend(function () {
        Container.call(this);
        var instance1 = new Sprite(fromFrame("bigAction_border1"))
            .setTransform(-165.05, -165.05);
        this.addChild(instance1);
    });

    lib.bigAction_gradient = Container.extend(function () {
        Container.call(this);
        var instance1 = new Sprite(fromFrame("bigAction_gradient1"))
            .setTransform(-165.05, -165.05);
        this.addChild(instance1);
    });

    lib.bigAction_background = Container.extend(function () {
        Container.call(this);
        var instance1 = new Sprite(fromFrame("bigAction_background1"))
            .setTransform(-165.05, -165.05);
        this.addChild(instance1);
    });

    lib.bigAction_content = Container.extend(function () {
        Container.call(this);
        var instance4 = new lib.bigAction_background();
        this[instance4.name = "background"] = instance4;
        var instance3 = new lib.bigAction_gradient();
        this[instance3.name = "gradient"] = instance3;
        var instance2 = new lib.bigAction_border();
        this[instance2.name = "border"] = instance2;
        var instance1 = new lib.bigAction_icon();
        this[instance1.name = "icon"] = instance1;
        this.addChild(instance4, instance3, instance2, instance1);
    });

    lib.bigActionButton = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 1,
            framerate: 30
        });
        var instance1 = new lib.bigAction_content()
            .setTransform(165.05, 165);
        this[instance1.name = "content"] = instance1;
        this.addChild(instance1);
    });

    lib.bigActionButton.assets = {
        "bigActionButton_atlas_1": "images/bigActionButton_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.bigActionButton,
        background: 0x0,
        width: 330,
        height: 330,
        framerate: 30,
        totalFrames: 1,
        library: lib
    };
}
