// jibo/resources/buttons/contentButton.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Container = PIXI.Container;
    var Graphics = PIXI.Graphics;
    var shapes = PIXI.animate.ShapesCache;

    lib.target = Container.extend(function () {
        Container.call(this);

    });

    lib.contentButton_1 = Container.extend(function () {
        Container.call(this);
        var instance1 = new Graphics()
            .drawCommands(shapes.contentButton[0])
            .setRenderable(false)
            .setTransform(-6.8, -8.85);
        var instance2 = this.target = new lib.target()
            .setTransform(-165, -165)
            .setMask(instance1);
        this.addChild(instance1, instance2);
    });

    lib.contentButton = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 1,
            framerate: 30
        });
        var instance1 = this.content = new lib.contentButton_1()
            .setTransform(165, 165);
        this.addChild(instance1);
    });

    lib.contentButton.assets = {
        "contentButton": "images/contentButton.shapes.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.contentButton,
        background: 0x0,
        width: 330,
        height: 330,
        framerate: 30,
        totalFrames: 1,
        library: lib
    };
}
