// jibo-cai-singleton/lib/debug/main/index.js
/**
 * @fileOverview
 *
 * Created on 3/7/16.
 * @author Siggi Orn <siggi@jibo.com>
 */
"use strict";
var path = require('path');
var findRoot = require('find-root');
var semverUtils = require('semver-utils');
var Version = (function () {
    function Version(name, semverString) {
        this.name = name;
        this.semver = Version.semverParse(semverString);
        this.major = this.semver.major;
        this.minor = this.semver.minor;
        this.patch = this.semver.patch;
    }
    Version.semverParse = function (input) {
        var parse = semverUtils.parse(input);
        return {
            semver: parse.semver,
            version: parse.version,
            major: parseInt(parse.major),
            minor: parseInt(parse.minor),
            patch: parseInt(parse.patch),
            release: parse.release
        };
    };
    /**
     * Checks to see if we fully match another version
     * @param {Version} other
     * @returns {boolean}
     */
    Version.prototype.majorMinorPatchMatch = function (other) {
        return other.major === this.major &&
            other.minor === this.minor &&
            other.patch === this.patch;
    };
    /**
     * Checks to see if we match another version on major and minor
     * @param {Version} other
     * @returns {boolean}
     */
    Version.prototype.majorMinorMatch = function (other) {
        return other.major === this.major &&
            other.minor === this.minor;
    };
    /**
     * Checks to see if we match another version on major, minor, and patch
     * @param {Version} other
     * @returns {boolean}
     */
    Version.prototype.exactMatch = function (other) {
        if (other.semver) {
            return other.semver.semver === this.semver.semver;
        }
        else {
            return this.majorMinorPatchMatch(other);
        }
    };
    Version.prototype.toString = function () {
        return this.name + '@' + this.semver.semver;
    };
    return Version;
}());
exports.Version = Version;
var Container = (function () {
    function Container(object, version) {
        this.object = object;
        this.version = version;
    }
    return Container;
}());
var Singleton = (function () {
    function Singleton() {
    }
    /**
     * Returns the root directory of the current module
     * @param {NodeModule} module
     * @returns {string}
     */
    Singleton.getProjectRoot = function (module) {
        return findRoot(module.filename);
    };
    /**
     * Reads version information from a package.json
     * @param {NodeModule} module The node module object for the project we
     * want to make into a singleton
     * @returns {Version}
     */
    Singleton.readVersion = function (module) {
        var root = Singleton.getProjectRoot(module);
        var packagePath = path.join(root, 'package.json');
        var versionString = require(packagePath).version;
        var nameString = require(packagePath).name;
        return new Version(nameString, versionString);
    };
    /**
     * Get's or creates the singleton class we want
     * @param {NodeModule} module The module of the project we want to make
     * into a singleton
     * @param {Function} creator A function that returns a new instance of
     * the class we want to make into a singleton
     */
    Singleton.getOrCreateSingleton = function (module, creator) {
        var incomingVersion = Singleton.readVersion(module);
        var container = singletons.get(incomingVersion.name);
        if (!container) {
            container = new Container(creator(), incomingVersion);
            container.object.__singletonVersion = incomingVersion;
            singletons.set(incomingVersion.name, container);
        }
        else if (!incomingVersion.majorMinorMatch(container.version)) {
            throw new Error("Incompatible versions: " +
                ("Incoming = '" + incomingVersion + "', existing= ") +
                ("" + container.version));
        }
        else if (!incomingVersion.exactMatch(container.version)) {
            console.warn("Patch difference in requested and loaded " +
                ("singletons. Loaded: '" + container.version + "', requested: ") +
                ("'" + incomingVersion + "'"));
        }
        return container.object;
    };
    return Singleton;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Singleton;
/**
 * This is where we store all the singletons
 */
if (!global['singletons4543']) {
    global['singletons4543'] = new Map();
}
var singletons = global['singletons4543'];
