// jibo-flow-core/lib/jibo-flow-core.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboFlowCore = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Activity {
    constructor(name, id, className, assetPack, options) {
        this.name = name;
        this.id = id;
        this.class = className;
        this.assetPack = assetPack;
        this.options = this.rejoinOptions(options);
    }
    rejoinOptions(options) {
        for (let prop in options) {
            let property = options[prop];
            if (Array.isArray(property)) {
                options[prop] = property.join('\n');
            }
        }
        return options;
    }
    set_transitions(transitions) {
        this.transitions = transitions;
    }
    set_exceptions(exceptions) {
        this.exceptions = exceptions;
    }
    set_procedure(procedure) {
        this.procedure = procedure;
    }
}
exports.default = Activity;

},{}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class FlowRoot {
    constructor(loader) {
        this.procedures = {};
        this.global_inits = {};
        this.loader = loader;
    }
    setStartInfoFromUri(uri) {
        this.startUri = this.loader.getUri(uri);
        this.startProcedureOrName = this.loader.getStartProcedureOrName(uri);
    }
    loadProcedure(currentUri, procedureObjectOrName) {
        return this.loader.loadProcedureRelative(currentUri, procedureObjectOrName);
    }
    getStartActivity() {
        return this.getStartProcedure().getBeginActivity();
    }
    getStartProcedure() {
        return this.loader.loadProcedure(this.startUri);
    }
    getStartProcedureOrName() {
        return this.startProcedureOrName;
    }
}
exports.default = FlowRoot;

},{}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const FlowRoot_1 = require("./FlowRoot");
const TreeLoader_1 = require("./loaders/TreeLoader");
const GojsLoader_1 = require("./loaders/GojsLoader");
class FlowRootFactory {
    static create(uri, loader) {
        if (loader === undefined) {
            if (typeof (uri) === 'function') {
                loader = TreeLoader_1.default;
            }
            else if (uri.endsWith(".flow")) {
                loader = new GojsLoader_1.default();
            }
            else {
                console.error('FlowRootFactory encountered an unknown file type: ' + uri);
                throw new Error('FlowRootFactory encountered an unknown file type: ' + uri);
            }
        }
        let flowRoot = new FlowRoot_1.default(loader);
        flowRoot.setStartInfoFromUri(uri);
        return flowRoot;
    }
}
exports.default = FlowRootFactory;

},{"./FlowRoot":2,"./loaders/GojsLoader":6,"./loaders/TreeLoader":7}],4:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Activity_1 = require("./Activity");
const path = require("path");
const findRoot = require("find-root");
class Procedure {
    constructor(name, uri, local_inits, param_inits) {
        this.name = name;
        this.uri = uri;
        this.activities = {};
        this.local_inits = local_inits;
        this.param_inits = param_inits;
        this.transitions = {};
        this.exceptions = {};
        this.connections = [];
    }
    loadPropertiesFromTreeFunction(tree) {
        let treeObj = tree();
        for (let activityId in treeObj) {
            let t = treeObj[activityId];
            if (typeof t === 'function') {
                t = t();
            }
            let activity = new Activity_1.default(t.name, t.id, t.class, t['asset-pack'], t.options);
            activity.transitions = t.transitions;
            activity.exceptions = t.exceptions;
            this.activities[t.id] = activity;
        }
        this.name = treeObj.meta.name;
        this.uri = treeObj.meta.uri;
        this.relativeUri = path.relative(findRoot('.'), this.uri);
        this.local_inits = {};
        this.param_inits = {};
        this.procedureConstructor = tree;
    }
    add_transition(transition) {
        let map = transition.value.startsWith('~') ? this.exceptions : this.transitions;
        if (!(transition.frm in map)) {
            map[transition.frm] = [];
        }
        map[transition.frm].push(transition);
        this.connections.push(transition);
    }
    add_activity(activity) {
        let transitions = activity.id in this.transitions ? this.transitions[activity.id] : [];
        let exceptions = activity.id in this.exceptions ? this.exceptions[activity.id] : [];
        activity.set_transitions(transitions);
        activity.set_exceptions(exceptions);
        activity.set_procedure(this);
        this.activities[activity.id] = activity;
    }
    getBeginActivity() {
        for (let key in this.activities) {
            let activity = this.activities[key];
            if (activity.class === 'Flow.Begin') {
                return activity;
            }
        }
        throw new Error("No Flow.Begin Activity in procedure " + this.name);
    }
    getAsyncBeginActivities() {
        let asyncs = [];
        for (let key in this.activities) {
            let activity = this.activities[key];
            if (activity.class === 'Flow.Begin-Parallel') {
                asyncs.push(activity);
            }
        }
        return asyncs;
    }
}
exports.default = Procedure;

},{"./Activity":1,"find-root":undefined,"path":undefined}],5:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Transition {
    constructor(frm, to, value) {
        this.frm = frm;
        this.to = to;
        this.value = value;
    }
}
exports.default = Transition;

},{}],6:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Activity_1 = require("../Activity");
const Procedure_1 = require("../Procedure");
const Transition_1 = require("../Transition");
const fs = require('fs');
const path = require('path');
class GojsLoader {
    constructor() {
        this.procedures = {};
    }
    getStartProcedureOrName(uri) {
        return path.parse(uri).name;
    }
    getUri(uri) {
        return uri;
    }
    getBase(uri) {
        return path.parse(uri).dir;
    }
    loadProcedureRelative(currentUri, name) {
        let baseUri = this.getBase(currentUri);
        let uri = path.join(baseUri, name + '.flow');
        return this.loadProcedure(uri);
    }
    loadProcedure(uri) {
        let name = this.getStartProcedureOrName(uri);
        if (!this.procedures.hasOwnProperty(name)) {
            this.procedures[name] = this.readProcedure(name, uri);
        }
        return this.procedures[name];
    }
    readProcedure(name, uri) {
        let data = fs.readFileSync(uri, 'utf8');
        let base = JSON.parse(data);
        let procedure = new Procedure_1.default(name, uri, {}, {});
        for (let i = 0; i < base.linkDataArray.length; i++) {
            let goLink = base.linkDataArray[i];
            let transitionValue = goLink.text === undefined ? "" : goLink.text;
            let transition = new Transition_1.default(goLink.from, goLink.to, transitionValue);
            procedure.add_transition(transition);
        }
        for (let i = 0; i < base.nodeDataArray.length; i++) {
            let goNode = base.nodeDataArray[i];
            let activity = new Activity_1.default(goNode.name, goNode.id, goNode.class, goNode['asset-pack'], goNode.options);
            procedure.add_activity(activity);
        }
        return procedure;
    }
}
exports.default = GojsLoader;

},{"../Activity":1,"../Procedure":4,"../Transition":5,"fs":undefined,"path":undefined}],7:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Procedure_1 = require("../Procedure");
class TreeLoader {
    static getStartProcedureOrName(data) {
        return data;
    }
    static getUri(data) {
        return data().meta.uri;
    }
    static loadProcedureRelative(uri, procedureObject) {
        let procedure = new Procedure_1.default();
        procedure.loadPropertiesFromTreeFunction(procedureObject);
        return procedure;
    }
}
exports.default = TreeLoader;

},{"../Procedure":4}],8:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GojsLoader_1 = require("./GojsLoader");
const TreeLoader_1 = require("./TreeLoader");
exports.default = {
    GojsLoader: GojsLoader_1.default,
    TreeLoader: TreeLoader_1.default
};

},{"./GojsLoader":6,"./TreeLoader":7}],9:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Activity_1 = require("./Activity");
exports.Activity = Activity_1.default;
const FlowRoot_1 = require("./FlowRoot");
exports.FlowRoot = FlowRoot_1.default;
const FlowRootFactory_1 = require("./FlowRootFactory");
exports.FlowRootFactory = FlowRootFactory_1.default;
const Procedure_1 = require("./Procedure");
exports.Procedure = Procedure_1.default;
const Transition_1 = require("./Transition");
exports.Transition = Transition_1.default;
const loaders_1 = require("./loaders");
exports.loaders = loaders_1.default;

},{"./Activity":1,"./FlowRoot":2,"./FlowRootFactory":3,"./Procedure":4,"./Transition":5,"./loaders":8}]},{},[9])(9)
});
