// jibo-interaction-memory/lib/jibo-interaction-memory.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.jiboInteractionMemory = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const InteractionMemory_1 = require("./im/InteractionMemory");
/**
 * Initialized the InteractionMemory Singleton. Requires a Jibo object
 * @param {jibo} jibo runtime
 * @param {StorageOptions} options for interaction memory
 */
function init(jibo, storageOptions) {
    exports._memory = new InteractionMemory_1.InteractionMemory(jibo, storageOptions);
}
exports.init = init;
/**
 * Store the event
 * @param {InteractionDescription} description
 * @returns {PromiseLike<TResult>|Promise<TResult>}
 */
function noteEvent(description) {
    return exports._memory.noteEvent(description);
}
exports.noteEvent = noteEvent;
/**
 *
 * @param {InteractionDescription} queryDesc - an InteractionDescription to be matched
 * @param compare - a function returning true when two InteractionDescriptions are considered a match, false otherwise
 * @param {Date|null} startTime - optional start time of search window; i.e., when we stop searching backwards in time
 * @param {Date} [refDate] - optional, a reference date to compute against. If not provided the current time (as per InteractionMemory's dateProvider) will be used.
 * @returns number of milliseconds since a matching event, or Number.POSITIVE_INFINITY to indicate no match
 */
function getTimeSinceLast(queryDesc, compare = InteractionMemory_1.InteractionMemory.compareUser, startTime, refDate) {
    return exports._memory.getTimeSinceLast(queryDesc, compare, startTime, refDate);
}
exports.getTimeSinceLast = getTimeSinceLast;
/**
 * Returns a static compare method to be used in interaction memory queries
 * Every person mentioned in the query should be included in the candidate
 * @returns compare function
 */
exports.compareUser = InteractionMemory_1.InteractionMemory.compareUser;
/**
 * Returns a static compare method to be used in interaction memory queries
 * Every property specified in the query must match; additional properties in the candidate are okay.
 * @returns compare function
 */
exports.compareQueryFields = InteractionMemory_1.InteractionMemory.compareQueryFields;

},{"./im/InteractionMemory":2}],2:[function(require,module,exports){
"use strict";
/**
 * @fileOverview
 *
 * Created on 4/28/16.
 * @author Andrew Stout <andrew.stout@jibo.com>
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const jibo_log_1 = require("jibo-log");
const _ = require("lodash/lodash.min");
class InteractionEvent {
    //timestamp: Date; // ms since the epoch
    constructor(description, timestamp) {
        this.description = description;
        this.timestamp = timestamp;
    }
}
exports.InteractionEvent = InteractionEvent;
class DefaultDateProvider {
    getDate() {
        return new Date();
    }
}
exports.DefaultDateProvider = DefaultDateProvider;
var Edge;
(function (Edge) {
    Edge[Edge["history"] = 'history'] = "history";
    Edge[Edge["stats"] = 'stats'] = "stats";
})(Edge || (Edge = {}));
exports.DEFAULT_STORAGE_OPTIONS = {
    storageFolder: '/jibo/interactionmemory'
};
/* tslint:disable:member-ordering */
class InteractionMemory {
    constructor(jibo, options) {
        this._dateProvider = new DefaultDateProvider();
        this.log = new jibo_log_1.Log('Jibo.InteractionMemory');
        this.log.info("Initializing interaction memory.");
        this.jibo = jibo;
        let storageOptions = options || exports.DEFAULT_STORAGE_OPTIONS;
        this._kbModel = this.jibo.kb.createModel(storageOptions.storageFolder);
    }
    /**
     * Set a date provider for Interaction Memory to use to create dates
     * @param {Object} dateProvider
     */
    setDateProvider(dateProvider) {
        this._dateProvider = dateProvider;
    }
    /**
     * create an InteractionEvent with a timestamp aligned with InteractionMemory's dateProvider.
     * @param description
     * @returns {InteractionEvent}
     */
    createEvent(description) {
        return new InteractionEvent(description, this._dateProvider.getDate());
    }
    _getOrInitializeHistory() {
        return this._getOrCreateNode(Edge.history)
            .then((node) => {
            // check existence of array, initialize if nonexistant
            if (!node.data.history) {
                node.data.history = [];
            }
            return node;
        }).catch((error) => {
            if (error.message) {
                error.message = 'Could not get or initialize history: ' + error.message;
            }
            throw error;
        });
    }
    _getOrInitializeStats() {
        return this._getOrCreateNode(Edge.stats)
            .then((node) => {
            // check existence of data, initialize if nonexistant
            if (!node.data.userSkillLaunchCounts) {
                node.data.userSkillLaunchCounts = {};
            }
            return node;
        }).catch((error) => {
            if (error.message) {
                error.message = 'Could not get or initialize stats: ' + error.message;
            }
            throw error;
        });
    }
    /**
     * Store the event
     * @param {InteractionDescription} description
     * @returns {Promise<InteractionEvent>}
     */
    noteEvent(description) {
        let event = new InteractionEvent(description, this._dateProvider.getDate());
        // store the event in the event history
        let phist = this._getOrInitializeHistory()
            .then((node) => {
            // push event
            // convert event timestamp to ms before saving
            event.timestamp = Number(event.timestamp);
            node.data.history.push(event);
            // as a paranoid safeguard, limit the size of the history.
            // if it's over size, dump the oldest - in batches, since that'll be a bit more performant
            if (node.data.history.length > InteractionMemory.HISTORY_SIZE_LIMIT) {
                node.data.history = node.data.history.slice(InteractionMemory.HISTORY_SIZE_DECREMENT);
            }
            return this._saveNode(node);
        }).catch((error) => {
            this.log.error("Could not note event: ", error);
        });
        // increment user skill launch counts (consider moving this to its own function)
        let pstats = this._getOrInitializeStats()
            .then((statsNode) => {
            if (description.personIDs && description.skillName) {
                for (let userID of description.personIDs) {
                    let u = statsNode.data.userSkillLaunchCounts[userID];
                    if (!u) {
                        statsNode.data.userSkillLaunchCounts[userID] = {};
                    }
                    let us = statsNode.data.userSkillLaunchCounts[userID][description.skillName];
                    if (!us) {
                        statsNode.data.userSkillLaunchCounts[userID][description.skillName] = 0;
                    }
                    statsNode.data.userSkillLaunchCounts[userID][description.skillName]++;
                }
            }
            return this._saveNode(statsNode);
        }).catch((error) => {
            this.log.error("Could not note event: ", error);
        });
        return Promise.all([phist, pstats])
            .then(() => {
            return event;
        });
    }
    /**
     * default EventDescription comparison function
     * if present, every person mentioned in the query should be included in the candidate
     * @param {InteractionDescription} query
     * @param {InteractionDescription} candidate
     * @returns {boolean}
     */
    static compareUser(query, candidate) {
        // if present, every person mentioned in the query should be included in the candidate
        if (query.personIDs && candidate.personIDs) {
            for (let person of query.personIDs) {
                if (candidate.personIDs.indexOf(person) === -1) {
                    return false;
                }
            }
            return true;
        }
        return false;
    }
    /**
     * every person mentioned in the query should be included in the candidate
     * skillName should match
     * both must be present
     * @param {InteractionDescription} query
     * @param {InteractionDescription} candidate
     * @returns {boolean}
     */
    static compareUserAndSkill(query, candidate) {
        // every person mentioned in the query should be included in the candidate
        // skillName should match
        // both must be present
        if (query.personIDs && candidate.personIDs && query.skillName && candidate.skillName) {
            for (let person of query.personIDs) {
                if (candidate.personIDs.indexOf(person) === -1) {
                    return false;
                }
            }
            return query.skillName === candidate.skillName;
        }
        else {
            return false;
        }
    }
    /**
     * Every property specified in the query must match; additional properties in the candidate are okay.
     * @param query
     * @param candidate
     * @returns {boolean}
     */
    static compareQueryFields(query, candidate) {
        let ret = true;
        Object.keys(query).forEach(function (key, index) {
            if (!(candidate[key] && _.isEqual(candidate[key], query[key]))) {
                ret = false;
            }
        });
        return ret;
    }
    /**
     *
     * @param {InteractionDescription} queryDesc - an InteractionDescription to be matched
     * @param compare - a function returning true when two InteractionDescriptions are considered a match, false otherwise
     * @param {Date|null} startTime - optional start time of search window; i.e., when we stop searching backwards in time
     * @param {Date} [refDate] - optional, a reference date to compute against. If not provided the current time (as per InteractionMemory's dateProvider) will be used.
     * @returns number of milliseconds since a matching event, or Number.POSITIVE_INFINITY to indicate no match
     */
    getTimeSinceLast(queryDesc, compare = InteractionMemory.compareUser, startTime, refDate = this._dateProvider.getDate()) {
        // stupid version: search (chronologically backwards) for matching event
        // eventually, a smarter version does some kind of database thing.
        const query = new InteractionEvent(queryDesc, refDate);
        return this._getOrInitializeHistory()
            .then((node) => {
            // search through history for matching event
            for (let ei = node.data.history.length - 1; ei >= 0; ei--) {
                let e = node.data.history[ei]; // timestamp comes out of storage as a number (ms)
                // we're progressing in reverse chronological order.
                if (startTime && e.timestamp < startTime.getTime()) {
                    // we have (progressing in reverse chronological order), reached events
                    // before the start time. Assuming order, we can stop iterating now.
                    break;
                }
                // check whether e matches the query
                if (compare(query.description, e.description)) {
                    const queryTimestamp = query.timestamp.getTime();
                    const eventTimestamp = e.timestamp;
                    this.log.info(`Matched. Query timestamp: ${queryTimestamp} Event timestamp: ${eventTimestamp}`);
                    // compute time since
                    return queryTimestamp - eventTimestamp;
                }
            }
            // if no match, return some indication
            return Number.POSITIVE_INFINITY;
        }).catch((error) => {
            // There's no valid number we can return here, throw an error.
            if (error.message) {
                error.message = 'Could not get time since last: ' + error.message;
            }
            throw error;
        });
    }
    /**
     * returns promise of number of milliseconds since last observed interaction event, or Number.POSITIVE_INFINITY to indicate no events in history
     * @param {Date} [refTime] Reference time (defaults to current time as set by dateProvider)
     * @returns {Promise<number>}
     */
    getTimeSinceLastInteraction(refTime) {
        const _refTime = refTime || this._dateProvider.getDate();
        return this._getOrInitializeHistory()
            .then((node) => {
            if (node.data.history.length > 0) {
                let lasttime = node.data.history[node.data.history.length - 1].timestamp;
                return _refTime.getTime() - lasttime; // timestamp comes out of storage as a number (ms)
            }
            else {
                return Number.POSITIVE_INFINITY;
            }
        }).catch((error) => {
            // There's no valid number we can return here, throw an error.
            if (error.message) {
                error.message = 'Could not get time since last interaction: ' + error.message;
            }
            throw error;
        });
    }
    /**
     * returns promise of number of times user userID has launched skill skillName
     * @param {string} userID
     * @param {string} skillName
     * @returns {Promise<number>}
     */
    getUserSkillCount(userID, skillName) {
        return this._getOrInitializeStats()
            .then((statsNode) => {
            let u = statsNode.data.userSkillLaunchCounts[userID];
            if (!u) {
                return 0;
            }
            else {
                let us = u[skillName];
                if (!us) {
                    return 0;
                }
                else {
                    return us;
                }
            }
        }).catch((error) => {
            if (error.message) {
                error.message = 'Could not get user skill count: ' + error.message;
            }
            throw error;
        });
    }
    /**
     * get all events matching a query
     * @intdocs
     * @param {EventFilter} filter Functor determining desired events
     * @param {Date} [startTime] start of chronological window of interest
     * @param {Date} [endTime] end of chronological window of interest
     */
    getEventsMatchingFilter(filter, startTime, endTime) {
        return __awaiter(this, void 0, void 0, function* () {
            // NOTE: like getTimeSinceLast, this implementation relies on the chronological ordering of events
            let filteredEvents = new Array();
            let node;
            try {
                node = yield this._getOrInitializeHistory();
            }
            catch (error) {
                if (error.message) {
                    error.message = "Could not get events matching filter: " + error.message;
                }
                throw error;
            }
            for (let ei = node.data.history.length - 1; ei >= 0; ei--) {
                let e = node.data.history[ei];
                // we're progressing in reverse chronological order.
                if (startTime && e.timestamp < startTime.getTime()) {
                    // we have (progressing in reverse chronological order), reached events
                    // before the start time. Assuming order, we can stop iterating now.
                    break;
                }
                if (endTime && e.timestamp > endTime.getTime()) {
                    // we haven't yet gotten to the time window of the query; keep iterating
                    continue;
                }
                if (filter(e)) {
                    filteredEvents.push(e); // This reverses the order relative to storage--this will be earliest first.
                }
            }
            return filteredEvents;
        });
    }
    /**
     * Compute the univariate mean and variance of univariate data set.
     * @param {number[]} data
     * @returns {UnivariateMeanAndVariance}
     * @intdocs
     * @static
     */
    static univariateMeanAndVariance(data) {
        // assumes uniformly-weighted samples
        // mu = sum(data)/data.length
        // sigma^2 = sum(data^2)/data.length - mu^2
        let sum_x = 0;
        let sum_xsq = 0;
        for (let x of data) {
            sum_x += x;
            sum_xsq += x * x;
        }
        let mu = sum_x / data.length;
        let sigmasq = sum_xsq / data.length - mu * mu;
        return { mu: mu, sigmasq: sigmasq };
    }
    /**
     * compute mean and variance for the time of day (as minutes after 0:00 midnight) of a set of events
     * @static
     * @param {InteractionEvent[]} events
     * @returns {UnivariateMeanAndVariance}
     * @intdocs
     * @memberOf InteractionMemory
     */
    static getTimeOfDaySummaryStatistics(events) {
        // apply accessor to get time data
        let accessor = (ie) => {
            let date = new Date(ie.timestamp);
            return date.getHours() * 60 + date.getMinutes();
        };
        let data = [];
        for (let ie of events) {
            data.push(accessor(ie));
        }
        // apply univariateMeanAndVariance
        return InteractionMemory.univariateMeanAndVariance(data);
    }
    /**
     * Clear the history.
     * @returns {Promise<void>}
     * @intdocs
     */
    _clear() {
        let phist = this._getOrCreateNode(Edge.history)
            .then((node) => {
            node.data.history = [];
            return this._saveNode(node);
        }).catch((error) => {
            this.log.error("Could not clear: ", error);
        });
        let pstats = this._getOrCreateNode(Edge.stats)
            .then((statsNode) => {
            statsNode.data.userSkillLaunchCounts = {};
            return this._saveNode(statsNode);
        }).catch((error) => {
            this.log.error("Could not clear: ", error);
        });
        return Promise.all([phist, pstats]).then(() => null);
    }
    _getHistory() {
        return this._getOrInitializeHistory()
            .then((node) => {
            // convert the history nodes back to InteractionEvents
            // timestamps come out of storage as a number (ms)
            // InteractionEvents have Date objects as their timestamp
            let history = [];
            for (let ei = node.data.history.length - 1; ei >= 0; ei--) {
                let e = node.data.history[ei];
                let event = new InteractionEvent(e.description, new Date(e.timestamp));
                history.push(event);
            }
            return history;
        }).catch((error) => {
            if (error.message) {
                error.message = 'Could not get history: ' + error.message;
            }
            throw error;
        });
    }
    /**
     * Load a history or stats node from the root node, or create one
     * and attach it to the root node if it doesn't already exist.
     * @param {string} edge Name of edge to follow or create. Also
     * serves as node type.
     * @returns {Promise<Node>}
     */
    _getOrCreateNode(edge) {
        let edgeName = Edge[edge];
        if (!edgeName) {
            throw new Error(`Invalid 'edge' name: ${edge}`);
        }
        return this._loadRoot()
            .then((rootNode) => {
            let edges = rootNode.getEdges(edgeName);
            if (edges.length > 0) {
                let id = edges[0];
                return this._loadNode(id);
            }
            else {
                // no history node yet, so create one
                let node = this._kbModel.createNode(edgeName); // node type = edge name
                return this._saveNode(node)
                    .then(() => {
                    rootNode.addEdges(node, edgeName);
                    return this._saveNode(rootNode)
                        .then(() => {
                        return node;
                    });
                });
            }
        }).catch((error) => {
            if (error.message) {
                error.message = 'Could not get or create node: ' + error.message;
            }
            throw error;
        });
    }
    /**
     * A promise wrapper around the jibo.kb.Model#_loadRoot call. Loads the
     * root node for the interaction memory KB model.
     * @returns {Promise<Node>} The root node for the interaction memory model.
     */
    _loadRoot() {
        return new Promise((res, rej) => {
            this._kbModel.loadRoot((err, rootNode) => {
                if (err) {
                    rej(err);
                }
                else {
                    res(rootNode);
                }
            });
        });
    }
    /**
     * A promise wrapper around the jibo.kb.Node#_save call.
     * @param {Node} The node to be saved.
     * @returns {Promise}
     */
    _saveNode(node) {
        return new Promise((res, rej) => {
            node.save((err) => {
                if (err) {
                    rej(err);
                }
                else {
                    res();
                }
            });
        });
    }
    /**
     * A promise wrapper around the jibo.kb.Model#_load call.
     * @param {string} id ID of node to load.
     * @returns {Promise<Node>}
     */
    _loadNode(id) {
        return new Promise((res, rej) => {
            this._kbModel.load(id, (err, node) => {
                if (err) {
                    rej(err);
                }
                else {
                    res(node);
                }
            });
        });
    }
}
InteractionMemory.HISTORY_SIZE_LIMIT = 3000; // limit to the size of the history, just as a paranoid safeguard.
// just because you're paranoid doesn't mean they're not out to get you
InteractionMemory.HISTORY_SIZE_DECREMENT = 100; // how many old events to remove at once
exports.InteractionMemory = InteractionMemory;
// For testing
if (global._imTest) {
    global._imTest.InteractionMemory = module.exports;
}

},{"jibo-log":undefined,"lodash/lodash.min":undefined}],3:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./InteractionMemory"));

},{"./InteractionMemory":2}],4:[function(require,module,exports){
"use strict";
// this will probably eventually need to be a singleton, but gonna worry about that later.
Object.defineProperty(exports, "__esModule", { value: true });
const im = require("./im");
exports.im = im;
const api = require("./api");
exports.api = api;

},{"./api":1,"./im":3}]},{},[4])(4)
});
