// jibo-anim-db-animations/timelines/alert_wifiDisconnected.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 66, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-text1"))
            .setTransform(-457.15, -71.9);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-dot2"))
            .setTransform(-63, -63.5);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-bar32"))
            .setTransform(-311.25, -83.85);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-bar22"))
            .setTransform(-227.6, -68.05);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-bar12"))
            .setTransform(-143, -50.5);
        this.addTimedChild(instance1);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 77, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-dot1"))
            .setTransform(-63, -63.5);
        this.addTimedChild(instance1);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 77, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-bar31"))
            .setTransform(-311.25, -83.85);
        this.addTimedChild(instance1);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 77, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-bar21"))
            .setTransform(-227.6, -68.05);
        this.addTimedChild(instance1);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 77, loop: false });
        var instance1 = new Sprite(fromFrame("alert-wifi-bar11"))
            .setTransform(-143, -50.5);
        this.addTimedChild(instance1);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance8 = new Graphic9(MovieClip.SYNCHED)
            .setTransform(0, -68.4);
        var instance7 = new Graphic8(MovieClip.SYNCHED)
            .setTransform(0, -166.6);
        var instance6 = new Graphic7(MovieClip.SYNCHED)
            .setTransform(0, -264.5);
        var instance5 = new Graphic6(MovieClip.SYNCHED)
            .setTransform(0, 62.8);
        var instance4 = new Graphic5(MovieClip.SYNCHED);
        var instance3 = new Graphic4(MovieClip.SYNCHED);
        var instance2 = new Graphic3(MovieClip.SYNCHED);
        var instance1 = new Graphic2(MovieClip.SYNCHED);
        var instance9 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance8, 13, 77)
            .addTimedChild(instance7, 13, 77)
            .addTimedChild(instance6, 13, 77)
            .addTimedChild(instance5, 13, 77)
            .addTimedChild(instance4, 0, 90, {
                "0": {
                    y: -68.4,
                    a: 0
                },
                "1": {
                    a: 0.01
                },
                "2": {
                    a: 0.05
                },
                "3": {
                    a: 0.13
                },
                "4": {
                    a: 0.23
                },
                "5": {
                    a: 0.38
                },
                "6": {
                    a: 0.55
                },
                "7": {
                    a: 0.72
                },
                "8": {
                    a: 0.85
                },
                "9": {
                    a: 0.94
                },
                "10": {
                    a: 0.98
                },
                "11": {
                    a: 1
                },
                "19": {
                    a: 0.98
                },
                "20": {
                    a: 0.9
                },
                "21": {
                    a: 0.75
                },
                "22": {
                    a: 0.53
                },
                "23": {
                    a: 0.3
                },
                "24": {
                    a: 0.13
                },
                "25": {
                    a: 0.03
                },
                "26": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 0, 90, {
                "0": {
                    y: -166.6,
                    a: 0
                },
                "1": {
                    a: 0.01
                },
                "2": {
                    a: 0.05
                },
                "3": {
                    a: 0.13
                },
                "4": {
                    a: 0.23
                },
                "5": {
                    a: 0.38
                },
                "6": {
                    a: 0.55
                },
                "7": {
                    a: 0.72
                },
                "8": {
                    a: 0.85
                },
                "9": {
                    a: 0.94
                },
                "10": {
                    a: 0.98
                },
                "11": {
                    a: 1
                },
                "17": {
                    a: 0.98
                },
                "18": {
                    a: 0.9
                },
                "19": {
                    a: 0.75
                },
                "20": {
                    a: 0.53
                },
                "21": {
                    a: 0.3
                },
                "22": {
                    a: 0.13
                },
                "23": {
                    a: 0.03
                },
                "24": {
                    a: 0
                }
            })
            .addTimedChild(instance2, 0, 90, {
                "0": {
                    y: -264.5,
                    a: 0
                },
                "1": {
                    a: 0.01
                },
                "2": {
                    a: 0.05
                },
                "3": {
                    a: 0.13
                },
                "4": {
                    a: 0.23
                },
                "5": {
                    a: 0.38
                },
                "6": {
                    a: 0.55
                },
                "7": {
                    a: 0.72
                },
                "8": {
                    a: 0.85
                },
                "9": {
                    a: 0.94
                },
                "10": {
                    a: 0.98
                },
                "11": {
                    a: 1
                },
                "14": {
                    a: 0.97
                },
                "15": {
                    a: 0.86
                },
                "16": {
                    a: 0.66
                },
                "17": {
                    a: 0.4
                },
                "18": {
                    a: 0.17
                },
                "19": {
                    a: 0.04
                },
                "20": {
                    a: 0
                }
            })
            .addTimedChild(instance1, 0, 90, {
                "0": {
                    y: 62.8,
                    sx: 0.873,
                    sy: 0.873,
                    a: 1,
                    c: [
                        0,
                        0.16,
                        0,
                        0.67,
                        0,
                        0.73
                    ]
                },
                "1": {
                    y: 62.787,
                    sx: 0.882,
                    sy: 0.882,
                    c: [
                        0.03,
                        0.16,
                        0.03,
                        0.66,
                        0.03,
                        0.71
                    ]
                },
                "2": {
                    y: 62.767,
                    sx: 0.919,
                    sy: 0.919,
                    c: [
                        0.13,
                        0.14,
                        0.13,
                        0.58,
                        0.13,
                        0.64
                    ]
                },
                "3": {
                    y: 62.768,
                    sx: 0.991,
                    sy: 0.991,
                    c: [
                        0.34,
                        0.11,
                        0.34,
                        0.45,
                        0.34,
                        0.49
                    ]
                },
                "4": {
                    y: 62.805,
                    sx: 1.036,
                    sy: 1.036,
                    c: [
                        0.47,
                        0.09,
                        0.47,
                        0.36,
                        0.47,
                        0.39
                    ]
                },
                "5": {
                    y: 62.8,
                    sx: 1.047,
                    sy: 1.047,
                    c: [
                        0.5,
                        0.08,
                        0.5,
                        0.34,
                        0.5,
                        0.37
                    ]
                },
                "6": {
                    y: 62.806,
                    sx: 1.045,
                    sy: 1.045,
                    c: [
                        0.53,
                        0.08,
                        0.53,
                        0.32,
                        0.53,
                        0.35
                    ]
                },
                "7": {
                    y: 62.801,
                    sx: 1.035,
                    sy: 1.035,
                    c: [
                        0.63,
                        0.06,
                        0.63,
                        0.25,
                        0.63,
                        0.27
                    ]
                },
                "8": {
                    y: 62.836,
                    sx: 1.015,
                    sy: 1.015,
                    c: [
                        0.84,
                        0.03,
                        0.84,
                        0.11,
                        0.84,
                        0.12
                    ]
                },
                "9": {
                    y: 62.793,
                    sx: 1.003,
                    sy: 1.003,
                    c: [
                        0.97,
                        0,
                        0.97,
                        0.02,
                        0.97,
                        0.02
                    ]
                },
                "10": {
                    y: 62.8,
                    sx: 1,
                    sy: 1,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "25": {
                    a: 0.97
                },
                "26": {
                    a: 0.83
                },
                "27": {
                    a: 0.55
                },
                "28": {
                    a: 0.26
                },
                "29": {
                    a: 0.09
                },
                "30": {
                    a: 0.02
                },
                "31": {
                    a: 0
                }
            })
            .addTimedChild(instance9, 24, 66, {
                "24": {
                    y: 214.45,
                    a: 0
                },
                "26": {
                    a: 0.02
                },
                "27": {
                    a: 0.04
                },
                "28": {
                    a: 0.07
                },
                "29": {
                    a: 0.13
                },
                "30": {
                    a: 0.2
                },
                "31": {
                    a: 0.3
                },
                "32": {
                    a: 0.42
                },
                "33": {
                    a: 0.55
                },
                "34": {
                    a: 0.69
                },
                "35": {
                    a: 0.8
                },
                "36": {
                    a: 0.88
                },
                "37": {
                    a: 0.94
                },
                "38": {
                    a: 0.98
                },
                "39": {
                    a: 1
                }
            });
    });

    lib.wifiDisconnected = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 90,
            framerate: 30,
            loop: false
        });
        var instance1 = new Graphic10(MovieClip.SYNCHED)
            .setTransform(648, 387.15);
        this.addTimedChild(instance1);
    });

    lib.wifiDisconnected.assets = {
        "wifiDisconnected_atlas_1": "images/wifiDisconnected_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.wifiDisconnected,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 90,
        library: lib
    };
}
