// jibo-anim-db-animations/timelines/alert_lowBattery.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;
    var Graphics = PIXI.Graphics;
    var shapes = PIXI.animate.ShapesCache;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 57, loop: false });
        var instance1 = new Sprite(fromFrame("batterylow-symbols-minus1"))
            .setTransform(-14.1, 155.6);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 57, loop: false });
        var instance1 = new Sprite(fromFrame("batterylow-symbols-plus1"))
            .setTransform(-13.95, -166.4);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("batterylow-cap1"));
        var instance2 = new Sprite(fromFrame("batterylow-cap2"));
        var instance3 = new Sprite(fromFrame("batterylow-cap3"));
        var instance4 = new Sprite(fromFrame("batterylow-cap4"));
        var instance5 = new Sprite(fromFrame("batterylow-cap5"))
            .setTransform(-55.3, -33.95, 1.087, 1.087);
        this.addTimedChild(instance1, 0, 61, {
                "0": {
                    x: -66.7,
                    y: -29.35,
                    sx: 1.087,
                    sy: 1.087
                }
            })
            .addTimedChild(instance2, 61, 2, {
                "61": {
                    x: -66.65,
                    y: -34.35,
                    sx: 1.087,
                    sy: 1.087
                }
            })
            .addTimedChild(instance3, 63, 1, {
                "63": {
                    x: -66.75,
                    y: -38.85,
                    sx: 1.087,
                    sy: 1.087
                }
            })
            .addTimedChild(instance4, 64, 1, {
                "64": {
                    x: -66.25,
                    y: -37.5,
                    sx: 1.087,
                    sy: 1.087
                }
            })
            .addTimedChild(instance5, 65, 8);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 86, loop: false });
        var instance1 = new Sprite(fromFrame("batterylow-top1"));
        var instance2 = new Sprite(fromFrame("batterylow-top2"));
        var instance3 = new Sprite(fromFrame("batterylow-top3"));
        var instance4 = new Sprite(fromFrame("batterylow-top4"));
        var instance5 = new Sprite(fromFrame("batterylow-top5"))
            .setTransform(-148.1, -151.9, 1.087, 1.087);
        this.addTimedChild(instance1, 0, 74, {
                "0": {
                    x: -153.15,
                    y: -34.8,
                    sx: 1.087,
                    sy: 1.087
                }
            })
            .addTimedChild(instance2, 74, 2, {
                "74": {
                    x: -152.35,
                    y: -47.8,
                    sx: 1.087,
                    sy: 1.087
                }
            })
            .addTimedChild(instance3, 76, 1, {
                "76": {
                    x: -152.35,
                    y: -74.4,
                    sx: 1.087,
                    sy: 1.087
                }
            })
            .addTimedChild(instance4, 77, 1, {
                "77": {
                    x: -147,
                    y: -103.95,
                    sx: 1.087,
                    sy: 1.087
                }
            })
            .addTimedChild(instance5, 78, 8);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 60, loop: false });
        var instance2 = new Graphics()
            .drawCommands(shapes.lowBattery[2]);
        var instance1 = new Graphics()
            .drawCommands(shapes.lowBattery[1]);
        this.addTimedChild(instance2)
            .addTimedChild(instance1);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("batterylow-container1"))
            .setTransform(-153.15, -249.2, 1.087, 1.087);
        this.addTimedChild(instance1);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic6(MovieClip.SYNCHED);
        var instance2 = new Graphic5(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
                "0": {
                    x: -9.2,
                    y: 440,
                    sx: 0.92,
                    sy: 0.92
                },
                "1": {
                    y: 438.85
                },
                "2": {
                    y: 435
                },
                "3": {
                    y: 427.75
                },
                "4": {
                    y: 416
                },
                "5": {
                    y: 398.1
                },
                "6": {
                    y: 370.9
                },
                "7": {
                    y: 328.7
                },
                "8": {
                    y: 260.2
                },
                "9": {
                    y: 164.9
                },
                "10": {
                    y: 96.75
                },
                "11": {
                    y: 64.3
                },
                "12": {
                    y: 50.45
                },
                "13": {
                    y: 46.85
                },
                "63": {
                    y: 51.7
                },
                "64": {
                    y: 67.55
                },
                "65": {
                    y: 96.75
                },
                "66": {
                    y: 141.35
                },
                "67": {
                    y: 201.95
                },
                "68": {
                    y: 274.25
                },
                "69": {
                    y: 346
                },
                "70": {
                    y: 402.4
                },
                "71": {
                    y: 435.45
                },
                "72": {
                    y: 445.65
                }
            })
            .addTimedChild(instance2, 13, 60, {
                "13": {
                    x: -9.2,
                    y: 606.6,
                    sx: 0.92,
                    sy: 0.92,
                    a: 1
                },
                "14": {
                    y: 600.45
                },
                "15": {
                    y: 595.4
                },
                "16": {
                    y: 591.25
                },
                "17": {
                    y: 587.8
                },
                "18": {
                    y: 584.95
                },
                "19": {
                    y: 582.55
                },
                "20": {
                    y: 580.6
                },
                "21": {
                    y: 579
                },
                "22": {
                    y: 577.7
                },
                "23": {
                    y: 576.7
                },
                "24": {
                    y: 575.95
                },
                "25": {
                    y: 575.45
                },
                "26": {
                    y: 575.15
                },
                "27": {
                    y: 575.05
                },
                "28": {
                    a: 0.96
                },
                "29": {
                    a: 0.84
                },
                "30": {
                    a: 0.61
                },
                "31": {
                    a: 0.31
                },
                "32": {
                    a: 0.07
                },
                "33": {
                    a: 0
                },
                "34": {
                    a: 0.04
                },
                "35": {
                    a: 0.16
                },
                "36": {
                    a: 0.39
                },
                "37": {
                    a: 0.69
                },
                "38": {
                    a: 0.93
                },
                "39": {
                    a: 1
                },
                "40": {
                    a: 0.96
                },
                "41": {
                    a: 0.84
                },
                "42": {
                    a: 0.61
                },
                "43": {
                    a: 0.31
                },
                "44": {
                    a: 0.07
                },
                "45": {
                    a: 0
                },
                "46": {
                    a: 0.04
                },
                "47": {
                    a: 0.16
                },
                "48": {
                    a: 0.39
                },
                "49": {
                    a: 0.69
                },
                "50": {
                    a: 0.93
                },
                "51": {
                    a: 1
                },
                "63": {
                    y: 576
                },
                "64": {
                    y: 579.1
                },
                "65": {
                    y: 584.8
                },
                "66": {
                    y: 593.55
                },
                "67": {
                    y: 605.4
                },
                "68": {
                    y: 619.55
                },
                "69": {
                    y: 633.6
                },
                "70": {
                    y: 644.65
                },
                "71": {
                    y: 651.1
                },
                "72": {
                    y: 653.1
                }
            });
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 86, loop: false });
        var instance2 = new Graphics()
            .drawCommands(shapes.lowBattery[0])
            .setRenderable(false);
        var instance3 = new Graphic7(MovieClip.SYNCHED);
        var instance1 = new Graphic4(MovieClip.SYNCHED);
        var instance6 = new Graphic3(MovieClip.SYNCHED);
        var instance5 = new Graphic2(MovieClip.SYNCHED);
        var instance4 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance2, 0, 73)
            .addTimedMask(instance3, {
                "0": instance2,
                "73": null
            })
            .addTimedChild(instance3, 0, 73, {
                "0": {
                    y: 0.45
                }
            })
            .addTimedChild(instance1, 0, 86, {
                "0": {
                    x: -9.2,
                    y: 215.15,
                    sx: 0.92,
                    sy: 0.92,
                    c: [
                        0,
                        0.7,
                        0,
                        0.75,
                        0,
                        0.76
                    ]
                },
                "1": {
                    y: 214,
                    c: [
                        0,
                        0.69,
                        0,
                        0.74,
                        0,
                        0.76
                    ]
                },
                "2": {
                    y: 210.1,
                    c: [
                        0.01,
                        0.69,
                        0.01,
                        0.74,
                        0.01,
                        0.75
                    ]
                },
                "3": {
                    y: 202.8,
                    c: [
                        0.03,
                        0.67,
                        0.03,
                        0.72,
                        0.03,
                        0.74
                    ]
                },
                "4": {
                    y: 190.95,
                    c: [
                        0.06,
                        0.65,
                        0.06,
                        0.7,
                        0.06,
                        0.71
                    ]
                },
                "5": {
                    y: 172.85,
                    c: [
                        0.11,
                        0.62,
                        0.11,
                        0.67,
                        0.11,
                        0.68
                    ]
                },
                "6": {
                    y: 145.4,
                    c: [
                        0.18,
                        0.58,
                        0.18,
                        0.62,
                        0.18,
                        0.63
                    ]
                },
                "7": {
                    y: 102.8,
                    c: [
                        0.28,
                        0.5,
                        0.28,
                        0.53,
                        0.28,
                        0.55
                    ]
                },
                "8": {
                    y: 33.65,
                    c: [
                        0.46,
                        0.38,
                        0.46,
                        0.4,
                        0.46,
                        0.41
                    ]
                },
                "9": {
                    y: -62.55,
                    c: [
                        0.7,
                        0.21,
                        0.7,
                        0.22,
                        0.7,
                        0.23
                    ]
                },
                "10": {
                    y: -131.35,
                    c: [
                        0.88,
                        0.09,
                        0.88,
                        0.09,
                        0.88,
                        0.1
                    ]
                },
                "11": {
                    y: -164.1,
                    c: [
                        0.96,
                        0.03,
                        0.96,
                        0.03,
                        0.96,
                        0.04
                    ]
                },
                "12": {
                    y: -178.05,
                    c: [
                        0.99,
                        0.01,
                        0.99,
                        0.01,
                        0.99,
                        0.01
                    ]
                },
                "13": {
                    y: -181.65,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "63": {
                    y: -176.7
                },
                "64": {
                    y: -160.55
                },
                "65": {
                    y: -130.85
                },
                "66": {
                    y: -85.4
                },
                "67": {
                    y: -23.65
                },
                "68": {
                    y: 49.95
                },
                "69": {
                    y: 123.05
                },
                "70": {
                    y: 180.5
                },
                "71": {
                    y: 214.15
                },
                "72": {
                    y: 224.55
                },
                "74": {
                    y: 208.55
                },
                "76": {
                    y: 176.05
                },
                "77": {
                    y: 125.9
                },
                "78": {
                    y: 105.95
                },
                "79": {
                    y: 77.4,
                    c: [
                        0.73,
                        0.18,
                        0.73,
                        0.2,
                        0.73,
                        0.2
                    ]
                },
                "80": {
                    y: 53.25,
                    c: [
                        0.51,
                        0.34,
                        0.51,
                        0.36,
                        0.51,
                        0.37
                    ]
                },
                "81": {
                    y: 33.5,
                    c: [
                        0.33,
                        0.47,
                        0.33,
                        0.5,
                        0.33,
                        0.51
                    ]
                },
                "82": {
                    y: 18.15,
                    c: [
                        0.18,
                        0.57,
                        0.18,
                        0.61,
                        0.18,
                        0.62
                    ]
                },
                "83": {
                    y: 7.2,
                    c: [
                        0.08,
                        0.64,
                        0.08,
                        0.68,
                        0.08,
                        0.7
                    ]
                },
                "84": {
                    y: 0.6,
                    c: [
                        0.02,
                        0.68,
                        0.02,
                        0.73,
                        0.02,
                        0.75
                    ]
                },
                "85": {
                    y: -1.6,
                    c: [
                        0,
                        0.7,
                        0,
                        0.75,
                        0,
                        0.76
                    ]
                }
            })
            .addTimedChild(instance6, 13, 73, {
                "13": {
                    x: -9.45,
                    y: -178.85,
                    sx: 0.92,
                    sy: 0.293,
                    r: 0,
                    a: 1,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "14": {
                    x: -9.461,
                    y: -184.772,
                    sy: 0.388,
                    r: -0.009
                },
                "15": {
                    x: -9.445,
                    y: -209.234,
                    sy: 0.774,
                    r: -0.053
                },
                "16": {
                    x: -9.45,
                    y: -218.35,
                    sy: 0.92,
                    r: -0.069
                },
                "18": {
                    x: -9.5,
                    y: -226.75,
                    r: -0.044
                },
                "19": {
                    y: -228.4,
                    r: 0.008
                },
                "21": {
                    y: -224.35,
                    r: 0.038
                },
                "22": {
                    y: -208.9,
                    r: 0.12
                },
                "23": {
                    x: -9.45,
                    y: -196.45,
                    sy: 0.895,
                    r: 0
                },
                "24": {
                    x: -9.457,
                    y: -196.724,
                    sx: 0.916,
                    sy: 0.905
                },
                "25": {
                    x: -9.463,
                    y: -198.001,
                    sx: 0.897,
                    sy: 0.948
                },
                "26": {
                    x: -9.449,
                    y: -199.696,
                    sx: 0.872,
                    sy: 1.006
                },
                "27": {
                    x: -9.45,
                    y: -200.1,
                    sx: 0.865,
                    sy: 1.02
                },
                "28": {
                    x: -9.418,
                    y: -198.81,
                    sx: 0.89,
                    sy: 0.976
                },
                "29": {
                    y: -197.895,
                    sx: 0.908,
                    sy: 0.943
                },
                "30": {
                    x: -9.45,
                    y: -197.2,
                    sx: 0.92,
                    sy: 0.92
                },
                "63": {
                    y: -192.25
                },
                "64": {
                    y: -176.1
                },
                "65": {
                    y: -146.4
                },
                "66": {
                    y: -100.95
                },
                "67": {
                    y: -39.2
                },
                "68": {
                    y: 34.4
                },
                "69": {
                    y: 107.5
                },
                "70": {
                    y: 164.95
                },
                "71": {
                    y: 198.6
                },
                "72": {
                    y: 209
                },
                "74": {
                    y: 194.6
                },
                "76": {
                    y: 164.5
                },
                "77": {
                    y: 114.35,
                    a: 0.79
                },
                "78": {
                    y: 84.05,
                    a: 0.41
                },
                "79": {
                    y: 55.5,
                    a: 0.57,
                    c: [
                        0.73,
                        0.18,
                        0.73,
                        0.2,
                        0.73,
                        0.2
                    ]
                },
                "80": {
                    y: 31.35,
                    a: 0.7,
                    c: [
                        0.51,
                        0.34,
                        0.51,
                        0.36,
                        0.51,
                        0.37
                    ]
                },
                "81": {
                    y: 11.6,
                    a: 0.81,
                    c: [
                        0.33,
                        0.47,
                        0.33,
                        0.5,
                        0.33,
                        0.51
                    ]
                },
                "82": {
                    y: -3.75,
                    a: 0.89,
                    c: [
                        0.18,
                        0.57,
                        0.18,
                        0.61,
                        0.18,
                        0.62
                    ]
                },
                "83": {
                    y: -14.7,
                    a: 0.95,
                    c: [
                        0.08,
                        0.64,
                        0.08,
                        0.68,
                        0.08,
                        0.7
                    ]
                },
                "84": {
                    y: -21.3,
                    a: 0.99,
                    c: [
                        0.02,
                        0.68,
                        0.02,
                        0.73,
                        0.02,
                        0.75
                    ]
                },
                "85": {
                    y: -23.5,
                    a: 1,
                    c: [
                        0,
                        0.7,
                        0,
                        0.75,
                        0,
                        0.76
                    ]
                }
            })
            .addTimedChild(instance5, 10, 57, {
                "10": {
                    x: -118.85,
                    y: 38.25,
                    a: 0
                },
                "11": {
                    a: 0.11
                },
                "12": {
                    a: 0.22
                },
                "13": {
                    a: 0.33
                },
                "14": {
                    a: 0.45
                },
                "15": {
                    a: 0.55
                },
                "16": {
                    a: 0.67
                },
                "17": {
                    a: 0.78
                },
                "18": {
                    a: 0.89
                },
                "19": {
                    a: 1
                },
                "60": {
                    a: 0.86
                },
                "61": {
                    a: 0.71
                },
                "62": {
                    a: 0.57
                },
                "63": {
                    a: 0.43
                },
                "64": {
                    a: 0.29
                },
                "65": {
                    a: 0.14
                },
                "66": {
                    a: 0
                }
            })
            .addTimedChild(instance4, 10, 57, {
                "10": {
                    x: -118.85,
                    y: 38.25,
                    a: 0
                },
                "11": {
                    a: 0.11
                },
                "12": {
                    a: 0.22
                },
                "13": {
                    a: 0.33
                },
                "14": {
                    a: 0.45
                },
                "15": {
                    a: 0.55
                },
                "16": {
                    a: 0.67
                },
                "17": {
                    a: 0.78
                },
                "18": {
                    a: 0.89
                },
                "19": {
                    a: 1
                },
                "60": {
                    a: 0.86
                },
                "61": {
                    a: 0.71
                },
                "62": {
                    a: 0.57
                },
                "63": {
                    a: 0.43
                },
                "64": {
                    a: 0.29
                },
                "65": {
                    a: 0.14
                },
                "66": {
                    a: 0
                }
            });
    });

    lib.lowBattery = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 87,
            framerate: 30,
            loop: false,
            labels: {
                HOLD_SAFE: 31
            }
        });
        var instance1 = new Graphic8(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 86, {
            "0": {
                x: 640,
                y: 360
            }
        });
    });

    lib.lowBattery.assets = {
        "lowBattery": "images/lowBattery.shapes.json",
        "lowBattery_atlas_1": "images/lowBattery_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.lowBattery,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 87,
        library: lib
    };
}
