// jibo-anim-db-animations/timelines/drawing_mondrian.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 6, loop: false });
        var instance1 = new Sprite(fromFrame("EYE1"));
        var instance2 = new Sprite(fromFrame("EYE2"));
        var instance3 = new Sprite(fromFrame("EYE3"));
        var instance4 = new Sprite(fromFrame("EYE4"));
        var instance5 = new Sprite(fromFrame("EYE5"));
        var instance6 = new Sprite(fromFrame("EYE6"))
            .setTransform(-257.3, -226.65);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -278.8,
                    y: -277.7
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -276.2,
                    y: -271.55
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -272.8,
                    y: -263.4
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -268.5,
                    y: -253.2
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -263.3,
                    y: -240.95
                }
            })
            .addTimedChild(instance6, 5, 1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 196, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-bottom1"))
            .setTransform(-195.85, -14);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 196, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-top1"))
            .setTransform(-177.2, -13.75);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 190, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-shadow1"))
            .setTransform(-176.3, -7.5);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 95, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 84);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 89, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 78);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 100, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 89);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 110, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 99);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 124, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 113);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 133, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 122);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 143, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 132);
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 153, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 142);
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 163, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 152);
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 180, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-line2"));
        var instance2 = new Sprite(fromFrame("modrian-line3"));
        var instance3 = new Sprite(fromFrame("modrian-line4"));
        var instance4 = new Sprite(fromFrame("modrian-line5"));
        var instance5 = new Sprite(fromFrame("modrian-line6"));
        var instance6 = new Sprite(fromFrame("modrian-line7"));
        var instance7 = new Sprite(fromFrame("modrian-line8"));
        var instance8 = new Sprite(fromFrame("modrian-line9"));
        var instance9 = new Sprite(fromFrame("modrian-line10"));
        var instance10 = new Sprite(fromFrame("modrian-line11"));
        var instance11 = new Sprite(fromFrame("modrian-line12"));
        var instance12 = new Sprite(fromFrame("modrian-line3"))
            .setTransform(-3.85, -110, 1.11, 1.11);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -3.85,
                    y: -110.05,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -3.85,
                    y: -109.95,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance11, 11, 165, {
                "11": {
                    x: -3.85,
                    y: -110,
                    sx: 1.11,
                    sy: 1.11
                }
            })
            .addTimedChild(instance12, 176, 4);
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 93, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-red1"));
        var instance2 = new Sprite(fromFrame("modrian-red3"));
        var instance3 = new Sprite(fromFrame("modrian-red5"));
        var instance4 = new Sprite(fromFrame("modrian-red7"));
        var instance5 = new Sprite(fromFrame("modrian-red9"));
        var instance6 = new Sprite(fromFrame("modrian-red11"));
        var instance7 = new Sprite(fromFrame("modrian-red13"));
        var instance8 = new Sprite(fromFrame("modrian-red15"));
        var instance9 = new Sprite(fromFrame("modrian-red17"));
        var instance10 = new Sprite(fromFrame("modrian-red19"));
        var instance11 = new Sprite(fromFrame("modrian-red21"));
        var instance12 = new Sprite(fromFrame("modrian-red23"));
        var instance13 = new Sprite(fromFrame("modrian-red25"));
        var instance14 = new Sprite(fromFrame("modrian-red27"));
        var instance15 = new Sprite(fromFrame("modrian-red29"));
        var instance16 = new Sprite(fromFrame("modrian-red31"));
        var instance17 = new Sprite(fromFrame("modrian-red33"));
        var instance18 = new Sprite(fromFrame("modrian-red35"));
        var instance19 = new Sprite(fromFrame("modrian-red37"))
            .setTransform(-69.5, -46.05);
        this.addTimedChild(instance1, 0, 2, {
                "0": {
                    x: -67.2,
                    y: -46.05
                }
            })
            .addTimedChild(instance2, 2, 2, {
                "2": {
                    x: -67.2,
                    y: -46.05
                }
            })
            .addTimedChild(instance3, 4, 2, {
                "4": {
                    x: -67.2,
                    y: -46.05
                }
            })
            .addTimedChild(instance4, 6, 2, {
                "6": {
                    x: -67.2,
                    y: -46.05
                }
            })
            .addTimedChild(instance5, 8, 2, {
                "8": {
                    x: -67.2,
                    y: -46.05
                }
            })
            .addTimedChild(instance6, 10, 2, {
                "10": {
                    x: -67.2,
                    y: -46.05
                }
            })
            .addTimedChild(instance7, 12, 2, {
                "12": {
                    x: -67.2,
                    y: -46.05
                }
            })
            .addTimedChild(instance8, 14, 2, {
                "14": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance9, 16, 2, {
                "16": {
                    x: -69.35,
                    y: -46.05
                }
            })
            .addTimedChild(instance10, 18, 2, {
                "18": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance11, 20, 2, {
                "20": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance12, 22, 2, {
                "22": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance13, 24, 2, {
                "24": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance14, 26, 2, {
                "26": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance15, 28, 2, {
                "28": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance16, 30, 2, {
                "30": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance17, 32, 2, {
                "32": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance18, 34, 2, {
                "34": {
                    x: -69.5,
                    y: -46.05
                }
            })
            .addTimedChild(instance19, 36, 57);
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 79, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-blue1"));
        var instance2 = new Sprite(fromFrame("modrian-blue3"));
        var instance3 = new Sprite(fromFrame("modrian-blue5"));
        var instance4 = new Sprite(fromFrame("modrian-blue7"));
        var instance5 = new Sprite(fromFrame("modrian-blue9"));
        var instance6 = new Sprite(fromFrame("modrian-blue11"));
        var instance7 = new Sprite(fromFrame("modrian-blue13"));
        var instance8 = new Sprite(fromFrame("modrian-blue15"))
            .setTransform(-27.15, -9.9);
        this.addTimedChild(instance1, 0, 2, {
                "0": {
                    x: -27.15,
                    y: -9.4
                }
            })
            .addTimedChild(instance2, 2, 2, {
                "2": {
                    x: -27.15,
                    y: -9.4
                }
            })
            .addTimedChild(instance3, 4, 2, {
                "4": {
                    x: -27.15,
                    y: -9.4
                }
            })
            .addTimedChild(instance4, 6, 2, {
                "6": {
                    x: -27.15,
                    y: -9.4
                }
            })
            .addTimedChild(instance5, 8, 2, {
                "8": {
                    x: -27.15,
                    y: -9.9
                }
            })
            .addTimedChild(instance6, 10, 2, {
                "10": {
                    x: -27.15,
                    y: -9.9
                }
            })
            .addTimedChild(instance7, 12, 2, {
                "12": {
                    x: -27.15,
                    y: -9.9
                }
            })
            .addTimedChild(instance8, 14, 65);
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 75, loop: false });
        var instance1 = new Sprite(fromFrame("modrian-yellow1"));
        var instance2 = new Sprite(fromFrame("modrian-yellow3"));
        var instance3 = new Sprite(fromFrame("modrian-yellow5"));
        var instance4 = new Sprite(fromFrame("modrian-yellow7"));
        var instance5 = new Sprite(fromFrame("modrian-yellow9"));
        var instance6 = new Sprite(fromFrame("modrian-yellow11"));
        var instance7 = new Sprite(fromFrame("modrian-yellow13"));
        var instance8 = new Sprite(fromFrame("modrian-yellow15"));
        var instance9 = new Sprite(fromFrame("modrian-yellow17"));
        var instance10 = new Sprite(fromFrame("modrian-yellow19"));
        var instance11 = new Sprite(fromFrame("modrian-yellow21"));
        var instance12 = new Sprite(fromFrame("modrian-yellow23"));
        var instance13 = new Sprite(fromFrame("modrian-yellow25"));
        var instance14 = new Sprite(fromFrame("modrian-yellow27"));
        var instance15 = new Sprite(fromFrame("modrian-yellow29"));
        var instance16 = new Sprite(fromFrame("modrian-yellow31"));
        var instance17 = new Sprite(fromFrame("modrian-yellow33"));
        var instance18 = new Sprite(fromFrame("modrian-yellow35"));
        var instance19 = new Sprite(fromFrame("modrian-yellow37"))
            .setTransform(-21.1, -65.5);
        this.addTimedChild(instance1, 0, 2, {
                "0": {
                    x: -8.5,
                    y: 57.45
                }
            })
            .addTimedChild(instance2, 2, 2, {
                "2": {
                    x: -16.8,
                    y: 50.5
                }
            })
            .addTimedChild(instance3, 4, 2, {
                "4": {
                    x: -16.95,
                    y: 47.4
                }
            })
            .addTimedChild(instance4, 6, 2, {
                "6": {
                    x: -16.95,
                    y: 44.05
                }
            })
            .addTimedChild(instance5, 8, 2, {
                "8": {
                    x: -17,
                    y: 38.25
                }
            })
            .addTimedChild(instance6, 10, 2, {
                "10": {
                    x: -17,
                    y: 33.4
                }
            })
            .addTimedChild(instance7, 12, 2, {
                "12": {
                    x: -17,
                    y: 26.85
                }
            })
            .addTimedChild(instance8, 14, 2, {
                "14": {
                    x: -17,
                    y: 20.6
                }
            })
            .addTimedChild(instance9, 16, 2, {
                "16": {
                    x: -17,
                    y: 12.3
                }
            })
            .addTimedChild(instance10, 18, 2, {
                "18": {
                    x: -17,
                    y: -0.8
                }
            })
            .addTimedChild(instance11, 20, 2, {
                "20": {
                    x: -17,
                    y: -9.5
                }
            })
            .addTimedChild(instance12, 22, 2, {
                "22": {
                    x: -17,
                    y: -18.25
                }
            })
            .addTimedChild(instance13, 24, 2, {
                "24": {
                    x: -17,
                    y: -30.2
                }
            })
            .addTimedChild(instance14, 26, 2, {
                "26": {
                    x: -17,
                    y: -39.75
                }
            })
            .addTimedChild(instance15, 28, 2, {
                "28": {
                    x: -17,
                    y: -44.65
                }
            })
            .addTimedChild(instance16, 30, 2, {
                "30": {
                    x: -18.3,
                    y: -47.55
                }
            })
            .addTimedChild(instance17, 32, 2, {
                "32": {
                    x: -18.6,
                    y: -62.6
                }
            })
            .addTimedChild(instance18, 34, 2, {
                "34": {
                    x: -18.6,
                    y: -65.5
                }
            })
            .addTimedChild(instance19, 36, 39);
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 180, loop: false });
        var instance13 = new Graphic17(MovieClip.SYNCHED);
        var instance12 = new Graphic16(MovieClip.SYNCHED);
        var instance10 = new Graphic15(MovieClip.SYNCHED);
        var instance1 = new Graphic14(MovieClip.SYNCHED);
        var instance2 = new Graphic13(MovieClip.SYNCHED);
        var instance3 = new Graphic12(MovieClip.SYNCHED);
        var instance4 = new Graphic11(MovieClip.SYNCHED);
        var instance5 = new Graphic10(MovieClip.SYNCHED);
        var instance6 = new Graphic9(MovieClip.SYNCHED);
        var instance7 = new Graphic8(MovieClip.SYNCHED);
        var instance8 = new Graphic7(MovieClip.SYNCHED);
        var instance11 = new Graphic6(MovieClip.SYNCHED);
        var instance9 = new Graphic5(MovieClip.SYNCHED);
        this.addTimedChild(instance13, 101, 75, {
                "101": {
                    x: -110.75,
                    y: 32
                }
            })
            .addTimedChild(instance12, 97, 79, {
                "97": {
                    x: 81.9,
                    y: 66.4
                }
            })
            .addTimedChild(instance10, 83, 93, {
                "83": {
                    x: -29.95,
                    y: 11.65
                }
            })
            .addTimedChild(instance1, 0, 180, {
                "0": {
                    x: -127.65,
                    y: 13.95
                }
            })
            .addTimedChild(instance2, 13, 163, {
                "13": {
                    x: -96.1,
                    y: 23.95
                }
            })
            .addTimedChild(instance3, 23, 153, {
                "23": {
                    x: 38.5,
                    y: 0.9
                }
            })
            .addTimedChild(instance4, 33, 143, {
                "33": {
                    x: 104.6,
                    y: 0.1
                }
            })
            .addTimedChild(instance5, 43, 133, {
                "43": {
                    x: 3.8,
                    y: 57.4,
                    sx: 0.901,
                    sy: 0.926,
                    r: -1.571
                }
            })
            .addTimedChild(instance6, 52, 124, {
                "52": {
                    x: -50,
                    y: -33.35,
                    sx: 0.901,
                    sy: 0.813,
                    r: -1.571
                }
            })
            .addTimedChild(instance7, 66, 110, {
                "66": {
                    x: 3.15,
                    y: 77.4,
                    sx: 0.901,
                    sy: 0.908,
                    r: -1.571
                }
            })
            .addTimedChild(instance8, 76, 100, {
                "76": {
                    x: -113.45,
                    y: 95.85,
                    sx: 0.901,
                    sy: 0.166,
                    r: -1.571
                }
            })
            .addTimedChild(instance11, 87, 89, {
                "87": {
                    x: 88.8,
                    y: -86.4,
                    sx: 0.901,
                    sy: 0.468,
                    r: -1.571
                }
            })
            .addTimedChild(instance9, 81, 95, {
                "81": {
                    x: 58.45,
                    y: 67.4,
                    sy: 0.093
                }
            });
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 201, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-blank1"))
            .setTransform(-232.5, -161.65);
        this.addTimedChild(instance1);
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 178, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-support1"))
            .setTransform(-15.9, -28.8);
        this.addTimedChild(instance1);
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 178, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-leg1"))
            .setTransform(-41.9, -74.65);
        this.addTimedChild(instance1);
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 178, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-leg1"))
            .setTransform(-41.9, -74.65);
        this.addTimedChild(instance1);
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 201, loop: false });
        var instance6 = new Graphic22(MovieClip.SYNCHED);
        var instance5 = new Graphic21(MovieClip.SYNCHED);
        var instance4 = new Graphic20(MovieClip.SYNCHED);
        var instance1 = new Graphic19(MovieClip.SYNCHED);
        var instance8 = new Graphic18(MovieClip.SYNCHED);
        var instance7 = new Graphic4(MovieClip.SYNCHED);
        var instance3 = new Graphic3(MovieClip.SYNCHED);
        var instance2 = new Graphic2(MovieClip.SYNCHED);
        this.addTimedChild(instance6, 8, 178, {
                "8": {
                    x: 66,
                    y: 61.45,
                    ky: 3.142
                },
                "9": {
                    x: 69.15,
                    y: 71.15
                },
                "10": {
                    x: 79.3,
                    y: 102.55
                },
                "11": {
                    x: 94.65,
                    y: 149.95
                },
                "12": {
                    x: 107.55,
                    y: 189.75
                },
                "13": {
                    x: 112,
                    y: 203.45
                },
                "14": {
                    x: 111.85,
                    y: 202.95
                },
                "15": {
                    x: 111.4,
                    y: 201.4
                },
                "16": {
                    x: 110.75,
                    y: 199.1
                },
                "17": {
                    x: 110.2,
                    y: 197.1
                },
                "18": {
                    x: 110,
                    y: 196.45
                },
                "175": {
                    x: 110.1,
                    y: 196.8
                },
                "176": {
                    x: 110.45,
                    y: 198
                },
                "177": {
                    x: 111,
                    y: 200
                },
                "178": {
                    x: 111.55,
                    y: 201.95
                },
                "179": {
                    x: 111.9,
                    y: 203.1
                },
                "180": {
                    x: 112,
                    y: 203.45
                },
                "181": {
                    x: 108.55,
                    y: 192.8
                },
                "182": {
                    x: 97,
                    y: 157.1
                },
                "183": {
                    x: 80.65,
                    y: 106.75
                },
                "184": {
                    x: 69.35,
                    y: 71.8
                },
                "185": {
                    x: 66,
                    y: 61.45
                }
            })
            .addTimedChild(instance5, 8, 178, {
                "8": {
                    x: -62.9,
                    y: 62.45
                },
                "9": {
                    x: -66.1,
                    y: 72.15
                },
                "10": {
                    x: -76.5,
                    y: 103.55
                },
                "11": {
                    x: -92.2,
                    y: 150.95
                },
                "12": {
                    x: -105.35,
                    y: 190.75
                },
                "13": {
                    x: -109.9,
                    y: 204.45
                },
                "14": {
                    x: -109.7,
                    y: 203.9
                },
                "15": {
                    x: -109.05,
                    y: 202.15
                },
                "16": {
                    x: -108.05,
                    y: 199.45
                },
                "17": {
                    x: -107.2,
                    y: 197.2
                },
                "18": {
                    x: -106.9,
                    y: 196.45
                },
                "175": {
                    x: -107.05,
                    y: 196.85
                },
                "176": {
                    x: -107.55,
                    y: 198.25
                },
                "177": {
                    x: -108.4,
                    y: 200.5
                },
                "178": {
                    x: -109.25,
                    y: 202.7
                },
                "179": {
                    x: -109.75,
                    y: 204.05
                },
                "180": {
                    x: -109.9,
                    y: 204.45
                },
                "181": {
                    x: -106.4,
                    y: 193.8
                },
                "182": {
                    x: -94.55,
                    y: 158.1
                },
                "183": {
                    x: -77.9,
                    y: 107.75
                },
                "184": {
                    x: -66.3,
                    y: 72.8
                },
                "185": {
                    x: -62.9,
                    y: 62.45
                }
            })
            .addTimedChild(instance4, 8, 178, {
                "8": {
                    y: -185.35
                },
                "9": {
                    y: -189.65
                },
                "10": {
                    y: -203.6
                },
                "11": {
                    y: -224.6
                },
                "12": {
                    y: -242.3
                },
                "13": {
                    y: -248.35
                },
                "14": {
                    y: -247.95
                },
                "15": {
                    y: -246.6
                },
                "16": {
                    y: -244.6
                },
                "17": {
                    y: -242.95
                },
                "18": {
                    y: -242.35
                },
                "175": {
                    y: -242.85
                },
                "176": {
                    y: -244.6
                },
                "177": {
                    y: -247.4
                },
                "178": {
                    y: -250.2
                },
                "179": {
                    y: -251.85
                },
                "180": {
                    y: -252.35
                },
                "181": {
                    y: -247.35
                },
                "182": {
                    y: -230.5
                },
                "183": {
                    y: -206.7
                },
                "184": {
                    y: -190.25
                },
                "185": {
                    y: -185.35
                }
            })
            .addTimedChild(instance1, 0, 201, {
                "0": {
                    x: 0,
                    y: -43.9,
                    r: 0.021,
                    c: [
                        0,
                        1,
                        0,
                        1,
                        0,
                        1
                    ]
                },
                "1": {
                    x: 0.005,
                    y: -43.897,
                    r: 0.014,
                    c: [
                        0.3,
                        0.69,
                        0.3,
                        0.69,
                        0.3,
                        0.69
                    ]
                },
                "2": {
                    x: 0.003,
                    y: -43.896,
                    r: 0.009,
                    c: [
                        0.55,
                        0.44,
                        0.55,
                        0.44,
                        0.55,
                        0.44
                    ]
                },
                "3": {
                    x: 0.013,
                    y: -43.893,
                    r: 0.005,
                    c: [
                        0.75,
                        0.25,
                        0.75,
                        0.25,
                        0.75,
                        0.25
                    ]
                },
                "4": {
                    x: -0.013,
                    y: -43.891,
                    r: 0.001,
                    c: [
                        0.89,
                        0.11,
                        0.89,
                        0.11,
                        0.89,
                        0.11
                    ]
                },
                "5": {
                    x: 0.018,
                    r: 0,
                    c: [
                        0.97,
                        0.03,
                        0.97,
                        0.03,
                        0.97,
                        0.03
                    ]
                },
                "6": {
                    x: 0,
                    y: -43.9,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "200": {
                    x: 0.013,
                    y: -43.893,
                    r: 0.005,
                    c: [
                        0.75,
                        0.25,
                        0.75,
                        0.25,
                        0.75,
                        0.25
                    ]
                }
            })
            .addTimedChild(instance8, 20, 180, {
                "20": {
                    x: 10.55,
                    y: -56.9,
                    a: 1
                },
                "189": {
                    a: 0.88
                },
                "190": {
                    a: 0.75
                },
                "191": {
                    a: 0.63
                },
                "192": {
                    a: 0.5
                },
                "193": {
                    a: 0.38
                },
                "194": {
                    a: 0.25
                },
                "195": {
                    a: 0.13
                },
                "196": {
                    a: 0
                }
            })
            .addTimedChild(instance7, 10, 190, {
                "10": {
                    y: -196.15,
                    a: 0
                },
                "11": {
                    a: 0.11
                },
                "12": {
                    a: 0.22
                },
                "13": {
                    a: 0.33
                },
                "14": {
                    a: 0.45
                },
                "15": {
                    a: 0.55
                },
                "16": {
                    a: 0.67
                },
                "17": {
                    a: 0.78
                },
                "18": {
                    a: 0.89
                },
                "19": {
                    a: 1
                },
                "175": {
                    a: 0.9
                },
                "176": {
                    a: 0.8
                },
                "177": {
                    a: 0.7
                },
                "178": {
                    a: 0.6
                },
                "179": {
                    a: 0.5
                },
                "180": {
                    a: 0.4
                },
                "181": {
                    a: 0.3
                },
                "182": {
                    a: 0.2
                },
                "183": {
                    a: 0.1
                },
                "184": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 4, 196, {
                "4": {
                    y: -190.45,
                    a: 0
                },
                "5": {
                    y: -191.7,
                    a: 0.05
                },
                "6": {
                    y: -195.8,
                    a: 0.2
                },
                "7": {
                    y: -202.6,
                    a: 0.45
                },
                "8": {
                    y: -210.2,
                    a: 0.73
                },
                "9": {
                    y: -215.65,
                    a: 0.93
                },
                "10": {
                    y: -217.45,
                    a: 1
                },
                "11": {
                    y: -217.15
                },
                "12": {
                    y: -216.25
                },
                "13": {
                    y: -214.75
                },
                "14": {
                    y: -213.05
                },
                "15": {
                    y: -211.85
                },
                "16": {
                    y: -211.45
                },
                "185": {
                    y: -211.65,
                    a: 0.98
                },
                "186": {
                    y: -212.2,
                    a: 0.93
                },
                "187": {
                    y: -213.25,
                    a: 0.82
                },
                "188": {
                    y: -214.7,
                    a: 0.67
                },
                "189": {
                    y: -216.5,
                    a: 0.5
                },
                "190": {
                    y: -218.25,
                    a: 0.32
                },
                "191": {
                    y: -219.7,
                    a: 0.17
                },
                "192": {
                    y: -220.7,
                    a: 0.07
                },
                "193": {
                    y: -221.3,
                    a: 0.02
                },
                "194": {
                    y: -221.45,
                    a: 0
                }
            })
            .addTimedChild(instance2, 4, 196, {
                "4": {
                    y: 102.45,
                    a: 0
                },
                "5": {
                    y: 103.85,
                    a: 0.05
                },
                "6": {
                    y: 108.4,
                    a: 0.2
                },
                "7": {
                    y: 115.95,
                    a: 0.45
                },
                "8": {
                    y: 124.4,
                    a: 0.73
                },
                "9": {
                    y: 130.4,
                    a: 0.93
                },
                "10": {
                    y: 132.4,
                    a: 1
                },
                "11": {
                    y: 132.2
                },
                "12": {
                    y: 131.6
                },
                "13": {
                    y: 130.6
                },
                "14": {
                    y: 129.45
                },
                "15": {
                    y: 128.65
                },
                "16": {
                    y: 128.4
                },
                "185": {
                    y: 128.6,
                    a: 0.98
                },
                "186": {
                    y: 129.15,
                    a: 0.93
                },
                "187": {
                    y: 130.2,
                    a: 0.82
                },
                "188": {
                    y: 131.65,
                    a: 0.67
                },
                "189": {
                    y: 133.45,
                    a: 0.5
                },
                "190": {
                    y: 135.2,
                    a: 0.32
                },
                "191": {
                    y: 136.65,
                    a: 0.17
                },
                "192": {
                    y: 137.65,
                    a: 0.07
                },
                "193": {
                    y: 138.25,
                    a: 0.02
                },
                "194": {
                    y: 138.4,
                    a: 0
                }
            });
    });

    var Graphic24 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 7, loop: false });
        var instance1 = new Sprite(fromFrame("EYE159"));
        var instance2 = new Sprite(fromFrame("EYE160"));
        var instance3 = new Sprite(fromFrame("EYE161"));
        var instance4 = new Sprite(fromFrame("EYE162"));
        var instance5 = new Sprite(fromFrame("EYE163"));
        var instance6 = new Sprite(fromFrame("EYE164"))
            .setTransform(-278.8, -277.7);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -264.1,
                    y: -236.5
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -270,
                    y: -251.3
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -275.85,
                    y: -268.85
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -277.35,
                    y: -273.25
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -278.85,
                    y: -277.75
                }
            })
            .addTimedChild(instance6, 5, 2);
    });

    lib.drawing_mondrian = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 213,
            framerate: 30,
            loop: false
        });
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        var instance3 = new Graphic24(MovieClip.SYNCHED)
            .setTransform(640.3, 359.3, 0.898, 0.898);
        var instance2 = new Graphic23(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 6, {
                "0": {
                    x: 640.3,
                    y: 359.3,
                    sx: 0.898,
                    sy: 0.898
                }
            })
            .addTimedChild(instance3, 206, 7)
            .addTimedChild(instance2, 6, 201, {
                "6": {
                    x: 640.45,
                    y: 359.95,
                    sx: 1,
                    sy: 1,
                    kx: 0,
                    ky: 0,
                    r: 0,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "196": {
                    x: 640.6,
                    y: 360,
                    r: -0.003
                },
                "197": {
                    x: 640.619,
                    y: 359.982,
                    r: -0.005
                },
                "198": {
                    x: 640.851,
                    y: 360.003,
                    r: -0.009
                },
                "199": {
                    x: 641.439,
                    y: 360.021,
                    r: -0.022
                },
                "200": {
                    x: 642.1,
                    y: 360.104,
                    r: -0.036
                },
                "201": {
                    x: 642.8,
                    y: 360.1,
                    r: -0.051
                },
                "202": {
                    x: 631.94,
                    y: 358.342,
                    sx: 0.999,
                    sy: 0.999,
                    r: 0.189,
                    c: [
                        0.96,
                        0.04,
                        0.96,
                        0.04,
                        0.96,
                        0.04
                    ]
                },
                "203": {
                    x: 602.636,
                    y: 333.341,
                    r: 1.05,
                    c: [
                        0.83,
                        0.17,
                        0.83,
                        0.17,
                        0.83,
                        0.17
                    ]
                },
                "204": {
                    x: 626.319,
                    y: 271.568,
                    kx: 3.632,
                    ky: 2.651,
                    r: 0,
                    c: [
                        0.58,
                        0.42,
                        0.58,
                        0.42,
                        0.58,
                        0.42
                    ]
                },
                "205": {
                    x: 690.582,
                    y: 316.559,
                    sx: 1,
                    sy: 1,
                    kx: 0,
                    ky: 0,
                    r: -1.602,
                    c: [
                        0.26,
                        0.74,
                        0.26,
                        0.74,
                        0.26,
                        0.74
                    ]
                },
                "206": {
                    x: 638.5,
                    y: 359.65,
                    r: 0.044,
                    c: [
                        0,
                        1,
                        0,
                        1,
                        0,
                        1
                    ]
                }
            });
    });

    lib.drawing_mondrian.assets = {
        "drawing_mondrian_atlas_1": "images/drawing_mondrian_atlas_1.json",
        "drawing_mondrian_atlas_2": "images/drawing_mondrian_atlas_2.json",
        "drawing_mondrian_atlas_3": "images/drawing_mondrian_atlas_3.json",
        "drawing_mondrian_atlas_4": "images/drawing_mondrian_atlas_4.json",
        "drawing_mondrian_atlas_5": "images/drawing_mondrian_atlas_5.json",
        "drawing_mondrian_atlas_6": "images/drawing_mondrian_atlas_6.json",
        "drawing_mondrian_atlas_7": "images/drawing_mondrian_atlas_7.json",
        "drawing_mondrian_atlas_8": "images/drawing_mondrian_atlas_8.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.drawing_mondrian,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 213,
        library: lib
    };
}
