// jibo-anim-db-animations/timelines/drawing_warhol.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 6, loop: false });
        var instance1 = new Sprite(fromFrame("EYE1"));
        var instance2 = new Sprite(fromFrame("EYE2"));
        var instance3 = new Sprite(fromFrame("EYE3"));
        var instance4 = new Sprite(fromFrame("EYE4"));
        var instance5 = new Sprite(fromFrame("EYE5"));
        var instance6 = new Sprite(fromFrame("EYE6"))
            .setTransform(-257.3, -226.65);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -278.8,
                    y: -277.7
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -276.2,
                    y: -271.55
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -272.8,
                    y: -263.4
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -268.5,
                    y: -253.2
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -263.3,
                    y: -240.95
                }
            })
            .addTimedChild(instance6, 5, 1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 230, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-bottom1"))
            .setTransform(-195.85, -14);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 230, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-top1"))
            .setTransform(-177.2, -13.75);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 224, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-shadow1"))
            .setTransform(-176.3, -7.5);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 71, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-eye1"));
        var instance2 = new Sprite(fromFrame("warhol-eye2"));
        var instance3 = new Sprite(fromFrame("warhol-eye3"));
        var instance4 = new Sprite(fromFrame("warhol-eye4"));
        var instance5 = new Sprite(fromFrame("warhol-eye5"));
        var instance6 = new Sprite(fromFrame("warhol-eye6"));
        var instance7 = new Sprite(fromFrame("warhol-eye7"));
        var instance8 = new Sprite(fromFrame("warhol-eye9"));
        var instance9 = new Sprite(fromFrame("warhol-eye10"));
        var instance10 = new Sprite(fromFrame("warhol-eye11"));
        var instance11 = new Sprite(fromFrame("warhol-eye12"));
        var instance12 = new Sprite(fromFrame("warhol-eye13"));
        var instance13 = new Sprite(fromFrame("warhol-eye14"));
        var instance14 = new Sprite(fromFrame("warhol-eye15"));
        var instance15 = new Sprite(fromFrame("warhol-eye16"));
        var instance16 = new Sprite(fromFrame("warhol-eye18"));
        var instance17 = new Sprite(fromFrame("warhol-eye20"));
        var instance18 = new Sprite(fromFrame("warhol-eye21"));
        var instance19 = new Sprite(fromFrame("warhol-eye22"));
        var instance20 = new Sprite(fromFrame("warhol-eye23"));
        var instance21 = new Sprite(fromFrame("warhol-eye24"));
        var instance22 = new Sprite(fromFrame("warhol-eye26"));
        var instance23 = new Sprite(fromFrame("warhol-eye28"));
        var instance24 = new Sprite(fromFrame("warhol-eye29"));
        var instance25 = new Sprite(fromFrame("warhol-eye30"));
        var instance26 = new Sprite(fromFrame("warhol-eye31"));
        var instance27 = new Sprite(fromFrame("warhol-eye32"));
        var instance28 = new Sprite(fromFrame("warhol-eye34"))
            .setTransform(-9.6, -9.6);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -8.6,
                    y: -7.2
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -8.6,
                    y: -8.3
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -8.6,
                    y: -9.55
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance6, 5, 1, {
                "5": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance7, 6, 2, {
                "6": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance15, 15, 2, {
                "15": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance16, 17, 2, {
                "17": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance17, 19, 1, {
                "19": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance18, 20, 1, {
                "20": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance19, 21, 1, {
                "21": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance20, 22, 1, {
                "22": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance21, 23, 2, {
                "23": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance22, 25, 2, {
                "25": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance23, 27, 1, {
                "27": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance24, 28, 1, {
                "28": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance25, 29, 1, {
                "29": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance26, 30, 1, {
                "30": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance27, 31, 2, {
                "31": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance28, 33, 38);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 146, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-head2"));
        var instance2 = new Sprite(fromFrame("warhol-head3"));
        var instance3 = new Sprite(fromFrame("warhol-head4"));
        var instance4 = new Sprite(fromFrame("warhol-head5"));
        var instance5 = new Sprite(fromFrame("warhol-head6"));
        var instance6 = new Sprite(fromFrame("warhol-head7"));
        var instance7 = new Sprite(fromFrame("warhol-head8"));
        var instance8 = new Sprite(fromFrame("warhol-head9"));
        var instance9 = new Sprite(fromFrame("warhol-head10"));
        var instance10 = new Sprite(fromFrame("warhol-head11"));
        var instance11 = new Sprite(fromFrame("warhol-head12"));
        var instance12 = new Sprite(fromFrame("warhol-head13"));
        var instance13 = new Sprite(fromFrame("warhol-head14"));
        var instance14 = new Sprite(fromFrame("warhol-head15"));
        var instance15 = new Sprite(fromFrame("warhol-head16"));
        var instance16 = new Sprite(fromFrame("warhol-head17"));
        var instance17 = new Sprite(fromFrame("warhol-head18"));
        var instance18 = new Sprite(fromFrame("warhol-head19"));
        var instance19 = new Sprite(fromFrame("warhol-head20"));
        var instance20 = new Sprite(fromFrame("warhol-head21"));
        var instance21 = new Sprite(fromFrame("warhol-head22"));
        var instance22 = new Sprite(fromFrame("warhol-head23"));
        var instance23 = new Sprite(fromFrame("warhol-head24"));
        var instance24 = new Sprite(fromFrame("warhol-head25"));
        var instance25 = new Sprite(fromFrame("warhol-head26"));
        var instance26 = new Sprite(fromFrame("warhol-head27"));
        var instance27 = new Sprite(fromFrame("warhol-head28"));
        var instance28 = new Sprite(fromFrame("warhol-head29"));
        var instance29 = new Sprite(fromFrame("warhol-head30"))
            .setTransform(-49.6, -49.6);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: 0.25,
                    y: -49.6
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance15, 15, 1, {
                "15": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance16, 16, 1, {
                "16": {
                    x: -14,
                    y: -49.6
                }
            })
            .addTimedChild(instance17, 17, 1, {
                "17": {
                    x: -33,
                    y: -49.6
                }
            })
            .addTimedChild(instance18, 18, 1, {
                "18": {
                    x: -44.65,
                    y: -49.6
                }
            })
            .addTimedChild(instance19, 19, 1, {
                "19": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance20, 20, 1, {
                "20": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance21, 21, 1, {
                "21": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance22, 22, 1, {
                "22": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance23, 23, 1, {
                "23": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance24, 24, 1, {
                "24": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance25, 25, 1, {
                "25": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance26, 26, 1, {
                "26": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance27, 27, 1, {
                "27": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance28, 28, 1, {
                "28": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance29, 29, 117);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 97, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-head-inside2"));
        var instance2 = new Sprite(fromFrame("warhol-head-inside4"));
        var instance3 = new Sprite(fromFrame("warhol-head-inside6"));
        var instance4 = new Sprite(fromFrame("warhol-head-inside8"));
        var instance5 = new Sprite(fromFrame("warhol-head-inside10"));
        var instance6 = new Sprite(fromFrame("warhol-head-inside12"));
        var instance7 = new Sprite(fromFrame("warhol-head-inside14"));
        var instance8 = new Sprite(fromFrame("warhol-head-inside16"));
        var instance9 = new Sprite(fromFrame("warhol-head-inside18"));
        var instance10 = new Sprite(fromFrame("warhol-head-inside20"));
        var instance11 = new Sprite(fromFrame("warhol-head-inside22"));
        var instance12 = new Sprite(fromFrame("warhol-head-inside24"));
        var instance13 = new Sprite(fromFrame("warhol-head-inside26"));
        var instance14 = new Sprite(fromFrame("warhol-head-inside28"));
        var instance15 = new Sprite(fromFrame("warhol-head-inside30"));
        var instance16 = new Sprite(fromFrame("warhol-head-inside32"));
        var instance17 = new Sprite(fromFrame("warhol-head-inside34"));
        var instance18 = new Sprite(fromFrame("warhol-head-inside36"));
        var instance19 = new Sprite(fromFrame("warhol-head-inside38"));
        var instance20 = new Sprite(fromFrame("warhol-head-inside40"));
        var instance21 = new Sprite(fromFrame("warhol-head-inside42"));
        var instance22 = new Sprite(fromFrame("warhol-head-inside44"));
        var instance23 = new Sprite(fromFrame("warhol-head-inside46"))
            .setTransform(-40.1, -40.1);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -33.9,
                    y: -40.05
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -38.2,
                    y: -40.1
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -39.85,
                    y: -40.1
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -40.05,
                    y: -40.1
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance11, 21, 2, {
                "21": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance12, 23, 2, {
                "23": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance13, 25, 2, {
                "25": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance14, 27, 2, {
                "27": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance15, 29, 2, {
                "29": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance16, 31, 2, {
                "31": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance17, 33, 2, {
                "33": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance18, 35, 2, {
                "35": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance19, 37, 2, {
                "37": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance20, 39, 2, {
                "39": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance21, 41, 2, {
                "41": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance22, 43, 2, {
                "43": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance23, 45, 52);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 88, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-shadow1"));
        var instance2 = new Sprite(fromFrame("warhol-shadow2"));
        var instance3 = new Sprite(fromFrame("warhol-shadow3"));
        var instance4 = new Sprite(fromFrame("warhol-shadow4"));
        var instance5 = new Sprite(fromFrame("warhol-shadow5"));
        var instance6 = new Sprite(fromFrame("warhol-shadow6"));
        var instance7 = new Sprite(fromFrame("warhol-shadow7"));
        var instance8 = new Sprite(fromFrame("warhol-shadow8"));
        var instance9 = new Sprite(fromFrame("warhol-shadow9"));
        var instance10 = new Sprite(fromFrame("warhol-shadow10"));
        var instance11 = new Sprite(fromFrame("warhol-shadow11"));
        var instance12 = new Sprite(fromFrame("warhol-shadow12"));
        var instance13 = new Sprite(fromFrame("warhol-shadow13"));
        var instance14 = new Sprite(fromFrame("warhol-shadow14"));
        var instance15 = new Sprite(fromFrame("warhol-shadow15"));
        var instance16 = new Sprite(fromFrame("warhol-shadow16"));
        var instance17 = new Sprite(fromFrame("warhol-shadow17"))
            .setTransform(-49.4, -30.3);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance6, 5, 1, {
                "5": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance7, 6, 1, {
                "6": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance8, 7, 1, {
                "7": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance9, 8, 1, {
                "8": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance10, 9, 1, {
                "9": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance11, 10, 1, {
                "10": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance12, 11, 1, {
                "11": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance13, 12, 1, {
                "12": {
                    x: -49.4,
                    y: -15.05
                }
            })
            .addTimedChild(instance14, 13, 1, {
                "13": {
                    x: -49.4,
                    y: -20.4
                }
            })
            .addTimedChild(instance15, 14, 1, {
                "14": {
                    x: -49.4,
                    y: -30.3
                }
            })
            .addTimedChild(instance16, 15, 1, {
                "15": {
                    x: -49.4,
                    y: -26.5
                }
            })
            .addTimedChild(instance17, 16, 72);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 125, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-body2"));
        var instance2 = new Sprite(fromFrame("warhol-body4"));
        var instance3 = new Sprite(fromFrame("warhol-body6"));
        var instance4 = new Sprite(fromFrame("warhol-body8"));
        var instance5 = new Sprite(fromFrame("warhol-body10"));
        var instance6 = new Sprite(fromFrame("warhol-body12"));
        var instance7 = new Sprite(fromFrame("warhol-body14"));
        var instance8 = new Sprite(fromFrame("warhol-body16"));
        var instance9 = new Sprite(fromFrame("warhol-body18"));
        var instance10 = new Sprite(fromFrame("warhol-body20"));
        var instance11 = new Sprite(fromFrame("warhol-body22"))
            .setTransform(-40.15, -4.1);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -34.85,
                    y: -4.1
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -35.95,
                    y: -4.1
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -33.45,
                    y: -4.1
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -37.4,
                    y: -4.1
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -38.25,
                    y: -4.1
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -38.55,
                    y: -4.1
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -38.65,
                    y: -4.1
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -40.15,
                    y: -4.1
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -39.45,
                    y: -4.1
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -40,
                    y: -4.1
                }
            })
            .addTimedChild(instance11, 21, 104);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 190, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-bg1"))
            .setTransform(-65.2, -65.2);
        this.addTimedChild(instance1);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 190, loop: false });
        var instance1 = new Graphic10(MovieClip.SYNCHED);
        var instance3 = new Graphic9(MovieClip.SYNCHED)
            .setTransform(1.05, 40.95)
            .setColorTransform(0, 1, 0, 1, 0, 1);
        var instance5 = new Graphic8(MovieClip.SYNCHED)
            .setTransform(2.05, 21.5)
            .setColorTransform(0, 0.38, 0, 0.06, 0, 0.3);
        var instance4 = new Graphic7(MovieClip.SYNCHED)
            .setTransform(1.1, -1.9)
            .setColorTransform(0, 0.38, 0, 0.06, 0, 0.3);
        var instance2 = new Graphic6(MovieClip.SYNCHED)
            .setTransform(1.1, -1.9)
            .setColorTransform(0, 1, 0, 1, 0, 1);
        var instance6 = new Graphic5(MovieClip.SYNCHED)
            .setTransform(0.75, -3.55)
            .setColorTransform(0, 1, 0, 1, 0, 1);
        this.addTimedChild(instance1, 0, 190, {
                "0": {
                    y: -0.3,
                    a: 0,
                    c: [
                        0,
                        0.2,
                        0,
                        0.37,
                        0,
                        0.64
                    ]
                },
                "1": {
                    a: 0.01
                },
                "2": {
                    a: 0.04
                },
                "3": {
                    a: 0.09
                },
                "4": {
                    a: 0.18
                },
                "5": {
                    a: 0.3
                },
                "6": {
                    a: 0.46
                },
                "7": {
                    a: 0.65
                },
                "8": {
                    a: 0.82
                },
                "9": {
                    a: 0.93
                },
                "10": {
                    a: 0.98
                },
                "11": {
                    a: 1
                }
            })
            .addTimedChild(instance3, 65, 125)
            .addTimedChild(instance5, 102, 88)
            .addTimedChild(instance4, 93, 97)
            .addTimedChild(instance2, 44, 146)
            .addTimedChild(instance6, 119, 71);
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 88, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-eye1"));
        var instance2 = new Sprite(fromFrame("warhol-eye2"));
        var instance3 = new Sprite(fromFrame("warhol-eye3"));
        var instance4 = new Sprite(fromFrame("warhol-eye4"));
        var instance5 = new Sprite(fromFrame("warhol-eye5"));
        var instance6 = new Sprite(fromFrame("warhol-eye6"));
        var instance7 = new Sprite(fromFrame("warhol-eye7"));
        var instance8 = new Sprite(fromFrame("warhol-eye9"));
        var instance9 = new Sprite(fromFrame("warhol-eye10"));
        var instance10 = new Sprite(fromFrame("warhol-eye11"));
        var instance11 = new Sprite(fromFrame("warhol-eye12"));
        var instance12 = new Sprite(fromFrame("warhol-eye13"));
        var instance13 = new Sprite(fromFrame("warhol-eye14"));
        var instance14 = new Sprite(fromFrame("warhol-eye15"));
        var instance15 = new Sprite(fromFrame("warhol-eye16"));
        var instance16 = new Sprite(fromFrame("warhol-eye18"));
        var instance17 = new Sprite(fromFrame("warhol-eye20"));
        var instance18 = new Sprite(fromFrame("warhol-eye21"));
        var instance19 = new Sprite(fromFrame("warhol-eye22"));
        var instance20 = new Sprite(fromFrame("warhol-eye23"));
        var instance21 = new Sprite(fromFrame("warhol-eye24"));
        var instance22 = new Sprite(fromFrame("warhol-eye26"));
        var instance23 = new Sprite(fromFrame("warhol-eye28"));
        var instance24 = new Sprite(fromFrame("warhol-eye29"));
        var instance25 = new Sprite(fromFrame("warhol-eye30"));
        var instance26 = new Sprite(fromFrame("warhol-eye31"));
        var instance27 = new Sprite(fromFrame("warhol-eye32"));
        var instance28 = new Sprite(fromFrame("warhol-eye34"))
            .setTransform(-9.6, -9.6);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -8.6,
                    y: -7.2
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -8.6,
                    y: -8.3
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -8.6,
                    y: -9.55
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance6, 5, 1, {
                "5": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance7, 6, 2, {
                "6": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance15, 15, 2, {
                "15": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance16, 17, 2, {
                "17": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance17, 19, 1, {
                "19": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance18, 20, 1, {
                "20": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance19, 21, 1, {
                "21": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance20, 22, 1, {
                "22": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance21, 23, 2, {
                "23": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance22, 25, 2, {
                "25": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance23, 27, 1, {
                "27": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance24, 28, 1, {
                "28": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance25, 29, 1, {
                "29": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance26, 30, 1, {
                "30": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance27, 31, 2, {
                "31": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance28, 33, 55);
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 163, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-head2"));
        var instance2 = new Sprite(fromFrame("warhol-head3"));
        var instance3 = new Sprite(fromFrame("warhol-head4"));
        var instance4 = new Sprite(fromFrame("warhol-head5"));
        var instance5 = new Sprite(fromFrame("warhol-head6"));
        var instance6 = new Sprite(fromFrame("warhol-head7"));
        var instance7 = new Sprite(fromFrame("warhol-head8"));
        var instance8 = new Sprite(fromFrame("warhol-head9"));
        var instance9 = new Sprite(fromFrame("warhol-head10"));
        var instance10 = new Sprite(fromFrame("warhol-head11"));
        var instance11 = new Sprite(fromFrame("warhol-head12"));
        var instance12 = new Sprite(fromFrame("warhol-head13"));
        var instance13 = new Sprite(fromFrame("warhol-head14"));
        var instance14 = new Sprite(fromFrame("warhol-head15"));
        var instance15 = new Sprite(fromFrame("warhol-head16"));
        var instance16 = new Sprite(fromFrame("warhol-head17"));
        var instance17 = new Sprite(fromFrame("warhol-head18"));
        var instance18 = new Sprite(fromFrame("warhol-head19"));
        var instance19 = new Sprite(fromFrame("warhol-head20"));
        var instance20 = new Sprite(fromFrame("warhol-head21"));
        var instance21 = new Sprite(fromFrame("warhol-head22"));
        var instance22 = new Sprite(fromFrame("warhol-head23"));
        var instance23 = new Sprite(fromFrame("warhol-head24"));
        var instance24 = new Sprite(fromFrame("warhol-head25"));
        var instance25 = new Sprite(fromFrame("warhol-head26"));
        var instance26 = new Sprite(fromFrame("warhol-head27"));
        var instance27 = new Sprite(fromFrame("warhol-head28"));
        var instance28 = new Sprite(fromFrame("warhol-head29"));
        var instance29 = new Sprite(fromFrame("warhol-head30"))
            .setTransform(-49.6, -49.6);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: 0.25,
                    y: -49.6
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance15, 15, 1, {
                "15": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance16, 16, 1, {
                "16": {
                    x: -14,
                    y: -49.6
                }
            })
            .addTimedChild(instance17, 17, 1, {
                "17": {
                    x: -33,
                    y: -49.6
                }
            })
            .addTimedChild(instance18, 18, 1, {
                "18": {
                    x: -44.65,
                    y: -49.6
                }
            })
            .addTimedChild(instance19, 19, 1, {
                "19": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance20, 20, 1, {
                "20": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance21, 21, 1, {
                "21": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance22, 22, 1, {
                "22": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance23, 23, 1, {
                "23": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance24, 24, 1, {
                "24": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance25, 25, 1, {
                "25": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance26, 26, 1, {
                "26": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance27, 27, 1, {
                "27": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance28, 28, 1, {
                "28": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance29, 29, 134);
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 114, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-head-inside2"));
        var instance2 = new Sprite(fromFrame("warhol-head-inside4"));
        var instance3 = new Sprite(fromFrame("warhol-head-inside6"));
        var instance4 = new Sprite(fromFrame("warhol-head-inside8"));
        var instance5 = new Sprite(fromFrame("warhol-head-inside10"));
        var instance6 = new Sprite(fromFrame("warhol-head-inside12"));
        var instance7 = new Sprite(fromFrame("warhol-head-inside14"));
        var instance8 = new Sprite(fromFrame("warhol-head-inside16"));
        var instance9 = new Sprite(fromFrame("warhol-head-inside18"));
        var instance10 = new Sprite(fromFrame("warhol-head-inside20"));
        var instance11 = new Sprite(fromFrame("warhol-head-inside22"));
        var instance12 = new Sprite(fromFrame("warhol-head-inside24"));
        var instance13 = new Sprite(fromFrame("warhol-head-inside26"));
        var instance14 = new Sprite(fromFrame("warhol-head-inside28"));
        var instance15 = new Sprite(fromFrame("warhol-head-inside30"));
        var instance16 = new Sprite(fromFrame("warhol-head-inside32"));
        var instance17 = new Sprite(fromFrame("warhol-head-inside34"));
        var instance18 = new Sprite(fromFrame("warhol-head-inside36"));
        var instance19 = new Sprite(fromFrame("warhol-head-inside38"));
        var instance20 = new Sprite(fromFrame("warhol-head-inside40"));
        var instance21 = new Sprite(fromFrame("warhol-head-inside42"));
        var instance22 = new Sprite(fromFrame("warhol-head-inside44"));
        var instance23 = new Sprite(fromFrame("warhol-head-inside46"))
            .setTransform(-40.1, -40.1);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -33.9,
                    y: -40.05
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -38.2,
                    y: -40.1
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -39.85,
                    y: -40.1
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -40.05,
                    y: -40.1
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance11, 21, 2, {
                "21": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance12, 23, 2, {
                "23": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance13, 25, 2, {
                "25": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance14, 27, 2, {
                "27": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance15, 29, 2, {
                "29": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance16, 31, 2, {
                "31": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance17, 33, 2, {
                "33": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance18, 35, 2, {
                "35": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance19, 37, 2, {
                "37": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance20, 39, 2, {
                "39": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance21, 41, 2, {
                "41": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance22, 43, 2, {
                "43": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance23, 45, 69);
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 105, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-shadow1"));
        var instance2 = new Sprite(fromFrame("warhol-shadow2"));
        var instance3 = new Sprite(fromFrame("warhol-shadow3"));
        var instance4 = new Sprite(fromFrame("warhol-shadow4"));
        var instance5 = new Sprite(fromFrame("warhol-shadow5"));
        var instance6 = new Sprite(fromFrame("warhol-shadow6"));
        var instance7 = new Sprite(fromFrame("warhol-shadow7"));
        var instance8 = new Sprite(fromFrame("warhol-shadow8"));
        var instance9 = new Sprite(fromFrame("warhol-shadow9"));
        var instance10 = new Sprite(fromFrame("warhol-shadow10"));
        var instance11 = new Sprite(fromFrame("warhol-shadow11"));
        var instance12 = new Sprite(fromFrame("warhol-shadow12"));
        var instance13 = new Sprite(fromFrame("warhol-shadow13"));
        var instance14 = new Sprite(fromFrame("warhol-shadow14"));
        var instance15 = new Sprite(fromFrame("warhol-shadow15"));
        var instance16 = new Sprite(fromFrame("warhol-shadow16"));
        var instance17 = new Sprite(fromFrame("warhol-shadow17"))
            .setTransform(-49.4, -30.3);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance6, 5, 1, {
                "5": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance7, 6, 1, {
                "6": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance8, 7, 1, {
                "7": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance9, 8, 1, {
                "8": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance10, 9, 1, {
                "9": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance11, 10, 1, {
                "10": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance12, 11, 1, {
                "11": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance13, 12, 1, {
                "12": {
                    x: -49.4,
                    y: -15.05
                }
            })
            .addTimedChild(instance14, 13, 1, {
                "13": {
                    x: -49.4,
                    y: -20.4
                }
            })
            .addTimedChild(instance15, 14, 1, {
                "14": {
                    x: -49.4,
                    y: -30.3
                }
            })
            .addTimedChild(instance16, 15, 1, {
                "15": {
                    x: -49.4,
                    y: -26.5
                }
            })
            .addTimedChild(instance17, 16, 89);
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 142, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-body2"));
        var instance2 = new Sprite(fromFrame("warhol-body4"));
        var instance3 = new Sprite(fromFrame("warhol-body6"));
        var instance4 = new Sprite(fromFrame("warhol-body8"));
        var instance5 = new Sprite(fromFrame("warhol-body10"));
        var instance6 = new Sprite(fromFrame("warhol-body12"));
        var instance7 = new Sprite(fromFrame("warhol-body14"));
        var instance8 = new Sprite(fromFrame("warhol-body16"));
        var instance9 = new Sprite(fromFrame("warhol-body18"));
        var instance10 = new Sprite(fromFrame("warhol-body20"));
        var instance11 = new Sprite(fromFrame("warhol-body22"))
            .setTransform(-40.15, -4.1);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -34.85,
                    y: -4.1
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -35.95,
                    y: -4.1
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -33.45,
                    y: -4.1
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -37.4,
                    y: -4.1
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -38.25,
                    y: -4.1
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -38.55,
                    y: -4.1
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -38.65,
                    y: -4.1
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -40.15,
                    y: -4.1
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -39.45,
                    y: -4.1
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -40,
                    y: -4.1
                }
            })
            .addTimedChild(instance11, 21, 121);
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 197, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-bg1"))
            .setTransform(-65.2, -65.2);
        this.addTimedChild(instance1);
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 197, loop: false });
        var instance1 = new Graphic17(MovieClip.SYNCHED);
        var instance3 = new Graphic16(MovieClip.SYNCHED)
            .setTransform(1.05, 40.95)
            .setColorTransform(0, 0.46, 0, 0.26, 0, 0.72);
        var instance5 = new Graphic15(MovieClip.SYNCHED)
            .setTransform(2.05, 21.5)
            .setColorTransform(0, 0.14, 0, 0.9, 0, 0.95);
        var instance4 = new Graphic14(MovieClip.SYNCHED)
            .setTransform(1.1, -1.9)
            .setColorTransform(0, 0.14, 0, 0.9, 0, 0.95);
        var instance2 = new Graphic13(MovieClip.SYNCHED)
            .setTransform(1.1, -1.9)
            .setColorTransform(0, 0.46, 0, 0.26, 0, 0.72);
        var instance6 = new Graphic12(MovieClip.SYNCHED)
            .setTransform(0.75, -3.55)
            .setColorTransform(0, 0.46, 0, 0.26, 0, 0.72);
        this.addTimedChild(instance1, 0, 197, {
                "0": {
                    y: -0.3,
                    a: 0,
                    c: [
                        0,
                        0.99,
                        0,
                        0.53,
                        0,
                        0.18
                    ]
                },
                "1": {
                    a: 0.01
                },
                "2": {
                    a: 0.04
                },
                "3": {
                    a: 0.09
                },
                "4": {
                    a: 0.18
                },
                "5": {
                    a: 0.3
                },
                "6": {
                    a: 0.46
                },
                "7": {
                    a: 0.65
                },
                "8": {
                    a: 0.82
                },
                "9": {
                    a: 0.93
                },
                "10": {
                    a: 0.98
                },
                "11": {
                    a: 1
                }
            })
            .addTimedChild(instance3, 55, 142)
            .addTimedChild(instance5, 92, 105)
            .addTimedChild(instance4, 83, 114)
            .addTimedChild(instance2, 34, 163)
            .addTimedChild(instance6, 109, 88);
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 98, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-eye1"));
        var instance2 = new Sprite(fromFrame("warhol-eye2"));
        var instance3 = new Sprite(fromFrame("warhol-eye3"));
        var instance4 = new Sprite(fromFrame("warhol-eye4"));
        var instance5 = new Sprite(fromFrame("warhol-eye5"));
        var instance6 = new Sprite(fromFrame("warhol-eye6"));
        var instance7 = new Sprite(fromFrame("warhol-eye7"));
        var instance8 = new Sprite(fromFrame("warhol-eye9"));
        var instance9 = new Sprite(fromFrame("warhol-eye10"));
        var instance10 = new Sprite(fromFrame("warhol-eye11"));
        var instance11 = new Sprite(fromFrame("warhol-eye12"));
        var instance12 = new Sprite(fromFrame("warhol-eye13"));
        var instance13 = new Sprite(fromFrame("warhol-eye14"));
        var instance14 = new Sprite(fromFrame("warhol-eye15"));
        var instance15 = new Sprite(fromFrame("warhol-eye16"));
        var instance16 = new Sprite(fromFrame("warhol-eye18"));
        var instance17 = new Sprite(fromFrame("warhol-eye20"));
        var instance18 = new Sprite(fromFrame("warhol-eye21"));
        var instance19 = new Sprite(fromFrame("warhol-eye22"));
        var instance20 = new Sprite(fromFrame("warhol-eye23"));
        var instance21 = new Sprite(fromFrame("warhol-eye24"));
        var instance22 = new Sprite(fromFrame("warhol-eye26"));
        var instance23 = new Sprite(fromFrame("warhol-eye28"));
        var instance24 = new Sprite(fromFrame("warhol-eye29"));
        var instance25 = new Sprite(fromFrame("warhol-eye30"));
        var instance26 = new Sprite(fromFrame("warhol-eye31"));
        var instance27 = new Sprite(fromFrame("warhol-eye32"));
        var instance28 = new Sprite(fromFrame("warhol-eye34"))
            .setTransform(-9.6, -9.6);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -8.6,
                    y: -7.2
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -8.6,
                    y: -8.3
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -8.6,
                    y: -9.55
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance6, 5, 1, {
                "5": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance7, 6, 2, {
                "6": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance15, 15, 2, {
                "15": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance16, 17, 2, {
                "17": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance17, 19, 1, {
                "19": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance18, 20, 1, {
                "20": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance19, 21, 1, {
                "21": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance20, 22, 1, {
                "22": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance21, 23, 2, {
                "23": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance22, 25, 2, {
                "25": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance23, 27, 1, {
                "27": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance24, 28, 1, {
                "28": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance25, 29, 1, {
                "29": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance26, 30, 1, {
                "30": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance27, 31, 2, {
                "31": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance28, 33, 65);
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 173, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-head2"));
        var instance2 = new Sprite(fromFrame("warhol-head3"));
        var instance3 = new Sprite(fromFrame("warhol-head4"));
        var instance4 = new Sprite(fromFrame("warhol-head5"));
        var instance5 = new Sprite(fromFrame("warhol-head6"));
        var instance6 = new Sprite(fromFrame("warhol-head7"));
        var instance7 = new Sprite(fromFrame("warhol-head8"));
        var instance8 = new Sprite(fromFrame("warhol-head9"));
        var instance9 = new Sprite(fromFrame("warhol-head10"));
        var instance10 = new Sprite(fromFrame("warhol-head11"));
        var instance11 = new Sprite(fromFrame("warhol-head12"));
        var instance12 = new Sprite(fromFrame("warhol-head13"));
        var instance13 = new Sprite(fromFrame("warhol-head14"));
        var instance14 = new Sprite(fromFrame("warhol-head15"));
        var instance15 = new Sprite(fromFrame("warhol-head16"));
        var instance16 = new Sprite(fromFrame("warhol-head17"));
        var instance17 = new Sprite(fromFrame("warhol-head18"));
        var instance18 = new Sprite(fromFrame("warhol-head19"));
        var instance19 = new Sprite(fromFrame("warhol-head20"));
        var instance20 = new Sprite(fromFrame("warhol-head21"));
        var instance21 = new Sprite(fromFrame("warhol-head22"));
        var instance22 = new Sprite(fromFrame("warhol-head23"));
        var instance23 = new Sprite(fromFrame("warhol-head24"));
        var instance24 = new Sprite(fromFrame("warhol-head25"));
        var instance25 = new Sprite(fromFrame("warhol-head26"));
        var instance26 = new Sprite(fromFrame("warhol-head27"));
        var instance27 = new Sprite(fromFrame("warhol-head28"));
        var instance28 = new Sprite(fromFrame("warhol-head29"));
        var instance29 = new Sprite(fromFrame("warhol-head30"))
            .setTransform(-49.6, -49.6);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: 0.25,
                    y: -49.6
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance15, 15, 1, {
                "15": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance16, 16, 1, {
                "16": {
                    x: -14,
                    y: -49.6
                }
            })
            .addTimedChild(instance17, 17, 1, {
                "17": {
                    x: -33,
                    y: -49.6
                }
            })
            .addTimedChild(instance18, 18, 1, {
                "18": {
                    x: -44.65,
                    y: -49.6
                }
            })
            .addTimedChild(instance19, 19, 1, {
                "19": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance20, 20, 1, {
                "20": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance21, 21, 1, {
                "21": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance22, 22, 1, {
                "22": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance23, 23, 1, {
                "23": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance24, 24, 1, {
                "24": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance25, 25, 1, {
                "25": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance26, 26, 1, {
                "26": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance27, 27, 1, {
                "27": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance28, 28, 1, {
                "28": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance29, 29, 144);
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 124, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-head-inside2"));
        var instance2 = new Sprite(fromFrame("warhol-head-inside4"));
        var instance3 = new Sprite(fromFrame("warhol-head-inside6"));
        var instance4 = new Sprite(fromFrame("warhol-head-inside8"));
        var instance5 = new Sprite(fromFrame("warhol-head-inside10"));
        var instance6 = new Sprite(fromFrame("warhol-head-inside12"));
        var instance7 = new Sprite(fromFrame("warhol-head-inside14"));
        var instance8 = new Sprite(fromFrame("warhol-head-inside16"));
        var instance9 = new Sprite(fromFrame("warhol-head-inside18"));
        var instance10 = new Sprite(fromFrame("warhol-head-inside20"));
        var instance11 = new Sprite(fromFrame("warhol-head-inside22"));
        var instance12 = new Sprite(fromFrame("warhol-head-inside24"));
        var instance13 = new Sprite(fromFrame("warhol-head-inside26"));
        var instance14 = new Sprite(fromFrame("warhol-head-inside28"));
        var instance15 = new Sprite(fromFrame("warhol-head-inside30"));
        var instance16 = new Sprite(fromFrame("warhol-head-inside32"));
        var instance17 = new Sprite(fromFrame("warhol-head-inside34"));
        var instance18 = new Sprite(fromFrame("warhol-head-inside36"));
        var instance19 = new Sprite(fromFrame("warhol-head-inside38"));
        var instance20 = new Sprite(fromFrame("warhol-head-inside40"));
        var instance21 = new Sprite(fromFrame("warhol-head-inside42"));
        var instance22 = new Sprite(fromFrame("warhol-head-inside44"));
        var instance23 = new Sprite(fromFrame("warhol-head-inside46"))
            .setTransform(-40.1, -40.1);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -33.9,
                    y: -40.05
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -38.2,
                    y: -40.1
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -39.85,
                    y: -40.1
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -40.05,
                    y: -40.1
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance11, 21, 2, {
                "21": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance12, 23, 2, {
                "23": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance13, 25, 2, {
                "25": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance14, 27, 2, {
                "27": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance15, 29, 2, {
                "29": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance16, 31, 2, {
                "31": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance17, 33, 2, {
                "33": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance18, 35, 2, {
                "35": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance19, 37, 2, {
                "37": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance20, 39, 2, {
                "39": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance21, 41, 2, {
                "41": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance22, 43, 2, {
                "43": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance23, 45, 79);
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 115, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-shadow1"));
        var instance2 = new Sprite(fromFrame("warhol-shadow2"));
        var instance3 = new Sprite(fromFrame("warhol-shadow3"));
        var instance4 = new Sprite(fromFrame("warhol-shadow4"));
        var instance5 = new Sprite(fromFrame("warhol-shadow5"));
        var instance6 = new Sprite(fromFrame("warhol-shadow6"));
        var instance7 = new Sprite(fromFrame("warhol-shadow7"));
        var instance8 = new Sprite(fromFrame("warhol-shadow8"));
        var instance9 = new Sprite(fromFrame("warhol-shadow9"));
        var instance10 = new Sprite(fromFrame("warhol-shadow10"));
        var instance11 = new Sprite(fromFrame("warhol-shadow11"));
        var instance12 = new Sprite(fromFrame("warhol-shadow12"));
        var instance13 = new Sprite(fromFrame("warhol-shadow13"));
        var instance14 = new Sprite(fromFrame("warhol-shadow14"));
        var instance15 = new Sprite(fromFrame("warhol-shadow15"));
        var instance16 = new Sprite(fromFrame("warhol-shadow16"));
        var instance17 = new Sprite(fromFrame("warhol-shadow17"))
            .setTransform(-49.4, -30.3);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance6, 5, 1, {
                "5": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance7, 6, 1, {
                "6": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance8, 7, 1, {
                "7": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance9, 8, 1, {
                "8": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance10, 9, 1, {
                "9": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance11, 10, 1, {
                "10": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance12, 11, 1, {
                "11": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance13, 12, 1, {
                "12": {
                    x: -49.4,
                    y: -15.05
                }
            })
            .addTimedChild(instance14, 13, 1, {
                "13": {
                    x: -49.4,
                    y: -20.4
                }
            })
            .addTimedChild(instance15, 14, 1, {
                "14": {
                    x: -49.4,
                    y: -30.3
                }
            })
            .addTimedChild(instance16, 15, 1, {
                "15": {
                    x: -49.4,
                    y: -26.5
                }
            })
            .addTimedChild(instance17, 16, 99);
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 152, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-body2"));
        var instance2 = new Sprite(fromFrame("warhol-body4"));
        var instance3 = new Sprite(fromFrame("warhol-body6"));
        var instance4 = new Sprite(fromFrame("warhol-body8"));
        var instance5 = new Sprite(fromFrame("warhol-body10"));
        var instance6 = new Sprite(fromFrame("warhol-body12"));
        var instance7 = new Sprite(fromFrame("warhol-body14"));
        var instance8 = new Sprite(fromFrame("warhol-body16"));
        var instance9 = new Sprite(fromFrame("warhol-body18"));
        var instance10 = new Sprite(fromFrame("warhol-body20"));
        var instance11 = new Sprite(fromFrame("warhol-body22"))
            .setTransform(-40.15, -4.1);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -34.85,
                    y: -4.1
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -35.95,
                    y: -4.1
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -33.45,
                    y: -4.1
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -37.4,
                    y: -4.1
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -38.25,
                    y: -4.1
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -38.55,
                    y: -4.1
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -38.65,
                    y: -4.1
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -40.15,
                    y: -4.1
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -39.45,
                    y: -4.1
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -40,
                    y: -4.1
                }
            })
            .addTimedChild(instance11, 21, 131);
    });

    var Graphic24 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 205, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-bg1"))
            .setTransform(-65.2, -65.2);
        this.addTimedChild(instance1);
    });

    var Graphic25 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 205, loop: false });
        var instance1 = new Graphic24(MovieClip.SYNCHED);
        var instance3 = new Graphic23(MovieClip.SYNCHED)
            .setTransform(1.05, 40.95)
            .setColorTransform(0, 0.53, 0, 0.83, 0, 0.25);
        var instance5 = new Graphic22(MovieClip.SYNCHED)
            .setTransform(2.05, 21.5)
            .setColorTransform(0, 0.2, 0, 0.35, 0, 0.63);
        var instance4 = new Graphic21(MovieClip.SYNCHED)
            .setTransform(1.1, -1.9)
            .setColorTransform(0, 0.2, 0, 0.35, 0, 0.63);
        var instance2 = new Graphic20(MovieClip.SYNCHED)
            .setTransform(1.1, -1.9)
            .setColorTransform(0, 0.53, 0, 0.83, 0, 0.25);
        var instance6 = new Graphic19(MovieClip.SYNCHED)
            .setTransform(0.75, -3.55)
            .setColorTransform(0, 0.53, 0, 0.83, 0, 0.25);
        this.addTimedChild(instance1, 0, 205, {
                "0": {
                    y: -0.3,
                    a: 0,
                    c: [
                        0,
                        0.88,
                        0,
                        0.09,
                        0,
                        0.31
                    ]
                },
                "1": {
                    a: 0.01
                },
                "2": {
                    a: 0.04
                },
                "3": {
                    a: 0.09
                },
                "4": {
                    a: 0.18
                },
                "5": {
                    a: 0.3
                },
                "6": {
                    a: 0.46
                },
                "7": {
                    a: 0.65
                },
                "8": {
                    a: 0.82
                },
                "9": {
                    a: 0.93
                },
                "10": {
                    a: 0.98
                },
                "11": {
                    a: 1
                }
            })
            .addTimedChild(instance3, 53, 152)
            .addTimedChild(instance5, 90, 115)
            .addTimedChild(instance4, 81, 124)
            .addTimedChild(instance2, 32, 173)
            .addTimedChild(instance6, 107, 98);
    });

    var Graphic26 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 116, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-eye1"));
        var instance2 = new Sprite(fromFrame("warhol-eye2"));
        var instance3 = new Sprite(fromFrame("warhol-eye3"));
        var instance4 = new Sprite(fromFrame("warhol-eye4"));
        var instance5 = new Sprite(fromFrame("warhol-eye5"));
        var instance6 = new Sprite(fromFrame("warhol-eye6"));
        var instance7 = new Sprite(fromFrame("warhol-eye7"));
        var instance8 = new Sprite(fromFrame("warhol-eye9"));
        var instance9 = new Sprite(fromFrame("warhol-eye10"));
        var instance10 = new Sprite(fromFrame("warhol-eye11"));
        var instance11 = new Sprite(fromFrame("warhol-eye12"));
        var instance12 = new Sprite(fromFrame("warhol-eye13"));
        var instance13 = new Sprite(fromFrame("warhol-eye14"));
        var instance14 = new Sprite(fromFrame("warhol-eye15"));
        var instance15 = new Sprite(fromFrame("warhol-eye16"));
        var instance16 = new Sprite(fromFrame("warhol-eye18"));
        var instance17 = new Sprite(fromFrame("warhol-eye20"));
        var instance18 = new Sprite(fromFrame("warhol-eye21"));
        var instance19 = new Sprite(fromFrame("warhol-eye22"));
        var instance20 = new Sprite(fromFrame("warhol-eye23"));
        var instance21 = new Sprite(fromFrame("warhol-eye24"));
        var instance22 = new Sprite(fromFrame("warhol-eye26"));
        var instance23 = new Sprite(fromFrame("warhol-eye28"));
        var instance24 = new Sprite(fromFrame("warhol-eye29"));
        var instance25 = new Sprite(fromFrame("warhol-eye30"));
        var instance26 = new Sprite(fromFrame("warhol-eye31"));
        var instance27 = new Sprite(fromFrame("warhol-eye32"));
        var instance28 = new Sprite(fromFrame("warhol-eye34"))
            .setTransform(-9.6, -9.6);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -8.6,
                    y: -7.2
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -8.6,
                    y: -8.3
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -8.6,
                    y: -9.55
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance6, 5, 1, {
                "5": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance7, 6, 2, {
                "6": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance15, 15, 2, {
                "15": {
                    x: -8.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance16, 17, 2, {
                "17": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance17, 19, 1, {
                "19": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance18, 20, 1, {
                "20": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance19, 21, 1, {
                "21": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance20, 22, 1, {
                "22": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance21, 23, 2, {
                "23": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance22, 25, 2, {
                "25": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance23, 27, 1, {
                "27": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance24, 28, 1, {
                "28": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance25, 29, 1, {
                "29": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance26, 30, 1, {
                "30": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance27, 31, 2, {
                "31": {
                    x: -9.6,
                    y: -9.6
                }
            })
            .addTimedChild(instance28, 33, 83);
    });

    var Graphic27 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 191, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-head2"));
        var instance2 = new Sprite(fromFrame("warhol-head3"));
        var instance3 = new Sprite(fromFrame("warhol-head4"));
        var instance4 = new Sprite(fromFrame("warhol-head5"));
        var instance5 = new Sprite(fromFrame("warhol-head6"));
        var instance6 = new Sprite(fromFrame("warhol-head7"));
        var instance7 = new Sprite(fromFrame("warhol-head8"));
        var instance8 = new Sprite(fromFrame("warhol-head9"));
        var instance9 = new Sprite(fromFrame("warhol-head10"));
        var instance10 = new Sprite(fromFrame("warhol-head11"));
        var instance11 = new Sprite(fromFrame("warhol-head12"));
        var instance12 = new Sprite(fromFrame("warhol-head13"));
        var instance13 = new Sprite(fromFrame("warhol-head14"));
        var instance14 = new Sprite(fromFrame("warhol-head15"));
        var instance15 = new Sprite(fromFrame("warhol-head16"));
        var instance16 = new Sprite(fromFrame("warhol-head17"));
        var instance17 = new Sprite(fromFrame("warhol-head18"));
        var instance18 = new Sprite(fromFrame("warhol-head19"));
        var instance19 = new Sprite(fromFrame("warhol-head20"));
        var instance20 = new Sprite(fromFrame("warhol-head21"));
        var instance21 = new Sprite(fromFrame("warhol-head22"));
        var instance22 = new Sprite(fromFrame("warhol-head23"));
        var instance23 = new Sprite(fromFrame("warhol-head24"));
        var instance24 = new Sprite(fromFrame("warhol-head25"));
        var instance25 = new Sprite(fromFrame("warhol-head26"));
        var instance26 = new Sprite(fromFrame("warhol-head27"));
        var instance27 = new Sprite(fromFrame("warhol-head28"));
        var instance28 = new Sprite(fromFrame("warhol-head29"));
        var instance29 = new Sprite(fromFrame("warhol-head30"))
            .setTransform(-49.6, -49.6);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: 0.25,
                    y: -49.6
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance15, 15, 1, {
                "15": {
                    x: 0.05,
                    y: -49.6
                }
            })
            .addTimedChild(instance16, 16, 1, {
                "16": {
                    x: -14,
                    y: -49.6
                }
            })
            .addTimedChild(instance17, 17, 1, {
                "17": {
                    x: -33,
                    y: -49.6
                }
            })
            .addTimedChild(instance18, 18, 1, {
                "18": {
                    x: -44.65,
                    y: -49.6
                }
            })
            .addTimedChild(instance19, 19, 1, {
                "19": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance20, 20, 1, {
                "20": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance21, 21, 1, {
                "21": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance22, 22, 1, {
                "22": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance23, 23, 1, {
                "23": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance24, 24, 1, {
                "24": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance25, 25, 1, {
                "25": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance26, 26, 1, {
                "26": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance27, 27, 1, {
                "27": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance28, 28, 1, {
                "28": {
                    x: -49.6,
                    y: -49.6
                }
            })
            .addTimedChild(instance29, 29, 162);
    });

    var Graphic28 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 142, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-head-inside2"));
        var instance2 = new Sprite(fromFrame("warhol-head-inside4"));
        var instance3 = new Sprite(fromFrame("warhol-head-inside6"));
        var instance4 = new Sprite(fromFrame("warhol-head-inside8"));
        var instance5 = new Sprite(fromFrame("warhol-head-inside10"));
        var instance6 = new Sprite(fromFrame("warhol-head-inside12"));
        var instance7 = new Sprite(fromFrame("warhol-head-inside14"));
        var instance8 = new Sprite(fromFrame("warhol-head-inside16"));
        var instance9 = new Sprite(fromFrame("warhol-head-inside18"));
        var instance10 = new Sprite(fromFrame("warhol-head-inside20"));
        var instance11 = new Sprite(fromFrame("warhol-head-inside22"));
        var instance12 = new Sprite(fromFrame("warhol-head-inside24"));
        var instance13 = new Sprite(fromFrame("warhol-head-inside26"));
        var instance14 = new Sprite(fromFrame("warhol-head-inside28"));
        var instance15 = new Sprite(fromFrame("warhol-head-inside30"));
        var instance16 = new Sprite(fromFrame("warhol-head-inside32"));
        var instance17 = new Sprite(fromFrame("warhol-head-inside34"));
        var instance18 = new Sprite(fromFrame("warhol-head-inside36"));
        var instance19 = new Sprite(fromFrame("warhol-head-inside38"));
        var instance20 = new Sprite(fromFrame("warhol-head-inside40"));
        var instance21 = new Sprite(fromFrame("warhol-head-inside42"));
        var instance22 = new Sprite(fromFrame("warhol-head-inside44"));
        var instance23 = new Sprite(fromFrame("warhol-head-inside46"))
            .setTransform(-40.1, -40.1);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -33.9,
                    y: -40.05
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -38.2,
                    y: -40.1
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -39.85,
                    y: -40.1
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -40.05,
                    y: -40.1
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance11, 21, 2, {
                "21": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance12, 23, 2, {
                "23": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance13, 25, 2, {
                "25": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance14, 27, 2, {
                "27": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance15, 29, 2, {
                "29": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance16, 31, 2, {
                "31": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance17, 33, 2, {
                "33": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance18, 35, 2, {
                "35": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance19, 37, 2, {
                "37": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance20, 39, 2, {
                "39": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance21, 41, 2, {
                "41": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance22, 43, 2, {
                "43": {
                    x: -40.1,
                    y: -40.1
                }
            })
            .addTimedChild(instance23, 45, 97);
    });

    var Graphic29 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 133, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-shadow1"));
        var instance2 = new Sprite(fromFrame("warhol-shadow2"));
        var instance3 = new Sprite(fromFrame("warhol-shadow3"));
        var instance4 = new Sprite(fromFrame("warhol-shadow4"));
        var instance5 = new Sprite(fromFrame("warhol-shadow5"));
        var instance6 = new Sprite(fromFrame("warhol-shadow6"));
        var instance7 = new Sprite(fromFrame("warhol-shadow7"));
        var instance8 = new Sprite(fromFrame("warhol-shadow8"));
        var instance9 = new Sprite(fromFrame("warhol-shadow9"));
        var instance10 = new Sprite(fromFrame("warhol-shadow10"));
        var instance11 = new Sprite(fromFrame("warhol-shadow11"));
        var instance12 = new Sprite(fromFrame("warhol-shadow12"));
        var instance13 = new Sprite(fromFrame("warhol-shadow13"));
        var instance14 = new Sprite(fromFrame("warhol-shadow14"));
        var instance15 = new Sprite(fromFrame("warhol-shadow15"));
        var instance16 = new Sprite(fromFrame("warhol-shadow16"));
        var instance17 = new Sprite(fromFrame("warhol-shadow17"))
            .setTransform(-49.4, -30.3);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance6, 5, 1, {
                "5": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance7, 6, 1, {
                "6": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance8, 7, 1, {
                "7": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance9, 8, 1, {
                "8": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance10, 9, 1, {
                "9": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance11, 10, 1, {
                "10": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance12, 11, 1, {
                "11": {
                    x: -49.4,
                    y: -12.4
                }
            })
            .addTimedChild(instance13, 12, 1, {
                "12": {
                    x: -49.4,
                    y: -15.05
                }
            })
            .addTimedChild(instance14, 13, 1, {
                "13": {
                    x: -49.4,
                    y: -20.4
                }
            })
            .addTimedChild(instance15, 14, 1, {
                "14": {
                    x: -49.4,
                    y: -30.3
                }
            })
            .addTimedChild(instance16, 15, 1, {
                "15": {
                    x: -49.4,
                    y: -26.5
                }
            })
            .addTimedChild(instance17, 16, 117);
    });

    var Graphic30 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 170, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-body2"));
        var instance2 = new Sprite(fromFrame("warhol-body4"));
        var instance3 = new Sprite(fromFrame("warhol-body6"));
        var instance4 = new Sprite(fromFrame("warhol-body8"));
        var instance5 = new Sprite(fromFrame("warhol-body10"));
        var instance6 = new Sprite(fromFrame("warhol-body12"));
        var instance7 = new Sprite(fromFrame("warhol-body14"));
        var instance8 = new Sprite(fromFrame("warhol-body16"));
        var instance9 = new Sprite(fromFrame("warhol-body18"));
        var instance10 = new Sprite(fromFrame("warhol-body20"));
        var instance11 = new Sprite(fromFrame("warhol-body22"))
            .setTransform(-40.15, -4.1);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -34.85,
                    y: -4.1
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -35.95,
                    y: -4.1
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -33.45,
                    y: -4.1
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -37.4,
                    y: -4.1
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -38.25,
                    y: -4.1
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -38.55,
                    y: -4.1
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -38.65,
                    y: -4.1
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -40.15,
                    y: -4.1
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -39.45,
                    y: -4.1
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -40,
                    y: -4.1
                }
            })
            .addTimedChild(instance11, 21, 149);
    });

    var Graphic31 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 214, loop: false });
        var instance1 = new Sprite(fromFrame("warhol-bg1"))
            .setTransform(-65.2, -65.2);
        this.addTimedChild(instance1);
    });

    var Graphic32 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 214, loop: false });
        var instance1 = new Graphic31(MovieClip.SYNCHED);
        var instance3 = new Graphic30(MovieClip.SYNCHED)
            .setTransform(1.05, 40.95)
            .setColorTransform(0, 0.97, 0, 0.75, 0, 0.2);
        var instance5 = new Graphic29(MovieClip.SYNCHED)
            .setTransform(2.05, 21.5)
            .setColorTransform(0, 0.9, 0, 0.09, 0, 0.32);
        var instance4 = new Graphic28(MovieClip.SYNCHED)
            .setTransform(1.1, -1.9)
            .setColorTransform(0, 0.9, 0, 0.09, 0, 0.32);
        var instance2 = new Graphic27(MovieClip.SYNCHED)
            .setTransform(1.1, -1.9)
            .setColorTransform(0, 0.97, 0, 0.75, 0, 0.2);
        var instance6 = new Graphic26(MovieClip.SYNCHED)
            .setTransform(0.75, -3.55)
            .setColorTransform(0, 0.97, 0, 0.75, 0, 0.2);
        this.addTimedChild(instance1, 0, 214, {
                "0": {
                    y: -0.3,
                    a: 0
                },
                "1": {
                    a: 0.01
                },
                "2": {
                    a: 0.04
                },
                "3": {
                    a: 0.09
                },
                "4": {
                    a: 0.18
                },
                "5": {
                    a: 0.3
                },
                "6": {
                    a: 0.46
                },
                "7": {
                    a: 0.65
                },
                "8": {
                    a: 0.82
                },
                "9": {
                    a: 0.93
                },
                "10": {
                    a: 0.98
                },
                "11": {
                    a: 1
                }
            })
            .addTimedChild(instance3, 44, 170)
            .addTimedChild(instance5, 81, 133)
            .addTimedChild(instance4, 72, 142)
            .addTimedChild(instance2, 23, 191)
            .addTimedChild(instance6, 98, 116);
    });

    var Graphic33 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 214, loop: false });
        var instance1 = new Graphic32(MovieClip.SYNCHED)
            .setTransform(-75.05, -49.65);
        var instance2 = new Graphic25(MovieClip.SYNCHED)
            .setTransform(54.95, -49.65);
        var instance3 = new Graphic18(MovieClip.SYNCHED)
            .setTransform(-75.05, 80.85);
        var instance4 = new Graphic11(MovieClip.SYNCHED)
            .setTransform(54.95, 80.85);
        this.addTimedChild(instance1)
            .addTimedChild(instance2, 9, 205)
            .addTimedChild(instance3, 17, 197)
            .addTimedChild(instance4, 24, 190);
    });

    var Graphic34 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 234, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-blank1"))
            .setTransform(-232.5, -161.65);
        this.addTimedChild(instance1);
    });

    var Graphic35 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 212, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-support1"))
            .setTransform(-15.9, -28.8);
        this.addTimedChild(instance1);
    });

    var Graphic36 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 212, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-leg1"))
            .setTransform(-41.9, -74.65);
        this.addTimedChild(instance1);
    });

    var Graphic37 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 212, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-leg1"))
            .setTransform(-41.9, -74.65);
        this.addTimedChild(instance1);
    });

    var Graphic38 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 234, loop: false });
        var instance6 = new Graphic37(MovieClip.SYNCHED);
        var instance5 = new Graphic36(MovieClip.SYNCHED);
        var instance4 = new Graphic35(MovieClip.SYNCHED);
        var instance1 = new Graphic34(MovieClip.SYNCHED);
        var instance8 = new Graphic33(MovieClip.SYNCHED);
        var instance7 = new Graphic4(MovieClip.SYNCHED);
        var instance3 = new Graphic3(MovieClip.SYNCHED);
        var instance2 = new Graphic2(MovieClip.SYNCHED);
        this.addTimedChild(instance6, 8, 212, {
                "8": {
                    x: 66,
                    y: 61.45,
                    ky: 3.142
                },
                "9": {
                    x: 69.15,
                    y: 71.15
                },
                "10": {
                    x: 79.3,
                    y: 102.55
                },
                "11": {
                    x: 94.65,
                    y: 149.95
                },
                "12": {
                    x: 107.55,
                    y: 189.75
                },
                "13": {
                    x: 112,
                    y: 203.45
                },
                "14": {
                    x: 111.85,
                    y: 202.95
                },
                "15": {
                    x: 111.4,
                    y: 201.4
                },
                "16": {
                    x: 110.75,
                    y: 199.1
                },
                "17": {
                    x: 110.2,
                    y: 197.1
                },
                "18": {
                    x: 110,
                    y: 196.45
                },
                "209": {
                    x: 110.1,
                    y: 196.8
                },
                "210": {
                    x: 110.45,
                    y: 198
                },
                "211": {
                    x: 111,
                    y: 200
                },
                "212": {
                    x: 111.55,
                    y: 201.95
                },
                "213": {
                    x: 111.9,
                    y: 203.1
                },
                "214": {
                    x: 112,
                    y: 203.45
                },
                "215": {
                    x: 108.55,
                    y: 192.8
                },
                "216": {
                    x: 97,
                    y: 157.1
                },
                "217": {
                    x: 80.65,
                    y: 106.75
                },
                "218": {
                    x: 69.35,
                    y: 71.8
                },
                "219": {
                    x: 66,
                    y: 61.45
                }
            })
            .addTimedChild(instance5, 8, 212, {
                "8": {
                    x: -62.9,
                    y: 62.45
                },
                "9": {
                    x: -66.1,
                    y: 72.15
                },
                "10": {
                    x: -76.5,
                    y: 103.55
                },
                "11": {
                    x: -92.2,
                    y: 150.95
                },
                "12": {
                    x: -105.35,
                    y: 190.75
                },
                "13": {
                    x: -109.9,
                    y: 204.45
                },
                "14": {
                    x: -109.7,
                    y: 203.9
                },
                "15": {
                    x: -109.05,
                    y: 202.15
                },
                "16": {
                    x: -108.05,
                    y: 199.45
                },
                "17": {
                    x: -107.2,
                    y: 197.2
                },
                "18": {
                    x: -106.9,
                    y: 196.45
                },
                "209": {
                    x: -107.05,
                    y: 196.85
                },
                "210": {
                    x: -107.55,
                    y: 198.25
                },
                "211": {
                    x: -108.4,
                    y: 200.5
                },
                "212": {
                    x: -109.25,
                    y: 202.7
                },
                "213": {
                    x: -109.75,
                    y: 204.05
                },
                "214": {
                    x: -109.9,
                    y: 204.45
                },
                "215": {
                    x: -106.4,
                    y: 193.8
                },
                "216": {
                    x: -94.55,
                    y: 158.1
                },
                "217": {
                    x: -77.9,
                    y: 107.75
                },
                "218": {
                    x: -66.3,
                    y: 72.8
                },
                "219": {
                    x: -62.9,
                    y: 62.45
                }
            })
            .addTimedChild(instance4, 8, 212, {
                "8": {
                    y: -185.35
                },
                "9": {
                    y: -189.65
                },
                "10": {
                    y: -203.6
                },
                "11": {
                    y: -224.6
                },
                "12": {
                    y: -242.3
                },
                "13": {
                    y: -248.35
                },
                "14": {
                    y: -247.95
                },
                "15": {
                    y: -246.6
                },
                "16": {
                    y: -244.6
                },
                "17": {
                    y: -242.95
                },
                "18": {
                    y: -242.35
                },
                "209": {
                    y: -242.85
                },
                "210": {
                    y: -244.6
                },
                "211": {
                    y: -247.4
                },
                "212": {
                    y: -250.2
                },
                "213": {
                    y: -251.85
                },
                "214": {
                    y: -252.35
                },
                "215": {
                    y: -247.35
                },
                "216": {
                    y: -230.5
                },
                "217": {
                    y: -206.7
                },
                "218": {
                    y: -190.25
                },
                "219": {
                    y: -185.35
                }
            })
            .addTimedChild(instance1, 0, 234, {
                "0": {
                    x: 0,
                    y: -43.9,
                    r: 0.021,
                    c: [
                        0,
                        1,
                        0,
                        1,
                        0,
                        1
                    ]
                },
                "1": {
                    x: 0.005,
                    y: -43.897,
                    r: 0.014,
                    c: [
                        0.3,
                        0.69,
                        0.3,
                        0.69,
                        0.3,
                        0.69
                    ]
                },
                "2": {
                    x: 0.003,
                    y: -43.896,
                    r: 0.009,
                    c: [
                        0.55,
                        0.44,
                        0.55,
                        0.44,
                        0.55,
                        0.44
                    ]
                },
                "3": {
                    x: 0.013,
                    y: -43.893,
                    r: 0.005,
                    c: [
                        0.75,
                        0.25,
                        0.75,
                        0.25,
                        0.75,
                        0.25
                    ]
                },
                "4": {
                    x: -0.013,
                    y: -43.891,
                    r: 0.001,
                    c: [
                        0.89,
                        0.11,
                        0.89,
                        0.11,
                        0.89,
                        0.11
                    ]
                },
                "5": {
                    x: 0.018,
                    r: 0,
                    c: [
                        0.97,
                        0.03,
                        0.97,
                        0.03,
                        0.97,
                        0.03
                    ]
                },
                "6": {
                    x: 0,
                    y: -43.9,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                }
            })
            .addTimedChild(instance8, 20, 214, {
                "20": {
                    x: 10.55,
                    y: -56.9,
                    a: 1
                },
                "223": {
                    a: 0.88
                },
                "224": {
                    a: 0.75
                },
                "225": {
                    a: 0.63
                },
                "226": {
                    a: 0.5
                },
                "227": {
                    a: 0.38
                },
                "228": {
                    a: 0.25
                },
                "229": {
                    a: 0.13
                },
                "230": {
                    a: 0
                }
            })
            .addTimedChild(instance7, 10, 224, {
                "10": {
                    y: -196.15,
                    a: 0
                },
                "11": {
                    a: 0.11
                },
                "12": {
                    a: 0.22
                },
                "13": {
                    a: 0.33
                },
                "14": {
                    a: 0.45
                },
                "15": {
                    a: 0.55
                },
                "16": {
                    a: 0.67
                },
                "17": {
                    a: 0.78
                },
                "18": {
                    a: 0.89
                },
                "19": {
                    a: 1
                },
                "209": {
                    a: 0.9
                },
                "210": {
                    a: 0.8
                },
                "211": {
                    a: 0.7
                },
                "212": {
                    a: 0.6
                },
                "213": {
                    a: 0.5
                },
                "214": {
                    a: 0.4
                },
                "215": {
                    a: 0.3
                },
                "216": {
                    a: 0.2
                },
                "217": {
                    a: 0.1
                },
                "218": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 4, 230, {
                "4": {
                    y: -190.45,
                    a: 0
                },
                "5": {
                    y: -191.7,
                    a: 0.05
                },
                "6": {
                    y: -195.8,
                    a: 0.2
                },
                "7": {
                    y: -202.6,
                    a: 0.45
                },
                "8": {
                    y: -210.2,
                    a: 0.73
                },
                "9": {
                    y: -215.65,
                    a: 0.93
                },
                "10": {
                    y: -217.45,
                    a: 1
                },
                "11": {
                    y: -217.15
                },
                "12": {
                    y: -216.25
                },
                "13": {
                    y: -214.75
                },
                "14": {
                    y: -213.05
                },
                "15": {
                    y: -211.85
                },
                "16": {
                    y: -211.45
                },
                "219": {
                    y: -211.65,
                    a: 0.98
                },
                "220": {
                    y: -212.2,
                    a: 0.93
                },
                "221": {
                    y: -213.25,
                    a: 0.82
                },
                "222": {
                    y: -214.7,
                    a: 0.67
                },
                "223": {
                    y: -216.5,
                    a: 0.5
                },
                "224": {
                    y: -218.25,
                    a: 0.32
                },
                "225": {
                    y: -219.7,
                    a: 0.17
                },
                "226": {
                    y: -220.7,
                    a: 0.07
                },
                "227": {
                    y: -221.3,
                    a: 0.02
                },
                "228": {
                    y: -221.45,
                    a: 0
                }
            })
            .addTimedChild(instance2, 4, 230, {
                "4": {
                    y: 102.45,
                    a: 0
                },
                "5": {
                    y: 103.85,
                    a: 0.05
                },
                "6": {
                    y: 108.4,
                    a: 0.2
                },
                "7": {
                    y: 115.95,
                    a: 0.45
                },
                "8": {
                    y: 124.4,
                    a: 0.73
                },
                "9": {
                    y: 130.4,
                    a: 0.93
                },
                "10": {
                    y: 132.4,
                    a: 1
                },
                "11": {
                    y: 132.2
                },
                "12": {
                    y: 131.6
                },
                "13": {
                    y: 130.6
                },
                "14": {
                    y: 129.45
                },
                "15": {
                    y: 128.65
                },
                "16": {
                    y: 128.4
                },
                "219": {
                    y: 128.6,
                    a: 0.98
                },
                "220": {
                    y: 129.15,
                    a: 0.93
                },
                "221": {
                    y: 130.2,
                    a: 0.82
                },
                "222": {
                    y: 131.65,
                    a: 0.67
                },
                "223": {
                    y: 133.45,
                    a: 0.5
                },
                "224": {
                    y: 135.2,
                    a: 0.32
                },
                "225": {
                    y: 136.65,
                    a: 0.17
                },
                "226": {
                    y: 137.65,
                    a: 0.07
                },
                "227": {
                    y: 138.25,
                    a: 0.02
                },
                "228": {
                    y: 138.4,
                    a: 0
                }
            });
    });

    var Graphic39 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 7, loop: false });
        var instance1 = new Sprite(fromFrame("EYE159"));
        var instance2 = new Sprite(fromFrame("EYE160"));
        var instance3 = new Sprite(fromFrame("EYE161"));
        var instance4 = new Sprite(fromFrame("EYE162"));
        var instance5 = new Sprite(fromFrame("EYE163"));
        var instance6 = new Sprite(fromFrame("EYE164"))
            .setTransform(-278.8, -277.7);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -264.1,
                    y: -236.5
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -270,
                    y: -251.3
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -275.85,
                    y: -268.85
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -277.35,
                    y: -273.25
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -278.85,
                    y: -277.75
                }
            })
            .addTimedChild(instance6, 5, 2);
    });

    lib.drawing_warhol = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 246,
            framerate: 30,
            loop: false
        });
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        var instance3 = new Graphic39(MovieClip.SYNCHED)
            .setTransform(640.3, 359.3, 0.898, 0.898);
        var instance2 = new Graphic38(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 6, {
                "0": {
                    x: 640.3,
                    y: 359.3,
                    sx: 0.898,
                    sy: 0.898
                }
            })
            .addTimedChild(instance3, 239, 7)
            .addTimedChild(instance2, 6, 234, {
                "6": {
                    x: 640.45,
                    y: 359.95,
                    sx: 1,
                    sy: 1,
                    kx: 0,
                    ky: 0,
                    r: 0,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "229": {
                    x: 640.6,
                    y: 360,
                    r: -0.003
                },
                "230": {
                    x: 640.619,
                    y: 359.982,
                    r: -0.005
                },
                "231": {
                    x: 640.851,
                    y: 360.003,
                    r: -0.009
                },
                "232": {
                    x: 641.439,
                    y: 360.021,
                    r: -0.022
                },
                "233": {
                    x: 642.1,
                    y: 360.104,
                    r: -0.036
                },
                "234": {
                    x: 642.8,
                    y: 360.1,
                    r: -0.051
                },
                "235": {
                    x: 631.94,
                    y: 358.342,
                    sx: 0.999,
                    sy: 0.999,
                    r: 0.189,
                    c: [
                        0.96,
                        0.04,
                        0.96,
                        0.04,
                        0.96,
                        0.04
                    ]
                },
                "236": {
                    x: 602.636,
                    y: 333.341,
                    r: 1.05,
                    c: [
                        0.83,
                        0.17,
                        0.83,
                        0.17,
                        0.83,
                        0.17
                    ]
                },
                "237": {
                    x: 626.319,
                    y: 271.568,
                    kx: 3.632,
                    ky: 2.651,
                    r: 0,
                    c: [
                        0.58,
                        0.42,
                        0.58,
                        0.42,
                        0.58,
                        0.42
                    ]
                },
                "238": {
                    x: 690.582,
                    y: 316.559,
                    sx: 1,
                    sy: 1,
                    kx: 0,
                    ky: 0,
                    r: -1.602,
                    c: [
                        0.26,
                        0.74,
                        0.26,
                        0.74,
                        0.26,
                        0.74
                    ]
                },
                "239": {
                    x: 638.5,
                    y: 359.65,
                    r: 0.044,
                    c: [
                        0,
                        1,
                        0,
                        1,
                        0,
                        1
                    ]
                }
            });
    });

    lib.drawing_warhol.assets = {
        "drawing_warhol_atlas_1": "images/drawing_warhol_atlas_1.json",
        "drawing_warhol_atlas_2": "images/drawing_warhol_atlas_2.json",
        "drawing_warhol_atlas_3": "images/drawing_warhol_atlas_3.json",
        "drawing_warhol_atlas_4": "images/drawing_warhol_atlas_4.json",
        "drawing_warhol_atlas_5": "images/drawing_warhol_atlas_5.json",
        "drawing_warhol_atlas_6": "images/drawing_warhol_atlas_6.json",
        "drawing_warhol_atlas_7": "images/drawing_warhol_atlas_7.json",
        "drawing_warhol_atlas_8": "images/drawing_warhol_atlas_8.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.drawing_warhol,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 246,
        library: lib
    };
}
