// jibo-anim-db-animations/timelines/drawing_haring.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 6, loop: false });
        var instance1 = new Sprite(fromFrame("EYE1"));
        var instance2 = new Sprite(fromFrame("EYE2"));
        var instance3 = new Sprite(fromFrame("EYE3"));
        var instance4 = new Sprite(fromFrame("EYE4"));
        var instance5 = new Sprite(fromFrame("EYE5"));
        var instance6 = new Sprite(fromFrame("EYE6"))
            .setTransform(-257.3, -226.65);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -278.8,
                    y: -277.7
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -276.2,
                    y: -271.55
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -272.8,
                    y: -263.4
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -268.5,
                    y: -253.2
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -263.3,
                    y: -240.95
                }
            })
            .addTimedChild(instance6, 5, 1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 259, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-bottom1"))
            .setTransform(-195.85, -14);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 259, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-top1"))
            .setTransform(-177.2, -13.75);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 253, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-shadow1"))
            .setTransform(-176.3, -7.5);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 243, loop: false });
        var instance1 = new Sprite(fromFrame("haring-frame2"));
        var instance2 = new Sprite(fromFrame("haring-frame4"));
        var instance3 = new Sprite(fromFrame("haring-frame6"));
        var instance4 = new Sprite(fromFrame("haring-frame8"));
        var instance5 = new Sprite(fromFrame("haring-frame9"));
        var instance6 = new Sprite(fromFrame("haring-frame10"));
        var instance7 = new Sprite(fromFrame("haring-frame11"));
        var instance8 = new Sprite(fromFrame("haring-frame12"));
        var instance9 = new Sprite(fromFrame("haring-frame13"));
        var instance10 = new Sprite(fromFrame("haring-frame14"));
        var instance11 = new Sprite(fromFrame("haring-frame15"));
        var instance12 = new Sprite(fromFrame("haring-frame16"));
        var instance13 = new Sprite(fromFrame("haring-frame17"));
        var instance14 = new Sprite(fromFrame("haring-frame18"));
        var instance15 = new Sprite(fromFrame("haring-frame19"));
        var instance16 = new Sprite(fromFrame("haring-frame20"));
        var instance17 = new Sprite(fromFrame("haring-frame21"));
        var instance18 = new Sprite(fromFrame("haring-frame22"));
        var instance19 = new Sprite(fromFrame("haring-frame23"));
        var instance20 = new Sprite(fromFrame("haring-frame24"));
        var instance21 = new Sprite(fromFrame("haring-frame25"));
        var instance22 = new Sprite(fromFrame("haring-frame26"));
        var instance23 = new Sprite(fromFrame("haring-frame27"));
        var instance24 = new Sprite(fromFrame("haring-frame28"));
        var instance25 = new Sprite(fromFrame("haring-frame29"));
        var instance26 = new Sprite(fromFrame("haring-frame30"));
        var instance27 = new Sprite(fromFrame("haring-frame31"));
        var instance28 = new Sprite(fromFrame("haring-frame32"));
        var instance29 = new Sprite(fromFrame("haring-frame33"));
        var instance30 = new Sprite(fromFrame("haring-frame34"));
        var instance31 = new Sprite(fromFrame("haring-frame35"));
        var instance32 = new Sprite(fromFrame("haring-frame36"));
        var instance33 = new Sprite(fromFrame("haring-frame37"));
        var instance34 = new Sprite(fromFrame("haring-frame38"));
        var instance35 = new Sprite(fromFrame("haring-frame39"));
        var instance36 = new Sprite(fromFrame("haring-frame40"));
        var instance37 = new Sprite(fromFrame("haring-frame41"));
        var instance38 = new Sprite(fromFrame("haring-frame43"));
        var instance39 = new Sprite(fromFrame("haring-frame45"))
            .setTransform(-186.85, -129.05);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -185,
                    y: -125.9
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -185,
                    y: -126.3
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -185,
                    y: -126.3
                }
            })
            .addTimedChild(instance4, 7, 1, {
                "7": {
                    x: -185,
                    y: -126.55
                }
            })
            .addTimedChild(instance5, 8, 1, {
                "8": {
                    x: -185.05,
                    y: -127
                }
            })
            .addTimedChild(instance6, 9, 1, {
                "9": {
                    x: -185.05,
                    y: -127.4
                }
            })
            .addTimedChild(instance7, 10, 1, {
                "10": {
                    x: -185.05,
                    y: -128.05
                }
            })
            .addTimedChild(instance8, 11, 1, {
                "11": {
                    x: -185.05,
                    y: -128.9
                }
            })
            .addTimedChild(instance9, 12, 1, {
                "12": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance10, 13, 1, {
                "13": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance11, 14, 1, {
                "14": {
                    x: -185,
                    y: -129.05
                }
            })
            .addTimedChild(instance12, 15, 1, {
                "15": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance13, 16, 1, {
                "16": {
                    x: -185,
                    y: -129.05
                }
            })
            .addTimedChild(instance14, 17, 1, {
                "17": {
                    x: -185,
                    y: -129.05
                }
            })
            .addTimedChild(instance15, 18, 1, {
                "18": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance16, 19, 1, {
                "19": {
                    x: -185,
                    y: -129.05
                }
            })
            .addTimedChild(instance17, 20, 1, {
                "20": {
                    x: -185,
                    y: -129.05
                }
            })
            .addTimedChild(instance18, 21, 1, {
                "21": {
                    x: -185,
                    y: -129.05
                }
            })
            .addTimedChild(instance19, 22, 1, {
                "22": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance20, 23, 1, {
                "23": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance21, 24, 1, {
                "24": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance22, 25, 1, {
                "25": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance23, 26, 1, {
                "26": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance24, 27, 1, {
                "27": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance25, 28, 1, {
                "28": {
                    x: -185,
                    y: -129.05
                }
            })
            .addTimedChild(instance26, 29, 1, {
                "29": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance27, 30, 1, {
                "30": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance28, 31, 1, {
                "31": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance29, 32, 1, {
                "32": {
                    x: -185,
                    y: -129.05
                }
            })
            .addTimedChild(instance30, 33, 1, {
                "33": {
                    x: -185.05,
                    y: -129.05
                }
            })
            .addTimedChild(instance31, 34, 1, {
                "34": {
                    x: -185.55,
                    y: -129.05
                }
            })
            .addTimedChild(instance32, 35, 1, {
                "35": {
                    x: -186.85,
                    y: -129.05
                }
            })
            .addTimedChild(instance33, 36, 1, {
                "36": {
                    x: -186.85,
                    y: -129.05
                }
            })
            .addTimedChild(instance34, 37, 1, {
                "37": {
                    x: -186.85,
                    y: -129.05
                }
            })
            .addTimedChild(instance35, 38, 1, {
                "38": {
                    x: -186.85,
                    y: -129.05
                }
            })
            .addTimedChild(instance36, 39, 1, {
                "39": {
                    x: -186.85,
                    y: -129.05
                }
            })
            .addTimedChild(instance37, 40, 2, {
                "40": {
                    x: -186.85,
                    y: -129.05
                }
            })
            .addTimedChild(instance38, 42, 2, {
                "42": {
                    x: -186.85,
                    y: -129.05
                }
            })
            .addTimedChild(instance39, 44, 199);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 63, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 55);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 66, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 58);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 69, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 61);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 72, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 64);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 77, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 69);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 81, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 73);
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 83, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 75);
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 85, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 77);
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 87, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 79);
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 89, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 81);
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 92, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 84);
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 95, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 87);
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 97, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 89);
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 98, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 90);
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 100, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 92);
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 103, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 95);
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 105, loop: false });
        var instance1 = new Sprite(fromFrame("haring-line12"));
        var instance2 = new Sprite(fromFrame("haring-line13"));
        var instance3 = new Sprite(fromFrame("haring-line14"));
        var instance4 = new Sprite(fromFrame("haring-line15"));
        var instance5 = new Sprite(fromFrame("haring-line16"));
        var instance6 = new Sprite(fromFrame("haring-line17"));
        var instance7 = new Sprite(fromFrame("haring-line18"));
        var instance8 = new Sprite(fromFrame("haring-line19"))
            .setTransform(-16.05, -3.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: 9.9,
                    y: -3.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: 6.5,
                    y: -3.5
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: 2.5,
                    y: -3.65
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -10,
                    y: -3.8
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -13,
                    y: -3.8
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -13.55,
                    y: -3.8
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -15.4,
                    y: -3.8
                }
            })
            .addTimedChild(instance8, 8, 97);
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 105, loop: false });
        var instance1 = new Graphic22(MovieClip.SYNCHED)
            .setTransform(-88.35, 82.15, 1, 1, -0.549);
        var instance2 = new Graphic21(MovieClip.SYNCHED)
            .setTransform(-99.35, 56.1, 0.656, 0.74, -0.348);
        var instance3 = new Graphic20(MovieClip.SYNCHED)
            .setTransform(-106.2, 29.35, 1, 1, -0.131);
        var instance4 = new Graphic19(MovieClip.SYNCHED)
            .setTransform(-103, 1.5, 0.656, 0.74, 0.039);
        var instance5 = new Graphic18(MovieClip.SYNCHED)
            .setTransform(-100.15, -27.6, 1, 1, 0.431);
        var instance6 = new Graphic17(MovieClip.SYNCHED)
            .setTransform(-79.15, -48.7, 0.656, 0.74, 0.72);
        var instance7 = new Graphic16(MovieClip.SYNCHED)
            .setTransform(-57.65, -69.05, 1, 1, 1.133);
        var instance8 = new Graphic15(MovieClip.SYNCHED)
            .setTransform(-29.95, -70.55, 0.656, 0.74, 1.426);
        var instance9 = new Graphic14(MovieClip.SYNCHED)
            .setTransform(-6.2, -78.05, 1, 1, 1.53);
        var instance10 = new Graphic13(MovieClip.SYNCHED)
            .setTransform(25.8, -72.45, 0.656, 0.74, 0, 4.491, 1.792);
        var instance11 = new Graphic12(MovieClip.SYNCHED)
            .setTransform(58.25, -66.05, 1, 1, 0, 4.064, 2.219);
        var instance12 = new Graphic11(MovieClip.SYNCHED)
            .setTransform(76.8, -43.8, 0.656, 0.74, 0, 3.786, 2.497);
        var instance13 = new Graphic10(MovieClip.SYNCHED)
            .setTransform(95.1, -22.1, 1, 1, 0, 3.533, 2.75);
        var instance14 = new Graphic9(MovieClip.SYNCHED)
            .setTransform(95.05, 4.2, 0.656, 0.74, -3.14);
        var instance15 = new Graphic8(MovieClip.SYNCHED)
            .setTransform(99.95, 28.4, 1, 1, -2.979);
        var instance16 = new Graphic7(MovieClip.SYNCHED)
            .setTransform(94.25, 52.25, 0.656, 0.74, -2.796);
        var instance17 = new Graphic6(MovieClip.SYNCHED)
            .setTransform(89.75, 75.6, 1, 1, -2.691);
        this.addTimedChild(instance1)
            .addTimedChild(instance2, 2, 103)
            .addTimedChild(instance3, 5, 100)
            .addTimedChild(instance4, 7, 98)
            .addTimedChild(instance5, 8, 97)
            .addTimedChild(instance6, 10, 95)
            .addTimedChild(instance7, 13, 92)
            .addTimedChild(instance8, 16, 89)
            .addTimedChild(instance9, 18, 87)
            .addTimedChild(instance10, 20, 85)
            .addTimedChild(instance11, 22, 83)
            .addTimedChild(instance12, 24, 81)
            .addTimedChild(instance13, 28, 77)
            .addTimedChild(instance14, 33, 72)
            .addTimedChild(instance15, 36, 69)
            .addTimedChild(instance16, 39, 66)
            .addTimedChild(instance17, 42, 63);
    });

    var Graphic24 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 110, loop: false });
        var instance1 = new Sprite(fromFrame("haring-eye2"));
        var instance2 = new Sprite(fromFrame("haring-eye4"));
        var instance3 = new Sprite(fromFrame("haring-eye6"));
        var instance4 = new Sprite(fromFrame("haring-eye8"));
        var instance5 = new Sprite(fromFrame("haring-eye10"));
        var instance6 = new Sprite(fromFrame("haring-eye12"));
        var instance7 = new Sprite(fromFrame("haring-eye14"));
        var instance8 = new Sprite(fromFrame("haring-eye16"));
        var instance9 = new Sprite(fromFrame("haring-eye18"))
            .setTransform(-15.5, -15.65);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -3.75,
                    y: -3.75
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -6.3,
                    y: -5.85
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -7.2,
                    y: -8
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -9.1,
                    y: -8.85
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -12.4,
                    y: -11.4
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -12.4,
                    y: -11.35
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -13.8,
                    y: -13.55
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -14.4,
                    y: -14
                }
            })
            .addTimedChild(instance9, 17, 93);
    });

    var Graphic25 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 161, loop: false });
        var instance1 = new Sprite(fromFrame("haring-jibo-lines-circle2"));
        var instance2 = new Sprite(fromFrame("haring-jibo-lines-circle4"));
        var instance3 = new Sprite(fromFrame("haring-jibo-lines-circle6"));
        var instance4 = new Sprite(fromFrame("haring-jibo-lines-circle7"));
        var instance5 = new Sprite(fromFrame("haring-jibo-lines-circle8"));
        var instance6 = new Sprite(fromFrame("haring-jibo-lines-circle9"));
        var instance7 = new Sprite(fromFrame("haring-jibo-lines-circle10"));
        var instance8 = new Sprite(fromFrame("haring-jibo-lines-circle11"));
        var instance9 = new Sprite(fromFrame("haring-jibo-lines-circle12"));
        var instance10 = new Sprite(fromFrame("haring-jibo-lines-circle13"));
        var instance11 = new Sprite(fromFrame("haring-jibo-lines-circle14"));
        var instance12 = new Sprite(fromFrame("haring-jibo-lines-circle15"));
        var instance13 = new Sprite(fromFrame("haring-jibo-lines-circle16"));
        var instance14 = new Sprite(fromFrame("haring-jibo-lines-circle17"));
        var instance15 = new Sprite(fromFrame("haring-jibo-lines-circle18"));
        var instance16 = new Sprite(fromFrame("haring-jibo-lines-circle20"));
        var instance17 = new Sprite(fromFrame("haring-jibo-lines-circle22"));
        var instance18 = new Sprite(fromFrame("haring-jibo-lines-circle24"))
            .setTransform(-57.25, -51.9);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -0.7,
                    y: -51.85
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -0.7,
                    y: -51.85
                }
            })
            .addTimedChild(instance3, 5, 1, {
                "5": {
                    x: -0.7,
                    y: -51.85
                }
            })
            .addTimedChild(instance4, 6, 1, {
                "6": {
                    x: -0.7,
                    y: -51.85
                }
            })
            .addTimedChild(instance5, 7, 1, {
                "7": {
                    x: -0.7,
                    y: -51.85
                }
            })
            .addTimedChild(instance6, 8, 1, {
                "8": {
                    x: -0.7,
                    y: -51.85
                }
            })
            .addTimedChild(instance7, 9, 1, {
                "9": {
                    x: -0.7,
                    y: -51.85
                }
            })
            .addTimedChild(instance8, 10, 1, {
                "10": {
                    x: -0.7,
                    y: -51.85
                }
            })
            .addTimedChild(instance9, 11, 1, {
                "11": {
                    x: -31.5,
                    y: -51.85
                }
            })
            .addTimedChild(instance10, 12, 1, {
                "12": {
                    x: -49.85,
                    y: -51.85
                }
            })
            .addTimedChild(instance11, 13, 1, {
                "13": {
                    x: -57,
                    y: -51.85
                }
            })
            .addTimedChild(instance12, 14, 1, {
                "14": {
                    x: -57.25,
                    y: -51.85
                }
            })
            .addTimedChild(instance13, 15, 1, {
                "15": {
                    x: -57.25,
                    y: -51.85
                }
            })
            .addTimedChild(instance14, 16, 1, {
                "16": {
                    x: -57.25,
                    y: -51.85
                }
            })
            .addTimedChild(instance15, 17, 2, {
                "17": {
                    x: -57.25,
                    y: -51.85
                }
            })
            .addTimedChild(instance16, 19, 2, {
                "19": {
                    x: -57.25,
                    y: -51.85
                }
            })
            .addTimedChild(instance17, 21, 2, {
                "21": {
                    x: -57.25,
                    y: -51.85
                }
            })
            .addTimedChild(instance18, 23, 138);
    });

    var Graphic26 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 205, loop: false });
        var instance1 = new Sprite(fromFrame("haring-jibo-lines-body2"));
        var instance2 = new Sprite(fromFrame("haring-jibo-lines-body3"));
        var instance3 = new Sprite(fromFrame("haring-jibo-lines-body4"));
        var instance4 = new Sprite(fromFrame("haring-jibo-lines-body5"));
        var instance5 = new Sprite(fromFrame("haring-jibo-lines-body6"));
        var instance6 = new Sprite(fromFrame("haring-jibo-lines-body7"));
        var instance7 = new Sprite(fromFrame("haring-jibo-lines-body8"));
        var instance8 = new Sprite(fromFrame("haring-jibo-lines-body9"));
        var instance9 = new Sprite(fromFrame("haring-jibo-lines-body10"));
        var instance10 = new Sprite(fromFrame("haring-jibo-lines-body11"));
        var instance11 = new Sprite(fromFrame("haring-jibo-lines-body12"));
        var instance12 = new Sprite(fromFrame("haring-jibo-lines-body13"));
        var instance13 = new Sprite(fromFrame("haring-jibo-lines-body14"));
        var instance14 = new Sprite(fromFrame("haring-jibo-lines-body15"));
        var instance15 = new Sprite(fromFrame("haring-jibo-lines-body16"));
        var instance16 = new Sprite(fromFrame("haring-jibo-lines-body17"));
        var instance17 = new Sprite(fromFrame("haring-jibo-lines-body18"));
        var instance18 = new Sprite(fromFrame("haring-jibo-lines-body19"));
        var instance19 = new Sprite(fromFrame("haring-jibo-lines-body20"));
        var instance20 = new Sprite(fromFrame("haring-jibo-lines-body21"));
        var instance21 = new Sprite(fromFrame("haring-jibo-lines-body22"));
        var instance22 = new Sprite(fromFrame("haring-jibo-lines-body23"));
        var instance23 = new Sprite(fromFrame("haring-jibo-lines-body24"));
        var instance24 = new Sprite(fromFrame("haring-jibo-lines-body25"));
        var instance25 = new Sprite(fromFrame("haring-jibo-lines-body26"));
        var instance26 = new Sprite(fromFrame("haring-jibo-lines-body27"));
        var instance27 = new Sprite(fromFrame("haring-jibo-lines-body28"));
        var instance28 = new Sprite(fromFrame("haring-jibo-lines-body29"));
        var instance29 = new Sprite(fromFrame("haring-jibo-lines-body30"));
        var instance30 = new Sprite(fromFrame("haring-jibo-lines-body31"));
        var instance31 = new Sprite(fromFrame("haring-jibo-lines-body32"));
        var instance32 = new Sprite(fromFrame("haring-jibo-lines-body33"));
        var instance33 = new Sprite(fromFrame("haring-jibo-lines-body34"));
        var instance34 = new Sprite(fromFrame("haring-jibo-lines-body35"));
        var instance35 = new Sprite(fromFrame("haring-jibo-lines-body36"));
        var instance36 = new Sprite(fromFrame("haring-jibo-lines-body37"));
        var instance37 = new Sprite(fromFrame("haring-jibo-lines-body38"));
        var instance38 = new Sprite(fromFrame("haring-jibo-lines-body39"));
        var instance39 = new Sprite(fromFrame("haring-jibo-lines-body40"))
            .setTransform(-84.6, -99.8);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -27.05,
                    y: 43.05
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -32.05,
                    y: 41.6
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -39.05,
                    y: 40.75
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -48.55,
                    y: 36.2
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -59.55,
                    y: 29.25
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -69.4,
                    y: 19.25
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -81.05,
                    y: -3.25
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -84.6,
                    y: -38.85
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -84.6,
                    y: -67.9
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -84.6,
                    y: -96.15
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance15, 15, 1, {
                "15": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance16, 16, 1, {
                "16": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance17, 17, 1, {
                "17": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance18, 18, 1, {
                "18": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance19, 19, 1, {
                "19": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance20, 20, 1, {
                "20": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance21, 21, 1, {
                "21": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance22, 22, 1, {
                "22": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance23, 23, 1, {
                "23": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance24, 24, 1, {
                "24": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance25, 25, 1, {
                "25": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance26, 26, 1, {
                "26": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance27, 27, 1, {
                "27": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance28, 28, 1, {
                "28": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance29, 29, 1, {
                "29": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance30, 30, 1, {
                "30": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance31, 31, 1, {
                "31": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance32, 32, 1, {
                "32": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance33, 33, 1, {
                "33": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance34, 34, 1, {
                "34": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance35, 35, 1, {
                "35": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance36, 36, 1, {
                "36": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance37, 37, 1, {
                "37": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance38, 38, 1, {
                "38": {
                    x: -84.6,
                    y: -99.8
                }
            })
            .addTimedChild(instance39, 39, 166);
    });

    var Graphic27 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 205, loop: false });
        var instance1 = new Graphic26(MovieClip.SYNCHED);
        var instance2 = new Graphic25(MovieClip.SYNCHED)
            .setTransform(-3.75, -25.05);
        this.addTimedChild(instance1)
            .addTimedChild(instance2, 44, 161);
    });

    var Graphic28 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 127, loop: false });
        var instance1 = new Sprite(fromFrame("haring-fill-black2"));
        var instance2 = new Sprite(fromFrame("haring-fill-black3"));
        var instance3 = new Sprite(fromFrame("haring-fill-black4"));
        var instance4 = new Sprite(fromFrame("haring-fill-black5"));
        var instance5 = new Sprite(fromFrame("haring-fill-black6"));
        var instance6 = new Sprite(fromFrame("haring-fill-black7"));
        var instance7 = new Sprite(fromFrame("haring-fill-black8"));
        var instance8 = new Sprite(fromFrame("haring-fill-black9"));
        var instance9 = new Sprite(fromFrame("haring-fill-black10"));
        var instance10 = new Sprite(fromFrame("haring-fill-black11"));
        var instance11 = new Sprite(fromFrame("haring-fill-black12"));
        var instance12 = new Sprite(fromFrame("haring-fill-black13"));
        var instance13 = new Sprite(fromFrame("haring-fill-black14"));
        var instance14 = new Sprite(fromFrame("haring-fill-black15"));
        var instance15 = new Sprite(fromFrame("haring-fill-black16"));
        var instance16 = new Sprite(fromFrame("haring-fill-black17"));
        var instance17 = new Sprite(fromFrame("haring-fill-black18"));
        var instance18 = new Sprite(fromFrame("haring-fill-black19"));
        var instance19 = new Sprite(fromFrame("haring-fill-black20"));
        var instance20 = new Sprite(fromFrame("haring-fill-black21"));
        var instance21 = new Sprite(fromFrame("haring-fill-black22"))
            .setTransform(-51.9, -46.55);
        this.addTimedChild(instance1, 1, 1, {
                "1": {
                    x: -46.55,
                    y: -46.35
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -49.5,
                    y: -46.55
                }
            })
            .addTimedChild(instance3, 3, 1, {
                "3": {
                    x: -50.55,
                    y: -46.55
                }
            })
            .addTimedChild(instance4, 4, 1, {
                "4": {
                    x: -51.05,
                    y: -46.55
                }
            })
            .addTimedChild(instance5, 5, 1, {
                "5": {
                    x: -51.8,
                    y: -46.55
                }
            })
            .addTimedChild(instance6, 6, 1, {
                "6": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance7, 7, 1, {
                "7": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance8, 8, 1, {
                "8": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance9, 9, 1, {
                "9": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance10, 10, 1, {
                "10": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance11, 11, 1, {
                "11": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance12, 12, 1, {
                "12": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance13, 13, 1, {
                "13": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance14, 14, 1, {
                "14": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance15, 15, 1, {
                "15": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance16, 16, 1, {
                "16": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance17, 17, 1, {
                "17": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance18, 18, 1, {
                "18": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance19, 19, 1, {
                "19": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance20, 20, 1, {
                "20": {
                    x: -51.9,
                    y: -46.55
                }
            })
            .addTimedChild(instance21, 21, 106);
    });

    var Graphic29 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 143, loop: false });
        var instance1 = new Sprite(fromFrame("haring-fill-white2"));
        var instance2 = new Sprite(fromFrame("haring-fill-white4"));
        var instance3 = new Sprite(fromFrame("haring-fill-white6"));
        var instance4 = new Sprite(fromFrame("haring-fill-white8"));
        var instance5 = new Sprite(fromFrame("haring-fill-white10"));
        var instance6 = new Sprite(fromFrame("haring-fill-white12"));
        var instance7 = new Sprite(fromFrame("haring-fill-white14"));
        var instance8 = new Sprite(fromFrame("haring-fill-white16"));
        var instance9 = new Sprite(fromFrame("haring-fill-white18"));
        var instance10 = new Sprite(fromFrame("haring-fill-white20"));
        var instance11 = new Sprite(fromFrame("haring-fill-white22"));
        var instance12 = new Sprite(fromFrame("haring-fill-white24"));
        var instance13 = new Sprite(fromFrame("haring-fill-white26"));
        var instance14 = new Sprite(fromFrame("haring-fill-white28"));
        var instance15 = new Sprite(fromFrame("haring-fill-white30"));
        var instance16 = new Sprite(fromFrame("haring-fill-white32"));
        var instance17 = new Sprite(fromFrame("haring-fill-white34"));
        var instance18 = new Sprite(fromFrame("haring-fill-white36"));
        var instance19 = new Sprite(fromFrame("haring-fill-white38"));
        var instance20 = new Sprite(fromFrame("haring-fill-white40"));
        var instance21 = new Sprite(fromFrame("haring-fill-white42"));
        var instance22 = new Sprite(fromFrame("haring-fill-white44"))
            .setTransform(-77, -95.7);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -32.85,
                    y: -95.7
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -48.55,
                    y: -95.7
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -55.35,
                    y: -95.7
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -61.45,
                    y: -95.7
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -71.75,
                    y: -95.7
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -71.7,
                    y: -95.7
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -76.85,
                    y: -95.7
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance11, 21, 2, {
                "21": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance12, 23, 2, {
                "23": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance13, 25, 2, {
                "25": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance14, 27, 2, {
                "27": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance15, 29, 2, {
                "29": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance16, 31, 2, {
                "31": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance17, 33, 2, {
                "33": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance18, 35, 2, {
                "35": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance19, 37, 2, {
                "37": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance20, 39, 2, {
                "39": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance21, 41, 2, {
                "41": {
                    x: -77,
                    y: -95.7
                }
            })
            .addTimedChild(instance22, 43, 100);
    });

    var Graphic30 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 117, loop: false });
        var instance1 = new Sprite(fromFrame("haring-specks2"));
        var instance2 = new Sprite(fromFrame("haring-specks4"));
        var instance3 = new Sprite(fromFrame("haring-specks6"));
        var instance4 = new Sprite(fromFrame("haring-specks8"));
        var instance5 = new Sprite(fromFrame("haring-specks10"));
        var instance6 = new Sprite(fromFrame("haring-specks11"));
        var instance7 = new Sprite(fromFrame("haring-specks12"));
        var instance8 = new Sprite(fromFrame("haring-specks13"));
        var instance9 = new Sprite(fromFrame("haring-specks14"));
        var instance10 = new Sprite(fromFrame("haring-specks15"));
        var instance11 = new Sprite(fromFrame("haring-specks16"));
        var instance12 = new Sprite(fromFrame("haring-specks17"));
        var instance13 = new Sprite(fromFrame("haring-specks18"));
        var instance14 = new Sprite(fromFrame("haring-specks19"));
        var instance15 = new Sprite(fromFrame("haring-specks20"));
        var instance16 = new Sprite(fromFrame("haring-specks21"));
        var instance17 = new Sprite(fromFrame("haring-specks22"));
        var instance18 = new Sprite(fromFrame("haring-specks23"));
        var instance19 = new Sprite(fromFrame("haring-specks24"));
        var instance20 = new Sprite(fromFrame("haring-specks25"));
        var instance21 = new Sprite(fromFrame("haring-specks27"));
        var instance22 = new Sprite(fromFrame("haring-specks29"));
        var instance23 = new Sprite(fromFrame("haring-specks31"))
            .setTransform(-181.85, -13.5);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -144.15,
                    y: -13.5
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -163.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance5, 9, 1, {
                "9": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance6, 10, 1, {
                "10": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance7, 11, 1, {
                "11": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance8, 12, 1, {
                "12": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance9, 13, 1, {
                "13": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance10, 14, 1, {
                "14": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance11, 15, 1, {
                "15": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance12, 16, 1, {
                "16": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance13, 17, 1, {
                "17": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance14, 18, 1, {
                "18": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance15, 19, 1, {
                "19": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance16, 20, 1, {
                "20": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance17, 21, 1, {
                "21": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance18, 22, 1, {
                "22": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance19, 23, 1, {
                "23": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance20, 24, 2, {
                "24": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance21, 26, 2, {
                "26": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance22, 28, 2, {
                "28": {
                    x: -181.85,
                    y: -13.5
                }
            })
            .addTimedChild(instance23, 30, 87);
    });

    var Graphic31 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 152, loop: false });
        var instance1 = new Sprite(fromFrame("haring-back-fill2"));
        var instance2 = new Sprite(fromFrame("haring-back-fill4"));
        var instance3 = new Sprite(fromFrame("haring-back-fill6"));
        var instance4 = new Sprite(fromFrame("haring-back-fill8"));
        var instance5 = new Sprite(fromFrame("haring-back-fill10"));
        var instance6 = new Sprite(fromFrame("haring-back-fill12"));
        var instance7 = new Sprite(fromFrame("haring-back-fill14"));
        var instance8 = new Sprite(fromFrame("haring-back-fill16"));
        var instance9 = new Sprite(fromFrame("haring-back-fill18"));
        var instance10 = new Sprite(fromFrame("haring-back-fill20"));
        var instance11 = new Sprite(fromFrame("haring-back-fill22"));
        var instance12 = new Sprite(fromFrame("haring-back-fill24"));
        var instance13 = new Sprite(fromFrame("haring-back-fill26"));
        var instance14 = new Sprite(fromFrame("haring-back-fill28"));
        var instance15 = new Sprite(fromFrame("haring-back-fill30"));
        var instance16 = new Sprite(fromFrame("haring-back-fill32"));
        var instance17 = new Sprite(fromFrame("haring-back-fill34"));
        var instance18 = new Sprite(fromFrame("haring-back-fill36"));
        var instance19 = new Sprite(fromFrame("haring-back-fill38"));
        var instance20 = new Sprite(fromFrame("haring-back-fill40"));
        var instance21 = new Sprite(fromFrame("haring-back-fill42"));
        var instance22 = new Sprite(fromFrame("haring-back-fill44"));
        var instance23 = new Sprite(fromFrame("haring-back-fill45"))
            .setTransform(-179.4, -19.45);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance11, 21, 2, {
                "21": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance12, 23, 2, {
                "23": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance13, 25, 2, {
                "25": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance14, 27, 2, {
                "27": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance15, 29, 2, {
                "29": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance16, 31, 2, {
                "31": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance17, 33, 2, {
                "33": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance18, 35, 2, {
                "35": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance19, 37, 2, {
                "37": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance20, 39, 2, {
                "39": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance21, 41, 2, {
                "41": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance22, 43, 1, {
                "43": {
                    x: -179.4,
                    y: -19.45
                }
            })
            .addTimedChild(instance23, 44, 108);
    });

    var Graphic32 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 170, loop: false });
        var instance1 = new Sprite(fromFrame("haring-back-line2"));
        var instance2 = new Sprite(fromFrame("haring-back-line4"));
        var instance3 = new Sprite(fromFrame("haring-back-line6"));
        var instance4 = new Sprite(fromFrame("haring-back-line7"));
        var instance5 = new Sprite(fromFrame("haring-back-line8"));
        var instance6 = new Sprite(fromFrame("haring-back-line9"));
        var instance7 = new Sprite(fromFrame("haring-back-line10"));
        var instance8 = new Sprite(fromFrame("haring-back-line11"));
        var instance9 = new Sprite(fromFrame("haring-back-line12"));
        var instance10 = new Sprite(fromFrame("haring-back-line13"));
        var instance11 = new Sprite(fromFrame("haring-back-line14"));
        var instance12 = new Sprite(fromFrame("haring-back-line15"));
        var instance13 = new Sprite(fromFrame("haring-back-line16"));
        var instance14 = new Sprite(fromFrame("haring-back-line17"));
        var instance15 = new Sprite(fromFrame("haring-back-line19"));
        var instance16 = new Sprite(fromFrame("haring-back-line21"))
            .setTransform(-181.45, -5.5);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -181.45,
                    y: -4.4
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -181.45,
                    y: -4.7
                }
            })
            .addTimedChild(instance3, 5, 1, {
                "5": {
                    x: -181.45,
                    y: -5.1
                }
            })
            .addTimedChild(instance4, 6, 1, {
                "6": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance5, 7, 1, {
                "7": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance6, 8, 1, {
                "8": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance7, 9, 1, {
                "9": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance8, 10, 1, {
                "10": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance9, 11, 1, {
                "11": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance10, 12, 1, {
                "12": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance11, 13, 1, {
                "13": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance12, 14, 1, {
                "14": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance13, 15, 1, {
                "15": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance14, 16, 2, {
                "16": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance15, 18, 2, {
                "18": {
                    x: -181.45,
                    y: -5.5
                }
            })
            .addTimedChild(instance16, 20, 150);
    });

    var Graphic33 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 216, loop: false });
        var instance1 = new Sprite(fromFrame("haring-bg2"));
        var instance2 = new Sprite(fromFrame("haring-bg4"));
        var instance3 = new Sprite(fromFrame("haring-bg6"));
        var instance4 = new Sprite(fromFrame("haring-bg8"));
        var instance5 = new Sprite(fromFrame("haring-bg10"));
        var instance6 = new Sprite(fromFrame("haring-bg12"));
        var instance7 = new Sprite(fromFrame("haring-bg14"));
        var instance8 = new Sprite(fromFrame("haring-bg16"));
        var instance9 = new Sprite(fromFrame("haring-bg18"));
        var instance10 = new Sprite(fromFrame("haring-bg20"));
        var instance11 = new Sprite(fromFrame("haring-bg22"));
        var instance12 = new Sprite(fromFrame("haring-bg24"));
        var instance13 = new Sprite(fromFrame("haring-bg26"));
        var instance14 = new Sprite(fromFrame("haring-bg28"));
        var instance15 = new Sprite(fromFrame("haring-bg30"));
        var instance16 = new Sprite(fromFrame("haring-bg32"));
        var instance17 = new Sprite(fromFrame("haring-bg34"));
        var instance18 = new Sprite(fromFrame("haring-bg36"));
        var instance19 = new Sprite(fromFrame("haring-bg38"));
        var instance20 = new Sprite(fromFrame("haring-bg40"));
        var instance21 = new Sprite(fromFrame("haring-bg42"));
        var instance22 = new Sprite(fromFrame("haring-bg44"));
        var instance23 = new Sprite(fromFrame("haring-bg46"));
        var instance24 = new Sprite(fromFrame("haring-bg48"));
        var instance25 = new Sprite(fromFrame("haring-bg50"));
        var instance26 = new Sprite(fromFrame("haring-bg52"));
        var instance27 = new Sprite(fromFrame("haring-bg54"));
        var instance28 = new Sprite(fromFrame("haring-bg56"));
        var instance29 = new Sprite(fromFrame("haring-bg58"));
        var instance30 = new Sprite(fromFrame("haring-bg60"));
        var instance31 = new Sprite(fromFrame("haring-bg62"))
            .setTransform(-179, -121.3);
        this.addTimedChild(instance1, 1, 2, {
                "1": {
                    x: -179,
                    y: -120.15
                }
            })
            .addTimedChild(instance2, 3, 2, {
                "3": {
                    x: -179,
                    y: -120.15
                }
            })
            .addTimedChild(instance3, 5, 2, {
                "5": {
                    x: -179,
                    y: -120.2
                }
            })
            .addTimedChild(instance4, 7, 2, {
                "7": {
                    x: -179,
                    y: -120.2
                }
            })
            .addTimedChild(instance5, 9, 2, {
                "9": {
                    x: -179,
                    y: -120.2
                }
            })
            .addTimedChild(instance6, 11, 2, {
                "11": {
                    x: -179,
                    y: -120.2
                }
            })
            .addTimedChild(instance7, 13, 2, {
                "13": {
                    x: -179,
                    y: -120.2
                }
            })
            .addTimedChild(instance8, 15, 2, {
                "15": {
                    x: -179,
                    y: -120.2
                }
            })
            .addTimedChild(instance9, 17, 2, {
                "17": {
                    x: -179,
                    y: -120.6
                }
            })
            .addTimedChild(instance10, 19, 2, {
                "19": {
                    x: -179,
                    y: -120.9
                }
            })
            .addTimedChild(instance11, 21, 2, {
                "21": {
                    x: -179,
                    y: -120.95
                }
            })
            .addTimedChild(instance12, 23, 2, {
                "23": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance13, 25, 2, {
                "25": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance14, 27, 2, {
                "27": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance15, 29, 2, {
                "29": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance16, 31, 2, {
                "31": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance17, 33, 2, {
                "33": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance18, 35, 2, {
                "35": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance19, 37, 2, {
                "37": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance20, 39, 2, {
                "39": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance21, 41, 2, {
                "41": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance22, 43, 2, {
                "43": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance23, 45, 2, {
                "45": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance24, 47, 2, {
                "47": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance25, 49, 2, {
                "49": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance26, 51, 2, {
                "51": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance27, 53, 2, {
                "53": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance28, 55, 2, {
                "55": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance29, 57, 2, {
                "57": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance30, 59, 2, {
                "59": {
                    x: -179,
                    y: -121.3
                }
            })
            .addTimedChild(instance31, 61, 155);
    });

    var Graphic34 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 243, loop: false });
        var instance2 = new Graphic33(MovieClip.SYNCHED)
            .setTransform(-0.3, 16.7);
        var instance4 = new Graphic32(MovieClip.SYNCHED)
            .setTransform(1.55, 100.45);
        var instance5 = new Graphic31(MovieClip.SYNCHED)
            .setTransform(2.05, 120.05);
        var instance8 = new Graphic30(MovieClip.SYNCHED)
            .setTransform(0.85, 119.75);
        var instance6 = new Graphic29(MovieClip.SYNCHED)
            .setTransform(-7.95, 41.9);
        var instance7 = new Graphic28(MovieClip.SYNCHED)
            .setTransform(-11.85, 12.65);
        var instance3 = new Graphic27(MovieClip.SYNCHED)
            .setTransform(-7.55, 37.75);
        var instance9 = new Graphic24(MovieClip.SYNCHED)
            .setTransform(-9.8, -0.05);
        var instance10 = new Graphic23(MovieClip.SYNCHED)
            .setTransform(-8.2, -7.8);
        var instance1 = new Graphic5(MovieClip.SYNCHED)
            .setTransform(-0.05, 16.45);
        this.addTimedChild(instance2, 27, 216)
            .addTimedChild(instance4, 73, 170)
            .addTimedChild(instance5, 91, 152)
            .addTimedChild(instance8, 126, 117)
            .addTimedChild(instance6, 100, 143)
            .addTimedChild(instance7, 116, 127)
            .addTimedChild(instance3, 38, 205)
            .addTimedChild(instance9, 133, 110)
            .addTimedChild(instance10, 138, 105)
            .addTimedChild(instance1);
    });

    var Graphic35 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 263, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-blank1"))
            .setTransform(-232.5, -161.65);
        this.addTimedChild(instance1);
    });

    var Graphic36 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 241, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-support1"))
            .setTransform(-15.9, -28.8);
        this.addTimedChild(instance1);
    });

    var Graphic37 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 241, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-leg1"))
            .setTransform(-41.9, -74.65);
        this.addTimedChild(instance1);
    });

    var Graphic38 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 241, loop: false });
        var instance1 = new Sprite(fromFrame("canvas-leg1"))
            .setTransform(-41.9, -74.65);
        this.addTimedChild(instance1);
    });

    var Graphic39 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 263, loop: false });
        var instance6 = new Graphic38(MovieClip.SYNCHED);
        var instance5 = new Graphic37(MovieClip.SYNCHED);
        var instance4 = new Graphic36(MovieClip.SYNCHED);
        var instance1 = new Graphic35(MovieClip.SYNCHED);
        var instance8 = new Graphic34(MovieClip.SYNCHED);
        var instance7 = new Graphic4(MovieClip.SYNCHED);
        var instance3 = new Graphic3(MovieClip.SYNCHED);
        var instance2 = new Graphic2(MovieClip.SYNCHED);
        this.addTimedChild(instance6, 8, 241, {
                "8": {
                    x: 66,
                    y: 61.45,
                    ky: 3.142
                },
                "9": {
                    x: 69.15,
                    y: 71.15
                },
                "10": {
                    x: 79.3,
                    y: 102.55
                },
                "11": {
                    x: 94.65,
                    y: 149.95
                },
                "12": {
                    x: 107.55,
                    y: 189.75
                },
                "13": {
                    x: 112,
                    y: 203.45
                },
                "14": {
                    x: 111.85,
                    y: 202.95
                },
                "15": {
                    x: 111.4,
                    y: 201.4
                },
                "16": {
                    x: 110.75,
                    y: 199.1
                },
                "17": {
                    x: 110.2,
                    y: 197.1
                },
                "18": {
                    x: 110,
                    y: 196.45
                },
                "238": {
                    x: 110.1,
                    y: 196.8
                },
                "239": {
                    x: 110.45,
                    y: 198
                },
                "240": {
                    x: 111,
                    y: 200
                },
                "241": {
                    x: 111.55,
                    y: 201.95
                },
                "242": {
                    x: 111.9,
                    y: 203.1
                },
                "243": {
                    x: 112,
                    y: 203.45
                },
                "244": {
                    x: 108.55,
                    y: 192.8
                },
                "245": {
                    x: 97,
                    y: 157.1
                },
                "246": {
                    x: 80.65,
                    y: 106.75
                },
                "247": {
                    x: 69.35,
                    y: 71.8
                },
                "248": {
                    x: 66,
                    y: 61.45
                }
            })
            .addTimedChild(instance5, 8, 241, {
                "8": {
                    x: -62.9,
                    y: 62.45
                },
                "9": {
                    x: -66.1,
                    y: 72.15
                },
                "10": {
                    x: -76.5,
                    y: 103.55
                },
                "11": {
                    x: -92.2,
                    y: 150.95
                },
                "12": {
                    x: -105.35,
                    y: 190.75
                },
                "13": {
                    x: -109.9,
                    y: 204.45
                },
                "14": {
                    x: -109.7,
                    y: 203.9
                },
                "15": {
                    x: -109.05,
                    y: 202.15
                },
                "16": {
                    x: -108.05,
                    y: 199.45
                },
                "17": {
                    x: -107.2,
                    y: 197.2
                },
                "18": {
                    x: -106.9,
                    y: 196.45
                },
                "238": {
                    x: -107.05,
                    y: 196.85
                },
                "239": {
                    x: -107.55,
                    y: 198.25
                },
                "240": {
                    x: -108.4,
                    y: 200.5
                },
                "241": {
                    x: -109.25,
                    y: 202.7
                },
                "242": {
                    x: -109.75,
                    y: 204.05
                },
                "243": {
                    x: -109.9,
                    y: 204.45
                },
                "244": {
                    x: -106.4,
                    y: 193.8
                },
                "245": {
                    x: -94.55,
                    y: 158.1
                },
                "246": {
                    x: -77.9,
                    y: 107.75
                },
                "247": {
                    x: -66.3,
                    y: 72.8
                },
                "248": {
                    x: -62.9,
                    y: 62.45
                }
            })
            .addTimedChild(instance4, 8, 241, {
                "8": {
                    y: -185.35
                },
                "9": {
                    y: -189.65
                },
                "10": {
                    y: -203.6
                },
                "11": {
                    y: -224.6
                },
                "12": {
                    y: -242.3
                },
                "13": {
                    y: -248.35
                },
                "14": {
                    y: -247.95
                },
                "15": {
                    y: -246.6
                },
                "16": {
                    y: -244.6
                },
                "17": {
                    y: -242.95
                },
                "18": {
                    y: -242.35
                },
                "238": {
                    y: -242.85
                },
                "239": {
                    y: -244.6
                },
                "240": {
                    y: -247.4
                },
                "241": {
                    y: -250.2
                },
                "242": {
                    y: -251.85
                },
                "243": {
                    y: -252.35
                },
                "244": {
                    y: -247.35
                },
                "245": {
                    y: -230.5
                },
                "246": {
                    y: -206.7
                },
                "247": {
                    y: -190.25
                },
                "248": {
                    y: -185.35
                }
            })
            .addTimedChild(instance1, 0, 263, {
                "0": {
                    x: 0,
                    y: -43.9,
                    r: 0.021,
                    c: [
                        0,
                        1,
                        0,
                        1,
                        0,
                        1
                    ]
                },
                "1": {
                    x: 0.005,
                    y: -43.897,
                    r: 0.014,
                    c: [
                        0.3,
                        0.69,
                        0.3,
                        0.69,
                        0.3,
                        0.69
                    ]
                },
                "2": {
                    x: 0.003,
                    y: -43.896,
                    r: 0.009,
                    c: [
                        0.55,
                        0.44,
                        0.55,
                        0.44,
                        0.55,
                        0.44
                    ]
                },
                "3": {
                    x: 0.013,
                    y: -43.893,
                    r: 0.005,
                    c: [
                        0.75,
                        0.25,
                        0.75,
                        0.25,
                        0.75,
                        0.25
                    ]
                },
                "4": {
                    x: -0.013,
                    y: -43.891,
                    r: 0.001,
                    c: [
                        0.89,
                        0.11,
                        0.89,
                        0.11,
                        0.89,
                        0.11
                    ]
                },
                "5": {
                    x: 0.018,
                    r: 0,
                    c: [
                        0.97,
                        0.03,
                        0.97,
                        0.03,
                        0.97,
                        0.03
                    ]
                },
                "6": {
                    x: 0,
                    y: -43.9,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                }
            })
            .addTimedChild(instance8, 20, 243, {
                "20": {
                    x: 10.55,
                    y: -56.9,
                    a: 1
                },
                "252": {
                    a: 0.88
                },
                "253": {
                    a: 0.75
                },
                "254": {
                    a: 0.63
                },
                "255": {
                    a: 0.5
                },
                "256": {
                    a: 0.38
                },
                "257": {
                    a: 0.25
                },
                "258": {
                    a: 0.13
                },
                "259": {
                    a: 0
                }
            })
            .addTimedChild(instance7, 10, 253, {
                "10": {
                    y: -196.15,
                    a: 0
                },
                "11": {
                    a: 0.11
                },
                "12": {
                    a: 0.22
                },
                "13": {
                    a: 0.33
                },
                "14": {
                    a: 0.45
                },
                "15": {
                    a: 0.55
                },
                "16": {
                    a: 0.67
                },
                "17": {
                    a: 0.78
                },
                "18": {
                    a: 0.89
                },
                "19": {
                    a: 1
                },
                "238": {
                    a: 0.9
                },
                "239": {
                    a: 0.8
                },
                "240": {
                    a: 0.7
                },
                "241": {
                    a: 0.6
                },
                "242": {
                    a: 0.5
                },
                "243": {
                    a: 0.4
                },
                "244": {
                    a: 0.3
                },
                "245": {
                    a: 0.2
                },
                "246": {
                    a: 0.1
                },
                "247": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 4, 259, {
                "4": {
                    y: -190.45,
                    a: 0
                },
                "5": {
                    y: -191.7,
                    a: 0.05
                },
                "6": {
                    y: -195.8,
                    a: 0.2
                },
                "7": {
                    y: -202.6,
                    a: 0.45
                },
                "8": {
                    y: -210.2,
                    a: 0.73
                },
                "9": {
                    y: -215.65,
                    a: 0.93
                },
                "10": {
                    y: -217.45,
                    a: 1
                },
                "11": {
                    y: -217.15
                },
                "12": {
                    y: -216.25
                },
                "13": {
                    y: -214.75
                },
                "14": {
                    y: -213.05
                },
                "15": {
                    y: -211.85
                },
                "16": {
                    y: -211.45
                },
                "248": {
                    y: -211.65,
                    a: 0.98
                },
                "249": {
                    y: -212.2,
                    a: 0.93
                },
                "250": {
                    y: -213.25,
                    a: 0.82
                },
                "251": {
                    y: -214.7,
                    a: 0.67
                },
                "252": {
                    y: -216.5,
                    a: 0.5
                },
                "253": {
                    y: -218.25,
                    a: 0.32
                },
                "254": {
                    y: -219.7,
                    a: 0.17
                },
                "255": {
                    y: -220.7,
                    a: 0.07
                },
                "256": {
                    y: -221.3,
                    a: 0.02
                },
                "257": {
                    y: -221.45,
                    a: 0
                }
            })
            .addTimedChild(instance2, 4, 259, {
                "4": {
                    y: 102.45,
                    a: 0
                },
                "5": {
                    y: 103.85,
                    a: 0.05
                },
                "6": {
                    y: 108.4,
                    a: 0.2
                },
                "7": {
                    y: 115.95,
                    a: 0.45
                },
                "8": {
                    y: 124.4,
                    a: 0.73
                },
                "9": {
                    y: 130.4,
                    a: 0.93
                },
                "10": {
                    y: 132.4,
                    a: 1
                },
                "11": {
                    y: 132.2
                },
                "12": {
                    y: 131.6
                },
                "13": {
                    y: 130.6
                },
                "14": {
                    y: 129.45
                },
                "15": {
                    y: 128.65
                },
                "16": {
                    y: 128.4
                },
                "248": {
                    y: 128.6,
                    a: 0.98
                },
                "249": {
                    y: 129.15,
                    a: 0.93
                },
                "250": {
                    y: 130.2,
                    a: 0.82
                },
                "251": {
                    y: 131.65,
                    a: 0.67
                },
                "252": {
                    y: 133.45,
                    a: 0.5
                },
                "253": {
                    y: 135.2,
                    a: 0.32
                },
                "254": {
                    y: 136.65,
                    a: 0.17
                },
                "255": {
                    y: 137.65,
                    a: 0.07
                },
                "256": {
                    y: 138.25,
                    a: 0.02
                },
                "257": {
                    y: 138.4,
                    a: 0
                }
            });
    });

    var Graphic40 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 7, loop: false });
        var instance1 = new Sprite(fromFrame("EYE159"));
        var instance2 = new Sprite(fromFrame("EYE160"));
        var instance3 = new Sprite(fromFrame("EYE161"));
        var instance4 = new Sprite(fromFrame("EYE162"));
        var instance5 = new Sprite(fromFrame("EYE163"));
        var instance6 = new Sprite(fromFrame("EYE164"))
            .setTransform(-278.8, -277.7);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -264.1,
                    y: -236.5
                }
            })
            .addTimedChild(instance2, 1, 1, {
                "1": {
                    x: -270,
                    y: -251.3
                }
            })
            .addTimedChild(instance3, 2, 1, {
                "2": {
                    x: -275.85,
                    y: -268.85
                }
            })
            .addTimedChild(instance4, 3, 1, {
                "3": {
                    x: -277.35,
                    y: -273.25
                }
            })
            .addTimedChild(instance5, 4, 1, {
                "4": {
                    x: -278.85,
                    y: -277.75
                }
            })
            .addTimedChild(instance6, 5, 2);
    });

    lib.drawing_haring = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 275,
            framerate: 30,
            loop: false
        });
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        var instance3 = new Graphic40(MovieClip.SYNCHED)
            .setTransform(640.3, 359.3, 0.898, 0.898);
        var instance2 = new Graphic39(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 6, {
                "0": {
                    x: 640.3,
                    y: 359.3,
                    sx: 0.898,
                    sy: 0.898
                }
            })
            .addTimedChild(instance3, 268, 7)
            .addTimedChild(instance2, 6, 263, {
                "6": {
                    x: 640.45,
                    y: 359.95,
                    sx: 1,
                    sy: 1,
                    kx: 0,
                    ky: 0,
                    r: 0,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "258": {
                    x: 640.6,
                    y: 360,
                    r: -0.003
                },
                "259": {
                    x: 640.619,
                    y: 359.982,
                    r: -0.005
                },
                "260": {
                    x: 640.851,
                    y: 360.003,
                    r: -0.009
                },
                "261": {
                    x: 641.439,
                    y: 360.021,
                    r: -0.022
                },
                "262": {
                    x: 642.1,
                    y: 360.104,
                    r: -0.036
                },
                "263": {
                    x: 642.8,
                    y: 360.1,
                    r: -0.051
                },
                "264": {
                    x: 631.94,
                    y: 358.342,
                    sx: 0.999,
                    sy: 0.999,
                    r: 0.189,
                    c: [
                        0.96,
                        0.04,
                        0.96,
                        0.04,
                        0.96,
                        0.04
                    ]
                },
                "265": {
                    x: 602.636,
                    y: 333.341,
                    r: 1.05,
                    c: [
                        0.83,
                        0.17,
                        0.83,
                        0.17,
                        0.83,
                        0.17
                    ]
                },
                "266": {
                    x: 626.319,
                    y: 271.568,
                    kx: 3.632,
                    ky: 2.651,
                    r: 0,
                    c: [
                        0.58,
                        0.42,
                        0.58,
                        0.42,
                        0.58,
                        0.42
                    ]
                },
                "267": {
                    x: 690.582,
                    y: 316.559,
                    sx: 1,
                    sy: 1,
                    kx: 0,
                    ky: 0,
                    r: -1.602,
                    c: [
                        0.26,
                        0.74,
                        0.26,
                        0.74,
                        0.26,
                        0.74
                    ]
                },
                "268": {
                    x: 638.5,
                    y: 359.65,
                    r: 0.044,
                    c: [
                        0,
                        1,
                        0,
                        1,
                        0,
                        1
                    ]
                }
            });
    });

    lib.drawing_haring.assets = {
        "drawing_haring_atlas_1": "images/drawing_haring_atlas_1.json",
        "drawing_haring_atlas_2": "images/drawing_haring_atlas_2.json",
        "drawing_haring_atlas_3": "images/drawing_haring_atlas_3.json",
        "drawing_haring_atlas_4": "images/drawing_haring_atlas_4.json",
        "drawing_haring_atlas_5": "images/drawing_haring_atlas_5.json",
        "drawing_haring_atlas_6": "images/drawing_haring_atlas_6.json",
        "drawing_haring_atlas_7": "images/drawing_haring_atlas_7.json",
        "drawing_haring_atlas_8": "images/drawing_haring_atlas_8.json",
        "drawing_haring_atlas_9": "images/drawing_haring_atlas_9.json",
        "drawing_haring_atlas_10": "images/drawing_haring_atlas_10.json",
        "drawing_haring_atlas_11": "images/drawing_haring_atlas_11.json",
        "drawing_haring_atlas_12": "images/drawing_haring_atlas_12.json",
        "drawing_haring_atlas_13": "images/drawing_haring_atlas_13.json",
        "drawing_haring_atlas_14": "images/drawing_haring_atlas_14.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.drawing_haring,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 275,
        library: lib
    };
}
