// jibo-anim-db-animations/timelines/dream_stars.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;
    var Graphics = PIXI.Graphics;
    var shapes = PIXI.animate.ShapesCache;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 390, loop: false });
        var instance1 = new Sprite(fromFrame("fade"))
            .setTransform(-640, -360);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic2(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic4(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic6(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic8(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic10(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic12(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic14(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 21, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 21, loop: false });
        var instance1 = new Graphic16(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 21, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic18(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic20(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic22(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic24 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic25 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic24(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic26 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic27 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic26(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic28 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic29 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic28(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic30 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic31 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic30(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic32 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic33 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic32(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic34 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 25, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic35 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 25, loop: false });
        var instance1 = new Graphic34(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 25, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "9": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "10": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "11": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "12": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "13": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "14": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "15": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "16": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "17": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "18": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "19": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "20": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "21": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic36 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic37 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic36(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic38 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic39 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic38(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic40 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic41 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic40(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic42 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic43 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic42(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic44 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("star2"))
            .setTransform(-8.75, -8.9);
        this.addTimedChild(instance1);
    });

    var Graphic45 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Graphic44(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 23, {
            "0": {
                sx: 0.381,
                sy: 0.381,
                r: -0.427,
                a: 1
            },
            "1": {
                sx: 0.441,
                sy: 0.441,
                r: -0.385
            },
            "2": {
                sx: 0.502,
                sy: 0.502,
                r: -0.345
            },
            "3": {
                sx: 0.562,
                sy: 0.562,
                r: -0.306
            },
            "4": {
                sx: 0.623,
                sy: 0.623,
                r: -0.263
            },
            "5": {
                sx: 0.684,
                sy: 0.684,
                r: -0.223
            },
            "6": {
                sx: 0.744,
                sy: 0.744,
                r: -0.184
            },
            "7": {
                sx: 0.805,
                sy: 0.805,
                r: -0.141
            },
            "8": {
                sx: 0.866,
                sy: 0.866,
                r: -0.101
            },
            "9": {
                sx: 0.926,
                sy: 0.926,
                r: -0.061
            },
            "10": {
                sx: 0.987,
                sy: 0.987,
                r: -0.022
            },
            "11": {
                sx: 1.026,
                sy: 1.026,
                r: 0.001,
                a: 0.89
            },
            "12": {
                sx: 1.064,
                sy: 1.064,
                r: 0.027,
                a: 0.78
            },
            "13": {
                sx: 1.102,
                sy: 1.102,
                r: 0.053,
                a: 0.67
            },
            "14": {
                sx: 1.141,
                sy: 1.141,
                r: 0.079,
                a: 0.55
            },
            "15": {
                sx: 1.179,
                sy: 1.179,
                r: 0.105,
                a: 0.45
            },
            "16": {
                sx: 1.217,
                sy: 1.217,
                r: 0.131,
                a: 0.33
            },
            "17": {
                sx: 1.256,
                sy: 1.256,
                r: 0.157,
                a: 0.22
            },
            "18": {
                sx: 1.294,
                sy: 1.294,
                r: 0.18,
                a: 0.11
            },
            "19": {
                sx: 1.333,
                sy: 1.333,
                r: 0.209,
                a: 0
            }
        });
    });

    var Graphic46 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 390, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.dream_stars[0]);
        this.addTimedChild(instance1);
    });

    var Graphic47 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 390, loop: false });
        var instance1 = new Graphic46(MovieClip.SYNCHED);
        var instance2 = new Graphic35(MovieClip.SYNCHED);
        var instance3 = new Graphic19(MovieClip.SYNCHED);
        var instance4 = new Graphic3(MovieClip.SYNCHED);
        var instance6 = new Graphic37(MovieClip.SYNCHED);
        var instance5 = new Graphic21(MovieClip.SYNCHED);
        var instance7 = new Graphic5(MovieClip.SYNCHED);
        var instance10 = new Graphic39(MovieClip.SYNCHED);
        var instance8 = new Graphic23(MovieClip.SYNCHED);
        var instance9 = new Graphic7(MovieClip.SYNCHED);
        var instance11 = new Graphic25(MovieClip.SYNCHED);
        var instance12 = new Graphic9(MovieClip.SYNCHED);
        var instance13 = new Graphic41(MovieClip.SYNCHED);
        var instance14 = new Graphic27(MovieClip.SYNCHED);
        var instance15 = new Graphic11(MovieClip.SYNCHED);
        var instance17 = new Graphic43(MovieClip.SYNCHED);
        var instance16 = new Graphic29(MovieClip.SYNCHED);
        var instance18 = new Graphic13(MovieClip.SYNCHED);
        var instance21 = new Graphic45(MovieClip.SYNCHED);
        var instance19 = new Graphic31(MovieClip.SYNCHED);
        var instance20 = new Graphic15(MovieClip.SYNCHED);
        var instance22 = new Graphic33(MovieClip.SYNCHED);
        var instance23 = new Graphic17(MovieClip.SYNCHED)
            .setTransform(149.45, -67.05, 1.14, 1.14, 0, 3.657, 2.626);
        this.addTimedChild(instance1, 0, 390, {
                "0": {
                    x: -76.45,
                    y: 28.7
                }
            })
            .addTimedChild(instance2, 22, 25, {
                "22": {
                    x: -427.45,
                    y: 272.95,
                    sx: 2.017,
                    sy: 2.017,
                    r: 1.37
                }
            })
            .addTimedChild(instance3, 33, 23, {
                "33": {
                    x: -676.7,
                    y: 314.45,
                    r: 1.37
                }
            })
            .addTimedChild(instance4, 48, 23, {
                "48": {
                    x: -155.15,
                    y: 211.35
                }
            })
            .addTimedChild(instance6, 82, 23, {
                "82": {
                    x: -165.95,
                    y: 8.3,
                    sx: 2.377,
                    sy: 2.377,
                    r: 1.37
                }
            })
            .addTimedChild(instance5, 71, 23, {
                "71": {
                    x: -204.4,
                    y: -270.15,
                    sx: 1.234,
                    sy: 1.234,
                    r: 1.37
                }
            })
            .addTimedChild(instance7, 94, 23, {
                "94": {
                    x: 400.15,
                    y: -270.2,
                    sx: 0.767,
                    sy: 0.767,
                    r: 1.37
                }
            })
            .addTimedChild(instance10, 136, 23, {
                "136": {
                    x: -18.25,
                    y: 534.45,
                    sx: 2.377,
                    sy: 2.377,
                    r: 1.37
                }
            })
            .addTimedChild(instance8, 119, 23, {
                "119": {
                    x: -804.4,
                    y: 16,
                    r: 1.37
                }
            })
            .addTimedChild(instance9, 122, 23, {
                "122": {
                    x: 175.6,
                    y: -276.3,
                    sx: 0.698,
                    sy: 0.698,
                    kx: 3.657,
                    ky: 2.626
                }
            })
            .addTimedChild(instance11, 149, 23, {
                "149": {
                    x: -892.05,
                    y: 476.1,
                    sx: 1.385,
                    sy: 1.385,
                    kx: 3.657,
                    ky: 2.626
                }
            })
            .addTimedChild(instance12, 166, 23, {
                "166": {
                    x: 149.45,
                    y: -67.05,
                    sx: 1.14,
                    sy: 1.14,
                    kx: 3.657,
                    ky: 2.626
                }
            })
            .addTimedChild(instance13, 227, 23, {
                "227": {
                    x: -427.45,
                    y: 272.95,
                    sx: 2.017,
                    sy: 2.017,
                    r: 1.37
                }
            })
            .addTimedChild(instance14, 236, 23, {
                "236": {
                    x: -676.7,
                    y: 314.45,
                    r: 1.37
                }
            })
            .addTimedChild(instance15, 251, 23, {
                "251": {
                    x: -155.15,
                    y: 211.35
                }
            })
            .addTimedChild(instance17, 285, 23, {
                "285": {
                    x: -165.95,
                    y: 8.3,
                    sx: 2.377,
                    sy: 2.377,
                    r: 1.37
                }
            })
            .addTimedChild(instance16, 274, 23, {
                "274": {
                    x: -204.4,
                    y: -270.15,
                    sx: 1.234,
                    sy: 1.234,
                    r: 1.37
                }
            })
            .addTimedChild(instance18, 297, 23, {
                "297": {
                    x: 400.15,
                    y: -270.2,
                    sx: 0.767,
                    sy: 0.767,
                    r: 1.37
                }
            })
            .addTimedChild(instance21, 339, 23, {
                "339": {
                    x: -18.25,
                    y: 534.45,
                    sx: 2.377,
                    sy: 2.377,
                    r: 1.37
                }
            })
            .addTimedChild(instance19, 322, 23, {
                "322": {
                    x: -804.4,
                    y: 16,
                    r: 1.37
                }
            })
            .addTimedChild(instance20, 325, 23, {
                "325": {
                    x: 175.6,
                    y: -276.3,
                    sx: 0.698,
                    sy: 0.698,
                    kx: 3.657,
                    ky: 2.626
                }
            })
            .addTimedChild(instance22, 352, 23, {
                "352": {
                    x: -892.05,
                    y: 476.1,
                    sx: 1.385,
                    sy: 1.385,
                    kx: 3.657,
                    ky: 2.626
                }
            })
            .addTimedChild(instance23, 369, 21);
    });

    var Graphic48 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 390, loop: false });
        var instance1 = new Graphic47(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 390, {
            "0": {
                x: -327.3,
                y: 294.05,
                sx: 1,
                sy: 1,
                r: 0
            },
            "1": {
                x: -327.284,
                y: 294.03,
                r: -0.001
            },
            "2": {
                x: -327.304,
                y: 294.034,
                r: -0.005
            },
            "3": {
                x: -327.302,
                y: 294.009
            },
            "4": {
                x: -327.311,
                y: 294.004,
                r: -0.009
            },
            "5": {
                x: -327.27,
                y: 293.99,
                r: -0.013
            },
            "6": {
                x: -327.254,
                y: 294.02,
                r: -0.014
            },
            "7": {
                x: -327.304,
                y: 293.956,
                r: -0.018
            },
            "8": {
                x: -327.257,
                y: 293.977,
                r: -0.022
            },
            "9": {
                x: -327.241,
                y: 293.958,
                r: -0.023
            },
            "10": {
                x: -327.23,
                y: 293.975,
                r: -0.027
            },
            "11": {
                x: -327.269,
                y: 293.943,
                r: -0.031
            },
            "12": {
                x: -327.248,
                y: 293.969
            },
            "13": {
                x: -327.237,
                y: 293.927,
                r: -0.035
            },
            "14": {
                x: -327.206,
                y: 293.886,
                r: -0.039
            },
            "15": {
                x: -327.235,
                y: 293.912,
                r: -0.04
            },
            "16": {
                x: -327.204,
                y: 293.911,
                r: -0.044
            },
            "17": {
                x: -327.218,
                y: 293.906,
                r: -0.048
            },
            "18": {
                x: -327.202,
                y: 293.877,
                r: -0.049
            },
            "19": {
                x: -327.211,
                y: 293.868,
                r: -0.053
            },
            "20": {
                x: -327.215,
                y: 293.854,
                r: -0.057
            },
            "21": {
                x: -327.189,
                y: 293.875,
                r: -0.058
            },
            "22": {
                x: -327.183,
                y: 293.802,
                r: -0.062
            },
            "23": {
                x: -327.186,
                y: 293.82,
                r: -0.065
            },
            "24": {
                x: -327.21,
                y: 293.791,
                r: -0.066
            },
            "25": {
                x: -327.194,
                y: 293.809,
                r: -0.07
            },
            "26": {
                x: -327.174,
                y: 293.823,
                r: -0.074
            },
            "27": {
                x: -327.148,
                y: 293.794,
                r: -0.075
            },
            "28": {
                x: -327.131,
                y: 293.793,
                r: -0.079
            },
            "29": {
                x: -327.2,
                y: 293.76,
                r: -0.08
            },
            "30": {
                x: -327.131,
                y: 293.793,
                r: -0.079
            },
            "31": {
                x: -327.2,
                y: 293.76,
                r: -0.08
            },
            "32": {
                x: -327.169,
                y: 293.765,
                r: -0.084
            },
            "33": {
                x: -327.137,
                y: 293.71,
                r: -0.088
            },
            "34": {
                x: -327.157,
                y: 293.727
            },
            "35": {
                x: -327.12,
                y: 293.758,
                r: -0.092
            },
            "36": {
                x: -327.178,
                y: 293.695,
                r: -0.096
            },
            "37": {
                x: -327.142,
                y: 293.707,
                r: -0.097
            },
            "38": {
                x: -327.136,
                y: 293.689,
                r: -0.101
            },
            "39": {
                x: -327.135,
                y: 293.667,
                r: -0.105
            },
            "40": {
                x: -327.104,
                y: 293.674,
                r: -0.106
            },
            "41": {
                x: -327.097,
                y: 293.637,
                r: -0.11
            },
            "42": {
                x: -327.13,
                y: 293.651,
                r: -0.114
            },
            "43": {
                x: -327.139,
                y: 293.609
            },
            "44": {
                x: -327.123,
                y: 293.613,
                r: -0.118
            },
            "45": {
                x: -327.096,
                y: 293.618,
                r: -0.122
            },
            "46": {
                x: -327.114,
                y: 293.616,
                r: -0.123
            },
            "47": {
                x: -327.083,
                y: 293.567,
                r: -0.127
            },
            "48": {
                x: -327.096,
                y: 293.603,
                r: -0.131
            },
            "49": {
                x: -327.105,
                y: 293.561,
                r: -0.132
            },
            "50": {
                x: -327.113,
                y: 293.543,
                r: -0.136
            },
            "51": {
                x: -327.08,
                y: 293.565,
                r: -0.14
            },
            "52": {
                x: -327.085,
                y: 293.518
            },
            "53": {
                x: -327.083,
                y: 293.541,
                r: -0.144
            },
            "54": {
                x: -327.08,
                y: 293.505,
                r: -0.148
            },
            "55": {
                x: -327.084,
                y: 293.508,
                r: -0.149
            },
            "56": {
                x: -327.082,
                y: 293.462,
                r: -0.153
            },
            "57": {
                x: -327.04,
                y: 293.51,
                r: -0.154
            },
            "58": {
                x: -327.068,
                y: 293.515,
                sx: 0.999,
                sy: 0.999,
                r: -0.158
            },
            "59": {
                x: -327.051,
                y: 293.466,
                r: -0.162
            },
            "60": {
                x: -327.055,
                y: 293.459,
                r: -0.163
            },
            "61": {
                x: -327.037,
                y: 293.45,
                r: -0.167
            },
            "62": {
                x: -327.054,
                y: 293.436,
                r: -0.171
            },
            "63": {
                x: -327.008,
                y: 293.43
            },
            "64": {
                x: -327.021,
                y: 293.412,
                r: -0.175
            },
            "65": {
                x: -327.033,
                y: 293.385,
                r: -0.179
            },
            "66": {
                x: -327.041,
                y: 293.424,
                r: -0.18
            },
            "67": {
                x: -326.993,
                y: 293.388,
                r: -0.184
            },
            "68": {
                x: -326.995,
                y: 293.351,
                r: -0.188
            },
            "69": {
                x: -326.994,
                y: 293.341,
                r: -0.189
            },
            "70": {
                x: -326.991,
                y: 293.35,
                r: -0.193
            },
            "71": {
                x: -327.042,
                y: 293.344,
                r: -0.197
            },
            "72": {
                x: -326.991,
                y: 293.334
            },
            "73": {
                x: -327.023,
                y: 293.329,
                r: -0.201
            },
            "74": {
                x: -327.005,
                y: 293.316,
                r: -0.205
            },
            "75": {
                x: -327.003,
                y: 293.255,
                r: -0.206
            },
            "76": {
                x: -327.039,
                y: 293.277,
                r: -0.21
            },
            "77": {
                x: -326.961,
                y: 293.254,
                r: -0.214
            },
            "78": {
                x: -327.005,
                y: 293.239,
                r: -0.215
            },
            "79": {
                x: -326.976,
                y: 293.257,
                r: -0.219
            },
            "80": {
                x: -326.987,
                y: 293.225,
                r: -0.223
            },
            "81": {
                x: -326.99,
                y: 293.2
            },
            "82": {
                x: -327.001,
                y: 293.259,
                r: -0.227
            },
            "83": {
                x: -326.99,
                y: 293.239,
                r: -0.228
            },
            "84": {
                x: -326.996,
                y: 293.184,
                r: -0.232
            },
            "85": {
                x: -326.947,
                r: -0.236
            },
            "86": {
                x: -326.944,
                y: 293.154,
                r: -0.237
            },
            "87": {
                x: -326.995,
                y: 293.144,
                r: -0.241
            },
            "88": {
                x: -326.936,
                y: 293.176,
                r: -0.245
            },
            "89": {
                x: -326.97,
                y: 293.152,
                r: -0.246
            },
            "90": {
                x: -326.96,
                y: 293.174,
                r: -0.25
            },
            "91": {
                x: -327.004,
                y: 293.09,
                r: -0.253
            },
            "92": {
                x: -326.993,
                y: 293.111,
                r: -0.254
            },
            "93": {
                x: -326.973,
                y: 293.074,
                r: -0.258
            },
            "94": {
                x: -326.948,
                y: 293.083,
                r: -0.262
            },
            "95": {
                x: -326.982,
                y: 293.059,
                r: -0.263
            },
            "96": {
                x: -326.961,
                y: 293.103,
                r: -0.267
            },
            "97": {
                x: -326.981,
                y: 293.047,
                r: -0.271
            },
            "98": {
                x: -326.964,
                y: 293.063,
                r: -0.272
            },
            "99": {
                x: -326.924,
                y: 292.999,
                r: -0.276
            },
            "100": {
                x: -326.938,
                y: 293.029,
                r: -0.28
            },
            "101": {
                x: -326.975,
                y: 292.991
            },
            "102": {
                x: -326.935,
                y: 293.017,
                r: -0.284
            },
            "103": {
                x: -326.934,
                y: 292.984,
                r: -0.288
            },
            "104": {
                x: -326.917,
                y: 292.95,
                r: -0.289
            },
            "105": {
                x: -326.961,
                y: 292.963,
                r: -0.293
            },
            "106": {
                x: -326.948,
                y: 292.915,
                r: -0.294
            },
            "107": {
                x: -326.938,
                y: 292.923,
                r: -0.298
            },
            "108": {
                x: -326.931,
                y: 292.916,
                r: -0.302
            },
            "109": {
                x: -326.909,
                y: 292.928
            },
            "110": {
                x: -326.893,
                y: 292.922,
                r: -0.306
            },
            "111": {
                x: -326.936,
                y: 292.906,
                r: -0.31
            },
            "112": {
                x: -326.913,
                y: 292.909,
                r: -0.311
            },
            "113": {
                x: -326.941,
                y: 292.889,
                r: -0.315
            },
            "114": {
                x: -326.91,
                y: 292.86,
                r: -0.319
            },
            "115": {
                x: -326.942,
                y: 292.867,
                r: -0.32
            },
            "116": {
                x: -326.906,
                y: 292.834,
                r: -0.324
            },
            "117": {
                x: -326.932,
                y: 292.835,
                r: -0.328
            },
            "118": {
                x: -326.91,
                y: 292.838,
                r: -0.329
            },
            "119": {
                x: -326.918,
                y: 292.841,
                r: -0.333
            },
            "120": {
                x: -326.88,
                y: 292.788,
                r: -0.337
            },
            "121": {
                x: -326.903,
                y: 292.786
            },
            "122": {
                x: -326.919,
                y: 292.769,
                r: -0.341
            },
            "123": {
                x: -326.916,
                y: 292.804,
                r: -0.345
            },
            "124": {
                x: -326.889,
                y: 292.752,
                r: -0.346
            },
            "125": {
                x: -326.936,
                y: 292.726,
                r: -0.35
            },
            "126": {
                x: -326.932,
                y: 292.692,
                r: -0.354
            },
            "127": {
                x: -326.859,
                y: 292.735,
                r: -0.355
            },
            "128": {
                x: -326.9,
                y: 292.695,
                r: -0.359
            },
            "129": {
                x: -326.887,
                y: 292.702,
                r: -0.363
            },
            "130": {
                x: -326.909,
                y: 292.7
            },
            "131": {
                x: -326.89,
                y: 292.692,
                r: -0.367
            },
            "132": {
                x: -326.917,
                y: 292.635,
                r: -0.368
            },
            "133": {
                x: -326.848,
                y: 292.677,
                r: -0.372
            },
            "134": {
                x: -326.873,
                y: 292.656,
                r: -0.376
            },
            "135": {
                x: -326.891,
                y: 292.649,
                r: -0.377
            },
            "136": {
                x: -326.916,
                y: 292.628,
                r: -0.381
            },
            "137": {
                x: -326.895,
                y: 292.591,
                r: -0.385
            },
            "138": {
                x: -326.858,
                y: 292.58
            },
            "139": {
                x: -326.878,
                y: 292.595,
                r: -0.389
            },
            "140": {
                x: -326.888,
                y: 292.6,
                r: -0.393
            },
            "141": {
                x: -326.855,
                y: 292.584,
                r: -0.394
            },
            "142": {
                x: -326.873,
                y: 292.58,
                r: -0.398
            },
            "143": {
                x: -326.878,
                y: 292.571,
                r: -0.402
            },
            "144": {
                x: -326.895,
                y: 292.505,
                r: -0.403
            },
            "145": {
                x: -326.844,
                y: 292.542,
                r: -0.407
            },
            "146": {
                x: -326.898,
                y: 292.524,
                r: -0.411
            },
            "147": {
                x: -326.869,
                y: 292.544,
                r: -0.412
            },
            "148": {
                x: -326.868,
                y: 292.471,
                r: -0.416
            },
            "149": {
                x: -326.861,
                y: 292.435,
                r: -0.42
            },
            "150": {
                x: -326.873,
                y: 292.464
            },
            "151": {
                x: -326.862,
                y: 292.473,
                r: -0.424
            },
            "152": {
                x: -326.863,
                y: 292.417,
                r: -0.428
            },
            "153": {
                x: -326.87,
                y: 292.442,
                r: -0.429
            },
            "154": {
                x: -326.853,
                y: 292.437,
                r: -0.433
            },
            "155": {
                x: -326.836,
                y: 292.423,
                r: -0.437
            },
            "156": {
                x: -326.843,
                y: 292.398,
                r: -0.438
            },
            "157": {
                x: -326.834,
                y: 292.423,
                r: -0.442
            },
            "158": {
                x: -326.861,
                y: 292.395,
                r: -0.446
            },
            "159": {
                x: -326.868,
                y: 292.32
            },
            "160": {
                x: -326.844,
                y: 292.382,
                r: -0.45
            },
            "161": {
                x: -326.896,
                y: 292.303,
                r: -0.451
            },
            "162": {
                x: -326.877,
                y: 292.36,
                r: -0.455
            },
            "163": {
                x: -326.898,
                y: 292.308,
                r: -0.459
            },
            "164": {
                x: -326.854,
                y: 292.324,
                r: -0.46
            },
            "165": {
                x: -326.861,
                y: 292.268,
                r: -0.464
            },
            "166": {
                x: -326.876,
                y: 292.302,
                r: -0.468
            },
            "167": {
                x: -326.892,
                y: 292.267,
                r: -0.469
            },
            "168": {
                x: -326.902,
                y: 292.247,
                r: -0.473
            },
            "169": {
                x: -326.853,
                y: 292.217,
                r: -0.477
            },
            "170": {
                x: -326.859,
                y: 292.233
            },
            "171": {
                x: -326.864,
                y: 292.249,
                r: -0.481
            },
            "172": {
                x: -326.873,
                y: 292.25,
                r: -0.485
            },
            "173": {
                x: -326.874,
                y: 292.211,
                r: -0.486
            },
            "174": {
                y: 292.213,
                r: -0.49
            },
            "175": {
                x: -326.868,
                y: 292.21,
                r: -0.494
            },
            "176": {
                x: -326.869,
                y: 292.212,
                r: -0.495
            },
            "177": {
                x: -326.872,
                y: 292.149,
                r: -0.499
            },
            "178": {
                x: -326.866,
                y: 292.127,
                r: -0.503
            },
            "179": {
                x: -326.867,
                y: 292.129
            },
            "180": {
                x: -326.855,
                y: 292.103,
                r: -0.507
            },
            "181": {
                x: -326.838,
                y: 292.122,
                r: -0.511
            },
            "182": {
                x: -326.898,
                y: 292.114,
                r: -0.512
            },
            "183": {
                x: -326.881,
                y: 292.074,
                r: -0.516
            },
            "184": {
                x: -326.859,
                y: 292.079,
                r: -0.52
            },
            "185": {
                x: -326.865,
                y: 292.076,
                r: -0.521
            },
            "186": {
                x: -326.842,
                y: 292.071,
                r: -0.525
            },
            "187": {
                x: -326.902,
                y: 292.063
            },
            "188": {
                x: -326.874,
                y: 292.054,
                r: -0.529
            },
            "189": {
                x: -326.896,
                y: 291.977,
                r: -0.533
            },
            "190": {
                x: -326.892,
                y: 292.024,
                r: -0.534
            },
            "191": {
                x: -326.863,
                y: 292.046,
                r: -0.538
            },
            "192": {
                x: -326.884,
                y: 292.009,
                r: -0.542
            },
            "193": {
                x: -326.844,
                y: 292.001,
                r: -0.543
            },
            "194": {
                x: -326.859,
                y: 291.959,
                r: -0.547
            },
            "195": {
                x: -326.87,
                y: 291.954,
                r: -0.551
            },
            "196": {
                x: -326.92,
                y: 291.946,
                r: -0.552
            },
            "197": {
                x: -326.881,
                y: 291.941,
                r: -0.556
            },
            "198": {
                x: -326.854,
                y: 291.92,
                r: -0.56
            },
            "199": {
                x: -326.9,
                y: 291.908
            },
            "200": {
                x: -326.905,
                y: 291.938,
                r: -0.564
            },
            "201": {
                x: -326.863,
                y: 291.854,
                r: -0.568
            },
            "202": {
                x: -326.904,
                y: 291.837,
                r: -0.569
            },
            "203": {
                x: -326.867,
                y: 291.849,
                r: -0.573
            },
            "204": {
                x: -326.871,
                y: 291.86,
                r: -0.577
            },
            "205": {
                x: -326.866,
                y: 291.838,
                r: -0.578
            },
            "206": {
                x: -326.914,
                y: 291.836,
                sx: 0.998,
                sy: 0.998,
                r: -0.582
            },
            "207": {
                x: -326.862,
                y: 291.824,
                r: -0.586
            },
            "208": {
                x: -326.917,
                y: 291.802
            },
            "209": {
                x: -326.865,
                y: 291.79,
                r: -0.59
            },
            "210": {
                x: -326.907,
                y: 291.764,
                r: -0.594
            },
            "211": {
                x: -326.902,
                y: 291.742,
                r: -0.595
            },
            "212": {
                x: -326.894,
                y: 291.807,
                r: -0.599
            },
            "213": {
                x: -326.893,
                y: 291.78,
                r: -0.6
            },
            "214": {
                x: -326.885,
                y: 291.745,
                r: -0.604
            },
            "215": {
                x: -326.921,
                y: 291.695,
                r: -0.608
            },
            "216": {
                x: -326.916,
                y: 291.714,
                r: -0.609
            },
            "217": {
                x: -326.902,
                y: 291.715,
                r: -0.613
            },
            "218": {
                x: -326.952,
                y: 291.7,
                r: -0.617
            },
            "219": {
                x: -326.887,
                y: 291.67
            },
            "220": {
                x: -326.922,
                y: 291.701,
                r: -0.621
            },
            "221": {
                x: -326.907,
                y: 291.624,
                r: -0.625
            },
            "222": {
                x: -326.892,
                y: 291.643,
                r: -0.626
            },
            "223": {
                x: -326.885,
                y: 291.605,
                r: -0.63
            },
            "224": {
                x: -326.915,
                y: 291.613,
                r: -0.634
            },
            "225": {
                x: -326.904,
                y: 291.628,
                r: -0.635
            },
            "226": {
                x: -326.928,
                y: 291.631,
                r: -0.639
            },
            "227": {
                x: -326.952,
                y: 291.626,
                r: -0.643
            },
            "228": {
                x: -326.906,
                y: 291.584
            },
            "229": {
                x: -326.929,
                y: 291.569,
                r: -0.647
            },
            "230": {
                x: -326.902,
                y: 291.554,
                r: -0.651
            },
            "231": {
                x: -326.941,
                y: 291.559,
                r: -0.652
            },
            "232": {
                x: -326.914,
                y: 291.534,
                r: -0.656
            },
            "233": {
                x: -326.941,
                y: 291.495,
                r: -0.66
            },
            "234": {
                x: -326.93,
                y: 291.55,
                r: -0.661
            },
            "235": {
                x: -326.901,
                y: 291.506,
                r: -0.665
            },
            "236": {
                x: -326.968,
                y: 291.508,
                r: -0.669
            },
            "237": {
                x: -326.952,
                y: 291.458,
                r: -0.67
            },
            "238": {
                x: -326.932,
                y: 291.495,
                r: -0.674
            },
            "239": {
                x: -326.921,
                y: 291.45
            },
            "240": {
                x: -326.987,
                y: 291.433,
                r: -0.678
            },
            "241": {
                x: -326.948,
                y: 291.461,
                r: -0.682
            },
            "242": {
                x: -326.936,
                y: 291.406,
                r: -0.683
            },
            "243": {
                x: -326.956,
                y: 291.425,
                r: -0.687
            },
            "244": {
                x: -326.971,
                y: 291.389,
                r: -0.691
            },
            "245": {
                x: -326.959,
                y: 291.434,
                r: -0.692
            },
            "246": {
                x: -326.964,
                y: 291.389,
                r: -0.696
            },
            "247": {
                x: -326.972,
                y: 291.38,
                r: -0.7
            },
            "248": {
                x: -326.971,
                y: 291.374
            },
            "249": {
                x: -326.979,
                y: 291.315,
                r: -0.704
            },
            "250": {
                x: -326.987,
                y: 291.346,
                r: -0.708
            },
            "251": {
                x: -326.971,
                y: 291.337,
                r: -0.709
            },
            "252": {
                x: -326.979,
                y: 291.358,
                r: -0.713
            },
            "253": {
                x: -326.991,
                y: 291.325,
                r: -0.717
            },
            "254": {
                x: -326.979,
                y: 291.311,
                r: -0.718
            },
            "255": {
                x: -326.981,
                y: 291.268,
                r: -0.722
            },
            "256": {
                x: -326.987,
                y: 291.271,
                r: -0.726
            },
            "257": {
                x: -327.016,
                y: 291.257,
                r: -0.727
            },
            "258": {
                x: -326.982,
                y: 291.25,
                r: -0.731
            },
            "259": {
                x: -326.983,
                y: 291.239,
                r: -0.735
            },
            "260": {
                x: -326.97,
                y: 291.215
            },
            "261": {
                x: -327.021,
                y: 291.244,
                r: -0.739
            },
            "262": {
                x: -327.017,
                y: 291.219,
                r: -0.743
            },
            "263": {
                x: -327.014,
                y: 291.194,
                r: -0.744
            },
            "264": {
                x: -327.013,
                y: 291.155,
                r: -0.748
            },
            "265": {
                x: -326.996,
                y: 291.186,
                r: -0.749
            },
            "266": {
                x: -327,
                y: 291.191,
                r: -0.753
            },
            "267": {
                x: -326.995,
                y: 291.137,
                r: -0.757
            },
            "268": {
                x: -327.032,
                y: 291.164
            },
            "269": {
                x: -327.035,
                y: 291.15,
                r: -0.761
            },
            "270": {
                x: -327.029,
                y: 291.137,
                r: -0.765
            },
            "271": {
                x: -327.021,
                y: 291.108,
                r: -0.766
            },
            "272": {
                x: -327.059,
                y: 291.081,
                r: -0.77
            },
            "273": {
                x: -327.056,
                y: 291.103,
                r: -0.774
            },
            "274": {
                x: -327.003,
                y: 291.069,
                r: -0.775
            },
            "275": {
                x: -327.045,
                y: 291.077,
                r: -0.779
            },
            "276": {
                x: -327.092,
                y: 291.031,
                r: -0.783
            },
            "277": {
                x: -327.079,
                y: 291.048,
                r: -0.784
            },
            "278": {
                x: -327.065,
                y: 291.042,
                r: -0.788
            },
            "279": {
                x: -327.07,
                y: 291.026,
                r: -0.792
            },
            "280": {
                x: -327.107,
                y: 290.993
            },
            "281": {
                x: -327.098,
                y: 291.022,
                r: -0.796
            },
            "282": {
                x: -327.092,
                y: 290.988,
                r: -0.8
            },
            "283": {
                x: -327.079,
                y: 291.005,
                r: -0.801
            },
            "284": {
                x: -327.078,
                y: 290.965,
                r: -0.805
            },
            "285": {
                x: -327.072,
                y: 290.971,
                r: -0.809
            },
            "286": {
                x: -327.055,
                y: 290.933,
                r: -0.81
            },
            "287": {
                x: -327.098,
                y: 290.93,
                r: -0.814
            },
            "288": {
                x: -327.086,
                y: 290.922,
                r: -0.818
            },
            "289": {
                x: -327.083,
                y: 290.928
            },
            "290": {
                x: -327.075,
                y: 290.915,
                r: -0.822
            },
            "291": {
                x: -327.107,
                y: 290.918,
                r: -0.823
            },
            "292": {
                x: -327.104,
                y: 290.891,
                r: -0.827
            },
            "293": {
                x: -327.141,
                y: 290.904,
                r: -0.831
            },
            "294": {
                x: -327.087,
                y: 290.911,
                r: -0.832
            },
            "295": {
                x: -327.129,
                y: 290.869,
                r: -0.836
            },
            "296": {
                x: -327.165,
                y: 290.823,
                r: -0.84
            },
            "297": {
                x: -327.151,
                y: 290.821,
                r: -0.841
            },
            "298": {
                x: -327.147,
                y: 290.815,
                r: -0.845
            },
            "299": {
                x: -327.096,
                y: 290.804,
                r: -0.849
            },
            "300": {
                x: -327.128,
                y: 290.798
            },
            "301": {
                x: -327.172,
                y: 290.782,
                r: -0.853
            },
            "302": {
                x: -327.162,
                y: 290.803,
                r: -0.857
            },
            "303": {
                x: -327.143,
                y: 290.796,
                r: -0.858
            },
            "304": {
                x: -327.146,
                y: 290.761,
                r: -0.862
            },
            "305": {
                x: -327.14,
                y: 290.767,
                r: -0.866
            },
            "306": {
                x: -327.175,
                y: 290.756,
                r: -0.867
            },
            "307": {
                x: -327.164,
                y: 290.707,
                r: -0.871
            },
            "308": {
                x: -327.161,
                y: 290.699,
                r: -0.875
            },
            "309": {
                x: -327.206,
                y: 290.687
            },
            "310": {
                x: -327.194,
                y: 290.719,
                r: -0.879
            },
            "311": {
                x: -327.185,
                y: 290.697,
                r: -0.883
            },
            "312": {
                x: -327.171,
                y: 290.685,
                r: -0.884
            },
            "313": {
                x: -327.217,
                y: 290.699,
                r: -0.888
            },
            "314": {
                x: -327.172,
                y: 290.661,
                r: -0.892
            },
            "315": {
                x: -327.203,
                y: 290.645,
                r: -0.893
            },
            "316": {
                x: -327.248,
                y: 290.649,
                r: -0.897
            },
            "317": {
                x: -327.183,
                y: 290.628,
                r: -0.898
            },
            "318": {
                x: -327.228,
                y: 290.622,
                r: -0.902
            },
            "319": {
                x: -327.232,
                y: 290.616,
                r: -0.906
            },
            "320": {
                x: -327.217,
                y: 290.595
            },
            "321": {
                x: -327.211,
                y: 290.62,
                r: -0.91
            },
            "322": {
                x: -327.255,
                y: 290.595,
                r: -0.914
            },
            "323": {
                x: -327.245,
                y: 290.569,
                r: -0.915
            },
            "324": {
                x: -327.248,
                y: 290.584,
                r: -0.919
            },
            "325": {
                x: -327.245,
                y: 290.545,
                r: -0.923
            },
            "326": {
                x: -327.28,
                y: 290.564,
                r: -0.924
            },
            "327": {
                x: -327.277,
                y: 290.516,
                r: -0.928
            },
            "328": {
                x: -327.274,
                y: 290.508,
                r: -0.932
            },
            "329": {
                x: -327.273,
                y: 290.481
            },
            "330": {
                x: -327.269,
                y: 290.513,
                r: -0.936
            },
            "331": {
                x: -327.27,
                y: 290.491,
                r: -0.94
            },
            "332": {
                x: -327.309,
                y: 290.455,
                r: -0.941
            },
            "333": {
                x: -327.255,
                y: 290.478,
                r: -0.945
            },
            "334": {
                x: -327.319,
                y: 290.44,
                r: -0.949
            },
            "335": {
                x: -327.308,
                y: 290.455,
                r: -0.95
            },
            "336": {
                x: -327.312,
                y: 290.408,
                r: -0.954
            },
            "337": {
                y: 290.457,
                r: -0.958
            },
            "338": {
                x: -327.346,
                y: 290.417,
                r: -0.959
            },
            "339": {
                x: -327.313,
                y: 290.405,
                r: -0.963
            },
            "340": {
                x: -327.367,
                y: 290.39,
                r: -0.967
            },
            "341": {
                x: -327.351,
                y: 290.4
            },
            "342": {
                x: -327.313,
                y: 290.374,
                r: -0.971
            },
            "343": {
                x: -327.347,
                y: 290.384,
                r: -0.972
            },
            "344": {
                x: -327.354,
                y: 290.394,
                r: -0.976
            },
            "345": {
                x: -327.375,
                y: 290.359,
                r: -0.98
            },
            "346": {
                x: -327.358,
                y: 290.309,
                r: -0.981
            },
            "347": {
                x: -327.37,
                y: 290.365,
                r: -0.985
            },
            "348": {
                x: -327.375,
                y: 290.306,
                r: -0.989
            },
            "349": {
                x: -327.368
            },
            "350": {
                x: -327.388,
                y: 290.292,
                r: -0.993
            },
            "351": {
                x: -327.403,
                y: 290.324,
                sx: 0.999,
                sy: 0.999,
                r: -0.997
            },
            "352": {
                x: -327.386,
                y: 290.274,
                r: -0.998
            },
            "353": {
                x: -327.4,
                y: 290.296,
                r: -1.002
            },
            "354": {
                x: -327.414,
                y: 290.259,
                r: -1.006
            },
            "355": {
                x: -327.367,
                y: 290.258,
                r: -1.007
            },
            "356": {
                x: -327.43,
                y: 290.261,
                r: -1.011
            },
            "357": {
                x: -327.397,
                y: 290.259,
                r: -1.015
            },
            "358": {
                x: -327.435,
                y: 290.205,
                r: -1.016
            },
            "359": {
                x: -327.457,
                y: 290.198,
                r: -1.02
            },
            "360": {
                x: -327.439,
                y: 290.231,
                r: -1.024
            },
            "361": {
                x: -327.431,
                y: 290.171
            },
            "362": {
                x: -327.452,
                y: 290.146,
                r: -1.028
            },
            "363": {
                x: -327.423,
                y: 290.17,
                r: -1.032
            },
            "364": {
                x: -327.511,
                y: 290.156,
                r: -1.033
            },
            "365": {
                x: -327.45,
                y: 290.169,
                r: -1.037
            },
            "366": {
                x: -327.475,
                y: 290.12,
                r: -1.041
            },
            "367": {
                x: -327.463,
                y: 290.155,
                r: -1.042
            },
            "368": {
                x: -327.492,
                y: 290.151,
                r: -1.046
            },
            "369": {
                x: -327.484,
                y: 290.091
            },
            "370": {
                x: -327.522,
                y: 290.126,
                r: -1.05
            },
            "371": {
                x: -327.505,
                y: 290.107,
                r: -1.054
            },
            "372": {
                x: -327.543,
                y: 290.093,
                r: -1.055
            },
            "373": {
                x: -327.525,
                y: 290.114,
                r: -1.059
            },
            "374": {
                x: -327.562,
                y: 290.08,
                r: -1.063
            },
            "375": {
                x: -327.564,
                y: 290.061,
                r: -1.064
            },
            "376": {
                x: -327.55,
                y: 290.068,
                r: -1.068
            },
            "377": {
                x: -327.536,
                y: 290.065,
                r: -1.072
            },
            "378": {
                x: -327.528,
                y: 290.046
            },
            "379": {
                x: -327.569,
                y: 290.038,
                r: -1.076
            },
            "380": {
                y: 289.98,
                r: -1.08
            },
            "381": {
                x: -327.56,
                y: 290.001,
                r: -1.081
            },
            "382": {
                x: -327.555,
                y: 290.029,
                r: -1.085
            },
            "383": {
                x: -327.604,
                y: 289.952,
                r: -1.089
            },
            "384": {
                x: -327.595,
                y: 289.973,
                r: -1.09
            },
            "385": {
                x: -327.603,
                y: 289.986,
                r: -1.094
            },
            "386": {
                x: -327.647,
                y: 289.945,
                r: -1.098
            },
            "387": {
                x: -327.598,
                y: 289.966,
                r: -1.099
            },
            "388": {
                x: -327.595,
                y: 289.96,
                r: -1.103
            },
            "389": {
                x: -327.652,
                y: 289.954,
                r: -1.107
            }
        });
    });

    var Graphic49 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("comet1"))
            .setTransform(-123.1, -3.5);
        this.addTimedChild(instance1);
    });

    var Graphic50 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Graphic49(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 16, {
            "0": {
                x: 337.35,
                y: 0.8,
                sx: 1,
                a: 0
            },
            "1": {
                x: 319.118,
                sx: 0.985,
                a: 0.19
            },
            "2": {
                x: 291.717,
                sx: 0.962,
                a: 0.48
            },
            "3": {
                x: 255.2,
                sx: 0.932,
                a: 0.86
            },
            "4": {
                x: 213.853,
                sx: 0.897,
                a: 0.79
            },
            "5": {
                x: 172.51,
                sx: 0.863,
                a: 0.71
            },
            "6": {
                x: 131.163,
                sx: 0.828,
                a: 0.64
            },
            "7": {
                x: 89.816,
                sx: 0.794,
                a: 0.57
            },
            "8": {
                x: 48.469,
                sx: 0.759,
                a: 0.5
            },
            "9": {
                x: 7.122,
                sx: 0.725,
                a: 0.43
            },
            "10": {
                x: -34.225,
                sx: 0.69,
                a: 0.36
            },
            "11": {
                x: -75.618,
                sx: 0.656,
                a: 0.29
            },
            "12": {
                x: -116.965,
                sx: 0.621,
                a: 0.21
            },
            "13": {
                x: -158.312,
                sx: 0.587,
                a: 0.14
            },
            "14": {
                x: -199.655,
                sx: 0.552,
                a: 0.07
            },
            "15": {
                x: -241,
                sx: 0.518,
                a: 0
            }
        });
    });

    var Graphic51 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("comet1"))
            .setTransform(-123.1, -3.5);
        this.addTimedChild(instance1);
    });

    var Graphic52 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Graphic51(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 16, {
            "0": {
                x: 337.35,
                y: 0.8,
                sx: 1,
                a: 0
            },
            "1": {
                x: 319.118,
                sx: 0.985,
                a: 0.19
            },
            "2": {
                x: 291.717,
                sx: 0.962,
                a: 0.48
            },
            "3": {
                x: 255.2,
                sx: 0.932,
                a: 0.86
            },
            "4": {
                x: 213.853,
                sx: 0.897,
                a: 0.79
            },
            "5": {
                x: 172.51,
                sx: 0.863,
                a: 0.71
            },
            "6": {
                x: 131.163,
                sx: 0.828,
                a: 0.64
            },
            "7": {
                x: 89.816,
                sx: 0.794,
                a: 0.57
            },
            "8": {
                x: 48.469,
                sx: 0.759,
                a: 0.5
            },
            "9": {
                x: 7.122,
                sx: 0.725,
                a: 0.43
            },
            "10": {
                x: -34.225,
                sx: 0.69,
                a: 0.36
            },
            "11": {
                x: -75.618,
                sx: 0.656,
                a: 0.29
            },
            "12": {
                x: -116.965,
                sx: 0.621,
                a: 0.21
            },
            "13": {
                x: -158.312,
                sx: 0.587,
                a: 0.14
            },
            "14": {
                x: -199.655,
                sx: 0.552,
                a: 0.07
            },
            "15": {
                x: -241,
                sx: 0.518,
                a: 0
            }
        });
    });

    lib.dream_stars = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 390,
            framerate: 30,
            loop: false
        });
        var instance2 = new Graphic48(MovieClip.SYNCHED);
        var instance3 = new Graphic50(MovieClip.SYNCHED);
        var instance4 = new Graphic52(MovieClip.SYNCHED);
        var instance1 = new Graphic1(MovieClip.SYNCHED)
            .setTransform(640.6, 360.55);
        this.addTimedChild(instance2, 0, 390, {
                "0": {
                    x: 972.05,
                    y: 69.25,
                    a: 1,
                    c: [
                        0,
                        0,
                        0,
                        0,
                        0,
                        0
                    ]
                },
                "3": {
                    c: [
                        0.01,
                        0,
                        0.01,
                        0,
                        0.01,
                        0
                    ]
                },
                "4": {
                    c: [
                        0.02,
                        0,
                        0.02,
                        0,
                        0.02,
                        0
                    ]
                },
                "5": {
                    c: [
                        0.03,
                        0,
                        0.03,
                        0,
                        0.03,
                        0
                    ]
                },
                "6": {
                    c: [
                        0.05,
                        0,
                        0.05,
                        0,
                        0.05,
                        0
                    ]
                },
                "7": {
                    c: [
                        0.07,
                        0,
                        0.07,
                        0,
                        0.07,
                        0
                    ]
                },
                "8": {
                    c: [
                        0.09,
                        0,
                        0.09,
                        0,
                        0.09,
                        0
                    ]
                },
                "9": {
                    c: [
                        0.12,
                        0,
                        0.12,
                        0,
                        0.12,
                        0
                    ]
                },
                "10": {
                    c: [
                        0.16,
                        0,
                        0.16,
                        0,
                        0.16,
                        0
                    ]
                },
                "11": {
                    c: [
                        0.2,
                        0,
                        0.2,
                        0,
                        0.2,
                        0
                    ]
                },
                "12": {
                    c: [
                        0.25,
                        0,
                        0.25,
                        0,
                        0.25,
                        0
                    ]
                },
                "13": {
                    c: [
                        0.3,
                        0,
                        0.3,
                        0,
                        0.3,
                        0
                    ]
                },
                "14": {
                    c: [
                        0.37,
                        0,
                        0.37,
                        0,
                        0.37,
                        0
                    ]
                },
                "15": {
                    c: [
                        0.44,
                        0,
                        0.44,
                        0,
                        0.44,
                        0
                    ]
                },
                "16": {
                    c: [
                        0.51,
                        0,
                        0.51,
                        0,
                        0.51,
                        0
                    ]
                },
                "17": {
                    c: [
                        0.59,
                        0,
                        0.59,
                        0,
                        0.59,
                        0
                    ]
                },
                "18": {
                    c: [
                        0.66,
                        0,
                        0.66,
                        0,
                        0.66,
                        0
                    ]
                },
                "19": {
                    c: [
                        0.72,
                        0,
                        0.72,
                        0,
                        0.72,
                        0
                    ]
                },
                "20": {
                    c: [
                        0.78,
                        0,
                        0.78,
                        0,
                        0.78,
                        0
                    ]
                },
                "21": {
                    c: [
                        0.83,
                        0,
                        0.83,
                        0,
                        0.83,
                        0
                    ]
                },
                "22": {
                    c: [
                        0.88,
                        0,
                        0.88,
                        0,
                        0.88,
                        0
                    ]
                },
                "23": {
                    c: [
                        0.91,
                        0,
                        0.91,
                        0,
                        0.91,
                        0
                    ]
                },
                "24": {
                    c: [
                        0.93,
                        0,
                        0.93,
                        0,
                        0.93,
                        0
                    ]
                },
                "25": {
                    c: [
                        0.96,
                        0,
                        0.96,
                        0,
                        0.96,
                        0
                    ]
                },
                "26": {
                    c: [
                        0.97,
                        0,
                        0.97,
                        0,
                        0.97,
                        0
                    ]
                },
                "27": {
                    c: [
                        0.98,
                        0,
                        0.98,
                        0,
                        0.98,
                        0
                    ]
                },
                "28": {
                    c: [
                        0.99,
                        0,
                        0.99,
                        0,
                        0.99,
                        0
                    ]
                },
                "29": {
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "374": {
                    a: 0.94
                },
                "375": {
                    a: 0.88
                },
                "376": {
                    a: 0.81
                },
                "377": {
                    a: 0.75
                },
                "378": {
                    a: 0.69
                },
                "379": {
                    a: 0.63
                },
                "380": {
                    a: 0.56
                },
                "381": {
                    a: 0.5
                },
                "382": {
                    a: 0.44
                },
                "383": {
                    a: 0.38
                },
                "384": {
                    a: 0.31
                },
                "385": {
                    a: 0.25
                },
                "386": {
                    a: 0.19
                },
                "387": {
                    a: 0.13
                },
                "388": {
                    a: 0.06
                },
                "389": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 122, 16, {
                "122": {
                    x: 966.85,
                    y: 116.9,
                    r: -0.589
                }
            })
            .addTimedChild(instance4, 324, 16, {
                "324": {
                    x: 302.8,
                    y: 287.75,
                    r: -0.589
                }
            })
            .addTimedChild(instance1);
    });

    lib.dream_stars.assets = {
        "fade": "images/fade.png",
        "dream_stars": "images/dream_stars.shapes.json",
        "dream_stars_atlas_1": "images/dream_stars_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.dream_stars,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 390,
        library: lib
    };
}
