// jibo-anim-db-animations/timelines/idle_pong.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 184, loop: false });
        var instance1 = new Sprite(fromFrame("pong-paddle1"))
            .setTransform(-0.15, -5.5);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 184, loop: false });
        var instance1 = new Sprite(fromFrame("pong-paddle1"))
            .setTransform(-0.15, -5.5);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 184, loop: false });
        var instance1 = new Sprite(fromFrame("pong-ball1"))
            .setTransform(-13.15, -3.25);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 184, loop: false });
        var instance1 = new Sprite(fromFrame("pong-line"))
            .setTransform(-4.1, -0.15);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 184, loop: false });
        var instance4 = new Graphic4(MovieClip.SYNCHED)
            .setTransform(574.55, -29.7);
        var instance3 = new Graphic3(MovieClip.SYNCHED);
        var instance2 = new Graphic2(MovieClip.SYNCHED);
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance4)
            .addTimedChild(instance3, 0, 184, {
                "0": {
                    x: 1133,
                    y: 76.3,
                    sx: 0.106,
                    sy: 0.141
                },
                "1": {
                    x: 1106.65,
                    y: 83.65,
                    sx: 0.126,
                    sy: 0.126
                },
                "2": {
                    x: 1055.8,
                    y: 88.45
                },
                "3": {
                    x: 1004.95,
                    y: 93.25
                },
                "4": {
                    x: 954.1,
                    y: 98
                },
                "5": {
                    x: 903.25,
                    y: 102.8
                },
                "6": {
                    x: 852.35,
                    y: 107.6
                },
                "7": {
                    x: 801.5,
                    y: 112.4
                },
                "8": {
                    x: 750.7,
                    y: 117.15
                },
                "9": {
                    x: 699.8,
                    y: 121.95
                },
                "10": {
                    x: 648.95,
                    y: 126.75
                },
                "11": {
                    x: 598.1,
                    y: 131.55
                },
                "12": {
                    x: 547.25,
                    y: 136.3
                },
                "13": {
                    x: 496.4,
                    y: 141.1
                },
                "14": {
                    x: 445.55,
                    y: 145.9
                },
                "15": {
                    x: 394.7,
                    y: 150.7
                },
                "16": {
                    x: 343.85,
                    y: 155.45
                },
                "17": {
                    x: 293,
                    y: 160.25
                },
                "18": {
                    x: 242.1,
                    y: 165.05
                },
                "19": {
                    x: 191.25,
                    y: 169.85
                },
                "20": {
                    x: 140.4,
                    y: 174.6
                },
                "21": {
                    x: 89.55,
                    y: 179.4
                },
                "22": {
                    x: 38.7,
                    y: 184.2
                },
                "23": {
                    x: 29.15,
                    y: 183.05,
                    sx: 0.091,
                    sy: 0.134
                },
                "24": {
                    x: 54,
                    y: 199.5,
                    sx: 0.126,
                    sy: 0.126
                },
                "25": {
                    x: 105,
                    y: 214.45
                },
                "26": {
                    x: 156,
                    y: 229.4
                },
                "27": {
                    x: 206.95,
                    y: 244.35
                },
                "28": {
                    x: 257.95,
                    y: 259.3
                },
                "29": {
                    x: 308.95,
                    y: 274.25
                },
                "30": {
                    x: 359.95,
                    y: 289.25
                },
                "31": {
                    x: 410.9,
                    y: 304.2
                },
                "32": {
                    x: 461.9,
                    y: 319.15
                },
                "33": {
                    x: 512.9,
                    y: 334.1
                },
                "34": {
                    x: 563.9,
                    y: 349.05
                },
                "35": {
                    x: 614.85,
                    y: 364
                },
                "36": {
                    x: 665.85,
                    y: 378.95
                },
                "37": {
                    x: 716.85,
                    y: 393.9
                },
                "38": {
                    x: 767.85,
                    y: 408.85
                },
                "39": {
                    x: 818.8,
                    y: 423.8
                },
                "40": {
                    x: 869.8,
                    y: 438.8
                },
                "41": {
                    x: 920.8,
                    y: 453.75
                },
                "42": {
                    x: 971.8,
                    y: 468.7
                },
                "43": {
                    x: 1022.75,
                    y: 483.65
                },
                "44": {
                    x: 1073.75,
                    y: 498.6
                },
                "45": {
                    x: 1124.75,
                    y: 513.55
                },
                "46": {
                    x: 1133.25,
                    y: 511.3,
                    sx: 0.102,
                    sy: 0.141
                },
                "47": {
                    x: 1109.45,
                    y: 513.55,
                    sx: 0.126,
                    sy: 0.126
                },
                "48": {
                    x: 1055.45,
                    y: 504.65
                },
                "49": {
                    x: 1001.5,
                    y: 495.7
                },
                "50": {
                    x: 947.5,
                    y: 486.8
                },
                "51": {
                    x: 893.5,
                    y: 477.85
                },
                "52": {
                    x: 839.5,
                    y: 468.95
                },
                "53": {
                    x: 785.55,
                    y: 460
                },
                "54": {
                    x: 731.55,
                    y: 451.1
                },
                "55": {
                    x: 677.55,
                    y: 442.15
                },
                "56": {
                    x: 623.55,
                    y: 433.25
                },
                "57": {
                    x: 569.55,
                    y: 424.3
                },
                "58": {
                    x: 515.6,
                    y: 415.4
                },
                "59": {
                    x: 461.6,
                    y: 406.5
                },
                "60": {
                    x: 407.6,
                    y: 397.55
                },
                "61": {
                    x: 353.65,
                    y: 388.65
                },
                "62": {
                    x: 299.65,
                    y: 379.7
                },
                "63": {
                    x: 245.65,
                    y: 370.8
                },
                "64": {
                    x: 191.65,
                    y: 361.85
                },
                "65": {
                    x: 137.7,
                    y: 352.95
                },
                "66": {
                    x: 83.7,
                    y: 344
                },
                "67": {
                    x: 29.7,
                    y: 335.1
                },
                "68": {
                    x: 30.65,
                    y: 333.95,
                    sx: 0.098,
                    sy: 0.134
                },
                "69": {
                    x: 52.3,
                    y: 331.7,
                    sx: 0.126,
                    sy: 0.126
                },
                "70": {
                    x: 102.8,
                    y: 316.65
                },
                "71": {
                    x: 153.3,
                    y: 301.6
                },
                "72": {
                    x: 203.75,
                    y: 286.55
                },
                "73": {
                    x: 254.25,
                    y: 271.5
                },
                "74": {
                    x: 304.75,
                    y: 256.45
                },
                "75": {
                    x: 355.25,
                    y: 241.4
                },
                "76": {
                    x: 405.7,
                    y: 226.35
                },
                "77": {
                    x: 456.2,
                    y: 211.3
                },
                "78": {
                    x: 506.7,
                    y: 196.25
                },
                "79": {
                    x: 557.2,
                    y: 181.2
                },
                "80": {
                    x: 607.65,
                    y: 166.1
                },
                "81": {
                    x: 658.15,
                    y: 151.05
                },
                "82": {
                    x: 708.65,
                    y: 136
                },
                "83": {
                    x: 759.15,
                    y: 120.95
                },
                "84": {
                    x: 809.6,
                    y: 105.9
                },
                "85": {
                    x: 860.1,
                    y: 90.85
                },
                "86": {
                    x: 910.6,
                    y: 75.8
                },
                "87": {
                    x: 961.1,
                    y: 60.75
                },
                "88": {
                    x: 1011.55,
                    y: 45.7
                },
                "89": {
                    x: 1062.05,
                    y: 30.65
                },
                "90": {
                    x: 1112.55,
                    y: 15.6
                },
                "91": {
                    x: 1133.15,
                    y: 12.55,
                    sx: 0.109,
                    sy: 0.147
                },
                "92": {
                    x: 1112.55,
                    y: 15.6,
                    sx: 0.126,
                    sy: 0.126
                },
                "93": {
                    x: 1055.9,
                    y: 26.95
                },
                "94": {
                    x: 999.25,
                    y: 38.35
                },
                "95": {
                    x: 942.6,
                    y: 49.7
                },
                "96": {
                    x: 885.95,
                    y: 61.1
                },
                "97": {
                    x: 829.35,
                    y: 72.45
                },
                "98": {
                    x: 772.65,
                    y: 83.85
                },
                "99": {
                    x: 716.05,
                    y: 95.2
                },
                "100": {
                    x: 659.4,
                    y: 106.6
                },
                "101": {
                    x: 602.75,
                    y: 117.95
                },
                "102": {
                    x: 546.1,
                    y: 129.35
                },
                "103": {
                    x: 489.45,
                    y: 140.7
                },
                "104": {
                    x: 432.8,
                    y: 152.1
                },
                "105": {
                    x: 376.2,
                    y: 163.45
                },
                "106": {
                    x: 319.5,
                    y: 174.85
                },
                "107": {
                    x: 262.9,
                    y: 186.2
                },
                "108": {
                    x: 206.25,
                    y: 197.6
                },
                "109": {
                    x: 149.6,
                    y: 208.95
                },
                "110": {
                    x: 92.95,
                    y: 220.35
                },
                "111": {
                    x: 36.3,
                    y: 231.7
                },
                "112": {
                    x: 28.15,
                    y: 228.65,
                    sx: 0.093,
                    sy: 0.147
                },
                "113": {
                    x: 52.3,
                    y: 231.7,
                    sx: 0.126,
                    sy: 0.126
                },
                "114": {
                    x: 108.4,
                    y: 228.7
                },
                "115": {
                    x: 164.55,
                    y: 225.7
                },
                "116": {
                    x: 220.6,
                    y: 222.7
                },
                "117": {
                    x: 276.75,
                    y: 219.7
                },
                "118": {
                    x: 332.85,
                    y: 216.7
                },
                "119": {
                    x: 388.95,
                    y: 213.7
                },
                "120": {
                    x: 445.05,
                    y: 210.7
                },
                "121": {
                    x: 501.15,
                    y: 207.7
                },
                "122": {
                    x: 557.25,
                    y: 204.7
                },
                "123": {
                    x: 613.4,
                    y: 201.75
                },
                "124": {
                    x: 669.5,
                    y: 198.75
                },
                "125": {
                    x: 725.6,
                    y: 195.75
                },
                "126": {
                    x: 781.7,
                    y: 192.75
                },
                "127": {
                    x: 837.8,
                    y: 189.75
                },
                "128": {
                    x: 893.9,
                    y: 186.75
                },
                "129": {
                    x: 950,
                    y: 183.75
                },
                "130": {
                    x: 1006.15,
                    y: 180.75
                },
                "131": {
                    x: 1062.25,
                    y: 177.75
                },
                "132": {
                    x: 1118.35,
                    y: 174.75
                },
                "133": {
                    x: 1130.95,
                    y: 173.25,
                    sx: 0.112,
                    sy: 0.136
                },
                "134": {
                    x: 1118.35,
                    y: 174.75,
                    sx: 0.126,
                    sy: 0.126
                },
                "135": {
                    x: 1069.05,
                    y: 179.3
                },
                "136": {
                    x: 1019.7,
                    y: 183.85
                },
                "137": {
                    x: 970.4,
                    y: 188.4
                },
                "138": {
                    x: 921.05,
                    y: 192.95
                },
                "139": {
                    x: 871.75,
                    y: 197.5
                },
                "140": {
                    x: 822.45,
                    y: 202
                },
                "141": {
                    x: 773.1,
                    y: 206.55
                },
                "142": {
                    x: 723.8,
                    y: 211.1
                },
                "143": {
                    x: 674.45,
                    y: 215.65
                },
                "144": {
                    x: 625.15,
                    y: 220.2
                },
                "145": {
                    x: 575.85,
                    y: 224.75
                },
                "146": {
                    x: 526.5,
                    y: 229.3
                },
                "147": {
                    x: 477.2,
                    y: 233.85
                },
                "148": {
                    x: 427.85,
                    y: 238.4
                },
                "149": {
                    x: 378.55,
                    y: 242.95
                },
                "150": {
                    x: 329.2,
                    y: 247.5
                },
                "151": {
                    x: 279.9,
                    y: 252
                },
                "152": {
                    x: 230.6,
                    y: 256.55
                },
                "153": {
                    x: 181.25,
                    y: 261.1
                },
                "154": {
                    x: 131.95,
                    y: 265.65
                },
                "155": {
                    x: 82.6,
                    y: 270.2
                },
                "156": {
                    x: 33.3,
                    y: 274.75
                },
                "157": {
                    x: 30.25,
                    y: 273.25,
                    sx: 0.115,
                    sy: 0.136
                },
                "158": {
                    x: 33.3,
                    y: 274.75,
                    sx: 0.126,
                    sy: 0.126
                },
                "159": {
                    x: 76.8,
                    y: 266.9
                },
                "160": {
                    x: 120.3,
                    y: 259.05
                },
                "161": {
                    x: 163.8,
                    y: 251.2
                },
                "162": {
                    x: 207.35,
                    y: 243.35
                },
                "163": {
                    x: 250.8,
                    y: 235.5
                },
                "164": {
                    x: 294.3,
                    y: 227.65
                },
                "165": {
                    x: 337.85,
                    y: 219.8
                },
                "166": {
                    x: 381.35,
                    y: 211.95
                },
                "167": {
                    x: 424.85,
                    y: 204.1
                },
                "168": {
                    x: 468.35,
                    y: 196.25
                },
                "169": {
                    x: 511.85,
                    y: 188.4
                },
                "170": {
                    x: 555.35,
                    y: 180.6
                },
                "171": {
                    x: 598.85,
                    y: 172.75
                },
                "172": {
                    x: 642.35,
                    y: 164.9
                },
                "173": {
                    x: 685.85,
                    y: 157.05
                },
                "174": {
                    x: 729.35,
                    y: 149.2
                },
                "175": {
                    x: 772.85,
                    y: 141.35
                },
                "176": {
                    x: 816.35,
                    y: 133.5
                },
                "177": {
                    x: 859.85,
                    y: 125.65
                },
                "178": {
                    x: 903.4,
                    y: 117.8
                },
                "179": {
                    x: 946.9,
                    y: 109.95
                },
                "180": {
                    x: 990.4,
                    y: 102.1
                },
                "181": {
                    x: 1033.9,
                    y: 94.25
                },
                "182": {
                    x: 1077.4,
                    y: 86.4
                },
                "183": {
                    x: 1120.9,
                    y: 78.55
                }
            })
            .addTimedChild(instance2, 0, 184, {
                "0": {
                    x: 1.95,
                    y: 238.45
                },
                "15": {
                    y: 236.05
                },
                "16": {
                    y: 227.95
                },
                "17": {
                    y: 212.75
                },
                "18": {
                    y: 189.65
                },
                "19": {
                    y: 161.1
                },
                "20": {
                    y: 133.9
                },
                "21": {
                    y: 114.3
                },
                "22": {
                    y: 103.6
                },
                "23": {
                    y: 100.45
                },
                "24": {
                    y: 101.1
                },
                "25": {
                    y: 103.45
                },
                "26": {
                    y: 107.7
                },
                "27": {
                    y: 112.55
                },
                "28": {
                    y: 115.6
                },
                "29": {
                    y: 116.45
                },
                "56": {
                    y: 118.2
                },
                "57": {
                    y: 123.95
                },
                "58": {
                    y: 134.55
                },
                "59": {
                    y: 150.7
                },
                "60": {
                    y: 172.85
                },
                "61": {
                    y: 200.05
                },
                "62": {
                    y: 229.15
                },
                "63": {
                    y: 255.7
                },
                "64": {
                    y: 276.45
                },
                "65": {
                    y: 290.35
                },
                "66": {
                    y: 298
                },
                "67": {
                    y: 300.3
                },
                "68": {
                    y: 299.85
                },
                "69": {
                    y: 298.35
                },
                "70": {
                    y: 295.45
                },
                "71": {
                    y: 291.2
                },
                "72": {
                    y: 286.55
                },
                "73": {
                    y: 282.9
                },
                "74": {
                    y: 280.9
                },
                "75": {
                    y: 280.3
                },
                "102": {
                    y: 279.15
                },
                "103": {
                    y: 275.4
                },
                "104": {
                    y: 268.4
                },
                "105": {
                    y: 257.65
                },
                "106": {
                    y: 243.15
                },
                "107": {
                    y: 226.2
                },
                "108": {
                    y: 209.5
                },
                "109": {
                    y: 196
                },
                "110": {
                    y: 186.8
                },
                "111": {
                    y: 181.8
                },
                "112": {
                    y: 180.3
                },
                "145": {
                    y: 180.85
                },
                "146": {
                    y: 182.7
                },
                "147": {
                    y: 186
                },
                "148": {
                    y: 191.15
                },
                "149": {
                    y: 198.15
                },
                "150": {
                    y: 206.75
                },
                "151": {
                    y: 215.95
                },
                "152": {
                    y: 224.35
                },
                "153": {
                    y: 230.9
                },
                "154": {
                    y: 235.3
                },
                "155": {
                    y: 237.7
                },
                "156": {
                    y: 238.45
                }
            })
            .addTimedChild(instance1, 0, 184, {
                "0": {
                    x: 1164.95,
                    y: 11.9
                },
                "25": {
                    y: 13.1
                },
                "26": {
                    y: 17
                },
                "27": {
                    y: 23.75
                },
                "28": {
                    y: 33.8
                },
                "29": {
                    y: 47.35
                },
                "30": {
                    y: 64.85
                },
                "31": {
                    y: 86.5
                },
                "32": {
                    y: 112.5
                },
                "33": {
                    y: 142.8
                },
                "34": {
                    y: 176.9
                },
                "35": {
                    y: 213.9
                },
                "36": {
                    y: 252.4
                },
                "37": {
                    y: 290.5
                },
                "38": {
                    y: 326.45
                },
                "39": {
                    y: 358.8
                },
                "40": {
                    y: 386.55
                },
                "41": {
                    y: 409.4
                },
                "42": {
                    y: 427.35
                },
                "43": {
                    y: 440.6
                },
                "44": {
                    y: 449.55
                },
                "45": {
                    y: 454.65
                },
                "46": {
                    y: 456.25
                },
                "79": {
                    y: 453.25
                },
                "80": {
                    y: 443.55
                },
                "81": {
                    y: 426.15
                },
                "82": {
                    y: 399.8
                },
                "83": {
                    y: 363.45
                },
                "84": {
                    y: 316.7
                },
                "85": {
                    y: 260.45
                },
                "86": {
                    y: 198.2
                },
                "87": {
                    y: 135.6
                },
                "88": {
                    y: 78.9
                },
                "89": {
                    y: 32.5
                },
                "90": {
                    y: -2.1
                },
                "91": {
                    y: -25.15
                },
                "92": {
                    y: -37.95
                },
                "93": {
                    y: -41.9
                },
                "94": {
                    y: -41.6
                },
                "95": {
                    y: -40.7
                },
                "96": {
                    y: -38.9
                },
                "97": {
                    y: -36.25
                },
                "98": {
                    y: -32.95
                },
                "99": {
                    y: -29.8
                },
                "100": {
                    y: -27.5
                },
                "101": {
                    y: -26.25
                },
                "102": {
                    y: -25.9
                },
                "116": {
                    y: -25.1
                },
                "117": {
                    y: -22.45
                },
                "118": {
                    y: -17.65
                },
                "119": {
                    y: -10.35
                },
                "120": {
                    y: -0.3
                },
                "121": {
                    y: 12.55
                },
                "122": {
                    y: 27.65
                },
                "123": {
                    y: 43.7
                },
                "124": {
                    y: 58.8
                },
                "125": {
                    y: 71.45
                },
                "126": {
                    y: 80.95
                },
                "127": {
                    y: 87.3
                },
                "128": {
                    y: 90.8
                },
                "129": {
                    y: 91.9
                },
                "172": {
                    y: 91.15
                },
                "173": {
                    y: 88.65
                },
                "174": {
                    y: 84.05
                },
                "175": {
                    y: 77
                },
                "176": {
                    y: 67.35
                },
                "177": {
                    y: 55.55
                },
                "178": {
                    y: 42.85
                },
                "179": {
                    y: 31.3
                },
                "180": {
                    y: 22.3
                },
                "181": {
                    y: 16.2
                },
                "182": {
                    y: 12.9
                },
                "183": {
                    y: 11.9
                }
            });
    });

    lib.idle_pong = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 184,
            framerate: 30
        });
        var instance1 = new Graphic5(MovieClip.SYNCHED)
            .setTransform(54.05, 66.25);
        this.addTimedChild(instance1);
    });

    lib.idle_pong.assets = {
        "idle_pong_atlas_1": "images/idle_pong_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.idle_pong,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 184,
        library: lib
    };
}
