// jibo-anim-db-animations/timelines/idle_paddleball.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 122, loop: false });
        var instance1 = new Sprite(fromFrame("paddle1"))
            .setTransform(-10.6, -32.8, 0.798, 0.798);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 112, loop: false });
        var instance1 = new Sprite(fromFrame("ball1"))
            .setTransform(-72.5, -72.5);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 3, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 4, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 4, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 4, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 4, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 4, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 4, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 7, loop: false });
        var instance1 = new Sprite(fromFrame("paddle-shadow1"))
            .setTransform(-28.5, -13.85);
        this.addTimedChild(instance1);
    });

    lib.idle_paddleball = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 123,
            framerate: 30,
            loop: false
        });
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        var instance3 = new Graphic3(MovieClip.SYNCHED);
        var instance4 = new Graphic4(MovieClip.SYNCHED);
        var instance5 = new Graphic5(MovieClip.SYNCHED);
        var instance6 = new Graphic6(MovieClip.SYNCHED);
        var instance7 = new Graphic7(MovieClip.SYNCHED);
        var instance8 = new Graphic8(MovieClip.SYNCHED);
        var instance9 = new Graphic9(MovieClip.SYNCHED);
        var instance10 = new Graphic10(MovieClip.SYNCHED);
        var instance11 = new Graphic11(MovieClip.SYNCHED);
        var instance12 = new Graphic12(MovieClip.SYNCHED);
        var instance13 = new Graphic13(MovieClip.SYNCHED);
        var instance14 = new Graphic14(MovieClip.SYNCHED);
        var instance15 = new Graphic15(MovieClip.SYNCHED);
        var instance16 = new Graphic16(MovieClip.SYNCHED);
        var instance2 = new Graphic2(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 122, {
                "0": {
                    x: 684,
                    y: 916.9,
                    sx: 1.254,
                    sy: 1.062,
                    r: -2.121,
                    a: 0.07,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "1": {
                    x: 678.815,
                    y: 908.068,
                    sx: 1.252,
                    sy: 1.06,
                    r: -2.104,
                    a: 0.1
                },
                "2": {
                    x: 659.362,
                    y: 877.513,
                    r: -2.043,
                    a: 0.19
                },
                "3": {
                    x: 624.953,
                    y: 816.442,
                    sy: 1.061,
                    r: -1.925,
                    a: 0.38
                },
                "4": {
                    x: 585.326,
                    y: 729.032,
                    sx: 1.253,
                    r: -1.768,
                    a: 0.62
                },
                "5": {
                    x: 557,
                    y: 647.299,
                    sy: 1.062,
                    r: -1.629,
                    a: 0.84
                },
                "6": {
                    x: 544.399,
                    y: 600.91,
                    sx: 1.254,
                    r: -1.553,
                    a: 0.96
                },
                "7": {
                    x: 540.95,
                    y: 586.6,
                    r: -1.529,
                    a: 1
                },
                "8": {
                    x: 538.9,
                    y: 581.5,
                    r: -1.508
                },
                "9": {
                    x: 529.45,
                    y: 578.45,
                    r: -1.421
                },
                "10": {
                    x: 526.2,
                    y: 591,
                    r: -1.387
                },
                "11": {
                    x: 528.15,
                    y: 608.7,
                    r: -1.421
                },
                "12": {
                    x: 534.05,
                    y: 638.5,
                    r: -1.499
                },
                "14": {
                    x: 536.45,
                    y: 647.3,
                    r: -1.525
                },
                "16": {
                    x: 542.4,
                    y: 664.15,
                    r: -1.58
                },
                "18": {
                    x: 547.8,
                    y: 668.55,
                    r: -1.628
                },
                "19": {
                    x: 558.75,
                    y: 636.8,
                    r: -1.727
                },
                "20": {
                    x: 549.4,
                    y: 617.1,
                    r: -1.654
                },
                "21": {
                    x: 540.95,
                    y: 586.6,
                    r: -1.529
                },
                "22": {
                    x: 538.9,
                    y: 581.5,
                    r: -1.508
                },
                "23": {
                    x: 529.45,
                    y: 578.45,
                    r: -1.421
                },
                "24": {
                    x: 526.2,
                    y: 591,
                    r: -1.387
                },
                "25": {
                    x: 528.15,
                    y: 608.7,
                    r: -1.421
                },
                "26": {
                    x: 534.05,
                    y: 638.5,
                    r: -1.499
                },
                "28": {
                    x: 536.45,
                    y: 647.3,
                    r: -1.525
                },
                "30": {
                    x: 542.4,
                    y: 664.15,
                    r: -1.58
                },
                "32": {
                    x: 547.8,
                    y: 668.55,
                    r: -1.628
                },
                "33": {
                    x: 558.75,
                    y: 636.8,
                    r: -1.727
                },
                "34": {
                    x: 549.4,
                    y: 617.1,
                    r: -1.654
                },
                "35": {
                    x: 540.95,
                    y: 586.6,
                    r: -1.529
                },
                "36": {
                    x: 538.9,
                    y: 581.5,
                    r: -1.508
                },
                "37": {
                    x: 529.45,
                    y: 578.45,
                    r: -1.421
                },
                "38": {
                    x: 526.2,
                    y: 591,
                    r: -1.387
                },
                "39": {
                    x: 528.15,
                    y: 608.7,
                    r: -1.421
                },
                "40": {
                    x: 534.05,
                    y: 638.5,
                    r: -1.499
                },
                "42": {
                    x: 536.45,
                    y: 647.3,
                    r: -1.525
                },
                "44": {
                    x: 542.4,
                    y: 664.15,
                    r: -1.58
                },
                "46": {
                    x: 547.8,
                    y: 668.55,
                    r: -1.628
                },
                "47": {
                    x: 558.75,
                    y: 636.8,
                    r: -1.727
                },
                "48": {
                    x: 549.4,
                    y: 617.1,
                    r: -1.654
                },
                "49": {
                    x: 540.95,
                    y: 586.6,
                    r: -1.529
                },
                "50": {
                    x: 538.9,
                    y: 581.5,
                    r: -1.508
                },
                "51": {
                    x: 529.45,
                    y: 578.45,
                    r: -1.421
                },
                "52": {
                    x: 526.2,
                    y: 591,
                    r: -1.387
                },
                "53": {
                    x: 528.15,
                    y: 608.7,
                    r: -1.421
                },
                "54": {
                    x: 534.05,
                    y: 638.5,
                    r: -1.499
                },
                "56": {
                    x: 536.45,
                    y: 647.3,
                    r: -1.525
                },
                "58": {
                    x: 542.4,
                    y: 664.15,
                    r: -1.58
                },
                "60": {
                    x: 547.8,
                    y: 668.55,
                    r: -1.628
                },
                "61": {
                    x: 558.75,
                    y: 636.8,
                    r: -1.727
                },
                "62": {
                    x: 549.4,
                    y: 617.1,
                    r: -1.654
                },
                "63": {
                    x: 540.95,
                    y: 586.6,
                    r: -1.529
                },
                "64": {
                    x: 538.9,
                    y: 581.5,
                    r: -1.508
                },
                "65": {
                    x: 529.45,
                    y: 578.45,
                    r: -1.421
                },
                "66": {
                    x: 526.2,
                    y: 591,
                    r: -1.387
                },
                "67": {
                    x: 528.15,
                    y: 608.7,
                    r: -1.421
                },
                "68": {
                    x: 534.05,
                    y: 638.5,
                    r: -1.499
                },
                "70": {
                    x: 536.45,
                    y: 647.3,
                    r: -1.525
                },
                "72": {
                    x: 542.4,
                    y: 664.15,
                    r: -1.58
                },
                "74": {
                    x: 547.8,
                    y: 668.55,
                    r: -1.628
                },
                "75": {
                    x: 558.75,
                    y: 636.8,
                    r: -1.727
                },
                "76": {
                    x: 549.4,
                    y: 617.1,
                    r: -1.654
                },
                "77": {
                    x: 540.95,
                    y: 586.6,
                    r: -1.529
                },
                "78": {
                    x: 538.9,
                    y: 581.5,
                    r: -1.508
                },
                "79": {
                    x: 529.45,
                    y: 578.45,
                    r: -1.421
                },
                "80": {
                    x: 526.2,
                    y: 591,
                    r: -1.387
                },
                "81": {
                    x: 528.15,
                    y: 608.7,
                    r: -1.421
                },
                "82": {
                    x: 534.05,
                    y: 638.5,
                    r: -1.499
                },
                "84": {
                    x: 536.45,
                    y: 647.3,
                    r: -1.525
                },
                "86": {
                    x: 542.4,
                    y: 664.15,
                    r: -1.58
                },
                "88": {
                    x: 547.8,
                    y: 668.55,
                    r: -1.628
                },
                "89": {
                    x: 558.75,
                    y: 636.8,
                    r: -1.727
                },
                "90": {
                    x: 549.4,
                    y: 617.1,
                    r: -1.654
                },
                "91": {
                    x: 540.95,
                    y: 586.6,
                    r: -1.529
                },
                "92": {
                    x: 538.9,
                    y: 581.5,
                    r: -1.508
                },
                "93": {
                    x: 529.45,
                    y: 578.45,
                    r: -1.421
                },
                "94": {
                    x: 526.2,
                    y: 591,
                    r: -1.387
                },
                "95": {
                    x: 528.15,
                    y: 608.7,
                    r: -1.421
                },
                "96": {
                    x: 534.05,
                    y: 638.5,
                    r: -1.499
                },
                "98": {
                    x: 536.45,
                    y: 647.3,
                    r: -1.525
                },
                "100": {
                    x: 542.4,
                    y: 664.15,
                    r: -1.58
                },
                "102": {
                    x: 547.8,
                    y: 668.55,
                    r: -1.628
                },
                "103": {
                    x: 558.75,
                    y: 636.8,
                    r: -1.727
                },
                "104": {
                    x: 549.4,
                    y: 617.1,
                    r: -1.654
                },
                "105": {
                    x: 540.95,
                    y: 586.6,
                    r: -1.529
                },
                "106": {
                    x: 538.9,
                    y: 581.5,
                    r: -1.508
                },
                "107": {
                    x: 529.45,
                    y: 578.45,
                    r: -1.421
                },
                "108": {
                    x: 526.2,
                    y: 591,
                    r: -1.387
                },
                "109": {
                    x: 528.15,
                    y: 608.7,
                    r: -1.421
                },
                "110": {
                    x: 534.05,
                    y: 638.5,
                    r: -1.499
                },
                "112": {
                    x: 536.45,
                    y: 647.3,
                    r: -1.525
                },
                "114": {
                    x: 537.053,
                    y: 651.159,
                    sx: 1.253,
                    r: -1.531,
                    c: [
                        0.98,
                        0,
                        0.98,
                        0,
                        0.98,
                        0
                    ]
                },
                "115": {
                    x: 538.4,
                    y: 662.391,
                    r: -1.544,
                    c: [
                        0.93,
                        0,
                        0.93,
                        0,
                        0.93,
                        0
                    ]
                },
                "116": {
                    x: 541.305,
                    y: 686.049,
                    sx: 1.254,
                    r: -1.571,
                    c: [
                        0.81,
                        0,
                        0.81,
                        0,
                        0.81,
                        0
                    ]
                },
                "117": {
                    x: 547.848,
                    y: 730.285,
                    sx: 1.253,
                    r: -1.624,
                    c: [
                        0.59,
                        0,
                        0.59,
                        0,
                        0.59,
                        0
                    ]
                },
                "118": {
                    x: 557.894,
                    y: 787.174,
                    sy: 1.061,
                    r: -1.694,
                    c: [
                        0.3,
                        0,
                        0.3,
                        0,
                        0.3,
                        0
                    ]
                },
                "119": {
                    x: 565.6,
                    y: 825.12,
                    r: -1.741,
                    c: [
                        0.11,
                        0,
                        0.11,
                        0,
                        0.11,
                        0
                    ]
                },
                "120": {
                    x: 568.761,
                    y: 841.104,
                    r: -1.76,
                    c: [
                        0.02,
                        0,
                        0.02,
                        0,
                        0.02,
                        0
                    ]
                },
                "121": {
                    x: 570.05,
                    y: 846.4,
                    sx: 1.254,
                    sy: 1.062,
                    r: -1.768,
                    c: [
                        0,
                        0,
                        0,
                        0,
                        0,
                        0
                    ]
                }
            })
            .addTimedChild(instance3, 9, 3, {
                "9": {
                    x: 657.5,
                    y: 497.45,
                    a: 0.54
                },
                "10": {
                    y: 510.7,
                    a: 0.3
                },
                "11": {
                    y: 527.05,
                    a: 0.08
                }
            })
            .addTimedChild(instance4, 19, 2, {
                "19": {
                    x: 657.5,
                    y: 515.9,
                    sx: 1,
                    sy: 1,
                    r: -0.078,
                    a: 0.34
                },
                "20": {
                    y: 509.7,
                    sx: 1.14,
                    sy: 1.14,
                    a: 0.54
                }
            })
            .addTimedChild(instance5, 22, 4, {
                "22": {
                    x: 657.5,
                    y: 489.65,
                    sx: 1.461,
                    sy: 1.165,
                    r: 0.025,
                    a: 0.54
                },
                "23": {
                    y: 497.45,
                    sx: 1.081,
                    sy: 1.018,
                    r: 0.099
                },
                "24": {
                    y: 510.7,
                    sx: 0.902,
                    sy: 0.809,
                    r: 0.118,
                    a: 0.3
                },
                "25": {
                    y: 527.05,
                    sx: 0.918,
                    sy: 0.932,
                    r: 0.052,
                    a: 0.08
                }
            })
            .addTimedChild(instance6, 33, 2, {
                "33": {
                    x: 657.5,
                    y: 515.9,
                    sx: 1,
                    sy: 1,
                    r: -0.078,
                    a: 0.34
                },
                "34": {
                    y: 509.7,
                    sx: 1.14,
                    sy: 1.14,
                    a: 0.54
                }
            })
            .addTimedChild(instance7, 36, 4, {
                "36": {
                    x: 657.5,
                    y: 489.65,
                    sx: 1.461,
                    sy: 1.165,
                    r: 0.025,
                    a: 0.54
                },
                "37": {
                    y: 497.45,
                    sx: 1.081,
                    sy: 1.018,
                    r: 0.099
                },
                "38": {
                    y: 510.7,
                    sx: 0.902,
                    sy: 0.809,
                    r: 0.118,
                    a: 0.3
                },
                "39": {
                    y: 527.05,
                    sx: 0.918,
                    sy: 0.932,
                    r: 0.052,
                    a: 0.08
                }
            })
            .addTimedChild(instance8, 47, 2, {
                "47": {
                    x: 657.5,
                    y: 515.9,
                    sx: 1,
                    sy: 1,
                    r: -0.078,
                    a: 0.34
                },
                "48": {
                    y: 509.7,
                    sx: 1.14,
                    sy: 1.14,
                    a: 0.54
                }
            })
            .addTimedChild(instance9, 50, 4, {
                "50": {
                    x: 657.5,
                    y: 489.65,
                    sx: 1.461,
                    sy: 1.165,
                    r: 0.025,
                    a: 0.54
                },
                "51": {
                    y: 497.45,
                    sx: 1.081,
                    sy: 1.018,
                    r: 0.099
                },
                "52": {
                    y: 510.7,
                    sx: 0.902,
                    sy: 0.809,
                    r: 0.118,
                    a: 0.3
                },
                "53": {
                    y: 527.05,
                    sx: 0.918,
                    sy: 0.932,
                    r: 0.052,
                    a: 0.08
                }
            })
            .addTimedChild(instance10, 61, 2, {
                "61": {
                    x: 657.5,
                    y: 515.9,
                    sx: 1,
                    sy: 1,
                    r: -0.078,
                    a: 0.34
                },
                "62": {
                    y: 509.7,
                    sx: 1.14,
                    sy: 1.14,
                    a: 0.54
                }
            })
            .addTimedChild(instance11, 64, 4, {
                "64": {
                    x: 657.5,
                    y: 489.65,
                    sx: 1.461,
                    sy: 1.165,
                    r: 0.025,
                    a: 0.54
                },
                "65": {
                    y: 497.45,
                    sx: 1.081,
                    sy: 1.018,
                    r: 0.099
                },
                "66": {
                    y: 510.7,
                    sx: 0.902,
                    sy: 0.809,
                    r: 0.118,
                    a: 0.3
                },
                "67": {
                    y: 527.05,
                    sx: 0.918,
                    sy: 0.932,
                    r: 0.052,
                    a: 0.08
                }
            })
            .addTimedChild(instance12, 75, 2, {
                "75": {
                    x: 657.5,
                    y: 515.9,
                    sx: 1,
                    sy: 1,
                    r: -0.078,
                    a: 0.34
                },
                "76": {
                    y: 509.7,
                    sx: 1.14,
                    sy: 1.14,
                    a: 0.54
                }
            })
            .addTimedChild(instance13, 78, 4, {
                "78": {
                    x: 657.5,
                    y: 489.65,
                    sx: 1.461,
                    sy: 1.165,
                    r: 0.025,
                    a: 0.54
                },
                "79": {
                    y: 497.45,
                    sx: 1.081,
                    sy: 1.018,
                    r: 0.099
                },
                "80": {
                    y: 510.7,
                    sx: 0.902,
                    sy: 0.809,
                    r: 0.118,
                    a: 0.3
                },
                "81": {
                    y: 527.05,
                    sx: 0.918,
                    sy: 0.932,
                    r: 0.052,
                    a: 0.08
                }
            })
            .addTimedChild(instance14, 89, 2, {
                "89": {
                    x: 657.5,
                    y: 515.9,
                    sx: 1,
                    sy: 1,
                    r: -0.078,
                    a: 0.34
                },
                "90": {
                    y: 509.7,
                    sx: 1.14,
                    sy: 1.14,
                    a: 0.54
                }
            })
            .addTimedChild(instance15, 92, 4, {
                "92": {
                    x: 657.5,
                    y: 489.65,
                    sx: 1.461,
                    sy: 1.165,
                    r: 0.025,
                    a: 0.54
                },
                "93": {
                    y: 497.45,
                    sx: 1.081,
                    sy: 1.018,
                    r: 0.099
                },
                "94": {
                    y: 510.7,
                    sx: 0.902,
                    sy: 0.809,
                    r: 0.118,
                    a: 0.3
                },
                "95": {
                    y: 527.05,
                    sx: 0.918,
                    sy: 0.932,
                    r: 0.052,
                    a: 0.08
                }
            })
            .addTimedChild(instance16, 103, 7, {
                "103": {
                    x: 657.5,
                    y: 515.9,
                    sx: 1,
                    sy: 1,
                    r: -0.078,
                    a: 0.34
                },
                "104": {
                    y: 509.7,
                    sx: 1.14,
                    sy: 1.14,
                    a: 0.54
                },
                "105": {
                    y: 488.9,
                    sx: 1.756,
                    r: -0.014
                },
                "106": {
                    y: 489.65,
                    sx: 1.461,
                    sy: 1.165,
                    r: 0.025
                },
                "107": {
                    y: 497.45,
                    sx: 1.081,
                    sy: 1.018,
                    r: 0.099
                },
                "108": {
                    y: 510.7,
                    sx: 0.902,
                    sy: 0.809,
                    r: 0.118,
                    a: 0.3
                },
                "109": {
                    y: 527.05,
                    sx: 0.918,
                    sy: 0.932,
                    r: 0.052,
                    a: 0.08
                }
            })
            .addTimedChild(instance2, 2, 112, {
                "2": {
                    x: 627.65,
                    y: 98.3,
                    sx: 0.195,
                    sy: 0.195
                },
                "3": {
                    x: 633.5,
                    y: 186.55,
                    sx: 0.155,
                    sy: 0.236
                },
                "4": {
                    y: 262.05
                },
                "5": {
                    y: 337.55
                },
                "6": {
                    y: 413.05
                },
                "7": {
                    x: 621.75,
                    y: 457.75,
                    sx: 0.236,
                    sy: 0.114
                },
                "8": {
                    x: 627.6,
                    y: 438.2,
                    sx: 0.196,
                    sy: 0.168
                },
                "9": {
                    x: 635.45,
                    y: 326.85,
                    sx: 0.142,
                    sy: 0.209
                },
                "10": {
                    y: 162.8
                },
                "11": {
                    x: 631.65,
                    y: 104.25,
                    sx: 0.168,
                    sy: 0.205
                },
                "12": {
                    x: 627.65,
                    y: 83.55,
                    sx: 0.195,
                    sy: 0.195
                },
                "14": {
                    y: 74.45,
                    sy: 0.192
                },
                "16": {
                    x: 626.7,
                    y: 82.75,
                    sx: 0.202,
                    sy: 0.189
                },
                "17": {
                    x: 627.65,
                    y: 98.3,
                    sx: 0.195,
                    sy: 0.195
                },
                "18": {
                    x: 633.5,
                    y: 186.55,
                    sx: 0.155,
                    sy: 0.236
                },
                "19": {
                    y: 360.6
                },
                "20": {
                    y: 413.05
                },
                "21": {
                    x: 621.75,
                    y: 457.75,
                    sx: 0.236,
                    sy: 0.114
                },
                "22": {
                    x: 627.6,
                    y: 438.2,
                    sx: 0.196,
                    sy: 0.168
                },
                "23": {
                    x: 635.45,
                    y: 326.85,
                    sx: 0.142,
                    sy: 0.209
                },
                "24": {
                    y: 162.8
                },
                "25": {
                    x: 631.65,
                    y: 104.25,
                    sx: 0.168,
                    sy: 0.205
                },
                "26": {
                    x: 627.65,
                    y: 83.55,
                    sx: 0.195,
                    sy: 0.195
                },
                "28": {
                    y: 74.45,
                    sy: 0.192
                },
                "30": {
                    x: 626.7,
                    y: 82.75,
                    sx: 0.202,
                    sy: 0.189
                },
                "31": {
                    x: 627.65,
                    y: 98.3,
                    sx: 0.195,
                    sy: 0.195
                },
                "32": {
                    x: 633.5,
                    y: 186.55,
                    sx: 0.155,
                    sy: 0.236
                },
                "33": {
                    y: 360.6
                },
                "34": {
                    y: 413.05
                },
                "35": {
                    x: 621.75,
                    y: 457.75,
                    sx: 0.236,
                    sy: 0.114
                },
                "36": {
                    x: 627.6,
                    y: 438.2,
                    sx: 0.196,
                    sy: 0.168
                },
                "37": {
                    x: 635.45,
                    y: 326.85,
                    sx: 0.142,
                    sy: 0.209
                },
                "38": {
                    y: 162.8
                },
                "39": {
                    x: 631.65,
                    y: 104.25,
                    sx: 0.168,
                    sy: 0.205
                },
                "40": {
                    x: 627.65,
                    y: 83.55,
                    sx: 0.195,
                    sy: 0.195
                },
                "42": {
                    y: 74.45,
                    sy: 0.192
                },
                "44": {
                    x: 626.7,
                    y: 82.75,
                    sx: 0.202,
                    sy: 0.189
                },
                "45": {
                    x: 627.65,
                    y: 98.3,
                    sx: 0.195,
                    sy: 0.195
                },
                "46": {
                    x: 633.5,
                    y: 186.55,
                    sx: 0.155,
                    sy: 0.236
                },
                "47": {
                    y: 360.6
                },
                "48": {
                    y: 413.05
                },
                "49": {
                    x: 621.75,
                    y: 457.75,
                    sx: 0.236,
                    sy: 0.114
                },
                "50": {
                    x: 627.6,
                    y: 438.2,
                    sx: 0.196,
                    sy: 0.168
                },
                "51": {
                    x: 635.45,
                    y: 326.85,
                    sx: 0.142,
                    sy: 0.209
                },
                "52": {
                    y: 162.8
                },
                "53": {
                    x: 631.65,
                    y: 104.25,
                    sx: 0.168,
                    sy: 0.205
                },
                "54": {
                    x: 627.65,
                    y: 83.55,
                    sx: 0.195,
                    sy: 0.195
                },
                "56": {
                    y: 74.45,
                    sy: 0.192
                },
                "58": {
                    x: 626.7,
                    y: 82.75,
                    sx: 0.202,
                    sy: 0.189
                },
                "59": {
                    x: 627.65,
                    y: 98.3,
                    sx: 0.195,
                    sy: 0.195
                },
                "60": {
                    x: 633.5,
                    y: 186.55,
                    sx: 0.155,
                    sy: 0.236
                },
                "61": {
                    y: 360.6
                },
                "62": {
                    y: 413.05
                },
                "63": {
                    x: 621.75,
                    y: 457.75,
                    sx: 0.236,
                    sy: 0.114
                },
                "64": {
                    x: 627.6,
                    y: 438.2,
                    sx: 0.196,
                    sy: 0.168
                },
                "65": {
                    x: 635.45,
                    y: 326.85,
                    sx: 0.142,
                    sy: 0.209
                },
                "66": {
                    y: 162.8
                },
                "67": {
                    x: 631.65,
                    y: 104.25,
                    sx: 0.168,
                    sy: 0.205
                },
                "68": {
                    x: 627.65,
                    y: 83.55,
                    sx: 0.195,
                    sy: 0.195
                },
                "70": {
                    y: 74.45,
                    sy: 0.192
                },
                "72": {
                    x: 626.7,
                    y: 82.75,
                    sx: 0.202,
                    sy: 0.189
                },
                "73": {
                    x: 627.65,
                    y: 98.3,
                    sx: 0.195,
                    sy: 0.195
                },
                "74": {
                    x: 633.5,
                    y: 186.55,
                    sx: 0.155,
                    sy: 0.236
                },
                "75": {
                    y: 360.6
                },
                "76": {
                    y: 413.05
                },
                "77": {
                    x: 621.75,
                    y: 457.75,
                    sx: 0.236,
                    sy: 0.114
                },
                "78": {
                    x: 627.6,
                    y: 438.2,
                    sx: 0.196,
                    sy: 0.168
                },
                "79": {
                    x: 635.45,
                    y: 326.85,
                    sx: 0.142,
                    sy: 0.209
                },
                "80": {
                    y: 162.8
                },
                "81": {
                    x: 631.65,
                    y: 104.25,
                    sx: 0.168,
                    sy: 0.205
                },
                "82": {
                    x: 627.65,
                    y: 83.55,
                    sx: 0.195,
                    sy: 0.195
                },
                "84": {
                    y: 74.45,
                    sy: 0.192
                },
                "86": {
                    x: 626.7,
                    y: 82.75,
                    sx: 0.202,
                    sy: 0.189
                },
                "87": {
                    x: 627.65,
                    y: 98.3,
                    sx: 0.195,
                    sy: 0.195
                },
                "88": {
                    x: 633.5,
                    y: 186.55,
                    sx: 0.155,
                    sy: 0.236
                },
                "89": {
                    y: 360.6
                },
                "90": {
                    y: 413.05
                },
                "91": {
                    x: 621.75,
                    y: 457.75,
                    sx: 0.236,
                    sy: 0.114
                },
                "92": {
                    x: 627.6,
                    y: 438.2,
                    sx: 0.196,
                    sy: 0.168
                },
                "93": {
                    x: 635.45,
                    y: 326.85,
                    sx: 0.142,
                    sy: 0.209
                },
                "94": {
                    y: 162.8
                },
                "95": {
                    x: 631.65,
                    y: 104.25,
                    sx: 0.168,
                    sy: 0.205
                },
                "96": {
                    x: 627.65,
                    y: 83.55,
                    sx: 0.195,
                    sy: 0.195
                },
                "98": {
                    y: 74.45,
                    sy: 0.192
                },
                "100": {
                    x: 626.7,
                    y: 82.75,
                    sx: 0.202,
                    sy: 0.189
                },
                "101": {
                    x: 627.65,
                    y: 98.3,
                    sx: 0.195,
                    sy: 0.195
                },
                "102": {
                    x: 633.5,
                    y: 186.55,
                    sx: 0.155,
                    sy: 0.236
                },
                "103": {
                    y: 360.6
                },
                "104": {
                    y: 413.05
                },
                "105": {
                    x: 621.75,
                    y: 457.75,
                    sx: 0.236,
                    sy: 0.114
                },
                "106": {
                    x: 627.6,
                    y: 438.2,
                    sx: 0.196,
                    sy: 0.168
                },
                "107": {
                    x: 635.45,
                    y: 326.85,
                    sx: 0.142,
                    sy: 0.209
                },
                "108": {
                    y: 162.8
                },
                "109": {
                    x: 631.65,
                    y: 104.25,
                    sx: 0.168,
                    sy: 0.205
                },
                "110": {
                    x: 627.65,
                    y: 83.55,
                    sx: 0.195,
                    sy: 0.195
                },
                "112": {
                    y: 74.45,
                    sy: 0.192
                }
            });
    });

    lib.idle_paddleball.assets = {
        "idle_paddleball_atlas_1": "images/idle_paddleball_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.idle_paddleball,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 123,
        library: lib
    };
}
