// jibo-anim-db-animations/timelines/jiboji_mothers_day_stork.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Sprite(fromFrame("stork-headnest1"))
            .setTransform(-44.2, -44.2);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 70, loop: false });
        var instance1 = new Sprite(fromFrame("stork-bundle1"))
            .setTransform(-71.45, -109.55);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 70, loop: false });
        var instance1 = new Sprite(fromFrame("stork-beak1"))
            .setTransform(-84.25, -6);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 54, loop: false });
        var instance1 = new Sprite(fromFrame("stork-wing4"));
        var instance2 = new Sprite(fromFrame("stork-wing3"));
        var instance3 = new Sprite(fromFrame("stork-wing10"));
        var instance4 = new Sprite(fromFrame("stork-wing11"));
        var instance5 = new Sprite(fromFrame("stork-wing1"));
        var instance6 = new Sprite(fromFrame("stork-wing9"));
        var instance7 = new Sprite(fromFrame("stork-wing2"));
        var instance8 = new Sprite(fromFrame("stork-wing3"));
        var instance9 = new Sprite(fromFrame("stork-wing4"));
        var instance10 = new Sprite(fromFrame("stork-wing6"));
        var instance11 = new Sprite(fromFrame("stork-wing5"));
        var instance12 = new Sprite(fromFrame("stork-wing6"));
        var instance13 = new Sprite(fromFrame("stork-wing7"));
        var instance14 = new Sprite(fromFrame("stork-wing4"));
        var instance15 = new Sprite(fromFrame("stork-wing3"));
        var instance16 = new Sprite(fromFrame("stork-wing10"));
        var instance17 = new Sprite(fromFrame("stork-wing11"));
        var instance18 = new Sprite(fromFrame("stork-wing1"));
        var instance19 = new Sprite(fromFrame("stork-wing9"));
        var instance20 = new Sprite(fromFrame("stork-wing2"));
        var instance21 = new Sprite(fromFrame("stork-wing3"));
        var instance22 = new Sprite(fromFrame("stork-wing4"));
        var instance23 = new Sprite(fromFrame("stork-wing6"));
        var instance24 = new Sprite(fromFrame("stork-wing5"));
        var instance25 = new Sprite(fromFrame("stork-wing6"));
        var instance26 = new Sprite(fromFrame("stork-wing7"));
        var instance27 = new Sprite(fromFrame("stork-wing4"));
        var instance28 = new Sprite(fromFrame("stork-wing3"));
        var instance29 = new Sprite(fromFrame("stork-wing10"));
        var instance30 = new Sprite(fromFrame("stork-wing11"));
        var instance31 = new Sprite(fromFrame("stork-wing1"));
        var instance32 = new Sprite(fromFrame("stork-wing9"));
        var instance33 = new Sprite(fromFrame("stork-wing2"));
        var instance34 = new Sprite(fromFrame("stork-wing3"));
        var instance35 = new Sprite(fromFrame("stork-wing4"))
            .setTransform(-75.2, 110.45);
        this.addTimedChild(instance1, 0, 2, {
                "0": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance3, 3, 2, {
                "3": {
                    x: -109.55,
                    y: -50.95
                }
            })
            .addTimedChild(instance4, 5, 1, {
                "5": {
                    x: -109.55,
                    y: -98.4
                }
            })
            .addTimedChild(instance5, 6, 2, {
                "6": {
                    x: -109.55,
                    y: -108.05
                }
            })
            .addTimedChild(instance6, 8, 2, {
                "8": {
                    x: -109.55,
                    y: -119.5
                }
            })
            .addTimedChild(instance7, 10, 1, {
                "10": {
                    x: -109.55,
                    y: -81.55
                }
            })
            .addTimedChild(instance8, 11, 1, {
                "11": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance9, 12, 1, {
                "12": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance10, 13, 1, {
                "13": {
                    x: -109.55,
                    y: 114.95
                }
            })
            .addTimedChild(instance11, 14, 3, {
                "14": {
                    x: -109.55,
                    y: 99.9
                }
            })
            .addTimedChild(instance12, 17, 1, {
                "17": {
                    x: -109.55,
                    y: 114.95
                }
            })
            .addTimedChild(instance13, 18, 1, {
                "18": {
                    x: -109.55,
                    y: 107.45
                }
            })
            .addTimedChild(instance14, 19, 2, {
                "19": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance15, 21, 1, {
                "21": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance16, 22, 2, {
                "22": {
                    x: -109.55,
                    y: -50.95
                }
            })
            .addTimedChild(instance17, 24, 1, {
                "24": {
                    x: -109.55,
                    y: -98.4
                }
            })
            .addTimedChild(instance18, 25, 2, {
                "25": {
                    x: -109.55,
                    y: -108.05
                }
            })
            .addTimedChild(instance19, 27, 2, {
                "27": {
                    x: -109.55,
                    y: -119.5
                }
            })
            .addTimedChild(instance20, 29, 1, {
                "29": {
                    x: -109.55,
                    y: -81.55
                }
            })
            .addTimedChild(instance21, 30, 1, {
                "30": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance22, 31, 1, {
                "31": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance23, 32, 1, {
                "32": {
                    x: -109.55,
                    y: 114.95
                }
            })
            .addTimedChild(instance24, 33, 3, {
                "33": {
                    x: -109.55,
                    y: 99.9
                }
            })
            .addTimedChild(instance25, 36, 1, {
                "36": {
                    x: -109.55,
                    y: 114.95
                }
            })
            .addTimedChild(instance26, 37, 1, {
                "37": {
                    x: -109.55,
                    y: 107.45
                }
            })
            .addTimedChild(instance27, 38, 2, {
                "38": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance28, 40, 1, {
                "40": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance29, 41, 2, {
                "41": {
                    x: -109.55,
                    y: -50.95
                }
            })
            .addTimedChild(instance30, 43, 1, {
                "43": {
                    x: -109.55,
                    y: -98.4
                }
            })
            .addTimedChild(instance31, 44, 2, {
                "44": {
                    x: -109.55,
                    y: -108.05
                }
            })
            .addTimedChild(instance32, 46, 2, {
                "46": {
                    x: -109.55,
                    y: -119.5
                }
            })
            .addTimedChild(instance33, 48, 1, {
                "48": {
                    x: -109.55,
                    y: -81.55
                }
            })
            .addTimedChild(instance34, 49, 1, {
                "49": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance35, 50, 4);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 54, loop: false });
        var instance1 = new Graphic4(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 54, {
            "0": {
                x: -8.8,
                y: -124.75,
                sx: 1,
                ky: 0,
                r: 0.078
            },
            "2": {
                x: 0.75,
                y: -112.4,
                r: 0.156
            },
            "3": {
                x: -17.15,
                y: -130.95,
                r: 0
            },
            "8": {
                y: -122.6
            },
            "10": {
                y: -130.95
            },
            "11": {
                x: -24.3,
                y: -136.95,
                r: -0.057
            },
            "12": {
                x: -25.8,
                y: -135
            },
            "13": {
                x: -17.15,
                y: -137.2,
                sx: 1.002,
                ky: -0.057,
                r: 0
            },
            "14": {
                y: -135.15,
                sx: 1.001,
                ky: -0.038
            },
            "15": {
                y: -130.95,
                sx: 1,
                ky: 0
            },
            "18": {
                x: -6.55,
                y: -119,
                r: 0.105
            },
            "19": {
                x: -8.8,
                y: -124.75,
                r: 0.078
            },
            "21": {
                x: 0.75,
                y: -112.4,
                r: 0.156
            },
            "22": {
                x: -17.15,
                y: -130.95,
                r: 0
            },
            "27": {
                y: -122.6
            },
            "29": {
                y: -130.95
            },
            "30": {
                x: -24.3,
                y: -136.95,
                r: -0.057
            },
            "31": {
                x: -25.8,
                y: -135
            },
            "32": {
                x: -17.15,
                y: -137.2,
                sx: 1.002,
                ky: -0.057,
                r: 0
            },
            "33": {
                y: -135.15,
                sx: 1.001,
                ky: -0.038
            },
            "34": {
                y: -130.95,
                sx: 1,
                ky: 0
            },
            "37": {
                x: -6.55,
                y: -119,
                r: 0.105
            },
            "38": {
                x: -8.8,
                y: -124.75,
                r: 0.078
            },
            "40": {
                x: 0.75,
                y: -112.4,
                r: 0.156
            },
            "41": {
                x: -17.15,
                y: -130.95,
                r: 0
            },
            "46": {
                y: -122.6
            },
            "48": {
                y: -130.95
            },
            "49": {
                x: -24.3,
                y: -136.95,
                r: -0.057
            },
            "50": {
                x: -25.8,
                y: -135
            }
        });
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 81, loop: false });
        var instance1 = new Sprite(fromFrame("stork-neck1"))
            .setTransform(-81.8, -6);
        this.addTimedChild(instance1);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 71, loop: false });
        var instance1 = new Sprite(fromFrame("stork-body1"))
            .setTransform(-118.3, -47.75);
        this.addTimedChild(instance1);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 59, loop: false });
        var instance1 = new Sprite(fromFrame("stork-tail1"))
            .setTransform(-79.25, -18.3);
        this.addTimedChild(instance1);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance3 = new Graphic8(MovieClip.SYNCHED);
        var instance2 = new Graphic7(MovieClip.SYNCHED);
        var instance1 = new Graphic6(MovieClip.SYNCHED);
        this.addTimedChild(instance3, 14, 59, {
                "14": {
                    x: 62.4,
                    y: -29.55
                },
                "15": {
                    x: 69
                },
                "16": {
                    x: 93
                },
                "17": {
                    x: 133.25
                },
                "18": {
                    x: 164.35
                },
                "19": {
                    x: 173.4
                },
                "20": {
                    x: 172.35
                },
                "21": {
                    x: 168.45
                },
                "22": {
                    x: 161.9
                },
                "23": {
                    x: 156.85
                },
                "24": {
                    x: 155.4
                },
                "63": {
                    x: 156.05
                },
                "64": {
                    x: 158.15
                },
                "65": {
                    x: 161.7
                },
                "66": {
                    x: 165.65
                },
                "67": {
                    x: 168.5
                },
                "68": {
                    x: 169.4
                },
                "69": {
                    x: 158
                },
                "70": {
                    x: 121.8
                },
                "71": {
                    x: 79.35
                },
                "72": {
                    x: 63.4
                }
            })
            .addTimedChild(instance2, 7, 71, {
                "7": {
                    x: 26.55,
                    y: -41.4,
                    sy: 0.132
                },
                "8": {
                    y: -37.006,
                    sy: 0.224
                },
                "9": {
                    y: -21.035,
                    sy: 0.559
                },
                "10": {
                    y: -2.003,
                    sy: 0.958
                },
                "11": {
                    y: 4.05,
                    sy: 1.084
                },
                "12": {
                    y: 3.887,
                    sy: 1.08
                },
                "13": {
                    y: 3.293,
                    sy: 1.068
                },
                "14": {
                    y: 2.259,
                    sy: 1.046
                },
                "15": {
                    y: 1.064,
                    sy: 1.021
                },
                "16": {
                    y: 0.254,
                    sy: 1.005
                },
                "17": {
                    y: 0.05,
                    sy: 1
                },
                "69": {
                    y: 0.651,
                    sy: 1.013
                },
                "70": {
                    y: 1.251,
                    sy: 1.025
                },
                "71": {
                    y: 1.852,
                    sy: 1.038
                },
                "72": {
                    y: 2.453,
                    sy: 1.05
                },
                "73": {
                    y: 3.05,
                    sy: 1.063
                },
                "74": {
                    y: -8.655,
                    sy: 0.817
                },
                "75": {
                    y: -20.361,
                    sy: 0.571
                },
                "76": {
                    y: -32.116,
                    sy: 0.325
                },
                "77": {
                    y: -43.9,
                    sy: 0.08
                }
            })
            .addTimedChild(instance1, 0, 81, {
                "0": {
                    x: -219.8,
                    y: -41.5,
                    sx: 0.195
                },
                "1": {
                    x: -218.044,
                    sx: 0.219
                },
                "2": {
                    x: -211.848,
                    sx: 0.3
                },
                "3": {
                    x: -200.192,
                    sx: 0.454
                },
                "4": {
                    x: -184.162,
                    sx: 0.666
                },
                "5": {
                    x: -169.612,
                    sx: 0.858
                },
                "6": {
                    x: -161.207,
                    sx: 0.968
                },
                "7": {
                    x: -158.8,
                    sx: 1
                },
                "74": {
                    x: -160.7,
                    sx: 0.976
                },
                "75": {
                    x: -166.615,
                    sx: 0.897
                },
                "76": {
                    x: -177.035,
                    sx: 0.76
                },
                "77": {
                    x: -190.358,
                    sx: 0.584
                },
                "78": {
                    x: -202.977,
                    sx: 0.417
                },
                "79": {
                    x: -211.194,
                    sx: 0.309
                },
                "80": {
                    x: -213.8,
                    sx: 0.274
                }
            });
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 54, loop: false });
        var instance1 = new Sprite(fromFrame("stork-wing4"));
        var instance2 = new Sprite(fromFrame("stork-wing3"));
        var instance3 = new Sprite(fromFrame("stork-wing10"));
        var instance4 = new Sprite(fromFrame("stork-wing11"));
        var instance5 = new Sprite(fromFrame("stork-wing1"));
        var instance6 = new Sprite(fromFrame("stork-wing9"));
        var instance7 = new Sprite(fromFrame("stork-wing2"));
        var instance8 = new Sprite(fromFrame("stork-wing3"));
        var instance9 = new Sprite(fromFrame("stork-wing4"));
        var instance10 = new Sprite(fromFrame("stork-wing6"));
        var instance11 = new Sprite(fromFrame("stork-wing5"));
        var instance12 = new Sprite(fromFrame("stork-wing6"));
        var instance13 = new Sprite(fromFrame("stork-wing7"));
        var instance14 = new Sprite(fromFrame("stork-wing4"));
        var instance15 = new Sprite(fromFrame("stork-wing3"));
        var instance16 = new Sprite(fromFrame("stork-wing10"));
        var instance17 = new Sprite(fromFrame("stork-wing11"));
        var instance18 = new Sprite(fromFrame("stork-wing1"));
        var instance19 = new Sprite(fromFrame("stork-wing9"));
        var instance20 = new Sprite(fromFrame("stork-wing2"));
        var instance21 = new Sprite(fromFrame("stork-wing3"));
        var instance22 = new Sprite(fromFrame("stork-wing4"));
        var instance23 = new Sprite(fromFrame("stork-wing6"));
        var instance24 = new Sprite(fromFrame("stork-wing5"));
        var instance25 = new Sprite(fromFrame("stork-wing6"));
        var instance26 = new Sprite(fromFrame("stork-wing7"));
        var instance27 = new Sprite(fromFrame("stork-wing4"));
        var instance28 = new Sprite(fromFrame("stork-wing3"));
        var instance29 = new Sprite(fromFrame("stork-wing10"));
        var instance30 = new Sprite(fromFrame("stork-wing11"));
        var instance31 = new Sprite(fromFrame("stork-wing1"));
        var instance32 = new Sprite(fromFrame("stork-wing9"));
        var instance33 = new Sprite(fromFrame("stork-wing2"));
        var instance34 = new Sprite(fromFrame("stork-wing3"));
        var instance35 = new Sprite(fromFrame("stork-wing4"))
            .setTransform(-75.2, 110.45);
        this.addTimedChild(instance1, 0, 2, {
                "0": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance2, 2, 1, {
                "2": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance3, 3, 2, {
                "3": {
                    x: -109.55,
                    y: -50.95
                }
            })
            .addTimedChild(instance4, 5, 1, {
                "5": {
                    x: -109.55,
                    y: -98.4
                }
            })
            .addTimedChild(instance5, 6, 2, {
                "6": {
                    x: -109.55,
                    y: -108.05
                }
            })
            .addTimedChild(instance6, 8, 2, {
                "8": {
                    x: -109.55,
                    y: -119.5
                }
            })
            .addTimedChild(instance7, 10, 1, {
                "10": {
                    x: -109.55,
                    y: -81.55
                }
            })
            .addTimedChild(instance8, 11, 1, {
                "11": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance9, 12, 1, {
                "12": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance10, 13, 1, {
                "13": {
                    x: -109.55,
                    y: 114.95
                }
            })
            .addTimedChild(instance11, 14, 3, {
                "14": {
                    x: -109.55,
                    y: 99.9
                }
            })
            .addTimedChild(instance12, 17, 1, {
                "17": {
                    x: -109.55,
                    y: 114.95
                }
            })
            .addTimedChild(instance13, 18, 1, {
                "18": {
                    x: -109.55,
                    y: 107.45
                }
            })
            .addTimedChild(instance14, 19, 2, {
                "19": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance15, 21, 1, {
                "21": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance16, 22, 2, {
                "22": {
                    x: -109.55,
                    y: -50.95
                }
            })
            .addTimedChild(instance17, 24, 1, {
                "24": {
                    x: -109.55,
                    y: -98.4
                }
            })
            .addTimedChild(instance18, 25, 2, {
                "25": {
                    x: -109.55,
                    y: -108.05
                }
            })
            .addTimedChild(instance19, 27, 2, {
                "27": {
                    x: -109.55,
                    y: -119.5
                }
            })
            .addTimedChild(instance20, 29, 1, {
                "29": {
                    x: -109.55,
                    y: -81.55
                }
            })
            .addTimedChild(instance21, 30, 1, {
                "30": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance22, 31, 1, {
                "31": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance23, 32, 1, {
                "32": {
                    x: -109.55,
                    y: 114.95
                }
            })
            .addTimedChild(instance24, 33, 3, {
                "33": {
                    x: -109.55,
                    y: 99.9
                }
            })
            .addTimedChild(instance25, 36, 1, {
                "36": {
                    x: -109.55,
                    y: 114.95
                }
            })
            .addTimedChild(instance26, 37, 1, {
                "37": {
                    x: -109.55,
                    y: 107.45
                }
            })
            .addTimedChild(instance27, 38, 2, {
                "38": {
                    x: -75.2,
                    y: 110.45
                }
            })
            .addTimedChild(instance28, 40, 1, {
                "40": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance29, 41, 2, {
                "41": {
                    x: -109.55,
                    y: -50.95
                }
            })
            .addTimedChild(instance30, 43, 1, {
                "43": {
                    x: -109.55,
                    y: -98.4
                }
            })
            .addTimedChild(instance31, 44, 2, {
                "44": {
                    x: -109.55,
                    y: -108.05
                }
            })
            .addTimedChild(instance32, 46, 2, {
                "46": {
                    x: -109.55,
                    y: -119.5
                }
            })
            .addTimedChild(instance33, 48, 1, {
                "48": {
                    x: -109.55,
                    y: -81.55
                }
            })
            .addTimedChild(instance34, 49, 1, {
                "49": {
                    x: -109.55,
                    y: 54.95
                }
            })
            .addTimedChild(instance35, 50, 4);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 54, loop: false });
        var instance1 = new Graphic10(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 54, {
            "0": {
                x: -8.8,
                y: -124.75,
                sx: 1,
                ky: 0,
                r: 0.078
            },
            "2": {
                x: 0.75,
                y: -112.4,
                r: 0.156
            },
            "3": {
                x: -17.15,
                y: -130.95,
                r: 0
            },
            "8": {
                y: -122.6
            },
            "10": {
                y: -130.95
            },
            "11": {
                x: -24.3,
                y: -136.95,
                r: -0.057
            },
            "12": {
                x: -25.8,
                y: -135
            },
            "13": {
                x: -17.15,
                y: -137.2,
                sx: 1.002,
                ky: -0.057,
                r: 0
            },
            "14": {
                y: -135.15,
                sx: 1.001,
                ky: -0.038
            },
            "15": {
                y: -130.95,
                sx: 1,
                ky: 0
            },
            "18": {
                x: -6.55,
                y: -119,
                r: 0.105
            },
            "19": {
                x: -8.8,
                y: -124.75,
                r: 0.078
            },
            "21": {
                x: 0.75,
                y: -112.4,
                r: 0.156
            },
            "22": {
                x: -17.15,
                y: -130.95,
                r: 0
            },
            "27": {
                y: -122.6
            },
            "29": {
                y: -130.95
            },
            "30": {
                x: -24.3,
                y: -136.95,
                r: -0.057
            },
            "31": {
                x: -25.8,
                y: -135
            },
            "32": {
                x: -17.15,
                y: -137.2,
                sx: 1.002,
                ky: -0.057,
                r: 0
            },
            "33": {
                y: -135.15,
                sx: 1.001,
                ky: -0.038
            },
            "34": {
                y: -130.95,
                sx: 1,
                ky: 0
            },
            "37": {
                x: -6.55,
                y: -119,
                r: 0.105
            },
            "38": {
                x: -8.8,
                y: -124.75,
                r: 0.078
            },
            "40": {
                x: 0.75,
                y: -112.4,
                r: 0.156
            },
            "41": {
                x: -17.15,
                y: -130.95,
                r: 0
            },
            "46": {
                y: -122.6
            },
            "48": {
                y: -130.95
            },
            "49": {
                x: -24.3,
                y: -136.95,
                r: -0.057
            },
            "50": {
                x: -25.8,
                y: -135
            }
        });
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 61, loop: false });
        var instance1 = new Sprite(fromFrame("stork-leg1"))
            .setTransform(-84.25, -6);
        this.addTimedChild(instance1);
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance7 = new Graphic12(MovieClip.SYNCHED);
        var instance6 = new Graphic11(MovieClip.SYNCHED);
        var instance2 = new Graphic9(MovieClip.SYNCHED)
            .setTransform(92.4, 28.25);
        var instance5 = new Graphic5(MovieClip.SYNCHED);
        var instance3 = new Graphic3(MovieClip.SYNCHED);
        var instance4 = new Graphic2(MovieClip.SYNCHED);
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance7, 12, 61, {
                "12": {
                    x: 131.25,
                    y: 36.55
                },
                "13": {
                    x: 139.95
                },
                "14": {
                    x: 171.55
                },
                "15": {
                    x: 224.4
                },
                "16": {
                    x: 265.35
                },
                "17": {
                    x: 277.25
                },
                "18": {
                    x: 276.4
                },
                "19": {
                    x: 273.4
                },
                "20": {
                    x: 268.3
                },
                "21": {
                    x: 264.4
                },
                "22": {
                    x: 263.25
                },
                "63": {
                    x: 266.05
                },
                "64": {
                    x: 268.85
                },
                "65": {
                    x: 271.65
                },
                "66": {
                    x: 274.45
                },
                "67": {
                    x: 277.25
                },
                "68": {
                    x: 248.05
                },
                "69": {
                    x: 218.85
                },
                "70": {
                    x: 189.65
                },
                "71": {
                    x: 160.45
                },
                "72": {
                    x: 131.25
                }
            })
            .addTimedChild(instance6, 12, 54, {
                "12": {
                    x: 153.35,
                    y: 13.15,
                    sy: 1.002,
                    kx: 0.059,
                    a: 1,
                    c: [
                        0.57,
                        0,
                        0.57,
                        0,
                        0.57,
                        0
                    ]
                },
                "13": {
                    x: 153.279,
                    y: 13.099,
                    kx: 0.057
                },
                "18": {
                    x: 153.35,
                    y: 13.15,
                    kx: 0.059
                },
                "63": {
                    x: 153.279,
                    y: 13.099,
                    kx: 0.057,
                    a: 0.85,
                    c: [
                        0.64,
                        0,
                        0.64,
                        0,
                        0.64,
                        0
                    ]
                },
                "64": {
                    a: 0.27,
                    c: [
                        0.89,
                        0,
                        0.89,
                        0,
                        0.89,
                        0
                    ]
                },
                "65": {
                    x: 153.35,
                    y: 13.15,
                    kx: 0.059,
                    a: 0,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                }
            })
            .addTimedChild(instance2)
            .addTimedChild(instance5, 12, 54, {
                "12": {
                    x: 189.3,
                    y: 13.15,
                    a: 1
                },
                "63": {
                    a: 0.85
                },
                "64": {
                    a: 0.27
                },
                "65": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 4, 70, {
                "4": {
                    x: -150,
                    y: -13.25,
                    sx: 0.47
                },
                "5": {
                    x: -155.64,
                    sx: 0.507
                },
                "6": {
                    x: -175.972,
                    sx: 0.642
                },
                "7": {
                    x: -210.025,
                    sx: 0.869
                },
                "8": {
                    x: -236.385,
                    sx: 1.045
                },
                "9": {
                    x: -244,
                    sx: 1.096
                },
                "10": {
                    x: -243.74,
                    sx: 1.092
                },
                "11": {
                    x: -242.687,
                    sx: 1.078
                },
                "12": {
                    x: -240.67,
                    sx: 1.053
                },
                "13": {
                    x: -238.381,
                    sx: 1.024
                },
                "14": {
                    x: -236.935,
                    sx: 1.005
                },
                "15": {
                    x: -236.5,
                    sx: 1
                },
                "63": {
                    x: -237.778,
                    sx: 1.016
                },
                "64": {
                    x: -239.003,
                    sx: 1.032
                },
                "65": {
                    x: -240.281,
                    sx: 1.048
                },
                "66": {
                    x: -241.51,
                    sx: 1.064
                },
                "67": {
                    x: -242.784,
                    sx: 1.08
                },
                "68": {
                    x: -244,
                    sx: 1.096
                },
                "69": {
                    x: -225.214,
                    sx: 0.971
                },
                "70": {
                    x: -206.427,
                    sx: 0.845
                },
                "71": {
                    x: -187.641,
                    sx: 0.72
                },
                "72": {
                    x: -168.855,
                    sx: 0.595
                },
                "73": {
                    x: -150,
                    sx: 0.47
                }
            })
            .addTimedChild(instance4, 12, 70, {
                "12": {
                    x: -270,
                    y: -13.7,
                    sx: 0.538,
                    sy: 0.538,
                    a: 0,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "13": {
                    x: -270.004,
                    y: -11.337,
                    sx: 0.571,
                    sy: 0.571,
                    a: 0.06
                },
                "14": {
                    x: -270.014,
                    y: -2.735,
                    sx: 0.691,
                    sy: 0.691,
                    a: 0.28
                },
                "15": {
                    x: -270.009,
                    y: 11.763,
                    sx: 0.891,
                    sy: 0.891,
                    a: 0.64
                },
                "16": {
                    x: -270.001,
                    y: 22.966,
                    sx: 1.046,
                    sy: 1.046,
                    a: 0.92
                },
                "17": {
                    x: -270,
                    y: 26.25,
                    sx: 1.091,
                    sy: 1.091,
                    a: 1
                },
                "18": {
                    x: -270.053,
                    y: 26.269,
                    sx: 1.086,
                    sy: 1.086
                },
                "19": {
                    x: -270.017,
                    y: 26.294,
                    sx: 1.066,
                    sy: 1.066
                },
                "20": {
                    x: -270.035,
                    y: 26.299,
                    sx: 1.033,
                    sy: 1.033
                },
                "21": {
                    x: -270.056,
                    y: 26.283,
                    sx: 1.007,
                    sy: 1.007
                },
                "22": {
                    x: -270,
                    y: 26.25,
                    sx: 1,
                    sy: 1
                },
                "23": {
                    y: 30.3
                },
                "24": {
                    y: 37.8
                },
                "25": {
                    y: 48.8
                },
                "26": {
                    y: 63.25
                },
                "27": {
                    y: 77.25,
                    sx: 0.902,
                    sy: 1.091
                },
                "28": {
                    x: -269.981,
                    y: 76.394,
                    sx: 0.908,
                    sy: 1.085
                },
                "29": {
                    x: -270.025,
                    y: 73.327,
                    sx: 0.928,
                    sy: 1.062
                },
                "30": {
                    x: -269.999,
                    y: 67.623,
                    sx: 0.965,
                    sy: 1.02
                },
                "31": {
                    x: -269.996,
                    y: 61.082,
                    sx: 1.007,
                    sy: 0.971
                },
                "32": {
                    x: -269.98,
                    y: 56.964,
                    sx: 1.034,
                    sy: 0.94
                },
                "33": {
                    x: -270,
                    y: 55.75,
                    sx: 1.042,
                    sy: 0.932
                },
                "34": {
                    x: -270.057,
                    y: 55.938,
                    sx: 1.041,
                    sy: 0.933
                },
                "35": {
                    x: -270.041,
                    y: 56.546,
                    sx: 1.038,
                    sy: 0.938
                },
                "36": {
                    x: -270.046,
                    y: 57.577,
                    sx: 1.032,
                    sy: 0.948
                },
                "37": {
                    x: -270.06,
                    y: 59.138,
                    sx: 1.023,
                    sy: 0.962
                },
                "38": {
                    y: 60.849,
                    sx: 1.013,
                    sy: 0.978
                },
                "39": {
                    x: -270.053,
                    y: 62.255,
                    sx: 1.006,
                    sy: 0.991
                },
                "40": {
                    x: -270.037,
                    y: 63.075,
                    sx: 1.001,
                    sy: 0.998
                },
                "41": {
                    x: -270,
                    y: 63.25,
                    sx: 1,
                    sy: 1
                },
                "72": {
                    y: 62.35
                },
                "73": {
                    y: 59.5
                },
                "74": {
                    y: 56.15
                },
                "75": {
                    y: 54.9
                },
                "76": {
                    y: 64.15,
                    c: [
                        0.93,
                        0,
                        0.93,
                        0,
                        0.93,
                        0
                    ]
                },
                "77": {
                    y: 94.35,
                    c: [
                        0.71,
                        0,
                        0.71,
                        0,
                        0.71,
                        0
                    ]
                },
                "78": {
                    y: 140.45,
                    c: [
                        0.38,
                        0,
                        0.38,
                        0,
                        0.38,
                        0
                    ]
                },
                "79": {
                    y: 179.1,
                    c: [
                        0.09,
                        0,
                        0.09,
                        0,
                        0.09,
                        0
                    ]
                },
                "80": {
                    y: 192.2,
                    c: [
                        0,
                        0,
                        0,
                        0,
                        0,
                        0
                    ]
                }
            })
            .addTimedChild(instance1, 0, 82, {
                "0": {
                    x: -144.95,
                    y: -19.25,
                    sx: 0.774,
                    sy: 0.774,
                    c: [
                        0,
                        1,
                        0,
                        1,
                        0,
                        1
                    ]
                },
                "1": {
                    x: -144.968,
                    y: -19.242,
                    sx: 0.779,
                    sy: 0.779,
                    c: [
                        0.02,
                        0.98,
                        0.02,
                        0.98,
                        0.02,
                        0.98
                    ]
                },
                "2": {
                    x: -144.978,
                    y: -19.241,
                    sx: 0.8,
                    sy: 0.8,
                    c: [
                        0.11,
                        0.89,
                        0.11,
                        0.89,
                        0.11,
                        0.89
                    ]
                },
                "3": {
                    x: -144.952,
                    y: -19.271,
                    sx: 0.839,
                    sy: 0.839,
                    c: [
                        0.29,
                        0.71,
                        0.29,
                        0.71,
                        0.29,
                        0.71
                    ]
                },
                "4": {
                    x: -144.946,
                    y: -19.274,
                    sx: 0.897,
                    sy: 0.897,
                    c: [
                        0.55,
                        0.45,
                        0.55,
                        0.45,
                        0.55,
                        0.45
                    ]
                },
                "5": {
                    x: -144.978,
                    y: -19.269,
                    sx: 0.955,
                    sy: 0.955,
                    c: [
                        0.8,
                        0.2,
                        0.8,
                        0.2,
                        0.8,
                        0.2
                    ]
                },
                "6": {
                    x: -144.972,
                    y: -19.231,
                    sx: 0.99,
                    sy: 0.99,
                    c: [
                        0.96,
                        0.04,
                        0.96,
                        0.04,
                        0.96,
                        0.04
                    ]
                },
                "7": {
                    x: -144.95,
                    y: -19.25,
                    sx: 1,
                    sy: 1,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "78": {
                    x: -144.974,
                    y: -19.278,
                    sx: 0.983,
                    sy: 0.983,
                    c: [
                        0.93,
                        0.07,
                        0.93,
                        0.07,
                        0.93,
                        0.07
                    ]
                },
                "79": {
                    x: -144.988,
                    y: -19.274,
                    sx: 0.929,
                    sy: 0.929,
                    c: [
                        0.71,
                        0.29,
                        0.71,
                        0.29,
                        0.71,
                        0.29
                    ]
                },
                "80": {
                    x: -144.972,
                    y: -19.265,
                    sx: 0.845,
                    sy: 0.845,
                    c: [
                        0.38,
                        0.62,
                        0.38,
                        0.62,
                        0.38,
                        0.62
                    ]
                },
                "81": {
                    x: -144.97,
                    y: -19.267,
                    sx: 0.775,
                    sy: 0.775,
                    c: [
                        0.09,
                        0.91,
                        0.09,
                        0.91,
                        0.09,
                        0.91
                    ]
                }
            });
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Sprite(fromFrame("triangle1"))
            .setTransform(-13.05, -11.7, 1.298, 1.298);
        this.addTimedChild(instance1);
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Graphic14(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 82, {
            "0": {
                sx: 0.329,
                sy: 0.329,
                r: -2.523,
                a: 1
            },
            "1": {
                sx: 0.332,
                sy: 0.332,
                r: -2.507
            },
            "2": {
                sx: 0.343,
                sy: 0.343,
                r: -2.459
            },
            "3": {
                sx: 0.362,
                sy: 0.362,
                r: -2.376
            },
            "4": {
                sx: 0.389,
                sy: 0.389,
                r: -2.258
            },
            "5": {
                sx: 0.425,
                sy: 0.425,
                r: -2.104
            },
            "6": {
                sx: 0.469,
                sy: 0.469,
                r: -1.908
            },
            "7": {
                sx: 0.521,
                sy: 0.521,
                r: -1.684
            },
            "8": {
                sx: 0.579,
                sy: 0.579,
                r: -1.431
            },
            "9": {
                sx: 0.642,
                sy: 0.642,
                r: -1.155
            },
            "10": {
                sx: 0.706,
                sy: 0.706,
                r: -0.871
            },
            "11": {
                sx: 0.77,
                sy: 0.77,
                r: -0.595
            },
            "12": {
                sx: 0.828,
                sy: 0.828,
                r: -0.341,
                a: 0.75
            },
            "13": {
                sx: 0.88,
                sy: 0.88,
                r: -0.114,
                a: 0.52
            },
            "14": {
                sx: 0.924,
                sy: 0.924,
                r: 0.07,
                a: 0.33
            },
            "15": {
                sx: 0.957,
                sy: 0.957,
                r: 0.215,
                a: 0.18
            },
            "16": {
                sx: 0.981,
                sy: 0.981,
                r: 0.319,
                a: 0.08
            },
            "17": {
                sx: 0.994,
                sy: 0.994,
                r: 0.38,
                a: 0.02
            },
            "18": {
                sx: 1,
                sy: 1,
                r: 0.401,
                a: 0
            }
        });
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot1"))
            .setTransform(-8.05, -8.05, 0.636, 0.636);
        this.addTimedChild(instance1);
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Graphic16(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 17, {
            "0": {
                sx: 0.375,
                sy: 0.375,
                a: 0
            },
            "1": {
                sx: 0.396,
                sy: 0.396,
                a: 0.01
            },
            "2": {
                sx: 0.465,
                sy: 0.465,
                a: 0.04
            },
            "3": {
                sx: 0.592,
                sy: 0.592,
                a: 0.08
            },
            "4": {
                sx: 0.792,
                sy: 0.792,
                a: 0.16
            },
            "5": {
                sx: 1.082,
                sy: 1.082,
                a: 0.27
            },
            "6": {
                sx: 1.475,
                sy: 1.475,
                a: 0.42
            },
            "7": {
                sx: 1.966,
                sy: 1.966,
                a: 0.61
            },
            "8": {
                sx: 2.505,
                sy: 2.505,
                a: 0.81
            },
            "9": {
                sx: 2.991,
                sy: 2.991,
                a: 1
            },
            "10": {
                sx: 3.194,
                sy: 3.194,
                a: 0.71
            },
            "11": {
                sx: 3.358,
                sy: 3.358,
                a: 0.47
            },
            "12": {
                sx: 3.486,
                sy: 3.486,
                a: 0.29
            },
            "13": {
                sx: 3.578,
                sy: 3.578,
                a: 0.16
            },
            "14": {
                sx: 3.64,
                sy: 3.64,
                a: 0.07
            },
            "15": {
                sx: 3.675,
                sy: 3.675,
                a: 0.02
            },
            "16": {
                sx: 3.686,
                sy: 3.686,
                a: 0
            }
        });
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Sprite(fromFrame("+1"))
            .setTransform(-20.5, -20.5, 0.978, 0.978);
        this.addTimedChild(instance1);
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Graphic18(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 82, {
            "0": {
                sx: 0.201,
                sy: 0.201,
                a: 0
            },
            "1": {
                sx: 0.222,
                sy: 0.222,
                a: 0.05
            },
            "2": {
                sx: 0.249,
                sy: 0.249,
                a: 0.11
            },
            "3": {
                sx: 0.281,
                sy: 0.281,
                a: 0.18
            },
            "4": {
                sx: 0.319,
                sy: 0.319,
                a: 0.26
            },
            "5": {
                sx: 0.362,
                sy: 0.362,
                a: 0.35
            },
            "6": {
                sx: 0.411,
                sy: 0.411,
                a: 0.46
            },
            "7": {
                sx: 0.465,
                sy: 0.465,
                a: 0.57
            },
            "8": {
                sx: 0.526,
                sy: 0.526,
                a: 0.7
            },
            "9": {
                sx: 0.591,
                sy: 0.591,
                a: 0.85
            },
            "10": {
                sx: 0.662,
                sy: 0.662,
                a: 1
            },
            "11": {
                sx: 0.697,
                sy: 0.697,
                a: 0.73
            },
            "12": {
                sx: 0.727,
                sy: 0.727,
                a: 0.51
            },
            "13": {
                sx: 0.751,
                sy: 0.751,
                a: 0.32
            },
            "14": {
                sx: 0.77,
                sy: 0.77,
                a: 0.18
            },
            "15": {
                sx: 0.783,
                sy: 0.783,
                a: 0.08
            },
            "16": {
                sx: 0.791,
                sy: 0.791,
                a: 0.02
            },
            "17": {
                sx: 0.794,
                sy: 0.794,
                a: 0
            }
        });
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Sprite(fromFrame("+1"))
            .setTransform(-20.5, -20.5, 0.978, 0.978);
        this.addTimedChild(instance1);
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Graphic20(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 82, {
            "0": {
                sx: 0.201,
                sy: 0.201,
                r: 0,
                a: 0
            },
            "1": {
                sx: 0.208,
                sy: 0.208,
                r: -0.005,
                a: 0.01
            },
            "2": {
                sx: 0.229,
                sy: 0.229,
                r: -0.027,
                a: 0.04
            },
            "3": {
                sx: 0.267,
                sy: 0.267,
                r: -0.062,
                a: 0.08
            },
            "4": {
                sx: 0.321,
                sy: 0.321,
                r: -0.114,
                a: 0.15
            },
            "5": {
                sx: 0.393,
                sy: 0.393,
                r: -0.184,
                a: 0.23
            },
            "6": {
                sx: 0.48,
                sy: 0.48,
                r: -0.267,
                a: 0.34
            },
            "7": {
                sx: 0.581,
                sy: 0.581,
                r: -0.363,
                a: 0.46
            },
            "8": {
                sx: 0.692,
                sy: 0.692,
                r: -0.468,
                a: 0.6
            },
            "9": {
                sx: 0.807,
                sy: 0.807,
                r: -0.581,
                a: 0.74
            },
            "10": {
                sx: 0.919,
                sy: 0.919,
                r: -0.687,
                a: 0.88
            },
            "11": {
                sx: 1.023,
                sy: 1.023,
                r: -0.786,
                a: 1
            },
            "12": {
                sx: 1.086,
                sy: 1.086,
                r: -0.81,
                a: 0.73
            },
            "13": {
                sx: 1.141,
                sy: 1.141,
                r: -0.831,
                a: 0.51
            },
            "14": {
                sx: 1.187,
                sy: 1.187,
                r: -0.849,
                a: 0.32
            },
            "15": {
                sx: 1.222,
                sy: 1.222,
                r: -0.862,
                a: 0.18
            },
            "16": {
                sx: 1.247,
                sy: 1.247,
                r: -0.871,
                a: 0.08
            },
            "17": {
                sx: 1.261,
                sy: 1.261,
                r: -0.875,
                a: 0.02
            },
            "18": {
                sx: 1.268,
                sy: 1.268,
                r: -0.877,
                a: 0
            }
        });
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot1"))
            .setTransform(-8.05, -8.05, 0.636, 0.636);
        this.addTimedChild(instance1);
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 82, loop: false });
        var instance1 = new Graphic22(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 17, {
            "0": {
                sx: 0.375,
                sy: 0.375,
                a: 0
            },
            "1": {
                sx: 0.396,
                sy: 0.396,
                a: 0.01
            },
            "2": {
                sx: 0.465,
                sy: 0.465,
                a: 0.04
            },
            "3": {
                sx: 0.592,
                sy: 0.592,
                a: 0.08
            },
            "4": {
                sx: 0.792,
                sy: 0.792,
                a: 0.16
            },
            "5": {
                sx: 1.082,
                sy: 1.082,
                a: 0.27
            },
            "6": {
                sx: 1.475,
                sy: 1.475,
                a: 0.42
            },
            "7": {
                sx: 1.966,
                sy: 1.966,
                a: 0.61
            },
            "8": {
                sx: 2.505,
                sy: 2.505,
                a: 0.81
            },
            "9": {
                sx: 2.991,
                sy: 2.991,
                a: 1
            },
            "10": {
                sx: 3.194,
                sy: 3.194,
                a: 0.71
            },
            "11": {
                sx: 3.358,
                sy: 3.358,
                a: 0.47
            },
            "12": {
                sx: 3.486,
                sy: 3.486,
                a: 0.29
            },
            "13": {
                sx: 3.578,
                sy: 3.578,
                a: 0.16
            },
            "14": {
                sx: 3.64,
                sy: 3.64,
                a: 0.07
            },
            "15": {
                sx: 3.675,
                sy: 3.675,
                a: 0.02
            },
            "16": {
                sx: 3.686,
                sy: 3.686,
                a: 0
            }
        });
    });

    lib.jiboji_mothers_day_stork = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 83,
            framerate: 30,
            loop: false
        });
        var instance6 = new Graphic23(MovieClip.SYNCHED)
            .setTransform(540.05, 414.55, 0.526, 0.526);
        var instance5 = new Graphic21(MovieClip.SYNCHED)
            .setTransform(645.7, 490)
            .setColorTransform(0, 0.57, 0, 0.93, 0, 0.95);
        var instance4 = new Graphic19(MovieClip.SYNCHED)
            .setTransform(798.3, 251.1);
        var instance3 = new Graphic17(MovieClip.SYNCHED)
            .setTransform(922.05, 303.6, 0.339, 0.339)
            .setColorTransform(0, 0.57, 0, 0.93, 0, 0.95);
        var instance2 = new Graphic15(MovieClip.SYNCHED)
            .setTransform(537, 258)
            .setColorTransform(0, 0.57, 0, 0.93, 0, 0.95);
        var instance1 = new Graphic13(MovieClip.SYNCHED);
        this.addTimedChild(instance6, 1, 82)
            .addTimedChild(instance5, 1, 82)
            .addTimedChild(instance4, 1, 82)
            .addTimedChild(instance3, 1, 82)
            .addTimedChild(instance2, 1, 82)
            .addTimedChild(instance1, 0, 82, {
                "0": {
                    x: 612.85,
                    y: 362.05
                }
            });
    });

    lib.jiboji_mothers_day_stork.assets = {
        "jiboji_mothers_day_stork_atlas_1": "images/jiboji_mothers_day_stork_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.jiboji_mothers_day_stork,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 83,
        library: lib
    };
}
