// jibo-anim-db-animations/timelines/personal_report_news_intro.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Container = PIXI.Container;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;
    var Text = PIXI.Text;
    var Graphics = PIXI.Graphics;
    var shapes = PIXI.animate.ShapesCache;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 64, loop: false });
        var instance1 = new Sprite(fromFrame("earth-cloud2"))
            .setTransform(-38.3, -30.25);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 64, loop: false });
        var instance1 = new Sprite(fromFrame("earth-cloud1"))
            .setTransform(-95.55, -40.5);
        this.addTimedChild(instance1);
    });

    lib.TopNews = Container.extend(function () {
        Container.call(this);
        var instance1 = new Text("Top News")
            .setStyle({
                fontFamily: "Proxima Nova Soft",
                fontSize: 120,
                fontWeight: "bold",
                fill: "#fff",
                letterSpacing: -6
            })
            .setTransform(2, 2);
        this.addChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 56, loop: false });
        var instance1 = new Sprite(fromFrame("Jibo_News_World_430_Height"));
        var instance2 = new Sprite(fromFrame("Bitmap 1"))
            .setTransform(-201, -256);
        this.addTimedChild(instance1, 0, 40, {
                "0": {
                    x: -314,
                    y: -271,
                    sx: 1.231,
                    sy: 1.231
                }
            })
            .addTimedChild(instance2, 40, 16);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 56, loop: false });
        var instance1 = new Graphic3(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 56, {
            "0": {
                x: 67.25,
                y: -2.05,
                a: 0.02
            },
            "1": {
                x: 64.45,
                a: 0.05
            },
            "2": {
                x: 61.6,
                a: 0.07
            },
            "3": {
                x: 58.8,
                a: 0.1
            },
            "4": {
                x: 55.95,
                a: 0.12
            },
            "5": {
                x: 53.1,
                a: 0.14
            },
            "6": {
                x: 50.3,
                a: 0.17
            },
            "7": {
                x: 47.45,
                a: 0.2
            },
            "8": {
                x: 44.65,
                a: 0.22
            },
            "9": {
                x: 41.8,
                a: 0.24
            },
            "10": {
                x: 38.95,
                a: 0.27
            },
            "11": {
                x: 36.15,
                a: 0.29
            },
            "12": {
                x: 33.3,
                a: 0.32
            },
            "13": {
                x: 30.45,
                a: 0.34
            },
            "14": {
                x: 27.65,
                a: 0.37
            },
            "15": {
                x: 24.8,
                a: 0.39
            },
            "16": {
                x: 22,
                a: 0.41
            },
            "17": {
                x: 19.15,
                a: 0.44
            },
            "18": {
                x: 16.3,
                a: 0.46
            },
            "19": {
                x: 13.5,
                a: 0.49
            },
            "20": {
                x: 10.65,
                a: 0.51
            },
            "21": {
                x: 7.85,
                a: 0.54
            },
            "22": {
                x: 5,
                a: 0.56
            },
            "23": {
                x: 2.15,
                a: 0.59
            },
            "24": {
                x: -0.65,
                a: 0.61
            },
            "25": {
                x: -3.5,
                a: 0.63
            },
            "26": {
                x: -6.3,
                a: 0.66
            },
            "27": {
                x: -9.15,
                a: 0.68
            },
            "28": {
                x: -12,
                a: 0.71
            },
            "29": {
                x: -14.8,
                a: 0.73
            },
            "30": {
                x: -17.65,
                a: 0.76
            },
            "31": {
                x: -20.5,
                a: 0.78
            },
            "32": {
                x: -23.3,
                a: 0.8
            },
            "33": {
                x: -26.15,
                a: 0.83
            },
            "34": {
                x: -28.95,
                a: 0.86
            },
            "35": {
                x: -31.8,
                a: 0.88
            },
            "36": {
                x: -34.65,
                a: 0.9
            },
            "37": {
                x: -37.45,
                a: 0.93
            },
            "38": {
                x: -40.3,
                a: 0.95
            },
            "39": {
                x: -43.1,
                a: 0.98
            },
            "40": {
                x: -45.95,
                a: 1
            }
        });
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("earth-water1"))
            .setTransform(-257, -257, 1.218, 1.218);
        this.addTimedChild(instance1);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance4 = new Graphics()
            .drawCommands(shapes.personal_report_news_intro[0])
            .setRenderable(false);
        var instance1 = new Graphic5(MovieClip.SYNCHED);
        var instance5 = new Graphic4(MovieClip.SYNCHED);
        var instance6 = new lib.TopNews();
        var instance3 = new Graphic2(MovieClip.SYNCHED);
        var instance2 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance4, 12, 40)
            .addTimedChild(instance1, 0, 73, {
                "0": {
                    x: 9.3,
                    y: -3,
                    sx: 0.821,
                    sy: 0.821,
                    a: 1,
                    c: [
                        0,
                        0.2,
                        0,
                        0.35,
                        0,
                        0.6
                    ]
                },
                "1": {
                    y: -2.966,
                    sx: 0.825,
                    sy: 0.825,
                    c: [
                        0.02,
                        0.2,
                        0.02,
                        0.35,
                        0.02,
                        0.59
                    ]
                },
                "2": {
                    x: 9.405,
                    y: -2.766,
                    sx: 0.839,
                    sy: 0.839,
                    c: [
                        0.09,
                        0.18,
                        0.09,
                        0.32,
                        0.09,
                        0.55
                    ]
                },
                "3": {
                    x: 9.471,
                    y: -2.402,
                    sx: 0.862,
                    sy: 0.862,
                    c: [
                        0.2,
                        0.16,
                        0.2,
                        0.28,
                        0.2,
                        0.49
                    ]
                },
                "4": {
                    x: 9.635,
                    y: -1.969,
                    sx: 0.895,
                    sy: 0.895,
                    c: [
                        0.35,
                        0.13,
                        0.35,
                        0.23,
                        0.35,
                        0.39
                    ]
                },
                "5": {
                    x: 9.815,
                    y: -1.408,
                    sx: 0.933,
                    sy: 0.933,
                    c: [
                        0.53,
                        0.09,
                        0.53,
                        0.16,
                        0.53,
                        0.28
                    ]
                },
                "6": {
                    x: 10,
                    y: -0.848,
                    sx: 0.971,
                    sy: 0.971,
                    c: [
                        0.71,
                        0.06,
                        0.71,
                        0.1,
                        0.71,
                        0.17
                    ]
                },
                "7": {
                    x: 10.115,
                    y: -0.416,
                    sx: 1.003,
                    sy: 1.003,
                    c: [
                        0.87,
                        0.03,
                        0.87,
                        0.05,
                        0.87,
                        0.08
                    ]
                },
                "8": {
                    x: 10.252,
                    y: -0.092,
                    sx: 1.024,
                    sy: 1.024,
                    c: [
                        0.96,
                        0.01,
                        0.96,
                        0.01,
                        0.96,
                        0.02
                    ]
                },
                "9": {
                    x: 10.3,
                    y: 0,
                    sx: 1.031,
                    sy: 1.031,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "10": {
                    x: 10.294,
                    sx: 1.03,
                    sy: 1.03
                },
                "11": {
                    x: 10.273,
                    sx: 1.028,
                    sy: 1.028
                },
                "12": {
                    x: 10.289,
                    sx: 1.025,
                    sy: 1.025
                },
                "13": {
                    x: 10.291,
                    sx: 1.02,
                    sy: 1.02
                },
                "14": {
                    x: 10.285,
                    sx: 1.015,
                    sy: 1.015
                },
                "15": {
                    x: 10.278,
                    sx: 1.009,
                    sy: 1.009
                },
                "16": {
                    x: 10.28,
                    sx: 1.004,
                    sy: 1.004
                },
                "17": {
                    x: 10.3,
                    sx: 1.001,
                    sy: 1.001
                },
                "18": {
                    sx: 1,
                    sy: 1
                },
                "65": {
                    x: 10.262,
                    y: -0.3,
                    sx: 0.982,
                    sy: 0.982,
                    a: 0.88
                },
                "66": {
                    x: 10.224,
                    y: -0.6,
                    sx: 0.964,
                    sy: 0.964,
                    a: 0.75
                },
                "67": {
                    x: 10.087,
                    y: -0.95,
                    sx: 0.945,
                    sy: 0.945,
                    a: 0.63
                },
                "68": {
                    x: 10.049,
                    y: -1.25,
                    sx: 0.927,
                    sy: 0.927,
                    a: 0.5
                },
                "69": {
                    x: 10.011,
                    y: -1.55,
                    sx: 0.909,
                    sy: 0.909,
                    a: 0.38
                },
                "70": {
                    x: 9.973,
                    y: -1.85,
                    sx: 0.891,
                    sy: 0.891,
                    a: 0.25
                },
                "71": {
                    x: 9.835,
                    y: -2.2,
                    sx: 0.872,
                    sy: 0.872,
                    a: 0.13
                },
                "72": {
                    x: 9.8,
                    y: -2.5,
                    sx: 0.854,
                    sy: 0.854,
                    a: 0
                }
            })
            .addTimedMask(instance5, {
                "12": instance4,
                "52": null
            })
            .addTimedChild(instance5, 12, 56, {
                "12": {
                    x: -0.45,
                    y: 0.5,
                    sx: 1,
                    sy: 1,
                    a: 1
                },
                "65": {
                    x: -0.442,
                    y: 0.492,
                    sx: 0.983,
                    sy: 0.983,
                    a: 0.67
                },
                "66": {
                    x: -0.485,
                    y: 0.483,
                    sx: 0.966,
                    sy: 0.966,
                    a: 0.33
                },
                "67": {
                    x: -0.45,
                    y: 0.5,
                    sx: 0.95,
                    sy: 0.95,
                    a: 0
                }
            })
            .addTimedChild(instance6, 25, 46, {
                "25": {
                    x: 48,
                    y: -80,
                    a: 0
                },
                "26": {
                    x: 42.4,
                    a: 0.1
                },
                "27": {
                    x: 37.3,
                    a: 0.19
                },
                "28": {
                    x: 32.6,
                    a: 0.27
                },
                "29": {
                    x: 28.25,
                    a: 0.35
                },
                "30": {
                    x: 24.2,
                    a: 0.42
                },
                "31": {
                    x: 20.45,
                    a: 0.48
                },
                "32": {
                    x: 17,
                    a: 0.54
                },
                "33": {
                    x: 13.8,
                    a: 0.6
                },
                "34": {
                    x: 10.8,
                    a: 0.65
                },
                "35": {
                    x: 8.05,
                    a: 0.7
                },
                "36": {
                    x: 5.5,
                    a: 0.75
                },
                "37": {
                    x: 3.2,
                    a: 0.79
                },
                "38": {
                    x: 1.1,
                    a: 0.82
                },
                "39": {
                    x: -0.8,
                    a: 0.86
                },
                "40": {
                    x: -2.55,
                    a: 0.89
                },
                "41": {
                    x: -4.05,
                    a: 0.91
                },
                "42": {
                    x: -5.35,
                    a: 0.94
                },
                "43": {
                    x: -6.45,
                    a: 0.96
                },
                "44": {
                    x: -7.35,
                    a: 0.97
                },
                "45": {
                    x: -8.1,
                    a: 0.98
                },
                "46": {
                    x: -8.6,
                    a: 0.99
                },
                "47": {
                    x: -8.9,
                    a: 1
                },
                "48": {
                    x: -9
                },
                "58": {
                    x: -11.85,
                    a: 0.92
                },
                "59": {
                    x: -14.7,
                    a: 0.85
                },
                "60": {
                    x: -17.55,
                    a: 0.77
                },
                "61": {
                    x: -20.4,
                    a: 0.69
                },
                "62": {
                    x: -23.25,
                    a: 0.62
                },
                "63": {
                    x: -26.1,
                    a: 0.54
                },
                "64": {
                    x: -28.9,
                    a: 0.46
                },
                "65": {
                    x: -31.75,
                    a: 0.38
                },
                "66": {
                    x: -34.6,
                    a: 0.31
                },
                "67": {
                    x: -37.45,
                    a: 0.23
                },
                "68": {
                    x: -40.3,
                    a: 0.15
                },
                "69": {
                    x: -43.15,
                    a: 0.08
                },
                "70": {
                    x: -46,
                    a: 0
                }
            })
            .addTimedChild(instance3, 4, 64, {
                "4": {
                    x: -279.9,
                    y: -118.8,
                    a: 0
                },
                "5": {
                    x: -279.7
                },
                "6": {
                    x: -279.1,
                    a: 0.01
                },
                "7": {
                    x: -278.05,
                    a: 0.02
                },
                "8": {
                    x: -276.55,
                    a: 0.04
                },
                "9": {
                    x: -274.65,
                    a: 0.06
                },
                "10": {
                    x: -272.3,
                    a: 0.09
                },
                "11": {
                    x: -269.5,
                    a: 0.13
                },
                "12": {
                    x: -266.25,
                    a: 0.16
                },
                "13": {
                    x: -262.6,
                    a: 0.21
                },
                "14": {
                    x: -258.6,
                    a: 0.25
                },
                "15": {
                    x: -254.2,
                    a: 0.3
                },
                "16": {
                    x: -249.55,
                    a: 0.36
                },
                "17": {
                    x: -244.6,
                    a: 0.42
                },
                "18": {
                    x: -239.55,
                    a: 0.48
                },
                "19": {
                    x: -234.4,
                    a: 0.54
                },
                "20": {
                    x: -229.25,
                    a: 0.6
                },
                "21": {
                    x: -224.15,
                    a: 0.66
                },
                "22": {
                    x: -219.3,
                    a: 0.72
                },
                "23": {
                    x: -214.75,
                    a: 0.78
                },
                "24": {
                    x: -210.55,
                    a: 0.82
                },
                "25": {
                    x: -206.75,
                    a: 0.87
                },
                "26": {
                    x: -203.5,
                    a: 0.91
                },
                "27": {
                    x: -200.8,
                    a: 0.94
                },
                "28": {
                    x: -198.65,
                    a: 0.97
                },
                "29": {
                    x: -197.1,
                    a: 0.98
                },
                "30": {
                    x: -196.2,
                    a: 1
                },
                "31": {
                    x: -195.9
                },
                "56": {
                    x: -194.75,
                    a: 0.99
                },
                "57": {
                    x: -191.15,
                    a: 0.95
                },
                "58": {
                    x: -185.05,
                    a: 0.88
                },
                "59": {
                    x: -176.55,
                    a: 0.79
                },
                "60": {
                    x: -165.85,
                    a: 0.68
                },
                "61": {
                    x: -153.6,
                    a: 0.55
                },
                "62": {
                    x: -140.65,
                    a: 0.41
                },
                "63": {
                    x: -128.1,
                    a: 0.28
                },
                "64": {
                    x: -117.15,
                    a: 0.16
                },
                "65": {
                    x: -108.8,
                    a: 0.07
                },
                "66": {
                    x: -103.65,
                    a: 0.02
                },
                "67": {
                    x: -101.9,
                    a: 0
                }
            })
            .addTimedChild(instance2, 4, 64, {
                "4": {
                    x: 212.1,
                    y: -219.8,
                    a: 0
                },
                "5": {
                    x: 211.9
                },
                "6": {
                    x: 211.35,
                    a: 0.01
                },
                "7": {
                    x: 210.4,
                    a: 0.02
                },
                "8": {
                    x: 209.1,
                    a: 0.04
                },
                "9": {
                    x: 207.35,
                    a: 0.06
                },
                "10": {
                    x: 205.2,
                    a: 0.09
                },
                "11": {
                    x: 202.65,
                    a: 0.13
                },
                "12": {
                    x: 199.75,
                    a: 0.16
                },
                "13": {
                    x: 196.45,
                    a: 0.21
                },
                "14": {
                    x: 192.8,
                    a: 0.25
                },
                "15": {
                    x: 188.85,
                    a: 0.3
                },
                "16": {
                    x: 184.6,
                    a: 0.36
                },
                "17": {
                    x: 180.2,
                    a: 0.42
                },
                "18": {
                    x: 175.6,
                    a: 0.48
                },
                "19": {
                    x: 170.9,
                    a: 0.54
                },
                "20": {
                    x: 166.25,
                    a: 0.6
                },
                "21": {
                    x: 161.7,
                    a: 0.66
                },
                "22": {
                    x: 157.3,
                    a: 0.72
                },
                "23": {
                    x: 153.15,
                    a: 0.78
                },
                "24": {
                    x: 149.35,
                    a: 0.82
                },
                "25": {
                    x: 145.95,
                    a: 0.87
                },
                "26": {
                    x: 143,
                    a: 0.91
                },
                "27": {
                    x: 140.5,
                    a: 0.94
                },
                "28": {
                    x: 138.6,
                    a: 0.97
                },
                "29": {
                    x: 137.2,
                    a: 0.98
                },
                "30": {
                    x: 136.4,
                    a: 1
                },
                "31": {
                    x: 136.1
                },
                "56": {
                    x: 134.95,
                    a: 0.99
                },
                "57": {
                    x: 131.35,
                    a: 0.95
                },
                "58": {
                    x: 125.25,
                    a: 0.88
                },
                "59": {
                    x: 116.75,
                    a: 0.79
                },
                "60": {
                    x: 106.05,
                    a: 0.68
                },
                "61": {
                    x: 93.8,
                    a: 0.55
                },
                "62": {
                    x: 80.85,
                    a: 0.41
                },
                "63": {
                    x: 68.3,
                    a: 0.28
                },
                "64": {
                    x: 57.35,
                    a: 0.16
                },
                "65": {
                    x: 49,
                    a: 0.07
                },
                "66": {
                    x: 43.85,
                    a: 0.02
                },
                "67": {
                    x: 42.1,
                    a: 0
                }
            });
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic7(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic9(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic11(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic13(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic15(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic17(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic19(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic21(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic24 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic23(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic25 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic26 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic25(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic27 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic28 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic27(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic29 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Sprite(fromFrame("line1"))
            .setTransform(-3.85, -15.5, 1.155, 1.155);
        this.addTimedChild(instance1);
    });

    var Graphic30 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance1 = new Graphic29(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 73, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 156.459,
                sy: 1.816
            },
            "2": {
                y: 140.404,
                sy: 1.707
            },
            "3": {
                y: 114.983,
                sy: 1.535
            },
            "4": {
                y: 83.631,
                sy: 1.323
            },
            "5": {
                y: 51.498,
                sy: 1.105
            },
            "6": {
                y: 24.145,
                sy: 0.919
            },
            "7": {
                y: 5.803,
                sy: 0.795
            },
            "8": {
                y: -1.25,
                sy: 0.747
            },
            "9": {
                y: -9.252,
                sy: 0.689,
                a: 0.84
            },
            "10": {
                y: -16.612,
                sy: 0.635,
                a: 0.69
            },
            "11": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "12": {
                y: -29.555,
                sy: 0.541,
                a: 0.43
            },
            "13": {
                y: -35.037,
                sy: 0.501,
                a: 0.32
            },
            "14": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "15": {
                y: -44.275,
                sy: 0.433,
                a: 0.14
            },
            "16": {
                y: -47.93,
                sy: 0.406,
                a: 0.06
            },
            "17": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic31 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 73, loop: false });
        var instance12 = new Graphic30(MovieClip.SYNCHED)
            .setTransform(-1.45, -226.15, 0.735, 0.735);
        var instance11 = new Graphic28(MovieClip.SYNCHED)
            .setTransform(125.15, -192.05, 0.735, 0.735, 0.61);
        var instance10 = new Graphic26(MovieClip.SYNCHED)
            .setTransform(-129.75, -188.05, 0.735, 0.735, 0, 0.636, 2.506);
        var instance9 = new Graphic24(MovieClip.SYNCHED)
            .setTransform(1, 217.15, 0.735, 0.735, 0, 3.142, 3.142);
        var instance8 = new Graphic22(MovieClip.SYNCHED)
            .setTransform(-120, 185.1, 0.735, 0.735, -2.575);
        var instance7 = new Graphic20(MovieClip.SYNCHED)
            .setTransform(132.4, 180.15, 0.735, 0.735, 0, 3.781, -0.64);
        var instance6 = new Graphic18(MovieClip.SYNCHED)
            .setTransform(222.4, -10, 0.735, 0.735, 0, 4.712, 1.571);
        var instance5 = new Graphic16(MovieClip.SYNCHED)
            .setTransform(191.4, 109.95, 0.735, 0.735, 0, 4.146, 2.137);
        var instance4 = new Graphic14(MovieClip.SYNCHED)
            .setTransform(193.2, -126.7, 0.735, 0.735, 0, -1.004, -2.137);
        var instance3 = new Graphic12(MovieClip.SYNCHED)
            .setTransform(-222.4, -10, 0.735, 0.735, 0, 1.571, 1.571);
        var instance2 = new Graphic10(MovieClip.SYNCHED)
            .setTransform(-195.45, 113.95, 0.735, 0.735, 0, 2.137, 1.004);
        var instance1 = new Graphic8(MovieClip.SYNCHED)
            .setTransform(-191.2, -123.7, 0.735, 0.735, -1.004);
        this.addTimedChild(instance12)
            .addTimedChild(instance11)
            .addTimedChild(instance10)
            .addTimedChild(instance9)
            .addTimedChild(instance8)
            .addTimedChild(instance7)
            .addTimedChild(instance6)
            .addTimedChild(instance5)
            .addTimedChild(instance4)
            .addTimedChild(instance3)
            .addTimedChild(instance2)
            .addTimedChild(instance1);
    });

    lib.personal_report_news_intro = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 74,
            framerate: 30,
            loop: false
        });
        var instance2 = new Graphic31(MovieClip.SYNCHED);
        var instance1 = new Graphic6(MovieClip.SYNCHED);
        this.addTimedChild(instance2, 0, 73, {
                "0": {
                    x: 640,
                    y: 362.7,
                    sx: 1.179,
                    sy: 1.179,
                    c: [
                        0,
                        0.08,
                        0,
                        0.95,
                        0,
                        0.98
                    ]
                }
            })
            .addTimedChild(instance1, 0, 73, {
                "0": {
                    x: 630.95,
                    y: 360.45
                }
            });
    });

    lib.personal_report_news_intro.assets = {
        "Jibo_News_World_430_Height": "images/Jibo_News_World_430_Height.png",
        "personal_report_news_intro": "images/personal_report_news_intro.shapes.json",
        "personal_report_news_intro_atlas_1": "images/personal_report_news_intro_atlas_1.json",
        "personal_report_news_intro_atlas_2": "images/personal_report_news_intro_atlas_2.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.personal_report_news_intro,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 74,
        library: lib
    };
}
