// jibo-anim-db-animations/timelines/send_01.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 15, loop: false });
        var instance1 = new Sprite(fromFrame("Envelope-Outside1"))
            .setTransform(-240, -184.7, 0.816, 0.816);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 28, loop: false });
        var instance1 = new Sprite(fromFrame("Envelope-Flap4"));
        var instance2 = new Sprite(fromFrame("Envelope-Flap1"))
            .setTransform(-240.05, -9.75, 0.816, 0.816);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -240.05,
                    y: -9.7,
                    sx: 0.816,
                    sy: 0.816
                }
            })
            .addTimedChild(instance2, 1, 27);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("Envelope-Flap2"));
        var instance2 = new Sprite(fromFrame("Envelope-Flap3"))
            .setTransform(-239.75, -209.6, 0.816, 0.816);
        this.addTimedChild(instance1, 0, 1, {
                "0": {
                    x: -238.45,
                    y: -171.8,
                    sx: 0.816,
                    sy: 0.816
                }
            })
            .addTimedChild(instance2, 1, 1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 27, loop: false });
        var instance1 = new Sprite(fromFrame("Envelope-Outside1"))
            .setTransform(-240, -184.7, 0.816, 0.816);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 42, loop: false });
        var instance2 = new Graphic3(MovieClip.SYNCHED);
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        var instance4 = new Graphic4(MovieClip.SYNCHED)
            .setTransform(0.35, 14.25, 0.329, 0.329);
        var instance3 = new Graphic2(MovieClip.SYNCHED);
        this.addTimedChild(instance2, 12, 2, {
                "12": {
                    x: 0.35,
                    y: -46.15,
                    sx: 0.329,
                    sy: 0.255
                },
                "13": {
                    y: -46.4,
                    sy: 0.088
                }
            })
            .addTimedChild(instance1, 0, 15, {
                "0": {
                    x: 0.35,
                    y: 14.25,
                    sx: 0.329,
                    sy: 0.329
                }
            })
            .addTimedChild(instance4, 15, 27)
            .addTimedChild(instance3, 14, 28, {
                "14": {
                    x: 0.35,
                    y: -46.2,
                    sx: 0.329,
                    sy: 0.065
                },
                "15": {
                    y: -46.85,
                    sy: 0.329
                }
            });
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 12, loop: false });
        var instance1 = new Sprite(fromFrame("Envelope-Inside1"))
            .setTransform(-240.15, -100.65, 0.816, 0.816);
        this.addTimedChild(instance1);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 30, loop: false });
        var instance1 = new Sprite(fromFrame("Envelope-Outside1"))
            .setTransform(-240, -184.7, 0.816, 0.816);
        this.addTimedChild(instance1);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 12, loop: false });
        var instance1 = new Sprite(fromFrame("Envelope-Flap2"))
            .setTransform(-238.45, -171.8, 0.816, 0.816);
        this.addTimedChild(instance1);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 30, loop: false });
        var instance1 = new Sprite(fromFrame("Envelope-Inside1"))
            .setTransform(-240.15, -100.65, 0.816, 0.816);
        this.addTimedChild(instance1);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 42, loop: false });
        var instance2 = new Graphic8(MovieClip.SYNCHED);
        var instance1 = new Graphic6(MovieClip.SYNCHED);
        var instance4 = new Graphic9(MovieClip.SYNCHED)
            .setTransform(0.35, -13.3, 0.329, 0.329);
        var instance3 = new Graphic7(MovieClip.SYNCHED)
            .setTransform(0.35, 14.25, 0.329, 0.329);
        this.addTimedChild(instance2, 0, 12, {
                "0": {
                    x: 0.35,
                    y: -46.85,
                    sx: 0.329,
                    sy: 0.329
                }
            })
            .addTimedChild(instance1, 0, 12, {
                "0": {
                    x: 0.35,
                    y: -13.3,
                    sx: 0.329,
                    sy: 0.329
                }
            })
            .addTimedChild(instance4, 12, 30)
            .addTimedChild(instance3, 12, 30);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 42, loop: false });
        var instance2 = new Graphic10(MovieClip.SYNCHED);
        var instance1 = new Graphic5(MovieClip.SYNCHED);
        this.addTimedChild(instance2, 0, 42, {
                "0": {
                    x: 1.1,
                    y: 788.6,
                    sx: 3.728,
                    sy: 3.728
                },
                "1": {
                    x: 1.05,
                    y: 764.962
                },
                "2": {
                    x: 1.15,
                    y: 690.662
                },
                "3": {
                    x: 1.25,
                    y: 564.262
                },
                "4": {
                    x: 1.4,
                    y: 393.912
                },
                "5": {
                    x: 1.55,
                    y: 204.312
                },
                "6": {
                    x: 1.65,
                    y: 34.412
                },
                "7": {
                    x: 1.75,
                    y: -79.088
                },
                "8": {
                    x: 1.85,
                    y: -117.7
                },
                "9": {
                    x: 1.8,
                    y: -116.952
                },
                "10": {
                    y: -114.952
                },
                "11": {
                    y: -112.452
                },
                "12": {
                    x: 1.85,
                    y: -111.45
                },
                "13": {
                    x: 1.75,
                    y: -105.202
                },
                "14": {
                    x: 1.6,
                    y: -85.552
                },
                "15": {
                    x: 1.35,
                    y: -53.552
                },
                "16": {
                    x: 1.1,
                    y: -16.752
                },
                "17": {
                    x: 0.9,
                    y: 11.498
                },
                "18": {
                    x: 0.85,
                    y: 21.55
                },
                "19": {
                    x: 0.95,
                    y: 18.25
                },
                "20": {
                    y: 8.15
                },
                "21": {
                    y: -8.25
                },
                "22": {
                    y: -27.05
                },
                "23": {
                    y: -41.55
                },
                "24": {
                    x: 0.85,
                    y: -46.6
                },
                "25": {
                    x: -1.2,
                    y: -46.601
                },
                "26": {
                    x: -7.95,
                    y: -46.501
                },
                "27": {
                    x: -19.3,
                    y: -46.351
                },
                "28": {
                    x: -33.8,
                    y: -46.101
                },
                "29": {
                    x: -48.15,
                    y: -45.851
                },
                "30": {
                    x: -58.2,
                    y: -45.701
                },
                "31": {
                    x: -61.8,
                    y: -45.6
                },
                "32": {
                    x: -34.651,
                    y: -45.551
                },
                "33": {
                    x: 50.249,
                    y: -45.501
                },
                "34": {
                    x: 195.649
                },
                "35": {
                    x: 396.849,
                    y: -45.451
                },
                "36": {
                    x: 635.799,
                    y: -45.351
                },
                "37": {
                    x: 878.399,
                    y: -45.301
                },
                "38": {
                    x: 1082.599,
                    y: -45.251
                },
                "39": {
                    x: 1214.499,
                    y: -45.201
                },
                "40": {
                    x: 1258.85,
                    y: -45.25
                }
            })
            .addTimedChild(instance1, 0, 42, {
                "0": {
                    x: 1.1,
                    y: 788.6,
                    sx: 3.728,
                    sy: 3.728
                },
                "1": {
                    x: 1.05,
                    y: 764.962
                },
                "2": {
                    x: 1.15,
                    y: 690.662
                },
                "3": {
                    x: 1.25,
                    y: 564.262
                },
                "4": {
                    x: 1.4,
                    y: 393.912
                },
                "5": {
                    x: 1.55,
                    y: 204.312
                },
                "6": {
                    x: 1.65,
                    y: 34.412
                },
                "7": {
                    x: 1.75,
                    y: -79.088
                },
                "8": {
                    x: 1.85,
                    y: -117.7
                },
                "9": {
                    x: 1.8,
                    y: -116.952
                },
                "10": {
                    y: -114.952
                },
                "11": {
                    y: -112.452
                },
                "12": {
                    x: 1.85,
                    y: -111.45
                },
                "13": {
                    x: 1.8,
                    y: -106.9
                },
                "14": {
                    x: 1.55,
                    y: -87.65,
                    sx: 3.727,
                    sy: 3.727
                },
                "15": {
                    x: 1.3,
                    y: -61.7,
                    sx: 3.728,
                    sy: 3.728
                },
                "16": {
                    x: 1.05,
                    y: -18.351
                },
                "17": {
                    x: 0.9,
                    y: 11.349
                },
                "18": {
                    x: 0.85,
                    y: 21.55
                },
                "19": {
                    x: 0.95,
                    y: 18.25
                },
                "20": {
                    y: 8.15
                },
                "21": {
                    y: -8.25
                },
                "22": {
                    y: -27.05
                },
                "23": {
                    y: -41.55
                },
                "24": {
                    x: 0.85,
                    y: -46.6
                },
                "25": {
                    x: -1.2,
                    y: -46.601
                },
                "26": {
                    x: -7.95,
                    y: -46.501
                },
                "27": {
                    x: -19.3,
                    y: -46.351
                },
                "28": {
                    x: -33.8,
                    y: -46.101
                },
                "29": {
                    x: -48.15,
                    y: -45.851
                },
                "30": {
                    x: -58.2,
                    y: -45.701
                },
                "31": {
                    x: -61.8,
                    y: -45.6
                },
                "32": {
                    x: -34.651,
                    y: -45.551
                },
                "33": {
                    x: 50.249,
                    y: -45.501
                },
                "34": {
                    x: 195.649
                },
                "35": {
                    x: 396.849,
                    y: -45.451
                },
                "36": {
                    x: 635.799,
                    y: -45.351
                },
                "37": {
                    x: 878.399,
                    y: -45.301
                },
                "38": {
                    x: 1082.599,
                    y: -45.251
                },
                "39": {
                    x: 1214.499,
                    y: -45.201
                },
                "40": {
                    x: 1258.85,
                    y: -45.25
                }
            });
    });

    lib.send01 = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 42,
            framerate: 30,
            loop: false
        });
        var instance1 = new Graphic11(MovieClip.SYNCHED)
            .setTransform(644.5, 338.4);
        this.addTimedChild(instance1);
    });

    lib.send01.assets = {
        "send01_atlas_1": "images/send01_atlas_1.json",
        "send01_atlas_2": "images/send01_atlas_2.json",
        "send01_atlas_3": "images/send01_atlas_3.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.send01,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 42,
        library: lib
    };
}
