// jibo-anim-db-animations/timelines/shutdown_01.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;
    var Graphics = PIXI.Graphics;
    var shapes = PIXI.animate.ShapesCache;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Sprite(fromFrame("ring"))
            .setTransform(-268.5, -267.5);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                sx: 0.099,
                sy: 0.099,
                a: 0
            },
            "1": {
                sx: 0.106,
                sy: 0.106,
                a: 0.02
            },
            "2": {
                sx: 0.127,
                sy: 0.127,
                a: 0.09
            },
            "3": {
                sx: 0.166,
                sy: 0.166,
                a: 0.22
            },
            "4": {
                sx: 0.224,
                sy: 0.224,
                a: 0.41
            },
            "5": {
                sx: 0.304,
                sy: 0.304,
                a: 0.67
            },
            "6": {
                sx: 0.404,
                sy: 0.404,
                a: 1
            },
            "7": {
                sx: 0.486,
                sy: 0.486,
                a: 0.89
            },
            "8": {
                sx: 0.575,
                sy: 0.575,
                a: 0.79
            },
            "9": {
                sx: 0.668,
                sy: 0.668,
                a: 0.66
            },
            "10": {
                sx: 0.762,
                sy: 0.762,
                a: 0.55
            },
            "11": {
                sx: 0.852,
                sy: 0.852,
                a: 0.43
            },
            "12": {
                sx: 0.934,
                sy: 0.934,
                a: 0.33
            },
            "13": {
                sx: 1.006,
                sy: 1.006,
                a: 0.23
            },
            "14": {
                sx: 1.066,
                sy: 1.066,
                a: 0.16
            },
            "15": {
                sx: 1.114,
                sy: 1.114,
                a: 0.1
            },
            "16": {
                sx: 1.149,
                sy: 1.149,
                a: 0.05
            },
            "17": {
                sx: 1.173,
                sy: 1.173,
                a: 0.02
            },
            "18": {
                sx: 1.187,
                sy: 1.187,
                a: 0
            },
            "19": {
                sx: 1.192,
                sy: 1.192
            }
        });
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic3(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic5(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic7(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic9(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic11(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic13(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic15(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic17(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic19(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic21(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic24 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic23(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic25 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.shutdown01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic26 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance1 = new Graphic25(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 47, {
            "0": {
                y: 161.85,
                sy: 1.852,
                a: 1
            },
            "1": {
                y: 157.581,
                sy: 1.823
            },
            "2": {
                y: 144.836,
                sy: 1.737
            },
            "3": {
                y: 124.355,
                sy: 1.598
            },
            "4": {
                y: 98.037,
                sy: 1.42
            },
            "5": {
                y: 69.13,
                sy: 1.224
            },
            "6": {
                y: 41.585,
                sy: 1.038
            },
            "7": {
                y: 19.158,
                sy: 0.886
            },
            "8": {
                y: 4.448,
                sy: 0.786
            },
            "9": {
                y: -1.25,
                sy: 0.747
            },
            "10": {
                y: -6.091,
                sy: 0.712,
                a: 0.9
            },
            "11": {
                y: -10.784,
                sy: 0.678,
                a: 0.81
            },
            "12": {
                y: -15.18,
                sy: 0.646,
                a: 0.72
            },
            "13": {
                y: -19.428,
                sy: 0.615,
                a: 0.64
            },
            "14": {
                y: -23.38,
                sy: 0.586,
                a: 0.55
            },
            "15": {
                y: -27.134,
                sy: 0.558,
                a: 0.48
            },
            "16": {
                y: -30.641,
                sy: 0.532,
                a: 0.41
            },
            "17": {
                y: -34,
                sy: 0.508,
                a: 0.34
            },
            "18": {
                y: -37.112,
                sy: 0.486,
                a: 0.28
            },
            "19": {
                y: -39.977,
                sy: 0.465,
                a: 0.22
            },
            "20": {
                y: -42.595,
                sy: 0.445,
                a: 0.17
            },
            "21": {
                y: -45.015,
                sy: 0.427,
                a: 0.12
            },
            "22": {
                y: -47.238,
                sy: 0.411,
                a: 0.07
            },
            "23": {
                y: -49.264,
                sy: 0.397,
                a: 0.04
            },
            "24": {
                y: -51.05,
                sy: 0.384,
                a: 0
            }
        });
    });

    var Graphic27 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 47, loop: false });
        var instance12 = new Graphic26(MovieClip.SYNCHED)
            .setTransform(-1.45, -226.15, 0.735, 0.735);
        var instance11 = new Graphic24(MovieClip.SYNCHED)
            .setTransform(125.15, -192.05, 0.735, 0.735, 0.61);
        var instance10 = new Graphic22(MovieClip.SYNCHED)
            .setTransform(-129.75, -188.05, 0.735, 0.735, 0, 0.636, 2.506);
        var instance9 = new Graphic20(MovieClip.SYNCHED)
            .setTransform(1, 217.15, 0.735, 0.735, 0, 3.142, 3.142);
        var instance8 = new Graphic18(MovieClip.SYNCHED)
            .setTransform(-120, 185.1, 0.735, 0.735, -2.575);
        var instance7 = new Graphic16(MovieClip.SYNCHED)
            .setTransform(132.4, 180.15, 0.735, 0.735, 0, 3.781, -0.64);
        var instance6 = new Graphic14(MovieClip.SYNCHED)
            .setTransform(222.4, -10, 0.735, 0.735, 0, 4.712, 1.571);
        var instance5 = new Graphic12(MovieClip.SYNCHED)
            .setTransform(191.4, 109.95, 0.735, 0.735, 0, 4.146, 2.137);
        var instance4 = new Graphic10(MovieClip.SYNCHED)
            .setTransform(193.2, -126.7, 0.735, 0.735, 0, -1.004, -2.137);
        var instance3 = new Graphic8(MovieClip.SYNCHED)
            .setTransform(-222.4, -10, 0.735, 0.735, 0, 1.571, 1.571);
        var instance2 = new Graphic6(MovieClip.SYNCHED)
            .setTransform(-195.45, 113.95, 0.735, 0.735, 0, 2.137, 1.004);
        var instance1 = new Graphic4(MovieClip.SYNCHED)
            .setTransform(-191.2, -123.7, 0.735, 0.735, -1.004);
        this.addTimedChild(instance12)
            .addTimedChild(instance11)
            .addTimedChild(instance10)
            .addTimedChild(instance9)
            .addTimedChild(instance8)
            .addTimedChild(instance7)
            .addTimedChild(instance6)
            .addTimedChild(instance5)
            .addTimedChild(instance4)
            .addTimedChild(instance3)
            .addTimedChild(instance2)
            .addTimedChild(instance1);
    });

    lib.shutdown01 = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 47,
            framerate: 30,
            loop: false
        });
        var instance2 = new Graphic27(MovieClip.SYNCHED)
            .setTransform(640, 371.5, 1.276, 1.276)
            .setColorTransform(0, 1, 0, 1, 0, 1);
        var instance1 = new Graphic2(MovieClip.SYNCHED)
            .setTransform(641.5, 368.45, 0.97, 0.97);
        this.addTimedChild(instance2)
            .addTimedChild(instance1);
    });

    lib.shutdown01.assets = {
        "shutdown01": "images/shutdown01.shapes.json",
        "shutdown01_atlas_1": "images/shutdown01_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.shutdown01,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 47,
        library: lib
    };
}
