// jibo-anim-db-animations/timelines/thinking_dots.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 25, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 48, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 43, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 22, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 19, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 30, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 8, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 20, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 15, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 41, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 12, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 9, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 22, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 24, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic24 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic25 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 35, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic26 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic27 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic28 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic29 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic30 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic31 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic32 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 11, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic33 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 19, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic34 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 24, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic35 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic36 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic37 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic38 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance8 = new Graphic34(MovieClip.SYNCHED);
        var instance7 = new Graphic31(MovieClip.SYNCHED);
        var instance6 = new Graphic27(MovieClip.SYNCHED);
        var instance19 = new Graphic35(MovieClip.SYNCHED);
        var instance15 = new Graphic28(MovieClip.SYNCHED);
        var instance11 = new Graphic25(MovieClip.SYNCHED);
        var instance5 = new Graphic22(MovieClip.SYNCHED);
        var instance9 = new Graphic18(MovieClip.SYNCHED);
        var instance10 = new Graphic16(MovieClip.SYNCHED);
        var instance4 = new Graphic12(MovieClip.SYNCHED);
        var instance3 = new Graphic9(MovieClip.SYNCHED);
        var instance25 = new Graphic32(MovieClip.SYNCHED);
        var instance29 = new Graphic36(MovieClip.SYNCHED);
        var instance24 = new Graphic29(MovieClip.SYNCHED);
        var instance16 = new Graphic23(MovieClip.SYNCHED);
        var instance17 = new Graphic19(MovieClip.SYNCHED);
        var instance18 = new Graphic17(MovieClip.SYNCHED);
        var instance14 = new Graphic14(MovieClip.SYNCHED);
        var instance36 = new Graphic37(MovieClip.SYNCHED);
        var instance33 = new Graphic33(MovieClip.SYNCHED);
        var instance35 = new Graphic30(MovieClip.SYNCHED);
        var instance30 = new Graphic26(MovieClip.SYNCHED);
        var instance34 = new Graphic24(MovieClip.SYNCHED);
        var instance31 = new Graphic20(MovieClip.SYNCHED);
        var instance27 = new Graphic15(MovieClip.SYNCHED);
        var instance37 = new Graphic21(MovieClip.SYNCHED);
        var instance26 = new Graphic13(MovieClip.SYNCHED);
        var instance20 = new Graphic10(MovieClip.SYNCHED);
        var instance13 = new Graphic8(MovieClip.SYNCHED);
        var instance12 = new Graphic6(MovieClip.SYNCHED);
        var instance2 = new Graphic3(MovieClip.SYNCHED);
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        var instance32 = new Graphic11(MovieClip.SYNCHED);
        var instance28 = new Graphic7(MovieClip.SYNCHED);
        var instance21 = new Graphic5(MovieClip.SYNCHED);
        var instance23 = new Graphic4(MovieClip.SYNCHED);
        var instance22 = new Graphic2(MovieClip.SYNCHED);
        this.addTimedChild(instance8, 0, 24, {
                "0": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "1": {
                    a: 0.13
                },
                "2": {
                    a: 0.25
                },
                "3": {
                    a: 0.38
                },
                "4": {
                    a: 0.5
                },
                "5": {
                    a: 0.63
                },
                "6": {
                    a: 0.75
                },
                "7": {
                    a: 0.88
                },
                "8": {
                    a: 1
                },
                "15": {
                    a: 0.89
                },
                "16": {
                    a: 0.78
                },
                "17": {
                    a: 0.67
                },
                "18": {
                    a: 0.55
                },
                "19": {
                    a: 0.45
                },
                "20": {
                    a: 0.33
                },
                "21": {
                    a: 0.22
                },
                "22": {
                    a: 0.11
                },
                "23": {
                    a: 0
                }
            })
            .addTimedChild(instance7, 0, 18, {
                "0": {
                    x: 203.85,
                    y: 127.25,
                    a: 1
                },
                "12": {
                    a: 0.83
                },
                "13": {
                    a: 0.67
                },
                "14": {
                    a: 0.5
                },
                "15": {
                    a: 0.33
                },
                "16": {
                    a: 0.17
                },
                "17": {
                    a: 0
                }
            })
            .addTimedChild(instance6, 0, 17, {
                "0": {
                    x: -552.95,
                    y: 174.85,
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance19, 37, 10, {
                "37": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "38": {
                    a: 0.2
                },
                "39": {
                    a: 0.4
                },
                "40": {
                    a: 0.6
                },
                "41": {
                    a: 0.8
                },
                "42": {
                    a: 1
                },
                "43": {
                    a: 0.75
                },
                "44": {
                    a: 0.5
                },
                "45": {
                    a: 0.25
                },
                "46": {
                    a: 0
                }
            })
            .addTimedChild(instance15, 22, 16, {
                "22": {
                    x: -552.95,
                    y: 174.85,
                    a: 0
                },
                "23": {
                    a: 0.17
                },
                "24": {
                    a: 0.33
                },
                "25": {
                    a: 0.5
                },
                "26": {
                    a: 0.67
                },
                "27": {
                    a: 0.83
                },
                "28": {
                    a: 1
                },
                "32": {
                    a: 0.83
                },
                "33": {
                    a: 0.67
                },
                "34": {
                    a: 0.5
                },
                "35": {
                    a: 0.33
                },
                "36": {
                    a: 0.17
                },
                "37": {
                    a: 0
                }
            })
            .addTimedChild(instance11, 7, 35, {
                "7": {
                    x: -27.35,
                    y: 273.35,
                    a: 0
                },
                "8": {
                    a: 0.14
                },
                "9": {
                    a: 0.29
                },
                "10": {
                    a: 0.43
                },
                "11": {
                    a: 0.57
                },
                "12": {
                    a: 0.71
                },
                "13": {
                    a: 0.86
                },
                "14": {
                    a: 1
                },
                "35": {
                    a: 0.86
                },
                "36": {
                    a: 0.71
                },
                "37": {
                    a: 0.57
                },
                "38": {
                    a: 0.43
                },
                "39": {
                    a: 0.29
                },
                "40": {
                    a: 0.14
                },
                "41": {
                    a: 0
                }
            })
            .addTimedChild(instance5, 0, 22, {
                "0": {
                    x: -313.3,
                    y: 315.75,
                    a: 1
                },
                "15": {
                    a: 0.86
                },
                "16": {
                    a: 0.71
                },
                "17": {
                    a: 0.57
                },
                "18": {
                    a: 0.43
                },
                "19": {
                    a: 0.29
                },
                "20": {
                    a: 0.14
                },
                "21": {
                    a: 0
                }
            })
            .addTimedChild(instance9, 1, 16, {
                "1": {
                    x: -312.6,
                    y: 79.85,
                    a: 0
                },
                "2": {
                    a: 0.2
                },
                "3": {
                    a: 0.4
                },
                "4": {
                    a: 0.6
                },
                "5": {
                    a: 0.8
                },
                "6": {
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance10, 7, 14, {
                "7": {
                    x: -405.2,
                    y: 267.35,
                    a: 0
                },
                "8": {
                    a: 0.17
                },
                "9": {
                    a: 0.33
                },
                "10": {
                    a: 0.5
                },
                "11": {
                    a: 0.67
                },
                "12": {
                    a: 0.83
                },
                "13": {
                    a: 1
                },
                "15": {
                    a: 0.83
                },
                "16": {
                    a: 0.67
                },
                "17": {
                    a: 0.5
                },
                "18": {
                    a: 0.33
                },
                "19": {
                    a: 0.17
                },
                "20": {
                    a: 0
                }
            })
            .addTimedChild(instance4, 0, 15, {
                "0": {
                    x: -75.35,
                    y: -61.2,
                    a: 1
                },
                "8": {
                    a: 0.86
                },
                "9": {
                    a: 0.71
                },
                "10": {
                    a: 0.57
                },
                "11": {
                    a: 0.43
                },
                "12": {
                    a: 0.29
                },
                "13": {
                    a: 0.14
                },
                "14": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 0, 8, {
                "0": {
                    x: -266.7,
                    y: -107.9,
                    a: 1
                },
                "1": {
                    a: 0.86
                },
                "2": {
                    a: 0.71
                },
                "3": {
                    a: 0.57
                },
                "4": {
                    a: 0.43
                },
                "5": {
                    a: 0.29
                },
                "6": {
                    a: 0.14
                },
                "7": {
                    a: 0
                }
            })
            .addTimedChild(instance25, 48, 11, {
                "48": {
                    x: 203.85,
                    y: 127.25,
                    a: 0
                },
                "49": {
                    a: 0.25
                },
                "50": {
                    a: 0.5
                },
                "51": {
                    a: 0.75
                },
                "52": {
                    a: 1
                },
                "54": {
                    a: 0.8
                },
                "55": {
                    a: 0.6
                },
                "56": {
                    a: 0.4
                },
                "57": {
                    a: 0.2
                },
                "58": {
                    a: 0
                }
            })
            .addTimedChild(instance29, 61, 10, {
                "61": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "62": {
                    a: 0.2
                },
                "63": {
                    a: 0.4
                },
                "64": {
                    a: 0.6
                },
                "65": {
                    a: 0.8
                },
                "66": {
                    a: 1
                },
                "67": {
                    a: 0.75
                },
                "68": {
                    a: 0.5
                },
                "69": {
                    a: 0.25
                },
                "70": {
                    a: 0
                }
            })
            .addTimedChild(instance24, 48, 17, {
                "48": {
                    x: -552.95,
                    y: 174.85,
                    a: 0
                },
                "49": {
                    a: 0.17
                },
                "50": {
                    a: 0.33
                },
                "51": {
                    a: 0.5
                },
                "52": {
                    a: 0.67
                },
                "53": {
                    a: 0.83
                },
                "54": {
                    a: 1
                },
                "56": {
                    a: 0.89
                },
                "57": {
                    a: 0.78
                },
                "58": {
                    a: 0.67
                },
                "59": {
                    a: 0.55
                },
                "60": {
                    a: 0.45
                },
                "61": {
                    a: 0.33
                },
                "62": {
                    a: 0.22
                },
                "63": {
                    a: 0.11
                },
                "64": {
                    a: 0
                }
            })
            .addTimedChild(instance16, 28, 24, {
                "28": {
                    x: -313.3,
                    y: 315.75,
                    a: 0
                },
                "29": {
                    a: 0.14
                },
                "30": {
                    a: 0.29
                },
                "31": {
                    a: 0.43
                },
                "32": {
                    a: 0.57
                },
                "33": {
                    a: 0.71
                },
                "34": {
                    a: 0.86
                },
                "35": {
                    a: 1
                },
                "44": {
                    a: 0.88
                },
                "45": {
                    a: 0.75
                },
                "46": {
                    a: 0.63
                },
                "47": {
                    a: 0.5
                },
                "48": {
                    a: 0.38
                },
                "49": {
                    a: 0.25
                },
                "50": {
                    a: 0.13
                },
                "51": {
                    a: 0
                }
            })
            .addTimedChild(instance17, 31, 16, {
                "31": {
                    x: -312.6,
                    y: 79.85,
                    a: 0
                },
                "32": {
                    a: 0.2
                },
                "33": {
                    a: 0.4
                },
                "34": {
                    a: 0.6
                },
                "35": {
                    a: 0.8
                },
                "36": {
                    a: 1
                },
                "39": {
                    a: 0.88
                },
                "40": {
                    a: 0.75
                },
                "41": {
                    a: 0.63
                },
                "42": {
                    a: 0.5
                },
                "43": {
                    a: 0.38
                },
                "44": {
                    a: 0.25
                },
                "45": {
                    a: 0.13
                },
                "46": {
                    a: 0
                }
            })
            .addTimedChild(instance18, 36, 23, {
                "36": {
                    x: -405.2,
                    y: 267.35,
                    a: 0
                },
                "37": {
                    a: 0.14
                },
                "38": {
                    a: 0.29
                },
                "39": {
                    a: 0.43
                },
                "40": {
                    a: 0.57
                },
                "41": {
                    a: 0.71
                },
                "42": {
                    a: 0.86
                },
                "43": {
                    a: 1
                },
                "51": {
                    a: 0.88
                },
                "52": {
                    a: 0.75
                },
                "53": {
                    a: 0.63
                },
                "54": {
                    a: 0.5
                },
                "55": {
                    a: 0.38
                },
                "56": {
                    a: 0.25
                },
                "57": {
                    a: 0.13
                },
                "58": {
                    a: 0
                }
            })
            .addTimedChild(instance14, 22, 18, {
                "22": {
                    x: 158.45,
                    y: -61.75,
                    a: 0
                },
                "23": {
                    a: 0.14
                },
                "24": {
                    a: 0.29
                },
                "25": {
                    a: 0.43
                },
                "26": {
                    a: 0.57
                },
                "27": {
                    a: 0.71
                },
                "28": {
                    a: 0.86
                },
                "29": {
                    a: 1
                },
                "32": {
                    a: 0.88
                },
                "33": {
                    a: 0.75
                },
                "34": {
                    a: 0.63
                },
                "35": {
                    a: 0.5
                },
                "36": {
                    a: 0.38
                },
                "37": {
                    a: 0.25
                },
                "38": {
                    a: 0.13
                },
                "39": {
                    a: 0
                }
            })
            .addTimedChild(instance36, 80, 10, {
                "80": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "81": {
                    a: 0.13
                },
                "82": {
                    a: 0.25
                }
            })
            .addTimedChild(instance33, 71, 19, {
                "71": {
                    x: 203.85,
                    y: 127.25,
                    a: 0
                },
                "72": {
                    a: 0.25
                },
                "73": {
                    a: 0.5
                },
                "74": {
                    a: 0.75
                },
                "75": {
                    a: 1
                }
            })
            .addTimedChild(instance35, 73, 17, {
                "73": {
                    x: -552.95,
                    y: 174.85,
                    a: 0
                },
                "74": {
                    a: 0.17
                },
                "75": {
                    a: 0.33
                },
                "76": {
                    a: 0.5
                },
                "77": {
                    a: 0.67
                },
                "78": {
                    a: 0.83
                },
                "79": {
                    a: 1
                }
            })
            .addTimedChild(instance30, 62, 16, {
                "62": {
                    x: -27.35,
                    y: 273.35,
                    a: 0
                },
                "63": {
                    a: 0.14
                },
                "64": {
                    a: 0.29
                },
                "65": {
                    a: 0.43
                },
                "66": {
                    a: 0.57
                },
                "67": {
                    a: 0.71
                },
                "68": {
                    a: 0.86
                },
                "69": {
                    a: 1
                },
                "71": {
                    a: 0.86
                },
                "72": {
                    a: 0.71
                },
                "73": {
                    a: 0.57
                },
                "74": {
                    a: 0.43
                },
                "75": {
                    a: 0.29
                },
                "76": {
                    a: 0.14
                },
                "77": {
                    a: 0
                }
            })
            .addTimedChild(instance34, 72, 18, {
                "72": {
                    x: -313.3,
                    y: 315.75,
                    a: 0
                },
                "73": {
                    a: 0.14
                },
                "74": {
                    a: 0.29
                },
                "75": {
                    a: 0.43
                },
                "76": {
                    a: 0.57
                },
                "77": {
                    a: 0.71
                },
                "78": {
                    a: 0.86
                },
                "79": {
                    a: 1
                }
            })
            .addTimedChild(instance31, 64, 12, {
                "64": {
                    x: -312.6,
                    y: 79.85,
                    a: 0
                },
                "65": {
                    a: 0.2
                },
                "66": {
                    a: 0.4
                },
                "67": {
                    a: 0.6
                },
                "68": {
                    a: 0.8
                },
                "69": {
                    a: 1
                },
                "72": {
                    a: 0.75
                },
                "73": {
                    a: 0.5
                },
                "74": {
                    a: 0.25
                },
                "75": {
                    a: 0
                }
            })
            .addTimedChild(instance27, 53, 16, {
                "53": {
                    x: 158.45,
                    y: -61.75,
                    a: 0
                },
                "54": {
                    a: 0.14
                },
                "55": {
                    a: 0.29
                },
                "56": {
                    a: 0.43
                },
                "57": {
                    a: 0.57
                },
                "58": {
                    a: 0.71
                },
                "59": {
                    a: 0.86
                },
                "60": {
                    a: 1
                },
                "63": {
                    a: 0.83
                },
                "64": {
                    a: 0.67
                },
                "65": {
                    a: 0.5
                },
                "66": {
                    a: 0.33
                },
                "67": {
                    a: 0.17
                },
                "68": {
                    a: 0
                }
            })
            .addTimedChild(instance37, 81, 9, {
                "81": {
                    x: -312.6,
                    y: 79.85,
                    a: 0
                },
                "82": {
                    a: 0.2
                }
            })
            .addTimedChild(instance26, 49, 41, {
                "49": {
                    x: -75.35,
                    y: -61.2,
                    a: 1
                },
                "62": {
                    a: 0.9
                },
                "63": {
                    a: 0.8
                },
                "64": {
                    a: 0.7
                },
                "65": {
                    a: 0.6
                },
                "66": {
                    a: 0.5
                },
                "67": {
                    a: 0.4
                },
                "68": {
                    a: 0.3
                },
                "69": {
                    a: 0.2
                },
                "70": {
                    a: 0.1
                },
                "71": {
                    a: 0
                },
                "72": {
                    a: 1
                }
            })
            .addTimedChild(instance20, 39, 17, {
                "39": {
                    x: -266.7,
                    y: -107.9,
                    a: 0
                },
                "40": {
                    a: 0.13
                },
                "41": {
                    a: 0.25
                },
                "42": {
                    a: 0.38
                },
                "43": {
                    a: 0.5
                },
                "44": {
                    a: 0.63
                },
                "45": {
                    a: 0.75
                },
                "46": {
                    a: 0.88
                },
                "47": {
                    a: 1
                },
                "49": {
                    a: 0.86
                },
                "50": {
                    a: 0.71
                },
                "51": {
                    a: 0.57
                },
                "52": {
                    a: 0.43
                },
                "53": {
                    a: 0.29
                },
                "54": {
                    a: 0.14
                },
                "55": {
                    a: 0
                }
            })
            .addTimedChild(instance13, 18, 30, {
                "18": {
                    x: -454.5,
                    y: 175,
                    a: 0
                },
                "19": {
                    a: 0.14
                },
                "20": {
                    a: 0.29
                },
                "21": {
                    a: 0.43
                },
                "22": {
                    a: 0.57
                },
                "23": {
                    a: 0.71
                },
                "24": {
                    a: 0.86
                },
                "25": {
                    a: 1
                },
                "41": {
                    a: 0.86
                },
                "42": {
                    a: 0.71
                },
                "43": {
                    a: 0.57
                },
                "44": {
                    a: 0.43
                },
                "45": {
                    a: 0.29
                },
                "46": {
                    a: 0.14
                },
                "47": {
                    a: 0
                }
            })
            .addTimedChild(instance12, 10, 18, {
                "10": {
                    x: -454.15,
                    y: -14.85,
                    a: 0
                },
                "11": {
                    a: 0.17
                },
                "12": {
                    a: 0.33
                },
                "13": {
                    a: 0.5
                },
                "14": {
                    a: 0.67
                },
                "15": {
                    a: 0.83
                },
                "16": {
                    a: 1
                },
                "19": {
                    a: 0.89
                },
                "20": {
                    a: 0.78
                },
                "21": {
                    a: 0.67
                },
                "22": {
                    a: 0.55
                },
                "23": {
                    a: 0.45
                },
                "24": {
                    a: 0.33
                },
                "25": {
                    a: 0.22
                },
                "26": {
                    a: 0.11
                },
                "27": {
                    a: 0
                }
            })
            .addTimedChild(instance2, 0, 17, {
                "0": {
                    x: -405.9,
                    y: -108.6,
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance1, 0, 25, {
                "0": {
                    x: -76.5,
                    y: 32.9,
                    a: 1
                },
                "15": {
                    a: 0.9
                },
                "16": {
                    a: 0.8
                },
                "17": {
                    a: 0.7
                },
                "18": {
                    a: 0.6
                },
                "19": {
                    a: 0.5
                },
                "20": {
                    a: 0.4
                },
                "21": {
                    a: 0.3
                },
                "22": {
                    a: 0.2
                },
                "23": {
                    a: 0.1
                },
                "24": {
                    a: 0
                }
            })
            .addTimedChild(instance32, 70, 20, {
                "70": {
                    x: -266.7,
                    y: -107.9,
                    a: 0
                },
                "71": {
                    a: 0.13
                },
                "72": {
                    a: 0.25
                },
                "73": {
                    a: 0.38
                },
                "74": {
                    a: 0.5
                },
                "75": {
                    a: 0.63
                },
                "76": {
                    a: 0.75
                },
                "77": {
                    a: 0.88
                },
                "78": {
                    a: 1
                },
                "81": {
                    a: 0.86
                },
                "82": {
                    a: 0.71
                }
            })
            .addTimedChild(instance28, 61, 19, {
                "61": {
                    x: -454.15,
                    y: -14.85,
                    a: 0
                },
                "62": {
                    a: 0.17
                },
                "63": {
                    a: 0.33
                },
                "64": {
                    a: 0.5
                },
                "65": {
                    a: 0.67
                },
                "66": {
                    a: 0.83
                },
                "67": {
                    a: 1
                },
                "70": {
                    a: 0.89
                },
                "71": {
                    a: 0.78
                },
                "72": {
                    a: 0.67
                },
                "73": {
                    a: 0.55
                },
                "74": {
                    a: 0.45
                },
                "75": {
                    a: 0.33
                },
                "76": {
                    a: 0.22
                },
                "77": {
                    a: 0.11
                },
                "78": {
                    a: 0
                }
            })
            .addTimedChild(instance21, 40, 22, {
                "40": {
                    x: -173.55,
                    y: -250.55,
                    a: 0
                },
                "41": {
                    a: 0.2
                },
                "42": {
                    a: 0.4
                },
                "43": {
                    a: 0.6
                },
                "44": {
                    a: 0.8
                },
                "45": {
                    a: 1
                },
                "55": {
                    a: 0.86
                },
                "56": {
                    a: 0.71
                },
                "57": {
                    a: 0.57
                },
                "58": {
                    a: 0.43
                },
                "59": {
                    a: 0.29
                },
                "60": {
                    a: 0.14
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance23, 47, 43, {
                "47": {
                    x: -405.9,
                    y: -108.6,
                    a: 0
                },
                "48": {
                    a: 0.11
                },
                "49": {
                    a: 0.22
                },
                "50": {
                    a: 0.33
                },
                "51": {
                    a: 0.45
                },
                "52": {
                    a: 0.55
                },
                "53": {
                    a: 0.67
                },
                "54": {
                    a: 0.78
                },
                "55": {
                    a: 0.89
                },
                "56": {
                    a: 1
                }
            })
            .addTimedChild(instance22, 42, 48, {
                "42": {
                    x: -76.5,
                    y: 32.9,
                    a: 0
                },
                "43": {
                    a: 0.1
                },
                "44": {
                    a: 0.2
                },
                "45": {
                    a: 0.3
                },
                "46": {
                    a: 0.4
                },
                "47": {
                    a: 0.5
                },
                "48": {
                    a: 0.6
                },
                "49": {
                    a: 0.7
                },
                "50": {
                    a: 0.8
                },
                "51": {
                    a: 0.9
                },
                "52": {
                    a: 1
                }
            });
    });

    var Graphic39 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 63, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic40 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 50, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic41 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 22, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic42 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 19, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic43 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic44 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic45 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 27, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic46 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic47 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic48 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 46, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic49 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 5, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic50 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic51 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic52 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic53 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic54 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 9, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic55 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 12, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic56 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 12, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic57 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic58 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic59 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic60 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 30, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic61 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic62 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 7, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic63 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic64 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 38, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic65 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 3, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic66 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic67 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 24, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic68 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic69 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 11, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic70 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 27, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic71 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic72 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic73 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 24, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic74 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 8, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic75 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance8 = new Graphic71(MovieClip.SYNCHED);
        var instance6 = new Graphic65(MovieClip.SYNCHED);
        var instance5 = new Graphic62(MovieClip.SYNCHED);
        var instance14 = new Graphic69(MovieClip.SYNCHED);
        var instance18 = new Graphic72(MovieClip.SYNCHED);
        var instance13 = new Graphic66(MovieClip.SYNCHED);
        var instance4 = new Graphic59(MovieClip.SYNCHED);
        var instance3 = new Graphic55(MovieClip.SYNCHED);
        var instance7 = new Graphic52(MovieClip.SYNCHED);
        var instance2 = new Graphic49(MovieClip.SYNCHED);
        var instance25 = new Graphic73(MovieClip.SYNCHED);
        var instance22 = new Graphic70(MovieClip.SYNCHED);
        var instance24 = new Graphic67(MovieClip.SYNCHED);
        var instance19 = new Graphic63(MovieClip.SYNCHED);
        var instance36 = new Graphic74(MovieClip.SYNCHED)
            .setTransform(109.9, -156.35)
            .setAlpha(0);
        var instance32 = new Graphic68(MovieClip.SYNCHED);
        var instance28 = new Graphic64(MovieClip.SYNCHED);
        var instance23 = new Graphic60(MovieClip.SYNCHED);
        var instance20 = new Graphic56(MovieClip.SYNCHED);
        var instance16 = new Graphic50(MovieClip.SYNCHED);
        var instance26 = new Graphic57(MovieClip.SYNCHED);
        var instance27 = new Graphic53(MovieClip.SYNCHED);
        var instance15 = new Graphic48(MovieClip.SYNCHED);
        var instance9 = new Graphic46(MovieClip.SYNCHED);
        var instance1 = new Graphic44(MovieClip.SYNCHED);
        var instance21 = new Graphic47(MovieClip.SYNCHED);
        var instance17 = new Graphic42(MovieClip.SYNCHED);
        var instance10 = new Graphic41(MovieClip.SYNCHED);
        var instance33 = new Graphic61(MovieClip.SYNCHED);
        var instance34 = new Graphic58(MovieClip.SYNCHED);
        var instance35 = new Graphic54(MovieClip.SYNCHED);
        var instance31 = new Graphic51(MovieClip.SYNCHED);
        var instance30 = new Graphic45(MovieClip.SYNCHED);
        var instance29 = new Graphic43(MovieClip.SYNCHED);
        var instance12 = new Graphic40(MovieClip.SYNCHED);
        var instance11 = new Graphic39(MovieClip.SYNCHED);
        this.addTimedChild(instance8, 2, 10, {
                "2": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "3": {
                    a: 0.2
                },
                "4": {
                    a: 0.4
                },
                "5": {
                    a: 0.6
                },
                "6": {
                    a: 0.8
                },
                "7": {
                    a: 1
                },
                "8": {
                    a: 0.75
                },
                "9": {
                    a: 0.5
                },
                "10": {
                    a: 0.25
                },
                "11": {
                    a: 0
                }
            })
            .addTimedChild(instance6, 0, 3, {
                "0": {
                    x: -552.95,
                    y: 174.85,
                    a: 0.33
                },
                "1": {
                    a: 0.17
                },
                "2": {
                    a: 0
                }
            })
            .addTimedChild(instance5, 0, 7, {
                "0": {
                    x: -27.35,
                    y: 273.35,
                    a: 0.86
                },
                "1": {
                    a: 0.71
                },
                "2": {
                    a: 0.57
                },
                "3": {
                    a: 0.43
                },
                "4": {
                    a: 0.29
                },
                "5": {
                    a: 0.14
                },
                "6": {
                    a: 0
                }
            })
            .addTimedChild(instance14, 13, 11, {
                "13": {
                    x: 203.85,
                    y: 127.25,
                    a: 0
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.5
                },
                "16": {
                    a: 0.75
                },
                "17": {
                    a: 1
                },
                "19": {
                    a: 0.8
                },
                "20": {
                    a: 0.6
                },
                "21": {
                    a: 0.4
                },
                "22": {
                    a: 0.2
                },
                "23": {
                    a: 0
                }
            })
            .addTimedChild(instance18, 26, 10, {
                "26": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "27": {
                    a: 0.2
                },
                "28": {
                    a: 0.4
                },
                "29": {
                    a: 0.6
                },
                "30": {
                    a: 0.8
                },
                "31": {
                    a: 1
                },
                "32": {
                    a: 0.75
                },
                "33": {
                    a: 0.5
                },
                "34": {
                    a: 0.25
                },
                "35": {
                    a: 0
                }
            })
            .addTimedChild(instance13, 13, 17, {
                "13": {
                    x: -552.95,
                    y: 174.85,
                    a: 0
                },
                "14": {
                    a: 0.17
                },
                "15": {
                    a: 0.33
                },
                "16": {
                    a: 0.5
                },
                "17": {
                    a: 0.67
                },
                "18": {
                    a: 0.83
                },
                "19": {
                    a: 1
                },
                "21": {
                    a: 0.89
                },
                "22": {
                    a: 0.78
                },
                "23": {
                    a: 0.67
                },
                "24": {
                    a: 0.55
                },
                "25": {
                    a: 0.45
                },
                "26": {
                    a: 0.33
                },
                "27": {
                    a: 0.22
                },
                "28": {
                    a: 0.11
                },
                "29": {
                    a: 0
                }
            })
            .addTimedChild(instance4, 0, 17, {
                "0": {
                    x: -313.3,
                    y: 315.75,
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 0, 12, {
                "0": {
                    x: -312.6,
                    y: 79.85,
                    a: 0.8
                },
                "1": {
                    a: 1
                },
                "4": {
                    a: 0.88
                },
                "5": {
                    a: 0.75
                },
                "6": {
                    a: 0.63
                },
                "7": {
                    a: 0.5
                },
                "8": {
                    a: 0.38
                },
                "9": {
                    a: 0.25
                },
                "10": {
                    a: 0.13
                },
                "11": {
                    a: 0
                }
            })
            .addTimedChild(instance7, 1, 23, {
                "1": {
                    x: -405.2,
                    y: 267.35,
                    a: 0
                },
                "2": {
                    a: 0.14
                },
                "3": {
                    a: 0.29
                },
                "4": {
                    a: 0.43
                },
                "5": {
                    a: 0.57
                },
                "6": {
                    a: 0.71
                },
                "7": {
                    a: 0.86
                },
                "8": {
                    a: 1
                },
                "16": {
                    a: 0.88
                },
                "17": {
                    a: 0.75
                },
                "18": {
                    a: 0.63
                },
                "19": {
                    a: 0.5
                },
                "20": {
                    a: 0.38
                },
                "21": {
                    a: 0.25
                },
                "22": {
                    a: 0.13
                },
                "23": {
                    a: 0
                }
            })
            .addTimedChild(instance2, 0, 5, {
                "0": {
                    x: 158.45,
                    y: -61.75,
                    a: 0.5
                },
                "1": {
                    a: 0.38
                },
                "2": {
                    a: 0.25
                },
                "3": {
                    a: 0.13
                },
                "4": {
                    a: 0
                }
            })
            .addTimedChild(instance25, 45, 24, {
                "45": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "46": {
                    a: 0.13
                },
                "47": {
                    a: 0.25
                },
                "48": {
                    a: 0.38
                },
                "49": {
                    a: 0.5
                },
                "50": {
                    a: 0.63
                },
                "51": {
                    a: 0.75
                },
                "52": {
                    a: 0.88
                },
                "53": {
                    a: 1
                },
                "60": {
                    a: 0.89
                },
                "61": {
                    a: 0.78
                },
                "62": {
                    a: 0.67
                },
                "63": {
                    a: 0.55
                },
                "64": {
                    a: 0.45
                },
                "65": {
                    a: 0.33
                },
                "66": {
                    a: 0.22
                },
                "67": {
                    a: 0.11
                },
                "68": {
                    a: 0
                }
            })
            .addTimedChild(instance22, 36, 27, {
                "36": {
                    x: 203.85,
                    y: 127.25,
                    a: 0
                },
                "37": {
                    a: 0.25
                },
                "38": {
                    a: 0.5
                },
                "39": {
                    a: 0.75
                },
                "40": {
                    a: 1
                },
                "57": {
                    a: 0.83
                },
                "58": {
                    a: 0.67
                },
                "59": {
                    a: 0.5
                },
                "60": {
                    a: 0.33
                },
                "61": {
                    a: 0.17
                },
                "62": {
                    a: 0
                }
            })
            .addTimedChild(instance24, 38, 24, {
                "38": {
                    x: -552.95,
                    y: 174.85,
                    a: 0
                },
                "39": {
                    a: 0.17
                },
                "40": {
                    a: 0.33
                },
                "41": {
                    a: 0.5
                },
                "42": {
                    a: 0.67
                },
                "43": {
                    a: 0.83
                },
                "44": {
                    a: 1
                },
                "54": {
                    a: 0.88
                },
                "55": {
                    a: 0.75
                },
                "56": {
                    a: 0.63
                },
                "57": {
                    a: 0.5
                },
                "58": {
                    a: 0.38
                },
                "59": {
                    a: 0.25
                },
                "60": {
                    a: 0.13
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance19, 27, 16, {
                "27": {
                    x: -27.35,
                    y: 273.35,
                    a: 0
                },
                "28": {
                    a: 0.14
                },
                "29": {
                    a: 0.29
                },
                "30": {
                    a: 0.43
                },
                "31": {
                    a: 0.57
                },
                "32": {
                    a: 0.71
                },
                "33": {
                    a: 0.86
                },
                "34": {
                    a: 1
                },
                "36": {
                    a: 0.86
                },
                "37": {
                    a: 0.71
                },
                "38": {
                    a: 0.57
                },
                "39": {
                    a: 0.43
                },
                "40": {
                    a: 0.29
                },
                "41": {
                    a: 0.14
                },
                "42": {
                    a: 0
                }
            })
            .addTimedChild(instance36, 82, 8)
            .addTimedChild(instance32, 67, 23, {
                "67": {
                    x: -552.95,
                    y: 174.85,
                    a: 0
                },
                "68": {
                    a: 0.17
                },
                "69": {
                    a: 0.33
                },
                "70": {
                    a: 0.5
                },
                "71": {
                    a: 0.67
                },
                "72": {
                    a: 0.83
                },
                "73": {
                    a: 1
                },
                "77": {
                    a: 0.83
                },
                "78": {
                    a: 0.67
                },
                "79": {
                    a: 0.5
                },
                "80": {
                    a: 0.33
                },
                "81": {
                    a: 0.17
                },
                "82": {
                    a: 0
                }
            })
            .addTimedChild(instance28, 52, 38, {
                "52": {
                    x: -27.35,
                    y: 273.35,
                    a: 0
                },
                "53": {
                    a: 0.14
                },
                "54": {
                    a: 0.29
                },
                "55": {
                    a: 0.43
                },
                "56": {
                    a: 0.57
                },
                "57": {
                    a: 0.71
                },
                "58": {
                    a: 0.86
                },
                "59": {
                    a: 1
                },
                "80": {
                    a: 0.86
                },
                "81": {
                    a: 0.71
                },
                "82": {
                    a: 0.57
                }
            })
            .addTimedChild(instance23, 37, 30, {
                "37": {
                    x: -313.3,
                    y: 315.75,
                    a: 0
                },
                "38": {
                    a: 0.14
                },
                "39": {
                    a: 0.29
                },
                "40": {
                    a: 0.43
                },
                "41": {
                    a: 0.57
                },
                "42": {
                    a: 0.71
                },
                "43": {
                    a: 0.86
                },
                "44": {
                    a: 1
                },
                "60": {
                    a: 0.86
                },
                "61": {
                    a: 0.71
                },
                "62": {
                    a: 0.57
                },
                "63": {
                    a: 0.43
                },
                "64": {
                    a: 0.29
                },
                "65": {
                    a: 0.14
                },
                "66": {
                    a: 0
                }
            })
            .addTimedChild(instance20, 29, 12, {
                "29": {
                    x: -312.6,
                    y: 79.85,
                    a: 0
                },
                "30": {
                    a: 0.2
                },
                "31": {
                    a: 0.4
                },
                "32": {
                    a: 0.6
                },
                "33": {
                    a: 0.8
                },
                "34": {
                    a: 1
                },
                "37": {
                    a: 0.75
                },
                "38": {
                    a: 0.5
                },
                "39": {
                    a: 0.25
                },
                "40": {
                    a: 0
                }
            })
            .addTimedChild(instance16, 18, 16, {
                "18": {
                    x: 158.45,
                    y: -61.75,
                    a: 0
                },
                "19": {
                    a: 0.14
                },
                "20": {
                    a: 0.29
                },
                "21": {
                    a: 0.43
                },
                "22": {
                    a: 0.57
                },
                "23": {
                    a: 0.71
                },
                "24": {
                    a: 0.86
                },
                "25": {
                    a: 1
                },
                "28": {
                    a: 0.83
                },
                "29": {
                    a: 0.67
                },
                "30": {
                    a: 0.5
                },
                "31": {
                    a: 0.33
                },
                "32": {
                    a: 0.17
                },
                "33": {
                    a: 0
                }
            })
            .addTimedChild(instance26, 46, 16, {
                "46": {
                    x: -312.6,
                    y: 79.85,
                    a: 0
                },
                "47": {
                    a: 0.2
                },
                "48": {
                    a: 0.4
                },
                "49": {
                    a: 0.6
                },
                "50": {
                    a: 0.8
                },
                "51": {
                    a: 1
                },
                "54": {
                    a: 0.88
                },
                "55": {
                    a: 0.75
                },
                "56": {
                    a: 0.63
                },
                "57": {
                    a: 0.5
                },
                "58": {
                    a: 0.38
                },
                "59": {
                    a: 0.25
                },
                "60": {
                    a: 0.13
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance27, 52, 14, {
                "52": {
                    x: -405.2,
                    y: 267.35,
                    a: 0
                },
                "53": {
                    a: 0.17
                },
                "54": {
                    a: 0.33
                },
                "55": {
                    a: 0.5
                },
                "56": {
                    a: 0.67
                },
                "57": {
                    a: 0.83
                },
                "58": {
                    a: 1
                },
                "60": {
                    a: 0.83
                },
                "61": {
                    a: 0.67
                },
                "62": {
                    a: 0.5
                },
                "63": {
                    a: 0.33
                },
                "64": {
                    a: 0.17
                },
                "65": {
                    a: 0
                }
            })
            .addTimedChild(instance15, 14, 46, {
                "14": {
                    x: -75.35,
                    y: -61.2,
                    a: 1
                },
                "27": {
                    a: 0.9
                },
                "28": {
                    a: 0.8
                },
                "29": {
                    a: 0.7
                },
                "30": {
                    a: 0.6
                },
                "31": {
                    a: 0.5
                },
                "32": {
                    a: 0.4
                },
                "33": {
                    a: 0.3
                },
                "34": {
                    a: 0.2
                },
                "35": {
                    a: 0.1
                },
                "36": {
                    a: 0
                },
                "37": {
                    a: 1
                },
                "53": {
                    a: 0.86
                },
                "54": {
                    a: 0.71
                },
                "55": {
                    a: 0.57
                },
                "56": {
                    a: 0.43
                },
                "57": {
                    a: 0.29
                },
                "58": {
                    a: 0.14
                },
                "59": {
                    a: 0
                }
            })
            .addTimedChild(instance9, 4, 17, {
                "4": {
                    x: -266.7,
                    y: -107.9,
                    a: 0
                },
                "5": {
                    a: 0.13
                },
                "6": {
                    a: 0.25
                },
                "7": {
                    a: 0.38
                },
                "8": {
                    a: 0.5
                },
                "9": {
                    a: 0.63
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.88
                },
                "12": {
                    a: 1
                },
                "14": {
                    a: 0.86
                },
                "15": {
                    a: 0.71
                },
                "16": {
                    a: 0.57
                },
                "17": {
                    a: 0.43
                },
                "18": {
                    a: 0.29
                },
                "19": {
                    a: 0.14
                },
                "20": {
                    a: 0
                }
            })
            .addTimedChild(instance1, 0, 13, {
                "0": {
                    x: -454.5,
                    y: 175,
                    a: 1
                },
                "6": {
                    a: 0.86
                },
                "7": {
                    a: 0.71
                },
                "8": {
                    a: 0.57
                },
                "9": {
                    a: 0.43
                },
                "10": {
                    a: 0.29
                },
                "11": {
                    a: 0.14
                },
                "12": {
                    a: 0
                }
            })
            .addTimedChild(instance21, 35, 18, {
                "35": {
                    x: -266.7,
                    y: -107.9,
                    a: 0
                },
                "36": {
                    a: 0.13
                },
                "37": {
                    a: 0.25
                },
                "38": {
                    a: 0.38
                },
                "39": {
                    a: 0.5
                },
                "40": {
                    a: 0.63
                },
                "41": {
                    a: 0.75
                },
                "42": {
                    a: 0.88
                },
                "43": {
                    a: 1
                },
                "46": {
                    a: 0.86
                },
                "47": {
                    a: 0.71
                },
                "48": {
                    a: 0.57
                },
                "49": {
                    a: 0.43
                },
                "50": {
                    a: 0.29
                },
                "51": {
                    a: 0.14
                },
                "52": {
                    a: 0
                }
            })
            .addTimedChild(instance17, 26, 19, {
                "26": {
                    x: -454.15,
                    y: -14.85,
                    a: 0
                },
                "27": {
                    a: 0.17
                },
                "28": {
                    a: 0.33
                },
                "29": {
                    a: 0.5
                },
                "30": {
                    a: 0.67
                },
                "31": {
                    a: 0.83
                },
                "32": {
                    a: 1
                },
                "35": {
                    a: 0.89
                },
                "36": {
                    a: 0.78
                },
                "37": {
                    a: 0.67
                },
                "38": {
                    a: 0.55
                },
                "39": {
                    a: 0.45
                },
                "40": {
                    a: 0.33
                },
                "41": {
                    a: 0.22
                },
                "42": {
                    a: 0.11
                },
                "43": {
                    a: 0
                }
            })
            .addTimedChild(instance10, 5, 22, {
                "5": {
                    x: -173.55,
                    y: -250.55,
                    a: 0
                },
                "6": {
                    a: 0.2
                },
                "7": {
                    a: 0.4
                },
                "8": {
                    a: 0.6
                },
                "9": {
                    a: 0.8
                },
                "10": {
                    a: 1
                },
                "20": {
                    a: 0.86
                },
                "21": {
                    a: 0.71
                },
                "22": {
                    a: 0.57
                },
                "23": {
                    a: 0.43
                },
                "24": {
                    a: 0.29
                },
                "25": {
                    a: 0.14
                },
                "26": {
                    a: 0
                }
            })
            .addTimedChild(instance33, 73, 17, {
                "73": {
                    x: -313.3,
                    y: 315.75,
                    a: 0
                },
                "74": {
                    a: 0.14
                },
                "75": {
                    a: 0.29
                },
                "76": {
                    a: 0.43
                },
                "77": {
                    a: 0.57
                },
                "78": {
                    a: 0.71
                },
                "79": {
                    a: 0.86
                },
                "80": {
                    a: 1
                }
            })
            .addTimedChild(instance34, 76, 14, {
                "76": {
                    x: -312.6,
                    y: 79.85,
                    a: 0
                },
                "77": {
                    a: 0.2
                },
                "78": {
                    a: 0.4
                },
                "79": {
                    a: 0.6
                },
                "80": {
                    a: 0.8
                },
                "81": {
                    a: 1
                }
            })
            .addTimedChild(instance35, 81, 9, {
                "81": {
                    x: -405.2,
                    y: 267.35,
                    a: 0
                },
                "82": {
                    a: 0.14
                }
            })
            .addTimedChild(instance31, 67, 23, {
                "67": {
                    x: 158.45,
                    y: -61.75,
                    a: 0
                },
                "68": {
                    a: 0.14
                },
                "69": {
                    a: 0.29
                },
                "70": {
                    a: 0.43
                },
                "71": {
                    a: 0.57
                },
                "72": {
                    a: 0.71
                },
                "73": {
                    a: 0.86
                },
                "74": {
                    a: 1
                },
                "77": {
                    a: 0.88
                },
                "78": {
                    a: 0.75
                },
                "79": {
                    a: 0.63
                },
                "80": {
                    a: 0.5
                },
                "81": {
                    a: 0.38
                },
                "82": {
                    a: 0.25
                }
            })
            .addTimedChild(instance30, 63, 27, {
                "63": {
                    x: -454.5,
                    y: 175,
                    a: 0
                },
                "64": {
                    a: 0.14
                },
                "65": {
                    a: 0.29
                },
                "66": {
                    a: 0.43
                },
                "67": {
                    a: 0.57
                },
                "68": {
                    a: 0.71
                },
                "69": {
                    a: 0.86
                },
                "70": {
                    a: 1
                }
            })
            .addTimedChild(instance29, 55, 18, {
                "55": {
                    x: -454.15,
                    y: -14.85,
                    a: 0
                },
                "56": {
                    a: 0.17
                },
                "57": {
                    a: 0.33
                },
                "58": {
                    a: 0.5
                },
                "59": {
                    a: 0.67
                },
                "60": {
                    a: 0.83
                },
                "61": {
                    a: 1
                },
                "64": {
                    a: 0.89
                },
                "65": {
                    a: 0.78
                },
                "66": {
                    a: 0.67
                },
                "67": {
                    a: 0.55
                },
                "68": {
                    a: 0.45
                },
                "69": {
                    a: 0.33
                },
                "70": {
                    a: 0.22
                },
                "71": {
                    a: 0.11
                },
                "72": {
                    a: 0
                }
            })
            .addTimedChild(instance12, 12, 50, {
                "12": {
                    x: -405.9,
                    y: -108.6,
                    a: 0
                },
                "13": {
                    a: 0.11
                },
                "14": {
                    a: 0.22
                },
                "15": {
                    a: 0.33
                },
                "16": {
                    a: 0.45
                },
                "17": {
                    a: 0.55
                },
                "18": {
                    a: 0.67
                },
                "19": {
                    a: 0.78
                },
                "20": {
                    a: 0.89
                },
                "21": {
                    a: 1
                },
                "54": {
                    a: 0.88
                },
                "55": {
                    a: 0.75
                },
                "56": {
                    a: 0.63
                },
                "57": {
                    a: 0.5
                },
                "58": {
                    a: 0.38
                },
                "59": {
                    a: 0.25
                },
                "60": {
                    a: 0.13
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance11, 7, 63, {
                "7": {
                    x: -76.5,
                    y: 32.9,
                    a: 0
                },
                "8": {
                    a: 0.1
                },
                "9": {
                    a: 0.2
                },
                "10": {
                    a: 0.3
                },
                "11": {
                    a: 0.4
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.6
                },
                "14": {
                    a: 0.7
                },
                "15": {
                    a: 0.8
                },
                "16": {
                    a: 0.9
                },
                "17": {
                    a: 1
                },
                "60": {
                    a: 0.9
                },
                "61": {
                    a: 0.8
                },
                "62": {
                    a: 0.7
                },
                "63": {
                    a: 0.6
                },
                "64": {
                    a: 0.5
                },
                "65": {
                    a: 0.4
                },
                "66": {
                    a: 0.3
                },
                "67": {
                    a: 0.2
                },
                "68": {
                    a: 0.1
                },
                "69": {
                    a: 0
                }
            });
    });

    var Graphic76 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot25"))
            .setTransform(-18.6, -17.9);
        this.addTimedChild(instance1);
    });

    var Graphic77 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot25"))
            .setTransform(-18.6, -17.9);
        this.addTimedChild(instance1);
    });

    var Graphic78 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot25"))
            .setTransform(-18.6, -17.9);
        this.addTimedChild(instance1);
    });

    var Graphic79 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot25"))
            .setTransform(-18.6, -17.9);
        this.addTimedChild(instance1);
    });

    var Graphic80 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 20, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic81 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic82 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic83 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic84 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 8, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic85 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic86 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 22, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic87 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic88 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic89 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 8, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic90 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 15, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic91 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic92 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic93 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic94 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic95 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic96 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 22, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic97 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic98 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 15, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic99 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic100 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic101 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic102 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic103 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic104 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic105 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 20, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic106 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic107 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic108 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic109 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance1 = new Sprite(fromFrame("dotgrid1"))
            .setTransform(-408.75, -322.1);
        this.addTimedChild(instance1);
    });

    var Graphic110 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 90, loop: false });
        var instance4 = new Graphic109(MovieClip.SYNCHED);
        var instance20 = new Graphic107(MovieClip.SYNCHED);
        var instance13 = new Graphic105(MovieClip.SYNCHED);
        var instance12 = new Graphic103(MovieClip.SYNCHED);
        var instance8 = new Graphic101(MovieClip.SYNCHED);
        var instance3 = new Graphic97(MovieClip.SYNCHED);
        var instance7 = new Graphic92(MovieClip.SYNCHED);
        var instance10 = new Graphic90(MovieClip.SYNCHED);
        var instance6 = new Graphic87(MovieClip.SYNCHED);
        var instance19 = new Graphic98(MovieClip.SYNCHED);
        var instance14 = new Graphic94(MovieClip.SYNCHED);
        var instance32 = new Graphic108(MovieClip.SYNCHED);
        var instance26 = new Graphic106(MovieClip.SYNCHED);
        var instance28 = new Graphic104(MovieClip.SYNCHED);
        var instance24 = new Graphic102(MovieClip.SYNCHED);
        var instance22 = new Graphic95(MovieClip.SYNCHED);
        var instance21 = new Graphic93(MovieClip.SYNCHED);
        var instance16 = new Graphic88(MovieClip.SYNCHED);
        var instance11 = new Graphic85(MovieClip.SYNCHED);
        var instance5 = new Graphic81(MovieClip.SYNCHED);
        var instance9 = new Graphic78(MovieClip.SYNCHED);
        var instance27 = new Graphic99(MovieClip.SYNCHED);
        var instance31 = new Graphic96(MovieClip.SYNCHED);
        var instance29 = new Graphic91(MovieClip.SYNCHED);
        var instance23 = new Graphic86(MovieClip.SYNCHED);
        var instance18 = new Graphic82(MovieClip.SYNCHED);
        var instance17 = new Graphic80(MovieClip.SYNCHED);
        var instance15 = new Graphic76(MovieClip.SYNCHED);
        var instance33 = new Graphic83(MovieClip.SYNCHED);
        var instance30 = new Graphic79(MovieClip.SYNCHED);
        var instance25 = new Graphic77(MovieClip.SYNCHED);
        var instance34 = new Graphic100(MovieClip.SYNCHED);
        var instance36 = new Graphic89(MovieClip.SYNCHED)
            .setTransform(-70.15, -313.2, 0.926, 0.926)
            .setAlpha(0);
        var instance35 = new Graphic84(MovieClip.SYNCHED)
            .setTransform(233.2, 211.75, 0.926, 0.926)
            .setAlpha(0);
        var instance2 = new Graphic75(MovieClip.SYNCHED)
            .setTransform(-187.65, -36.85, 0.926, 0.926, 0, 0, 3.142);
        var instance1 = new Graphic38(MovieClip.SYNCHED)
            .setTransform(-187.65, -36.85, 0.926, 0.926, 0, 0, 3.142);
        this.addTimedChild(instance4, 0, 90, {
                "0": {
                    x: -28.5,
                    y: -6.65
                }
            })
            .addTimedChild(instance20, 31, 17, {
                "31": {
                    x: 189.45,
                    y: -225.7,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "32": {
                    a: 0.17
                },
                "33": {
                    a: 0.33
                },
                "34": {
                    a: 0.5
                },
                "35": {
                    a: 0.67
                },
                "36": {
                    a: 0.83
                },
                "37": {
                    a: 1
                },
                "42": {
                    a: 0.83
                },
                "43": {
                    a: 0.67
                },
                "44": {
                    a: 0.5
                },
                "45": {
                    a: 0.33
                },
                "46": {
                    a: 0.17
                },
                "47": {
                    a: 0
                }
            })
            .addTimedChild(instance13, 13, 20, {
                "13": {
                    x: -161.25,
                    y: -138.3,
                    sx: 0.977,
                    sy: 0.977,
                    a: 0
                },
                "14": {
                    a: 0.13
                },
                "15": {
                    a: 0.25
                },
                "16": {
                    a: 0.38
                },
                "17": {
                    a: 0.5
                },
                "18": {
                    a: 0.63
                },
                "19": {
                    a: 0.75
                },
                "20": {
                    a: 0.88
                },
                "21": {
                    a: 1
                },
                "25": {
                    a: 0.88
                },
                "26": {
                    a: 0.75
                },
                "27": {
                    a: 0.63
                },
                "28": {
                    a: 0.5
                },
                "29": {
                    a: 0.38
                },
                "30": {
                    a: 0.25
                },
                "31": {
                    a: 0.13
                },
                "32": {
                    a: 0
                }
            })
            .addTimedChild(instance12, 10, 13, {
                "10": {
                    x: -247.8,
                    y: -138.3,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "11": {
                    a: 0.17
                },
                "12": {
                    a: 0.33
                },
                "13": {
                    a: 0.5
                },
                "14": {
                    a: 0.67
                },
                "15": {
                    a: 0.83
                },
                "16": {
                    a: 1
                },
                "17": {
                    a: 0.83
                },
                "18": {
                    a: 0.67
                },
                "19": {
                    a: 0.5
                },
                "20": {
                    a: 0.33
                },
                "21": {
                    a: 0.17
                },
                "22": {
                    a: 0
                }
            })
            .addTimedChild(instance8, 4, 13, {
                "4": {
                    x: -204.3,
                    y: 124.5,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "5": {
                    a: 0.17
                },
                "6": {
                    a: 0.33
                },
                "7": {
                    a: 0.5
                },
                "8": {
                    a: 0.67
                },
                "9": {
                    a: 0.83
                },
                "10": {
                    a: 1
                },
                "13": {
                    a: 0.75
                },
                "14": {
                    a: 0.5
                },
                "15": {
                    a: 0.25
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 0, 17, {
                "0": {
                    x: 276.05,
                    y: 80.3,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "1": {
                    a: 0.17
                },
                "2": {
                    a: 0.33
                },
                "3": {
                    a: 0.5
                },
                "4": {
                    a: 0.67
                },
                "5": {
                    a: 0.83
                },
                "6": {
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance7, 3, 16, {
                "3": {
                    x: 233.55,
                    y: -139.1,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "4": {
                    a: 0.17
                },
                "5": {
                    a: 0.33
                },
                "6": {
                    a: 0.5
                },
                "7": {
                    a: 0.67
                },
                "8": {
                    a: 0.83
                },
                "9": {
                    a: 1
                },
                "12": {
                    a: 0.86
                },
                "13": {
                    a: 0.71
                },
                "14": {
                    a: 0.57
                },
                "15": {
                    a: 0.43
                },
                "16": {
                    a: 0.29
                },
                "17": {
                    a: 0.14
                },
                "18": {
                    a: 0
                }
            })
            .addTimedChild(instance10, 7, 15, {
                "7": {
                    x: -248.6,
                    y: 212.25,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "8": {
                    a: 0.2
                },
                "9": {
                    a: 0.4
                },
                "10": {
                    a: 0.6
                },
                "11": {
                    a: 0.8
                },
                "12": {
                    a: 1
                },
                "16": {
                    a: 0.83
                },
                "17": {
                    a: 0.67
                },
                "18": {
                    a: 0.5
                },
                "19": {
                    a: 0.33
                },
                "20": {
                    a: 0.17
                },
                "21": {
                    a: 0
                }
            })
            .addTimedChild(instance6, 2, 13, {
                "2": {
                    x: -70.15,
                    y: -313.2,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "3": {
                    a: 0.2
                },
                "4": {
                    a: 0.4
                },
                "5": {
                    a: 0.6
                },
                "6": {
                    a: 0.8
                },
                "7": {
                    a: 1
                },
                "9": {
                    a: 0.83
                },
                "10": {
                    a: 0.67
                },
                "11": {
                    a: 0.5
                },
                "12": {
                    a: 0.33
                },
                "13": {
                    a: 0.17
                },
                "14": {
                    a: 0
                }
            })
            .addTimedChild(instance19, 31, 15, {
                "31": {
                    x: 276.05,
                    y: 80.3,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "32": {
                    a: 0.2
                },
                "33": {
                    a: 0.4
                },
                "34": {
                    a: 0.6
                },
                "35": {
                    a: 0.8
                },
                "36": {
                    a: 1
                },
                "40": {
                    a: 0.83
                },
                "41": {
                    a: 0.67
                },
                "42": {
                    a: 0.5
                },
                "43": {
                    a: 0.33
                },
                "44": {
                    a: 0.17
                },
                "45": {
                    a: 0
                }
            })
            .addTimedChild(instance14, 22, 14, {
                "22": {
                    x: 145.7,
                    y: -6.55,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "23": {
                    a: 0.25
                },
                "24": {
                    a: 0.5
                },
                "25": {
                    a: 0.75
                },
                "26": {
                    a: 1
                },
                "31": {
                    a: 0.8
                },
                "32": {
                    a: 0.6
                },
                "33": {
                    a: 0.4
                },
                "34": {
                    a: 0.2
                },
                "35": {
                    a: 0
                }
            })
            .addTimedChild(instance32, 58, 10, {
                "58": {
                    x: 189.45,
                    y: -225.7,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "59": {
                    a: 0.2
                },
                "60": {
                    a: 0.4
                },
                "61": {
                    a: 0.6
                },
                "62": {
                    a: 0.8
                },
                "63": {
                    a: 1
                },
                "64": {
                    a: 0.75
                },
                "65": {
                    a: 0.5
                },
                "66": {
                    a: 0.25
                },
                "67": {
                    a: 0
                }
            })
            .addTimedChild(instance26, 49, 13, {
                "49": {
                    x: 102.05,
                    y: -268.85,
                    sx: 0.977,
                    sy: 0.977,
                    a: 0
                },
                "50": {
                    a: 0.2
                },
                "51": {
                    a: 0.4
                },
                "52": {
                    a: 0.6
                },
                "53": {
                    a: 0.8
                },
                "54": {
                    a: 1
                },
                "58": {
                    a: 0.75
                },
                "59": {
                    a: 0.5
                },
                "60": {
                    a: 0.25
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance28, 55, 13, {
                "55": {
                    x: -247.8,
                    y: -138.3,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "56": {
                    a: 0.17
                },
                "57": {
                    a: 0.33
                },
                "58": {
                    a: 0.5
                },
                "59": {
                    a: 0.67
                },
                "60": {
                    a: 0.83
                },
                "61": {
                    a: 1
                },
                "64": {
                    a: 0.75
                },
                "65": {
                    a: 0.5
                },
                "66": {
                    a: 0.25
                },
                "67": {
                    a: 0
                }
            })
            .addTimedChild(instance24, 46, 13, {
                "46": {
                    x: -204.3,
                    y: 124.5,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "47": {
                    a: 0.2
                },
                "48": {
                    a: 0.4
                },
                "49": {
                    a: 0.6
                },
                "50": {
                    a: 0.8
                },
                "51": {
                    a: 1
                },
                "55": {
                    a: 0.75
                },
                "56": {
                    a: 0.5
                },
                "57": {
                    a: 0.25
                },
                "58": {
                    a: 0
                }
            })
            .addTimedChild(instance22, 39, 13, {
                "39": {
                    x: 145.7,
                    y: -6.55,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "40": {
                    a: 0.17
                },
                "41": {
                    a: 0.33
                },
                "42": {
                    a: 0.5
                },
                "43": {
                    a: 0.67
                },
                "44": {
                    a: 0.83
                },
                "45": {
                    a: 1
                },
                "47": {
                    a: 0.8
                },
                "48": {
                    a: 0.6
                },
                "49": {
                    a: 0.4
                },
                "50": {
                    a: 0.2
                },
                "51": {
                    a: 0
                }
            })
            .addTimedChild(instance21, 33, 18, {
                "33": {
                    x: -29.45,
                    y: 214.55,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "34": {
                    a: 0.11
                },
                "35": {
                    a: 0.22
                },
                "36": {
                    a: 0.33
                },
                "37": {
                    a: 0.45
                },
                "38": {
                    a: 0.55
                },
                "39": {
                    a: 0.67
                },
                "40": {
                    a: 0.78
                },
                "41": {
                    a: 0.89
                },
                "42": {
                    a: 1
                },
                "44": {
                    a: 0.86
                },
                "45": {
                    a: 0.71
                },
                "46": {
                    a: 0.57
                },
                "47": {
                    a: 0.43
                },
                "48": {
                    a: 0.29
                },
                "49": {
                    a: 0.14
                },
                "50": {
                    a: 0
                }
            })
            .addTimedChild(instance16, 24, 13, {
                "24": {
                    x: -70.15,
                    y: -313.2,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "25": {
                    a: 0.2
                },
                "26": {
                    a: 0.4
                },
                "27": {
                    a: 0.6
                },
                "28": {
                    a: 0.8
                },
                "29": {
                    a: 1
                },
                "33": {
                    a: 0.75
                },
                "34": {
                    a: 0.5
                },
                "35": {
                    a: 0.25
                },
                "36": {
                    a: 0
                }
            })
            .addTimedChild(instance11, 10, 16, {
                "10": {
                    x: -333.8,
                    y: 80.15,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "11": {
                    a: 0.14
                },
                "12": {
                    a: 0.29
                },
                "13": {
                    a: 0.43
                },
                "14": {
                    a: 0.57
                },
                "15": {
                    a: 0.71
                },
                "16": {
                    a: 0.86
                },
                "17": {
                    a: 1
                },
                "18": {
                    a: 0.88
                },
                "19": {
                    a: 0.75
                },
                "20": {
                    a: 0.63
                },
                "21": {
                    a: 0.5
                },
                "22": {
                    a: 0.38
                },
                "23": {
                    a: 0.25
                },
                "24": {
                    a: 0.13
                },
                "25": {
                    a: 0
                }
            })
            .addTimedChild(instance5, 2, 17, {
                "2": {
                    x: 233.2,
                    y: 211.75,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "3": {
                    a: 0.14
                },
                "4": {
                    a: 0.29
                },
                "5": {
                    a: 0.43
                },
                "6": {
                    a: 0.57
                },
                "7": {
                    a: 0.71
                },
                "8": {
                    a: 0.86
                },
                "9": {
                    a: 1
                },
                "13": {
                    a: 0.83
                },
                "14": {
                    a: 0.67
                },
                "15": {
                    a: 0.5
                },
                "16": {
                    a: 0.33
                },
                "17": {
                    a: 0.17
                },
                "18": {
                    a: 0
                }
            })
            .addTimedChild(instance9, 5, 17, {
                "5": {
                    x: -247.35,
                    y: -51.15,
                    sx: 0.547,
                    sy: 0.547,
                    a: 0
                },
                "6": {
                    x: -247.286,
                    y: -51.137,
                    sx: 0.608,
                    sy: 0.608,
                    a: 0.11
                },
                "7": {
                    x: -247.261,
                    y: -51.122,
                    sx: 0.67,
                    sy: 0.67,
                    a: 0.22
                },
                "8": {
                    x: -247.19,
                    y: -51.108,
                    sx: 0.732,
                    sy: 0.732,
                    a: 0.33
                },
                "9": {
                    x: -247.172,
                    y: -51.095,
                    sx: 0.794,
                    sy: 0.794,
                    a: 0.45
                },
                "10": {
                    x: -247.154,
                    y: -51.131,
                    sx: 0.856,
                    sy: 0.856,
                    a: 0.55
                },
                "11": {
                    x: -247.09,
                    y: -51.118,
                    sx: 0.917,
                    sy: 0.917,
                    a: 0.67
                },
                "12": {
                    x: -247.065,
                    y: -51.103,
                    sx: 0.979,
                    sy: 0.979,
                    a: 0.78
                },
                "13": {
                    x: -247.044,
                    y: -51.089,
                    sx: 1.041,
                    sy: 1.041,
                    a: 0.89
                },
                "14": {
                    x: -246.95,
                    y: -51.1,
                    sx: 1.103,
                    sy: 1.103,
                    a: 1
                },
                "15": {
                    x: -246.909,
                    y: -51.137,
                    sx: 1.138,
                    sy: 1.138,
                    a: 0.86
                },
                "16": {
                    x: -246.872,
                    y: -51.124,
                    sx: 1.174,
                    sy: 1.174,
                    a: 0.71
                },
                "17": {
                    x: -246.832,
                    y: -51.111,
                    sx: 1.209,
                    sy: 1.209,
                    a: 0.57
                },
                "18": {
                    x: -246.741,
                    y: -51.097,
                    sx: 1.244,
                    sy: 1.244,
                    a: 0.43
                },
                "19": {
                    x: -246.751,
                    y: -51.134,
                    sx: 1.28,
                    sy: 1.28,
                    a: 0.29
                },
                "20": {
                    x: -246.714,
                    y: -51.121,
                    sx: 1.315,
                    sy: 1.315,
                    a: 0.14
                },
                "21": {
                    x: -246.8,
                    y: -51.15,
                    sx: 1.35,
                    sy: 1.35,
                    a: 0
                }
            })
            .addTimedChild(instance27, 54, 17, {
                "54": {
                    x: 276.05,
                    y: 80.3,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "55": {
                    a: 0.13
                },
                "56": {
                    a: 0.25
                },
                "57": {
                    a: 0.38
                },
                "58": {
                    a: 0.5
                },
                "59": {
                    a: 0.63
                },
                "60": {
                    a: 0.75
                },
                "61": {
                    a: 0.88
                },
                "62": {
                    a: 1
                },
                "64": {
                    a: 0.86
                },
                "65": {
                    a: 0.71
                },
                "66": {
                    a: 0.57
                },
                "67": {
                    a: 0.43
                },
                "68": {
                    a: 0.29
                },
                "69": {
                    a: 0.14
                },
                "70": {
                    a: 0
                }
            })
            .addTimedChild(instance31, 58, 22, {
                "58": {
                    x: 145.7,
                    y: -6.55,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "59": {
                    a: 0.11
                },
                "60": {
                    a: 0.22
                },
                "61": {
                    a: 0.33
                },
                "62": {
                    a: 0.45
                },
                "63": {
                    a: 0.55
                },
                "64": {
                    a: 0.67
                },
                "65": {
                    a: 0.78
                },
                "66": {
                    a: 0.89
                },
                "67": {
                    a: 1
                },
                "69": {
                    a: 0.91
                },
                "70": {
                    a: 0.82
                },
                "71": {
                    a: 0.73
                },
                "72": {
                    a: 0.64
                },
                "73": {
                    a: 0.55
                },
                "74": {
                    a: 0.45
                },
                "75": {
                    a: 0.36
                },
                "76": {
                    a: 0.27
                },
                "77": {
                    a: 0.18
                },
                "78": {
                    a: 0.09
                },
                "79": {
                    a: 0
                }
            })
            .addTimedChild(instance29, 57, 13, {
                "57": {
                    x: -248.6,
                    y: 212.25,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "58": {
                    a: 0.2
                },
                "59": {
                    a: 0.4
                },
                "60": {
                    a: 0.6
                },
                "61": {
                    a: 0.8
                },
                "62": {
                    a: 1
                },
                "64": {
                    a: 0.83
                },
                "65": {
                    a: 0.67
                },
                "66": {
                    a: 0.5
                },
                "67": {
                    a: 0.33
                },
                "68": {
                    a: 0.17
                },
                "69": {
                    a: 0
                }
            })
            .addTimedChild(instance23, 41, 22, {
                "41": {
                    x: -116.75,
                    y: -138.45,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "42": {
                    a: 0.11
                },
                "43": {
                    a: 0.22
                },
                "44": {
                    a: 0.33
                },
                "45": {
                    a: 0.45
                },
                "46": {
                    a: 0.55
                },
                "47": {
                    a: 0.67
                },
                "48": {
                    a: 0.78
                },
                "49": {
                    a: 0.89
                },
                "50": {
                    a: 1
                },
                "53": {
                    a: 0.9
                },
                "54": {
                    a: 0.8
                },
                "55": {
                    a: 0.7
                },
                "56": {
                    a: 0.6
                },
                "57": {
                    a: 0.5
                },
                "58": {
                    a: 0.4
                },
                "59": {
                    a: 0.3
                },
                "60": {
                    a: 0.2
                },
                "61": {
                    a: 0.1
                },
                "62": {
                    a: 0
                }
            })
            .addTimedChild(instance18, 31, 16, {
                "31": {
                    x: 233.2,
                    y: 211.75,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "32": {
                    a: 0.14
                },
                "33": {
                    a: 0.29
                },
                "34": {
                    a: 0.43
                },
                "35": {
                    a: 0.57
                },
                "36": {
                    a: 0.71
                },
                "37": {
                    a: 0.86
                },
                "38": {
                    a: 1
                },
                "40": {
                    a: 0.86
                },
                "41": {
                    a: 0.71
                },
                "42": {
                    a: 0.57
                },
                "43": {
                    a: 0.43
                },
                "44": {
                    a: 0.29
                },
                "45": {
                    a: 0.14
                },
                "46": {
                    a: 0
                }
            })
            .addTimedChild(instance17, 31, 20, {
                "31": {
                    x: -28.9,
                    y: -6.3,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "32": {
                    a: 0.11
                },
                "33": {
                    a: 0.22
                },
                "34": {
                    a: 0.33
                },
                "35": {
                    a: 0.45
                },
                "36": {
                    a: 0.55
                },
                "37": {
                    a: 0.67
                },
                "38": {
                    a: 0.78
                },
                "39": {
                    a: 0.89
                },
                "40": {
                    a: 1
                },
                "41": {
                    a: 0.9
                },
                "42": {
                    a: 0.8
                },
                "43": {
                    a: 0.7
                },
                "44": {
                    a: 0.6
                },
                "45": {
                    a: 0.5
                },
                "46": {
                    a: 0.4
                },
                "47": {
                    a: 0.3
                },
                "48": {
                    a: 0.2
                },
                "49": {
                    a: 0.1
                },
                "50": {
                    a: 0
                }
            })
            .addTimedChild(instance15, 24, 17, {
                "24": {
                    x: 102.9,
                    y: -181.45,
                    sx: 0.547,
                    sy: 0.547,
                    a: 0
                },
                "25": {
                    x: 102.943,
                    y: -181.43,
                    sx: 0.608,
                    sy: 0.608,
                    a: 0.11
                },
                "26": {
                    x: 102.98,
                    y: -181.452,
                    sx: 0.67,
                    sy: 0.67,
                    a: 0.22
                },
                "27": {
                    x: 103.02,
                    y: -181.426,
                    sx: 0.732,
                    sy: 0.732,
                    a: 0.33
                },
                "28": {
                    x: 103.061,
                    y: -181.453,
                    sx: 0.794,
                    sy: 0.794,
                    a: 0.45
                },
                "29": {
                    x: 103.102,
                    y: -181.43,
                    sx: 0.856,
                    sy: 0.856,
                    a: 0.55
                },
                "30": {
                    x: 103.144,
                    y: -181.41,
                    sx: 0.917,
                    sy: 0.917,
                    a: 0.67
                },
                "31": {
                    x: 103.182,
                    y: -181.432,
                    sx: 0.979,
                    sy: 0.979,
                    a: 0.78
                },
                "32": {
                    x: 103.221,
                    y: -181.406,
                    sx: 1.041,
                    sy: 1.041,
                    a: 0.89
                },
                "33": {
                    x: 103.3,
                    y: -181.4,
                    sx: 1.103,
                    sy: 1.103,
                    a: 1
                },
                "34": {
                    x: 103.309,
                    y: -181.41,
                    sx: 1.138,
                    sy: 1.138,
                    a: 0.86
                },
                "35": {
                    x: 103.369,
                    y: -181.423,
                    sx: 1.174,
                    sy: 1.174,
                    a: 0.71
                },
                "36": {
                    x: 103.377,
                    y: -181.433,
                    sx: 1.209,
                    sy: 1.209,
                    a: 0.57
                },
                "37": {
                    x: 103.436,
                    y: -181.392,
                    sx: 1.244,
                    sy: 1.244,
                    a: 0.43
                },
                "38": {
                    x: 103.494,
                    y: -181.402,
                    sx: 1.28,
                    sy: 1.28,
                    a: 0.29
                },
                "39": {
                    x: 103.504,
                    y: -181.415,
                    sx: 1.315,
                    sy: 1.315,
                    a: 0.14
                },
                "40": {
                    x: 103.5,
                    y: -181.45,
                    sx: 1.35,
                    sy: 1.35,
                    a: 0
                }
            })
            .addTimedChild(instance33, 66, 14, {
                "66": {
                    x: -203.9,
                    y: -138.3,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "67": {
                    a: 0.2
                },
                "68": {
                    a: 0.4
                },
                "69": {
                    a: 0.6
                },
                "70": {
                    a: 0.8
                },
                "71": {
                    a: 1
                },
                "75": {
                    a: 0.8
                },
                "76": {
                    a: 0.6
                },
                "77": {
                    a: 0.4
                },
                "78": {
                    a: 0.2
                },
                "79": {
                    a: 0
                }
            })
            .addTimedChild(instance30, 58, 17, {
                "58": {
                    x: -247.35,
                    y: -51.15,
                    sx: 0.547,
                    sy: 0.547,
                    a: 0
                },
                "59": {
                    x: -247.286,
                    y: -51.137,
                    sx: 0.608,
                    sy: 0.608,
                    a: 0.11
                },
                "60": {
                    x: -247.261,
                    y: -51.122,
                    sx: 0.67,
                    sy: 0.67,
                    a: 0.22
                },
                "61": {
                    x: -247.19,
                    y: -51.108,
                    sx: 0.732,
                    sy: 0.732,
                    a: 0.33
                },
                "62": {
                    x: -247.172,
                    y: -51.095,
                    sx: 0.794,
                    sy: 0.794,
                    a: 0.45
                },
                "63": {
                    x: -247.154,
                    y: -51.131,
                    sx: 0.856,
                    sy: 0.856,
                    a: 0.55
                },
                "64": {
                    x: -247.09,
                    y: -51.118,
                    sx: 0.917,
                    sy: 0.917,
                    a: 0.67
                },
                "65": {
                    x: -247.065,
                    y: -51.103,
                    sx: 0.979,
                    sy: 0.979,
                    a: 0.78
                },
                "66": {
                    x: -247.044,
                    y: -51.089,
                    sx: 1.041,
                    sy: 1.041,
                    a: 0.89
                },
                "67": {
                    x: -246.95,
                    y: -51.1,
                    sx: 1.103,
                    sy: 1.103,
                    a: 1
                },
                "68": {
                    x: -246.909,
                    y: -51.137,
                    sx: 1.138,
                    sy: 1.138,
                    a: 0.86
                },
                "69": {
                    x: -246.872,
                    y: -51.124,
                    sx: 1.174,
                    sy: 1.174,
                    a: 0.71
                },
                "70": {
                    x: -246.832,
                    y: -51.111,
                    sx: 1.209,
                    sy: 1.209,
                    a: 0.57
                },
                "71": {
                    x: -246.741,
                    y: -51.097,
                    sx: 1.244,
                    sy: 1.244,
                    a: 0.43
                },
                "72": {
                    x: -246.751,
                    y: -51.134,
                    sx: 1.28,
                    sy: 1.28,
                    a: 0.29
                },
                "73": {
                    x: -246.714,
                    y: -51.121,
                    sx: 1.315,
                    sy: 1.315,
                    a: 0.14
                },
                "74": {
                    x: -246.8,
                    y: -51.15,
                    sx: 1.35,
                    sy: 1.35,
                    a: 0
                }
            })
            .addTimedChild(instance25, 48, 17, {
                "48": {
                    x: 56.8,
                    y: 125.4,
                    sx: 0.547,
                    sy: 0.547,
                    a: 0
                },
                "49": {
                    x: 56.827,
                    y: 125.438,
                    sx: 0.608,
                    sy: 0.608,
                    a: 0.11
                },
                "50": {
                    x: 56.901,
                    y: 125.421,
                    sx: 0.67,
                    sy: 0.67,
                    a: 0.22
                },
                "51": {
                    x: 56.925,
                    y: 125.405,
                    sx: 0.732,
                    sy: 0.732,
                    a: 0.33
                },
                "52": {
                    x: 57.001,
                    y: 125.442,
                    sx: 0.794,
                    sy: 0.794,
                    a: 0.45
                },
                "53": {
                    x: 56.977,
                    y: 125.428,
                    sx: 0.856,
                    sy: 0.856,
                    a: 0.55
                },
                "54": {
                    x: 57.003,
                    y: 125.466,
                    sx: 0.917,
                    sy: 0.917,
                    a: 0.67
                },
                "55": {
                    x: 57.077,
                    y: 125.449,
                    sx: 0.979,
                    sy: 0.979,
                    a: 0.78
                },
                "56": {
                    x: 57.102,
                    y: 125.483,
                    sx: 1.041,
                    sy: 1.041,
                    a: 0.89
                },
                "57": {
                    x: 57.2,
                    y: 125.45,
                    sx: 1.103,
                    sy: 1.103,
                    a: 1
                },
                "58": {
                    x: 57.232,
                    y: 125.468,
                    sx: 1.138,
                    sy: 1.138,
                    a: 0.86
                },
                "59": {
                    x: 57.265,
                    y: 125.438,
                    sx: 1.174,
                    sy: 1.174,
                    a: 0.71
                },
                "60": {
                    x: 57.297,
                    y: 125.456,
                    sx: 1.209,
                    sy: 1.209,
                    a: 0.57
                },
                "61": {
                    x: 57.329,
                    y: 125.474,
                    sx: 1.244,
                    sy: 1.244,
                    a: 0.43
                },
                "62": {
                    x: 57.361,
                    y: 125.492,
                    sx: 1.28,
                    sy: 1.28,
                    a: 0.29
                },
                "63": {
                    x: 57.394,
                    y: 125.461,
                    sx: 1.315,
                    sy: 1.315,
                    a: 0.14
                },
                "64": {
                    x: 57.4,
                    y: 125.4,
                    sx: 1.35,
                    sy: 1.35,
                    a: 0
                }
            })
            .addTimedChild(instance34, 80, 10, {
                "80": {
                    x: 276.05,
                    y: 80.3,
                    sx: 0.926,
                    sy: 0.926,
                    a: 0
                },
                "81": {
                    a: 0.17
                },
                "82": {
                    a: 0.33
                }
            })
            .addTimedChild(instance36, 82, 8)
            .addTimedChild(instance35, 82, 8)
            .addTimedChild(instance2)
            .addTimedChild(instance1);
    });

    lib.thinking_dots = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 90,
            framerate: 30,
            loop: false
        });
        var instance1 = new Graphic110(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 90, {
            "0": {
                x: 668.9,
                y: 365.95,
                a: 0
            },
            "1": {
                a: 0.02
            },
            "2": {
                a: 0.06
            },
            "3": {
                a: 0.16
            },
            "4": {
                a: 0.34
            },
            "5": {
                a: 0.58
            },
            "6": {
                a: 0.79
            },
            "7": {
                a: 0.92
            },
            "8": {
                a: 0.98
            },
            "9": {
                a: 1
            },
            "83": {
                a: 0.86
            },
            "84": {
                a: 0.71
            },
            "85": {
                a: 0.57
            },
            "86": {
                a: 0.43
            },
            "87": {
                a: 0.29
            },
            "88": {
                a: 0.14
            },
            "89": {
                a: 0
            }
        });
    });

    lib.thinking_dots.assets = {
        "thinking_dots_atlas_1": "images/thinking_dots_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.thinking_dots,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 90,
        library: lib
    };
}
