// jibo-anim-db-animations/timelines/thinking_dots2.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Container = PIXI.Container;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 32, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 50, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 22, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic4 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 19, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic5 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 8, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic6 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic7 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic8 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic9 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 46, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic10 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 5, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic11 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic12 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic13 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 11, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic14 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 12, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic15 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 12, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic16 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic17 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic18 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 26, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic19 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 7, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic20 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic21 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 11, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic22 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 3, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic23 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic24 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 24, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic25 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic26 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic27 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic28 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 63, loop: false });
        var instance8 = new Graphic25(MovieClip.SYNCHED);
        var instance6 = new Graphic22(MovieClip.SYNCHED);
        var instance5 = new Graphic19(MovieClip.SYNCHED);
        var instance16 = new Graphic26(MovieClip.SYNCHED);
        var instance12 = new Graphic23(MovieClip.SYNCHED);
        var instance4 = new Graphic17(MovieClip.SYNCHED);
        var instance3 = new Graphic14(MovieClip.SYNCHED);
        var instance7 = new Graphic12(MovieClip.SYNCHED);
        var instance2 = new Graphic10(MovieClip.SYNCHED);
        var instance23 = new Graphic27(MovieClip.SYNCHED);
        var instance22 = new Graphic24(MovieClip.SYNCHED);
        var instance17 = new Graphic20(MovieClip.SYNCHED);
        var instance26 = new Graphic21(MovieClip.SYNCHED);
        var instance21 = new Graphic18(MovieClip.SYNCHED);
        var instance19 = new Graphic15(MovieClip.SYNCHED);
        var instance14 = new Graphic11(MovieClip.SYNCHED);
        var instance24 = new Graphic16(MovieClip.SYNCHED);
        var instance25 = new Graphic13(MovieClip.SYNCHED);
        var instance13 = new Graphic9(MovieClip.SYNCHED);
        var instance9 = new Graphic7(MovieClip.SYNCHED);
        var instance1 = new Graphic6(MovieClip.SYNCHED);
        var instance20 = new Graphic8(MovieClip.SYNCHED);
        var instance15 = new Graphic4(MovieClip.SYNCHED);
        var instance10 = new Graphic3(MovieClip.SYNCHED);
        var instance27 = new Graphic5(MovieClip.SYNCHED);
        var instance11 = new Graphic2(MovieClip.SYNCHED);
        var instance18 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance8, 2, 10, {
                "2": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "3": {
                    a: 0.2
                },
                "4": {
                    a: 0.4
                },
                "5": {
                    a: 0.6
                },
                "6": {
                    a: 0.8
                },
                "7": {
                    a: 1
                },
                "8": {
                    a: 0.75
                },
                "9": {
                    a: 0.5
                },
                "10": {
                    a: 0.25
                },
                "11": {
                    a: 0
                }
            })
            .addTimedChild(instance6, 0, 3, {
                "0": {
                    x: -287.4,
                    y: 195.5,
                    a: 0.33
                },
                "1": {
                    a: 0.17
                },
                "2": {
                    a: 0
                }
            })
            .addTimedChild(instance5, 0, 7, {
                "0": {
                    x: -65.95,
                    y: 240.75,
                    a: 0.86
                },
                "1": {
                    a: 0.71
                },
                "2": {
                    a: 0.57
                },
                "3": {
                    a: 0.43
                },
                "4": {
                    a: 0.29
                },
                "5": {
                    a: 0.14
                },
                "6": {
                    a: 0
                }
            })
            .addTimedChild(instance16, 26, 10, {
                "26": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "27": {
                    a: 0.2
                },
                "28": {
                    a: 0.4
                },
                "29": {
                    a: 0.6
                },
                "30": {
                    a: 0.8
                },
                "31": {
                    a: 1
                },
                "32": {
                    a: 0.75
                },
                "33": {
                    a: 0.5
                },
                "34": {
                    a: 0.25
                },
                "35": {
                    a: 0
                }
            })
            .addTimedChild(instance12, 13, 17, {
                "13": {
                    x: -287.4,
                    y: 195.5,
                    a: 0
                },
                "14": {
                    a: 0.17
                },
                "15": {
                    a: 0.33
                },
                "16": {
                    a: 0.5
                },
                "17": {
                    a: 0.67
                },
                "18": {
                    a: 0.83
                },
                "19": {
                    a: 1
                },
                "21": {
                    a: 0.89
                },
                "22": {
                    a: 0.78
                },
                "23": {
                    a: 0.67
                },
                "24": {
                    a: 0.55
                },
                "25": {
                    a: 0.45
                },
                "26": {
                    a: 0.33
                },
                "27": {
                    a: 0.22
                },
                "28": {
                    a: 0.11
                },
                "29": {
                    a: 0
                }
            })
            .addTimedChild(instance4, 0, 17, {
                "0": {
                    x: -108.25,
                    y: 284.85,
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 0, 12, {
                "0": {
                    x: -24.3,
                    y: 198.25,
                    a: 0.8
                },
                "1": {
                    a: 1
                },
                "4": {
                    a: 0.88
                },
                "5": {
                    a: 0.75
                },
                "6": {
                    a: 0.63
                },
                "7": {
                    a: 0.5
                },
                "8": {
                    a: 0.38
                },
                "9": {
                    a: 0.25
                },
                "10": {
                    a: 0.13
                },
                "11": {
                    a: 0
                }
            })
            .addTimedChild(instance7, 1, 23, {
                "1": {
                    x: -196.65,
                    y: 240.75,
                    a: 0
                },
                "2": {
                    a: 0.14
                },
                "3": {
                    a: 0.29
                },
                "4": {
                    a: 0.43
                },
                "5": {
                    a: 0.57
                },
                "6": {
                    a: 0.71
                },
                "7": {
                    a: 0.86
                },
                "8": {
                    a: 1
                },
                "16": {
                    a: 0.88
                },
                "17": {
                    a: 0.75
                },
                "18": {
                    a: 0.63
                },
                "19": {
                    a: 0.5
                },
                "20": {
                    a: 0.38
                },
                "21": {
                    a: 0.25
                },
                "22": {
                    a: 0.13
                },
                "23": {
                    a: 0
                }
            })
            .addTimedChild(instance2, 0, 5, {
                "0": {
                    x: 155,
                    y: 21.5,
                    a: 0.5
                },
                "1": {
                    a: 0.38
                },
                "2": {
                    a: 0.25
                },
                "3": {
                    a: 0.13
                },
                "4": {
                    a: 0
                }
            })
            .addTimedChild(instance23, 45, 18, {
                "45": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "46": {
                    a: 0.13
                },
                "47": {
                    a: 0.25
                },
                "48": {
                    a: 0.38
                },
                "49": {
                    a: 0.5
                },
                "50": {
                    a: 0.63
                },
                "51": {
                    a: 0.75
                },
                "52": {
                    a: 0.88
                },
                "53": {
                    a: 1
                },
                "60": {
                    a: 0.89
                },
                "61": {
                    a: 0.78
                },
                "62": {
                    a: 0.67
                }
            })
            .addTimedChild(instance22, 38, 24, {
                "38": {
                    x: -287.4,
                    y: 195.5,
                    a: 0
                },
                "39": {
                    a: 0.17
                },
                "40": {
                    a: 0.33
                },
                "41": {
                    a: 0.5
                },
                "42": {
                    a: 0.67
                },
                "43": {
                    a: 0.83
                },
                "44": {
                    a: 1
                },
                "54": {
                    a: 0.88
                },
                "55": {
                    a: 0.75
                },
                "56": {
                    a: 0.63
                },
                "57": {
                    a: 0.5
                },
                "58": {
                    a: 0.38
                },
                "59": {
                    a: 0.25
                },
                "60": {
                    a: 0.13
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance17, 27, 16, {
                "27": {
                    x: -65.95,
                    y: 240.75,
                    a: 0
                },
                "28": {
                    a: 0.14
                },
                "29": {
                    a: 0.29
                },
                "30": {
                    a: 0.43
                },
                "31": {
                    a: 0.57
                },
                "32": {
                    a: 0.71
                },
                "33": {
                    a: 0.86
                },
                "34": {
                    a: 1
                },
                "36": {
                    a: 0.86
                },
                "37": {
                    a: 0.71
                },
                "38": {
                    a: 0.57
                },
                "39": {
                    a: 0.43
                },
                "40": {
                    a: 0.29
                },
                "41": {
                    a: 0.14
                },
                "42": {
                    a: 0
                }
            })
            .addTimedChild(instance26, 52, 11, {
                "52": {
                    x: -65.95,
                    y: 240.75,
                    a: 0
                },
                "53": {
                    a: 0.14
                },
                "54": {
                    a: 0.29
                },
                "55": {
                    a: 0.43
                },
                "56": {
                    a: 0.57
                },
                "57": {
                    a: 0.71
                },
                "58": {
                    a: 0.86
                },
                "59": {
                    a: 1
                }
            })
            .addTimedChild(instance21, 37, 26, {
                "37": {
                    x: -108.25,
                    y: 284.85,
                    a: 0
                },
                "38": {
                    a: 0.14
                },
                "39": {
                    a: 0.29
                },
                "40": {
                    a: 0.43
                },
                "41": {
                    a: 0.57
                },
                "42": {
                    a: 0.71
                },
                "43": {
                    a: 0.86
                },
                "44": {
                    a: 1
                },
                "60": {
                    a: 0.86
                },
                "61": {
                    a: 0.71
                },
                "62": {
                    a: 0.57
                }
            })
            .addTimedChild(instance19, 29, 12, {
                "29": {
                    x: -24.3,
                    y: 198.25,
                    a: 0
                },
                "30": {
                    a: 0.2
                },
                "31": {
                    a: 0.4
                },
                "32": {
                    a: 0.6
                },
                "33": {
                    a: 0.8
                },
                "34": {
                    a: 1
                },
                "37": {
                    a: 0.75
                },
                "38": {
                    a: 0.5
                },
                "39": {
                    a: 0.25
                },
                "40": {
                    a: 0
                }
            })
            .addTimedChild(instance14, 18, 16, {
                "18": {
                    x: 155,
                    y: 21.5,
                    a: 0
                },
                "19": {
                    a: 0.14
                },
                "20": {
                    a: 0.29
                },
                "21": {
                    a: 0.43
                },
                "22": {
                    a: 0.57
                },
                "23": {
                    a: 0.71
                },
                "24": {
                    a: 0.86
                },
                "25": {
                    a: 1
                },
                "28": {
                    a: 0.83
                },
                "29": {
                    a: 0.67
                },
                "30": {
                    a: 0.5
                },
                "31": {
                    a: 0.33
                },
                "32": {
                    a: 0.17
                },
                "33": {
                    a: 0
                }
            })
            .addTimedChild(instance24, 46, 16, {
                "46": {
                    x: -24.3,
                    y: 198.25,
                    a: 0
                },
                "47": {
                    a: 0.2
                },
                "48": {
                    a: 0.4
                },
                "49": {
                    a: 0.6
                },
                "50": {
                    a: 0.8
                },
                "51": {
                    a: 1
                },
                "54": {
                    a: 0.88
                },
                "55": {
                    a: 0.75
                },
                "56": {
                    a: 0.63
                },
                "57": {
                    a: 0.5
                },
                "58": {
                    a: 0.38
                },
                "59": {
                    a: 0.25
                },
                "60": {
                    a: 0.13
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance25, 52, 11, {
                "52": {
                    x: -196.65,
                    y: 240.75,
                    a: 0
                },
                "53": {
                    a: 0.17
                },
                "54": {
                    a: 0.33
                },
                "55": {
                    a: 0.5
                },
                "56": {
                    a: 0.67
                },
                "57": {
                    a: 0.83
                },
                "58": {
                    a: 1
                },
                "60": {
                    a: 0.83
                },
                "61": {
                    a: 0.67
                },
                "62": {
                    a: 0.5
                }
            })
            .addTimedChild(instance13, 14, 46, {
                "14": {
                    x: 110.85,
                    y: -22.6,
                    a: 1
                },
                "27": {
                    a: 0.9
                },
                "28": {
                    a: 0.8
                },
                "29": {
                    a: 0.7
                },
                "30": {
                    a: 0.6
                },
                "31": {
                    a: 0.5
                },
                "32": {
                    a: 0.4
                },
                "33": {
                    a: 0.3
                },
                "34": {
                    a: 0.2
                },
                "35": {
                    a: 0.1
                },
                "36": {
                    a: 0
                },
                "37": {
                    a: 1
                },
                "53": {
                    a: 0.86
                },
                "54": {
                    a: 0.71
                },
                "55": {
                    a: 0.57
                },
                "56": {
                    a: 0.43
                },
                "57": {
                    a: 0.29
                },
                "58": {
                    a: 0.14
                },
                "59": {
                    a: 0
                }
            })
            .addTimedChild(instance9, 4, 17, {
                "4": {
                    x: -329.1,
                    y: -153.85,
                    a: 0
                },
                "5": {
                    a: 0.13
                },
                "6": {
                    a: 0.25
                },
                "7": {
                    a: 0.38
                },
                "8": {
                    a: 0.5
                },
                "9": {
                    a: 0.63
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.88
                },
                "12": {
                    a: 1
                },
                "14": {
                    a: 0.86
                },
                "15": {
                    a: 0.71
                },
                "16": {
                    a: 0.57
                },
                "17": {
                    a: 0.43
                },
                "18": {
                    a: 0.29
                },
                "19": {
                    a: 0.14
                },
                "20": {
                    a: 0
                }
            })
            .addTimedChild(instance1, 0, 13, {
                "0": {
                    x: -461.35,
                    y: 154.4,
                    a: 1
                },
                "6": {
                    a: 0.86
                },
                "7": {
                    a: 0.71
                },
                "8": {
                    a: 0.57
                },
                "9": {
                    a: 0.43
                },
                "10": {
                    a: 0.29
                },
                "11": {
                    a: 0.14
                },
                "12": {
                    a: 0
                }
            })
            .addTimedChild(instance20, 35, 18, {
                "35": {
                    x: -329.1,
                    y: -153.85,
                    a: 0
                },
                "36": {
                    a: 0.13
                },
                "37": {
                    a: 0.25
                },
                "38": {
                    a: 0.38
                },
                "39": {
                    a: 0.5
                },
                "40": {
                    a: 0.63
                },
                "41": {
                    a: 0.75
                },
                "42": {
                    a: 0.88
                },
                "43": {
                    a: 1
                },
                "46": {
                    a: 0.86
                },
                "47": {
                    a: 0.71
                },
                "48": {
                    a: 0.57
                },
                "49": {
                    a: 0.43
                },
                "50": {
                    a: 0.29
                },
                "51": {
                    a: 0.14
                },
                "52": {
                    a: 0
                }
            })
            .addTimedChild(instance15, 26, 19, {
                "26": {
                    x: -417.35,
                    y: -22.6,
                    a: 0
                },
                "27": {
                    a: 0.17
                },
                "28": {
                    a: 0.33
                },
                "29": {
                    a: 0.5
                },
                "30": {
                    a: 0.67
                },
                "31": {
                    a: 0.83
                },
                "32": {
                    a: 1
                },
                "35": {
                    a: 0.89
                },
                "36": {
                    a: 0.78
                },
                "37": {
                    a: 0.67
                },
                "38": {
                    a: 0.55
                },
                "39": {
                    a: 0.45
                },
                "40": {
                    a: 0.33
                },
                "41": {
                    a: 0.22
                },
                "42": {
                    a: 0.11
                },
                "43": {
                    a: 0
                }
            })
            .addTimedChild(instance10, 5, 22, {
                "5": {
                    x: -240.5,
                    y: -240.35,
                    a: 0
                },
                "6": {
                    a: 0.2
                },
                "7": {
                    a: 0.4
                },
                "8": {
                    a: 0.6
                },
                "9": {
                    a: 0.8
                },
                "10": {
                    a: 1
                },
                "20": {
                    a: 0.86
                },
                "21": {
                    a: 0.71
                },
                "22": {
                    a: 0.57
                },
                "23": {
                    a: 0.43
                },
                "24": {
                    a: 0.29
                },
                "25": {
                    a: 0.14
                },
                "26": {
                    a: 0
                }
            })
            .addTimedChild(instance27, 55, 8, {
                "55": {
                    x: -417.35,
                    y: -22.6,
                    a: 0
                },
                "56": {
                    a: 0.17
                },
                "57": {
                    a: 0.33
                },
                "58": {
                    a: 0.5
                },
                "59": {
                    a: 0.67
                },
                "60": {
                    a: 0.83
                },
                "61": {
                    a: 1
                }
            })
            .addTimedChild(instance11, 12, 50, {
                "12": {
                    x: -417.35,
                    y: -111.2,
                    a: 0
                },
                "13": {
                    a: 0.11
                },
                "14": {
                    a: 0.22
                },
                "15": {
                    a: 0.33
                },
                "16": {
                    a: 0.45
                },
                "17": {
                    a: 0.55
                },
                "18": {
                    a: 0.67
                },
                "19": {
                    a: 0.78
                },
                "20": {
                    a: 0.89
                },
                "21": {
                    a: 1
                },
                "54": {
                    a: 0.88
                },
                "55": {
                    a: 0.75
                },
                "56": {
                    a: 0.63
                },
                "57": {
                    a: 0.5
                },
                "58": {
                    a: 0.38
                },
                "59": {
                    a: 0.25
                },
                "60": {
                    a: 0.13
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance18, 29, 32, {
                "29": {
                    x: 65.9,
                    y: 20.85,
                    a: 0
                },
                "30": {
                    a: 0.17
                },
                "31": {
                    a: 0.33
                },
                "32": {
                    a: 0.5
                },
                "33": {
                    a: 0.67
                },
                "34": {
                    a: 0.83
                },
                "35": {
                    a: 1
                },
                "54": {
                    a: 0.86
                },
                "55": {
                    a: 0.71
                },
                "56": {
                    a: 0.57
                },
                "57": {
                    a: 0.43
                },
                "58": {
                    a: 0.29
                },
                "59": {
                    a: 0.14
                },
                "60": {
                    a: 0
                }
            });
    });

    var Graphic29 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic30 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic31 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic32 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 22, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic33 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic34 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic35 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 30, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic36 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 8, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic37 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic38 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 15, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic39 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic40 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 18, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic41 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic42 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic43 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 23, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic44 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic45 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic46 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 22, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic47 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 24, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic48 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 35, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    lib.Graphic49 = Container.extend(function () {
        Container.call(this);
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addChild(instance1);
    });

    var Graphic50 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic51 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic52 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 15, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic53 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 24, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic54 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 10, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic55 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 2, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic56 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 63, loop: false });
        var instance7 = new Graphic53(MovieClip.SYNCHED);
        var instance6 = new Graphic50(MovieClip.SYNCHED);
        var instance18 = new Graphic54(MovieClip.SYNCHED);
        var instance14 = new Graphic51(MovieClip.SYNCHED);
        var instance10 = new Graphic48(MovieClip.SYNCHED);
        var instance5 = new Graphic46(MovieClip.SYNCHED);
        var instance8 = new Graphic44(MovieClip.SYNCHED);
        var instance9 = new Graphic42(MovieClip.SYNCHED);
        var instance4 = new Graphic38(MovieClip.SYNCHED);
        var instance3 = new Graphic36(MovieClip.SYNCHED);
        var instance26 = new Graphic55(MovieClip.SYNCHED);
        var instance22 = new Graphic52(MovieClip.SYNCHED);
        var instance15 = new Graphic47(MovieClip.SYNCHED);
        var instance16 = new Graphic45(MovieClip.SYNCHED);
        var instance17 = new Graphic43(MovieClip.SYNCHED);
        var instance13 = new Graphic40(MovieClip.SYNCHED);
        var instance27 = new lib.Graphic49()
            .setTransform(-65.95, 240.75)
            .setAlpha(0);
        var instance24 = new Graphic41(MovieClip.SYNCHED);
        var instance23 = new Graphic39(MovieClip.SYNCHED);
        var instance19 = new Graphic37(MovieClip.SYNCHED);
        var instance12 = new Graphic35(MovieClip.SYNCHED);
        var instance11 = new Graphic33(MovieClip.SYNCHED);
        var instance2 = new Graphic30(MovieClip.SYNCHED);
        var instance1 = new Graphic29(MovieClip.SYNCHED);
        var instance25 = new Graphic34(MovieClip.SYNCHED);
        var instance20 = new Graphic32(MovieClip.SYNCHED);
        var instance21 = new Graphic31(MovieClip.SYNCHED);
        this.addTimedChild(instance7, 0, 24, {
                "0": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "1": {
                    a: 0.13
                },
                "2": {
                    a: 0.25
                },
                "3": {
                    a: 0.38
                },
                "4": {
                    a: 0.5
                },
                "5": {
                    a: 0.63
                },
                "6": {
                    a: 0.75
                },
                "7": {
                    a: 0.88
                },
                "8": {
                    a: 1
                },
                "15": {
                    a: 0.89
                },
                "16": {
                    a: 0.78
                },
                "17": {
                    a: 0.67
                },
                "18": {
                    a: 0.55
                },
                "19": {
                    a: 0.45
                },
                "20": {
                    a: 0.33
                },
                "21": {
                    a: 0.22
                },
                "22": {
                    a: 0.11
                },
                "23": {
                    a: 0
                }
            })
            .addTimedChild(instance6, 0, 17, {
                "0": {
                    x: -287.4,
                    y: 195.5,
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance18, 37, 10, {
                "37": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "38": {
                    a: 0.2
                },
                "39": {
                    a: 0.4
                },
                "40": {
                    a: 0.6
                },
                "41": {
                    a: 0.8
                },
                "42": {
                    a: 1
                },
                "43": {
                    a: 0.75
                },
                "44": {
                    a: 0.5
                },
                "45": {
                    a: 0.25
                },
                "46": {
                    a: 0
                }
            })
            .addTimedChild(instance14, 22, 16, {
                "22": {
                    x: -287.4,
                    y: 195.5,
                    a: 0
                },
                "23": {
                    a: 0.17
                },
                "24": {
                    a: 0.33
                },
                "25": {
                    a: 0.5
                },
                "26": {
                    a: 0.67
                },
                "27": {
                    a: 0.83
                },
                "28": {
                    a: 1
                },
                "32": {
                    a: 0.83
                },
                "33": {
                    a: 0.67
                },
                "34": {
                    a: 0.5
                },
                "35": {
                    a: 0.33
                },
                "36": {
                    a: 0.17
                },
                "37": {
                    a: 0
                }
            })
            .addTimedChild(instance10, 7, 35, {
                "7": {
                    x: -65.95,
                    y: 240.75,
                    a: 0
                },
                "8": {
                    a: 0.14
                },
                "9": {
                    a: 0.29
                },
                "10": {
                    a: 0.43
                },
                "11": {
                    a: 0.57
                },
                "12": {
                    a: 0.71
                },
                "13": {
                    a: 0.86
                },
                "14": {
                    a: 1
                },
                "35": {
                    a: 0.86
                },
                "36": {
                    a: 0.71
                },
                "37": {
                    a: 0.57
                },
                "38": {
                    a: 0.43
                },
                "39": {
                    a: 0.29
                },
                "40": {
                    a: 0.14
                },
                "41": {
                    a: 0
                }
            })
            .addTimedChild(instance5, 0, 22, {
                "0": {
                    x: -108.25,
                    y: 284.85,
                    a: 1
                },
                "15": {
                    a: 0.86
                },
                "16": {
                    a: 0.71
                },
                "17": {
                    a: 0.57
                },
                "18": {
                    a: 0.43
                },
                "19": {
                    a: 0.29
                },
                "20": {
                    a: 0.14
                },
                "21": {
                    a: 0
                }
            })
            .addTimedChild(instance8, 1, 16, {
                "1": {
                    x: -24.3,
                    y: 198.25,
                    a: 0
                },
                "2": {
                    a: 0.2
                },
                "3": {
                    a: 0.4
                },
                "4": {
                    a: 0.6
                },
                "5": {
                    a: 0.8
                },
                "6": {
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance9, 7, 14, {
                "7": {
                    x: -196.65,
                    y: 240.75,
                    a: 0
                },
                "8": {
                    a: 0.17
                },
                "9": {
                    a: 0.33
                },
                "10": {
                    a: 0.5
                },
                "11": {
                    a: 0.67
                },
                "12": {
                    a: 0.83
                },
                "13": {
                    a: 1
                },
                "15": {
                    a: 0.83
                },
                "16": {
                    a: 0.67
                },
                "17": {
                    a: 0.5
                },
                "18": {
                    a: 0.33
                },
                "19": {
                    a: 0.17
                },
                "20": {
                    a: 0
                }
            })
            .addTimedChild(instance4, 0, 15, {
                "0": {
                    x: 110.85,
                    y: -22.6,
                    a: 1
                },
                "8": {
                    a: 0.86
                },
                "9": {
                    a: 0.71
                },
                "10": {
                    a: 0.57
                },
                "11": {
                    a: 0.43
                },
                "12": {
                    a: 0.29
                },
                "13": {
                    a: 0.14
                },
                "14": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 0, 8, {
                "0": {
                    x: -329.1,
                    y: -153.85,
                    a: 1
                },
                "1": {
                    a: 0.86
                },
                "2": {
                    a: 0.71
                },
                "3": {
                    a: 0.57
                },
                "4": {
                    a: 0.43
                },
                "5": {
                    a: 0.29
                },
                "6": {
                    a: 0.14
                },
                "7": {
                    a: 0
                }
            })
            .addTimedChild(instance26, 61, 2, {
                "61": {
                    x: 109.9,
                    y: -156.35,
                    a: 0
                },
                "62": {
                    a: 0.2
                }
            })
            .addTimedChild(instance22, 48, 15, {
                "48": {
                    x: -287.4,
                    y: 195.5,
                    a: 0
                },
                "49": {
                    a: 0.17
                },
                "50": {
                    a: 0.33
                },
                "51": {
                    a: 0.5
                },
                "52": {
                    a: 0.67
                },
                "53": {
                    a: 0.83
                },
                "54": {
                    a: 1
                },
                "56": {
                    a: 0.89
                },
                "57": {
                    a: 0.78
                },
                "58": {
                    a: 0.67
                },
                "59": {
                    a: 0.55
                },
                "60": {
                    a: 0.45
                },
                "61": {
                    a: 0.33
                },
                "62": {
                    a: 0.22
                }
            })
            .addTimedChild(instance15, 28, 24, {
                "28": {
                    x: -108.25,
                    y: 284.85,
                    a: 0
                },
                "29": {
                    a: 0.14
                },
                "30": {
                    a: 0.29
                },
                "31": {
                    a: 0.43
                },
                "32": {
                    a: 0.57
                },
                "33": {
                    a: 0.71
                },
                "34": {
                    a: 0.86
                },
                "35": {
                    a: 1
                },
                "44": {
                    a: 0.88
                },
                "45": {
                    a: 0.75
                },
                "46": {
                    a: 0.63
                },
                "47": {
                    a: 0.5
                },
                "48": {
                    a: 0.38
                },
                "49": {
                    a: 0.25
                },
                "50": {
                    a: 0.13
                },
                "51": {
                    a: 0
                }
            })
            .addTimedChild(instance16, 31, 16, {
                "31": {
                    x: -24.3,
                    y: 198.25,
                    a: 0
                },
                "32": {
                    a: 0.2
                },
                "33": {
                    a: 0.4
                },
                "34": {
                    a: 0.6
                },
                "35": {
                    a: 0.8
                },
                "36": {
                    a: 1
                },
                "39": {
                    a: 0.88
                },
                "40": {
                    a: 0.75
                },
                "41": {
                    a: 0.63
                },
                "42": {
                    a: 0.5
                },
                "43": {
                    a: 0.38
                },
                "44": {
                    a: 0.25
                },
                "45": {
                    a: 0.13
                },
                "46": {
                    a: 0
                }
            })
            .addTimedChild(instance17, 36, 23, {
                "36": {
                    x: -196.65,
                    y: 240.75,
                    a: 0
                },
                "37": {
                    a: 0.14
                },
                "38": {
                    a: 0.29
                },
                "39": {
                    a: 0.43
                },
                "40": {
                    a: 0.57
                },
                "41": {
                    a: 0.71
                },
                "42": {
                    a: 0.86
                },
                "43": {
                    a: 1
                },
                "51": {
                    a: 0.88
                },
                "52": {
                    a: 0.75
                },
                "53": {
                    a: 0.63
                },
                "54": {
                    a: 0.5
                },
                "55": {
                    a: 0.38
                },
                "56": {
                    a: 0.25
                },
                "57": {
                    a: 0.13
                },
                "58": {
                    a: 0
                }
            })
            .addTimedChild(instance13, 22, 18, {
                "22": {
                    x: 155,
                    y: 21.5,
                    a: 0
                },
                "23": {
                    a: 0.14
                },
                "24": {
                    a: 0.29
                },
                "25": {
                    a: 0.43
                },
                "26": {
                    a: 0.57
                },
                "27": {
                    a: 0.71
                },
                "28": {
                    a: 0.86
                },
                "29": {
                    a: 1
                },
                "32": {
                    a: 0.88
                },
                "33": {
                    a: 0.75
                },
                "34": {
                    a: 0.63
                },
                "35": {
                    a: 0.5
                },
                "36": {
                    a: 0.38
                },
                "37": {
                    a: 0.25
                },
                "38": {
                    a: 0.13
                },
                "39": {
                    a: 0
                }
            })
            .addTimedChild(instance27, 62, 1)
            .addTimedChild(instance24, 53, 10, {
                "53": {
                    x: 155,
                    y: 21.5,
                    a: 0
                },
                "54": {
                    a: 0.14
                },
                "55": {
                    a: 0.29
                },
                "56": {
                    a: 0.43
                },
                "57": {
                    a: 0.57
                },
                "58": {
                    a: 0.71
                },
                "59": {
                    a: 0.86
                },
                "60": {
                    a: 1
                }
            })
            .addTimedChild(instance23, 49, 14, {
                "49": {
                    x: 110.85,
                    y: -22.6,
                    a: 1
                },
                "62": {
                    a: 0.9
                }
            })
            .addTimedChild(instance19, 39, 17, {
                "39": {
                    x: -329.1,
                    y: -153.85,
                    a: 0
                },
                "40": {
                    a: 0.13
                },
                "41": {
                    a: 0.25
                },
                "42": {
                    a: 0.38
                },
                "43": {
                    a: 0.5
                },
                "44": {
                    a: 0.63
                },
                "45": {
                    a: 0.75
                },
                "46": {
                    a: 0.88
                },
                "47": {
                    a: 1
                },
                "49": {
                    a: 0.86
                },
                "50": {
                    a: 0.71
                },
                "51": {
                    a: 0.57
                },
                "52": {
                    a: 0.43
                },
                "53": {
                    a: 0.29
                },
                "54": {
                    a: 0.14
                },
                "55": {
                    a: 0
                }
            })
            .addTimedChild(instance12, 18, 30, {
                "18": {
                    x: -461.35,
                    y: 154.4,
                    a: 0
                },
                "19": {
                    a: 0.14
                },
                "20": {
                    a: 0.29
                },
                "21": {
                    a: 0.43
                },
                "22": {
                    a: 0.57
                },
                "23": {
                    a: 0.71
                },
                "24": {
                    a: 0.86
                },
                "25": {
                    a: 1
                },
                "41": {
                    a: 0.86
                },
                "42": {
                    a: 0.71
                },
                "43": {
                    a: 0.57
                },
                "44": {
                    a: 0.43
                },
                "45": {
                    a: 0.29
                },
                "46": {
                    a: 0.14
                },
                "47": {
                    a: 0
                }
            })
            .addTimedChild(instance11, 10, 18, {
                "10": {
                    x: -417.35,
                    y: -22.6,
                    a: 0
                },
                "11": {
                    a: 0.17
                },
                "12": {
                    a: 0.33
                },
                "13": {
                    a: 0.5
                },
                "14": {
                    a: 0.67
                },
                "15": {
                    a: 0.83
                },
                "16": {
                    a: 1
                },
                "19": {
                    a: 0.89
                },
                "20": {
                    a: 0.78
                },
                "21": {
                    a: 0.67
                },
                "22": {
                    a: 0.55
                },
                "23": {
                    a: 0.45
                },
                "24": {
                    a: 0.33
                },
                "25": {
                    a: 0.22
                },
                "26": {
                    a: 0.11
                },
                "27": {
                    a: 0
                }
            })
            .addTimedChild(instance2, 0, 17, {
                "0": {
                    x: -417.35,
                    y: -111.2,
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance1, 0, 16, {
                "0": {
                    x: 65.9,
                    y: 20.85,
                    a: 1
                },
                "9": {
                    a: 0.86
                },
                "10": {
                    a: 0.71
                },
                "11": {
                    a: 0.57
                },
                "12": {
                    a: 0.43
                },
                "13": {
                    a: 0.29
                },
                "14": {
                    a: 0.14
                },
                "15": {
                    a: 0
                }
            })
            .addTimedChild(instance25, 61, 2, {
                "61": {
                    x: -417.35,
                    y: -22.6,
                    a: 0
                },
                "62": {
                    a: 0.17
                }
            })
            .addTimedChild(instance20, 40, 22, {
                "40": {
                    x: -240.5,
                    y: -240.35,
                    a: 0
                },
                "41": {
                    a: 0.2
                },
                "42": {
                    a: 0.4
                },
                "43": {
                    a: 0.6
                },
                "44": {
                    a: 0.8
                },
                "45": {
                    a: 1
                },
                "55": {
                    a: 0.86
                },
                "56": {
                    a: 0.71
                },
                "57": {
                    a: 0.57
                },
                "58": {
                    a: 0.43
                },
                "59": {
                    a: 0.29
                },
                "60": {
                    a: 0.14
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance21, 47, 16, {
                "47": {
                    x: -417.35,
                    y: -111.2,
                    a: 0
                },
                "48": {
                    a: 0.11
                },
                "49": {
                    a: 0.22
                },
                "50": {
                    a: 0.33
                },
                "51": {
                    a: 0.45
                },
                "52": {
                    a: 0.55
                },
                "53": {
                    a: 0.67
                },
                "54": {
                    a: 0.78
                },
                "55": {
                    a: 0.89
                },
                "56": {
                    a: 1
                }
            });
    });

    var Graphic57 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 45, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic58 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 15, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic59 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 60, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic60 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot25"))
            .setTransform(-18.6, -17.9);
        this.addTimedChild(instance1);
    });

    var Graphic61 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 5, loop: false });
        var instance1 = new Sprite(fromFrame("dot25"))
            .setTransform(-18.6, -17.9);
        this.addTimedChild(instance1);
    });

    var Graphic62 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic63 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 15, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic64 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 5, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic65 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic66 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic67 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot25"))
            .setTransform(-18.6, -17.9);
        this.addTimedChild(instance1);
    });

    var Graphic68 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot25"))
            .setTransform(-18.6, -17.9);
        this.addTimedChild(instance1);
    });

    var Graphic69 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic70 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 8, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic71 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic72 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic73 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic74 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot23"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic75 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 16, loop: false });
        var instance1 = new Sprite(fromFrame("dot22"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic76 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 14, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic77 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 13, loop: false });
        var instance1 = new Sprite(fromFrame("dot24"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic78 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 17, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic79 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 5, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-15.85, -15.85);
        this.addTimedChild(instance1);
    });

    var Graphic80 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 63, loop: false });
        var instance1 = new Sprite(fromFrame("dots21"))
            .setTransform(-280.2, -282.65, 0.89, 0.89);
        this.addTimedChild(instance1);
    });

    var Graphic81 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 63, loop: false });
        var instance4 = new Graphic80(MovieClip.SYNCHED);
        var instance18 = new Graphic78(MovieClip.SYNCHED);
        var instance14 = new Graphic76(MovieClip.SYNCHED);
        var instance12 = new Graphic75(MovieClip.SYNCHED);
        var instance6 = new Graphic71(MovieClip.SYNCHED);
        var instance13 = new Graphic73(MovieClip.SYNCHED);
        var instance11 = new Graphic69(MovieClip.SYNCHED);
        var instance8 = new Graphic65(MovieClip.SYNCHED);
        var instance3 = new Graphic62(MovieClip.SYNCHED);
        var instance9 = new Graphic60(MovieClip.SYNCHED);
        var instance15 = new Graphic72(MovieClip.SYNCHED);
        var instance19 = new Graphic77(MovieClip.SYNCHED);
        var instance16 = new Graphic67(MovieClip.SYNCHED);
        var instance17 = new Graphic63(MovieClip.SYNCHED);
        var instance26 = new Graphic79(MovieClip.SYNCHED);
        var instance22 = new Graphic74(MovieClip.SYNCHED);
        var instance23 = new Graphic70(MovieClip.SYNCHED);
        var instance21 = new Graphic68(MovieClip.SYNCHED);
        var instance20 = new Graphic66(MovieClip.SYNCHED);
        var instance25 = new Graphic64(MovieClip.SYNCHED);
        var instance24 = new Graphic61(MovieClip.SYNCHED);
        var instance7 = new Graphic59(MovieClip.SYNCHED);
        var instance10 = new Graphic58(MovieClip.SYNCHED);
        var instance5 = new Graphic57(MovieClip.SYNCHED);
        var instance2 = new Graphic56(MovieClip.SYNCHED)
            .setTransform(110.15, 1.5);
        var instance1 = new Graphic28(MovieClip.SYNCHED)
            .setTransform(-242.05, 1.3, 1, 1, 0, 0, 3.142);
        this.addTimedChild(instance4, 0, 63, {
                "0": {
                    x: -43.6,
                    y: 24.8
                }
            })
            .addTimedChild(instance18, 31, 17, {
                "31": {
                    x: 176.15,
                    y: -242.85,
                    a: 0
                },
                "32": {
                    a: 0.17
                },
                "33": {
                    a: 0.33
                },
                "34": {
                    a: 0.5
                },
                "35": {
                    a: 0.67
                },
                "36": {
                    a: 0.83
                },
                "37": {
                    a: 1
                },
                "42": {
                    a: 0.83
                },
                "43": {
                    a: 0.67
                },
                "44": {
                    a: 0.5
                },
                "45": {
                    a: 0.33
                },
                "46": {
                    a: 0.17
                },
                "47": {
                    a: 0
                }
            })
            .addTimedChild(instance14, 22, 14, {
                "22": {
                    x: 176.15,
                    y: 64.9,
                    a: 0
                },
                "23": {
                    a: 0.25
                },
                "24": {
                    a: 0.5
                },
                "25": {
                    a: 0.75
                },
                "26": {
                    a: 1
                },
                "31": {
                    a: 0.8
                },
                "32": {
                    a: 0.6
                },
                "33": {
                    a: 0.4
                },
                "34": {
                    a: 0.2
                },
                "35": {
                    a: 0
                }
            })
            .addTimedChild(instance12, 10, 16, {
                "10": {
                    x: -349.15,
                    y: 64.9,
                    a: 0
                },
                "11": {
                    a: 0.33
                },
                "12": {
                    a: 0.67
                },
                "13": {
                    a: 1
                },
                "22": {
                    a: 0.75
                },
                "23": {
                    a: 0.5
                },
                "24": {
                    a: 0.25
                },
                "25": {
                    a: 0
                }
            })
            .addTimedChild(instance6, 2, 13, {
                "2": {
                    x: -132.5,
                    y: -243.75,
                    a: 0
                },
                "3": {
                    a: 0.2
                },
                "4": {
                    a: 0.4
                },
                "5": {
                    a: 0.6
                },
                "6": {
                    a: 0.8
                },
                "7": {
                    a: 1
                },
                "9": {
                    a: 0.83
                },
                "10": {
                    a: 0.67
                },
                "11": {
                    a: 0.5
                },
                "12": {
                    a: 0.33
                },
                "13": {
                    a: 0.17
                },
                "14": {
                    a: 0
                }
            })
            .addTimedChild(instance13, 16, 13, {
                "16": {
                    x: -177.55,
                    y: -149.05,
                    a: 0
                },
                "17": {
                    a: 0.2
                },
                "18": {
                    a: 0.4
                },
                "19": {
                    a: 0.6
                },
                "20": {
                    a: 0.8
                },
                "21": {
                    a: 1
                },
                "25": {
                    a: 0.75
                },
                "26": {
                    a: 0.5
                },
                "27": {
                    a: 0.25
                },
                "28": {
                    a: 0
                }
            })
            .addTimedChild(instance11, 10, 13, {
                "10": {
                    x: -263.25,
                    y: -154.8,
                    a: 0
                },
                "11": {
                    a: 0.17
                },
                "12": {
                    a: 0.33
                },
                "13": {
                    a: 0.5
                },
                "14": {
                    a: 0.67
                },
                "15": {
                    a: 0.83
                },
                "16": {
                    a: 1
                },
                "17": {
                    a: 0.83
                },
                "18": {
                    a: 0.67
                },
                "19": {
                    a: 0.5
                },
                "20": {
                    a: 0.33
                },
                "21": {
                    a: 0.17
                },
                "22": {
                    a: 0
                }
            })
            .addTimedChild(instance8, 4, 13, {
                "4": {
                    x: -218.8,
                    y: 152.3,
                    a: 0
                },
                "5": {
                    a: 0.17
                },
                "6": {
                    a: 0.33
                },
                "7": {
                    a: 0.5
                },
                "8": {
                    a: 0.67
                },
                "9": {
                    a: 0.83
                },
                "10": {
                    a: 1
                },
                "13": {
                    a: 0.75
                },
                "14": {
                    a: 0.5
                },
                "15": {
                    a: 0.25
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance3, 0, 17, {
                "0": {
                    x: 176.1,
                    y: 110.55,
                    a: 0
                },
                "1": {
                    a: 0.17
                },
                "2": {
                    a: 0.33
                },
                "3": {
                    a: 0.5
                },
                "4": {
                    a: 0.67
                },
                "5": {
                    a: 0.83
                },
                "6": {
                    a: 1
                },
                "9": {
                    a: 0.88
                },
                "10": {
                    a: 0.75
                },
                "11": {
                    a: 0.63
                },
                "12": {
                    a: 0.5
                },
                "13": {
                    a: 0.38
                },
                "14": {
                    a: 0.25
                },
                "15": {
                    a: 0.13
                },
                "16": {
                    a: 0
                }
            })
            .addTimedChild(instance9, 5, 17, {
                "5": {
                    x: -264.5,
                    y: -65.6,
                    sx: 0.59,
                    sy: 0.59,
                    a: 0
                },
                "6": {
                    x: -264.418,
                    y: -65.57,
                    sx: 0.657,
                    sy: 0.657,
                    a: 0.11
                },
                "7": {
                    x: -264.395,
                    y: -65.592,
                    sx: 0.723,
                    sy: 0.723,
                    a: 0.22
                },
                "8": {
                    x: -264.355,
                    y: -65.561,
                    sx: 0.79,
                    sy: 0.79,
                    a: 0.33
                },
                "9": {
                    x: -264.282,
                    y: -65.583,
                    sx: 0.857,
                    sy: 0.857,
                    a: 0.45
                },
                "10": {
                    x: -264.251,
                    y: -65.553,
                    sx: 0.924,
                    sy: 0.924,
                    a: 0.55
                },
                "11": {
                    x: -264.177,
                    y: -65.575,
                    sx: 0.99,
                    sy: 0.99,
                    a: 0.67
                },
                "12": {
                    x: -264.138,
                    y: -65.544,
                    sx: 1.057,
                    sy: 1.057,
                    a: 0.78
                },
                "13": {
                    x: -264.064,
                    y: -65.566,
                    sx: 1.124,
                    sy: 1.124,
                    a: 0.89
                },
                "14": {
                    x: -264.05,
                    y: -65.55,
                    sx: 1.19,
                    sy: 1.19,
                    a: 1
                },
                "15": {
                    x: -264.007,
                    y: -65.549,
                    sx: 1.229,
                    sy: 1.229,
                    a: 0.86
                },
                "16": {
                    x: -264.014,
                    sx: 1.267,
                    sy: 1.267,
                    a: 0.71
                },
                "17": {
                    x: -263.971,
                    y: -65.548,
                    sx: 1.305,
                    sy: 1.305,
                    a: 0.57
                },
                "18": {
                    x: -263.928,
                    sx: 1.343,
                    sy: 1.343,
                    a: 0.43
                },
                "19": {
                    x: -263.935,
                    y: -65.547,
                    sx: 1.381,
                    sy: 1.381,
                    a: 0.29
                },
                "20": {
                    x: -263.896,
                    y: -65.548,
                    sx: 1.419,
                    sy: 1.419,
                    a: 0.14
                },
                "21": {
                    x: -263.85,
                    y: -65.6,
                    sx: 1.457,
                    sy: 1.457,
                    a: 0
                }
            })
            .addTimedChild(instance15, 24, 13, {
                "24": {
                    x: -132.5,
                    y: -243.75,
                    a: 0
                },
                "25": {
                    a: 0.2
                },
                "26": {
                    a: 0.4
                },
                "27": {
                    a: 0.6
                },
                "28": {
                    a: 0.8
                },
                "29": {
                    a: 1
                },
                "33": {
                    a: 0.75
                },
                "34": {
                    a: 0.5
                },
                "35": {
                    a: 0.25
                },
                "36": {
                    a: 0
                }
            })
            .addTimedChild(instance19, 39, 13, {
                "39": {
                    x: 176.15,
                    y: 64.9,
                    a: 0
                },
                "40": {
                    a: 0.17
                },
                "41": {
                    a: 0.33
                },
                "42": {
                    a: 0.5
                },
                "43": {
                    a: 0.67
                },
                "44": {
                    a: 0.83
                },
                "45": {
                    a: 1
                },
                "47": {
                    a: 0.8
                },
                "48": {
                    a: 0.6
                },
                "49": {
                    a: 0.4
                },
                "50": {
                    a: 0.2
                },
                "51": {
                    a: 0
                }
            })
            .addTimedChild(instance16, 28, 17, {
                "28": {
                    x: 132.5,
                    y: 155.85,
                    sx: 0.59,
                    sy: 0.59,
                    a: 0
                },
                "29": {
                    x: 132.538,
                    y: 155.879,
                    sx: 0.657,
                    sy: 0.657,
                    a: 0.11
                },
                "30": {
                    x: 132.579,
                    y: 155.862,
                    sx: 0.723,
                    sy: 0.723,
                    a: 0.22
                },
                "31": {
                    x: 132.663,
                    y: 155.886,
                    sx: 0.79,
                    sy: 0.79,
                    a: 0.33
                },
                "32": {
                    x: 132.704,
                    y: 155.87,
                    sx: 0.857,
                    sy: 0.857,
                    a: 0.45
                },
                "33": {
                    x: 132.742,
                    y: 155.898,
                    sx: 0.924,
                    sy: 0.924,
                    a: 0.55
                },
                "34": {
                    x: 132.783,
                    y: 155.882,
                    sx: 0.99,
                    sy: 0.99,
                    a: 0.67
                },
                "35": {
                    x: 132.867,
                    y: 155.906,
                    sx: 1.057,
                    sy: 1.057,
                    a: 0.78
                },
                "36": {
                    x: 132.908,
                    y: 155.889,
                    sx: 1.124,
                    sy: 1.124,
                    a: 0.89
                },
                "37": {
                    x: 132.95,
                    y: 155.9,
                    sx: 1.19,
                    sy: 1.19,
                    a: 1
                },
                "38": {
                    x: 133.008,
                    y: 155.943,
                    sx: 1.229,
                    sy: 1.229,
                    a: 0.86
                },
                "39": {
                    x: 133.016,
                    y: 155.936,
                    sx: 1.267,
                    sy: 1.267,
                    a: 0.71
                },
                "40": {
                    x: 133.074,
                    y: 155.93,
                    sx: 1.305,
                    sy: 1.305,
                    a: 0.57
                },
                "41": {
                    x: 133.083,
                    y: 155.923,
                    sx: 1.343,
                    sy: 1.343,
                    a: 0.43
                },
                "42": {
                    x: 133.141,
                    y: 155.966,
                    sx: 1.381,
                    sy: 1.381,
                    a: 0.29
                },
                "43": {
                    x: 133.151,
                    y: 155.962,
                    sx: 1.419,
                    sy: 1.419,
                    a: 0.14
                },
                "44": {
                    x: 133.15,
                    y: 155.85,
                    sx: 1.457,
                    sy: 1.457,
                    a: 0
                }
            })
            .addTimedChild(instance17, 31, 15, {
                "31": {
                    x: 176.1,
                    y: 110.55,
                    a: 0
                },
                "32": {
                    a: 0.2
                },
                "33": {
                    a: 0.4
                },
                "34": {
                    a: 0.6
                },
                "35": {
                    a: 0.8
                },
                "36": {
                    a: 1
                },
                "40": {
                    a: 0.83
                },
                "41": {
                    a: 0.67
                },
                "42": {
                    a: 0.5
                },
                "43": {
                    a: 0.33
                },
                "44": {
                    a: 0.17
                },
                "45": {
                    a: 0
                }
            })
            .addTimedChild(instance26, 58, 5, {
                "58": {
                    x: 176.15,
                    y: -242.85,
                    a: 0
                },
                "59": {
                    a: 0.2
                },
                "60": {
                    a: 0.4
                },
                "61": {
                    a: 0.6
                },
                "62": {
                    a: 0.8
                }
            })
            .addTimedChild(instance22, 49, 14, {
                "49": {
                    x: 177.95,
                    y: -273.15,
                    a: 0
                },
                "50": {
                    a: 0.2
                },
                "51": {
                    a: 0.4
                },
                "52": {
                    a: 0.6
                },
                "53": {
                    a: 0.8
                },
                "54": {
                    a: 1
                },
                "58": {
                    a: 0.75
                },
                "59": {
                    a: 0.5
                },
                "60": {
                    a: 0.25
                },
                "61": {
                    a: 0
                }
            })
            .addTimedChild(instance23, 55, 8, {
                "55": {
                    x: -263.25,
                    y: -154.8,
                    a: 0
                },
                "56": {
                    a: 0.17
                },
                "57": {
                    a: 0.33
                },
                "58": {
                    a: 0.5
                },
                "59": {
                    a: 0.67
                },
                "60": {
                    a: 0.83
                },
                "61": {
                    a: 1
                }
            })
            .addTimedChild(instance21, 46, 17, {
                "46": {
                    x: -267.55,
                    y: 239.95,
                    sx: 0.59,
                    sy: 0.59,
                    a: 0
                },
                "47": {
                    x: -267.513,
                    y: 239.942,
                    sx: 0.657,
                    sy: 0.657,
                    a: 0.11
                },
                "48": {
                    x: -267.435,
                    y: 239.94,
                    sx: 0.723,
                    sy: 0.723,
                    a: 0.22
                },
                "49": {
                    x: -267.39,
                    y: 239.925,
                    sx: 0.79,
                    sy: 0.79,
                    a: 0.33
                },
                "50": {
                    x: -267.362,
                    y: 239.974,
                    sx: 0.857,
                    sy: 0.857,
                    a: 0.45
                },
                "51": {
                    x: -267.326,
                    y: 239.965,
                    sx: 0.924,
                    sy: 0.924,
                    a: 0.55
                },
                "52": {
                    x: -267.247,
                    y: 239.964,
                    sx: 0.99,
                    sy: 0.99,
                    a: 0.67
                },
                "53": {
                    x: -267.203,
                    y: 239.948,
                    sx: 1.057,
                    sy: 1.057,
                    a: 0.78
                },
                "54": {
                    x: -267.174,
                    y: 239.947,
                    sx: 1.124,
                    sy: 1.124,
                    a: 0.89
                },
                "55": {
                    x: -267.1,
                    y: 240,
                    sx: 1.19,
                    sy: 1.19,
                    a: 1
                },
                "56": {
                    x: -267.105,
                    y: 240.037,
                    sx: 1.229,
                    sy: 1.229,
                    a: 0.86
                },
                "57": {
                    x: -267.059,
                    y: 240.024,
                    sx: 1.267,
                    sy: 1.267,
                    a: 0.71
                },
                "58": {
                    x: -267.014,
                    y: 240.01,
                    sx: 1.305,
                    sy: 1.305,
                    a: 0.57
                },
                "59": {
                    x: -267.019,
                    y: 240.047,
                    sx: 1.343,
                    sy: 1.343,
                    a: 0.43
                },
                "60": {
                    x: -266.974,
                    y: 240.034,
                    sx: 1.381,
                    sy: 1.381,
                    a: 0.29
                },
                "61": {
                    x: -266.932,
                    y: 240.024,
                    sx: 1.419,
                    sy: 1.419,
                    a: 0.14
                },
                "62": {
                    x: -266.9,
                    y: 239.95,
                    sx: 1.457,
                    sy: 1.457,
                    a: 0
                }
            })
            .addTimedChild(instance20, 46, 13, {
                "46": {
                    x: -218.8,
                    y: 152.3,
                    a: 0
                },
                "47": {
                    a: 0.2
                },
                "48": {
                    a: 0.4
                },
                "49": {
                    a: 0.6
                },
                "50": {
                    a: 0.8
                },
                "51": {
                    a: 1
                },
                "55": {
                    a: 0.75
                },
                "56": {
                    a: 0.5
                },
                "57": {
                    a: 0.25
                },
                "58": {
                    a: 0
                }
            })
            .addTimedChild(instance25, 58, 5, {
                "58": {
                    x: 176.1,
                    y: 110.55,
                    a: 0
                },
                "59": {
                    a: 0.33
                },
                "60": {
                    a: 0.67
                },
                "61": {
                    a: 1
                }
            })
            .addTimedChild(instance24, 58, 5, {
                "58": {
                    x: -264.5,
                    y: -65.6,
                    sx: 0.59,
                    sy: 0.59,
                    a: 0
                },
                "59": {
                    x: -264.418,
                    y: -65.57,
                    sx: 0.657,
                    sy: 0.657,
                    a: 0.11
                },
                "60": {
                    x: -264.395,
                    y: -65.592,
                    sx: 0.723,
                    sy: 0.723,
                    a: 0.22
                },
                "61": {
                    x: -264.355,
                    y: -65.561,
                    sx: 0.79,
                    sy: 0.79,
                    a: 0.33
                },
                "62": {
                    x: -264.282,
                    y: -65.583,
                    sx: 0.857,
                    sy: 0.857,
                    a: 0.45
                }
            })
            .addTimedChild(instance7, 3, 60, {
                "3": {
                    x: 260.6,
                    y: -153.7,
                    a: 0
                },
                "4": {
                    a: 0.17
                },
                "5": {
                    a: 0.33
                },
                "6": {
                    a: 0.5
                },
                "7": {
                    a: 0.67
                },
                "8": {
                    a: 0.83
                },
                "9": {
                    a: 1
                },
                "12": {
                    a: 0.86
                },
                "13": {
                    a: 0.71
                },
                "14": {
                    a: 0.57
                },
                "15": {
                    a: 0.43
                },
                "16": {
                    a: 0.29
                },
                "17": {
                    a: 0.14
                },
                "18": {
                    a: 0
                },
                "35": {
                    x: -38.7,
                    y: 241.75
                },
                "36": {
                    a: 0.14
                },
                "37": {
                    a: 0.29
                },
                "38": {
                    a: 0.43
                },
                "39": {
                    a: 0.57
                },
                "40": {
                    a: 0.71
                },
                "41": {
                    a: 0.86
                },
                "42": {
                    a: 1
                },
                "44": {
                    a: 0.86
                },
                "45": {
                    a: 0.71
                },
                "46": {
                    a: 0.57
                },
                "47": {
                    a: 0.43
                },
                "48": {
                    a: 0.29
                },
                "49": {
                    a: 0.14
                },
                "50": {
                    a: 0
                }
            })
            .addTimedChild(instance10, 7, 15, {
                "7": {
                    x: -261.55,
                    y: 242.75,
                    a: 0
                },
                "8": {
                    a: 0.2
                },
                "9": {
                    a: 0.4
                },
                "10": {
                    a: 0.6
                },
                "11": {
                    a: 0.8
                },
                "12": {
                    a: 1
                },
                "16": {
                    a: 0.83
                },
                "17": {
                    a: 0.67
                },
                "18": {
                    a: 0.5
                },
                "19": {
                    a: 0.33
                },
                "20": {
                    a: 0.17
                },
                "21": {
                    a: 0
                }
            })
            .addTimedChild(instance5, 2, 45, {
                "2": {
                    x: 176.15,
                    y: 242.15,
                    a: 0
                },
                "3": {
                    a: 0.14
                },
                "4": {
                    a: 0.29
                },
                "5": {
                    a: 0.43
                },
                "6": {
                    a: 0.57
                },
                "7": {
                    a: 0.71
                },
                "8": {
                    a: 0.86
                },
                "9": {
                    a: 1
                },
                "13": {
                    a: 0.83
                },
                "14": {
                    a: 0.67
                },
                "15": {
                    a: 0.5
                },
                "16": {
                    a: 0.33
                },
                "17": {
                    a: 0.17
                },
                "18": {
                    a: 0
                },
                "32": {
                    a: 0.14
                },
                "33": {
                    a: 0.29
                },
                "34": {
                    a: 0.43
                },
                "35": {
                    a: 0.57
                },
                "36": {
                    a: 0.71
                },
                "37": {
                    a: 0.86
                },
                "38": {
                    a: 1
                },
                "40": {
                    a: 0.86
                },
                "41": {
                    a: 0.71
                },
                "42": {
                    a: 0.57
                },
                "43": {
                    a: 0.43
                },
                "44": {
                    a: 0.29
                },
                "45": {
                    a: 0.14
                },
                "46": {
                    a: 0
                }
            })
            .addTimedChild(instance2)
            .addTimedChild(instance1);
    });

    lib.thinking_dots2 = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 64,
            framerate: 30,
            loop: false
        });
        var instance1 = new Graphic81(MovieClip.SYNCHED);
        this.addTimedChild(instance1, 0, 63, {
            "0": {
                x: 683.5,
                y: 342.1,
                sx: 1.124,
                sy: 1.124,
                a: 0
            },
            "1": {
                a: 0.14
            },
            "2": {
                a: 0.29
            },
            "3": {
                a: 0.43
            },
            "4": {
                a: 0.57
            },
            "5": {
                a: 0.71
            },
            "6": {
                a: 0.86
            },
            "7": {
                a: 1
            },
            "55": {
                a: 0.88
            },
            "56": {
                a: 0.75
            },
            "57": {
                a: 0.63
            },
            "58": {
                a: 0.5
            },
            "59": {
                a: 0.38
            },
            "60": {
                a: 0.25
            },
            "61": {
                a: 0.13
            },
            "62": {
                a: 0
            }
        });
    });

    lib.thinking_dots2.assets = {
        "thinking_dots2_atlas_1": "images/thinking_dots2_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.thinking_dots2,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 64,
        library: lib
    };
}
