// jibo-anim-db-animations/timelines/thinking_dots3.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 45, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-56.75, -57.55);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 45, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-56.75, -57.55);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 45, loop: false });
        var instance1 = new Sprite(fromFrame("dot21"))
            .setTransform(-56.75, -57.55);
        this.addTimedChild(instance1);
    });

    lib.thinking_dots3 = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 45,
            framerate: 30
        });
        var instance3 = new Graphic3(MovieClip.SYNCHED);
        var instance2 = new Graphic2(MovieClip.SYNCHED);
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance3, 0, 45, {
                "0": {
                    x: 461.6,
                    y: 359.8,
                    sx: 1,
                    sy: 1,
                    c: [
                        0,
                        0.14,
                        0,
                        0.14,
                        0,
                        0.17
                    ]
                },
                "1": {
                    x: 461.594,
                    y: 359.794,
                    sx: 0.983,
                    sy: 0.983
                },
                "2": {
                    x: 461.551,
                    y: 359.86,
                    sx: 0.92,
                    sy: 0.92
                },
                "3": {
                    x: 461.55,
                    y: 359.85,
                    sx: 0.899,
                    sy: 0.899
                },
                "4": {
                    x: 461.547,
                    y: 359.852,
                    sx: 0.903,
                    sy: 0.903,
                    c: [
                        0,
                        0.14,
                        0,
                        0.15,
                        0,
                        0.18
                    ]
                },
                "5": {
                    x: 461.572,
                    y: 359.865,
                    sx: 0.919,
                    sy: 0.919,
                    c: [
                        0,
                        0.13,
                        0,
                        0.19,
                        0,
                        0.23
                    ]
                },
                "6": {
                    x: 461.595,
                    y: 359.826,
                    sx: 0.954,
                    sy: 0.954,
                    c: [
                        0,
                        0.12,
                        0,
                        0.29,
                        0,
                        0.34
                    ]
                },
                "7": {
                    x: 461.628,
                    y: 359.836,
                    sx: 1.012,
                    sy: 1.012,
                    c: [
                        0,
                        0.09,
                        0,
                        0.44,
                        0,
                        0.52
                    ]
                },
                "8": {
                    x: 461.644,
                    y: 359.778,
                    sx: 1.077,
                    sy: 1.077,
                    c: [
                        0,
                        0.07,
                        0,
                        0.62,
                        0,
                        0.72
                    ]
                },
                "9": {
                    x: 461.661,
                    y: 359.766,
                    sx: 1.119,
                    sy: 1.119,
                    c: [
                        0,
                        0.05,
                        0,
                        0.74,
                        0,
                        0.85
                    ]
                },
                "10": {
                    x: 461.701,
                    y: 359.75,
                    sx: 1.139,
                    sy: 1.139,
                    c: [
                        0,
                        0.05,
                        0,
                        0.79,
                        0,
                        0.92
                    ]
                },
                "11": {
                    x: 461.7,
                    y: 359.8,
                    sx: 1.145,
                    sy: 1.145,
                    c: [
                        0,
                        0.05,
                        0,
                        0.81,
                        0,
                        0.93
                    ]
                },
                "12": {
                    x: 461.682,
                    y: 359.772,
                    sx: 1.14,
                    sy: 1.14,
                    c: [
                        0,
                        0.05,
                        0,
                        0.8,
                        0,
                        0.92
                    ]
                },
                "13": {
                    x: 461.671,
                    y: 359.788,
                    sx: 1.124,
                    sy: 1.124,
                    c: [
                        0,
                        0.05,
                        0,
                        0.75,
                        0,
                        0.87
                    ]
                },
                "14": {
                    x: 461.663,
                    y: 359.765,
                    sx: 1.09,
                    sy: 1.09,
                    c: [
                        0,
                        0.07,
                        0,
                        0.66,
                        0,
                        0.76
                    ]
                },
                "15": {
                    x: 461.631,
                    y: 359.805,
                    sx: 1.032,
                    sy: 1.032,
                    c: [
                        0,
                        0.09,
                        0,
                        0.5,
                        0,
                        0.58
                    ]
                },
                "16": {
                    x: 461.589,
                    y: 359.866,
                    sx: 0.967,
                    sy: 0.967,
                    c: [
                        0,
                        0.11,
                        0,
                        0.32,
                        0,
                        0.38
                    ]
                },
                "17": {
                    x: 461.571,
                    y: 359.848,
                    sx: 0.924,
                    sy: 0.924,
                    c: [
                        0,
                        0.13,
                        0,
                        0.21,
                        0,
                        0.25
                    ]
                },
                "18": {
                    x: 461.576,
                    y: 359.879,
                    sx: 0.904,
                    sy: 0.904,
                    c: [
                        0,
                        0.14,
                        0,
                        0.15,
                        0,
                        0.18
                    ]
                },
                "19": {
                    x: 461.55,
                    y: 359.85,
                    sx: 0.899,
                    sy: 0.899,
                    c: [
                        0,
                        0.14,
                        0,
                        0.14,
                        0,
                        0.17
                    ]
                },
                "20": {
                    x: 461.568,
                    y: 359.86,
                    sx: 0.9,
                    sy: 0.9
                },
                "21": {
                    x: 461.532,
                    y: 359.859,
                    sx: 0.905,
                    sy: 0.905
                },
                "22": {
                    x: 461.571,
                    y: 359.832,
                    sx: 0.916,
                    sy: 0.916
                },
                "23": {
                    x: 461.576,
                    y: 359.821,
                    sx: 0.934,
                    sy: 0.934
                },
                "24": {
                    x: 461.578,
                    y: 359.778,
                    sx: 0.958,
                    sy: 0.958
                },
                "25": {
                    x: 461.6,
                    y: 359.783,
                    sx: 0.979,
                    sy: 0.979
                },
                "26": {
                    x: 461.613,
                    y: 359.778,
                    sx: 0.992,
                    sy: 0.992
                },
                "27": {
                    x: 461.632,
                    y: 359.798,
                    sx: 0.998,
                    sy: 0.998
                },
                "28": {
                    x: 461.6,
                    y: 359.8,
                    sx: 1,
                    sy: 1
                }
            })
            .addTimedChild(instance2, 0, 45, {
                "0": {
                    x: 641.6,
                    y: 359.8,
                    sx: 1,
                    sy: 1,
                    c: [
                        0,
                        0.14,
                        0,
                        0.14,
                        0,
                        0.17
                    ]
                },
                "8": {
                    x: 641.592,
                    y: 359.822,
                    sx: 1.001,
                    sy: 1.001,
                    c: [
                        0,
                        0.14,
                        0,
                        0.15,
                        0,
                        0.18
                    ]
                },
                "9": {
                    x: 641.59,
                    y: 359.797,
                    sx: 1.006,
                    sy: 1.006,
                    c: [
                        0,
                        0.13,
                        0,
                        0.16,
                        0,
                        0.2
                    ]
                },
                "10": {
                    x: 641.607,
                    y: 359.803,
                    sx: 1.015,
                    sy: 1.015,
                    c: [
                        0,
                        0.13,
                        0,
                        0.2,
                        0,
                        0.25
                    ]
                },
                "11": {
                    x: 641.621,
                    y: 359.795,
                    sx: 1.03,
                    sy: 1.03,
                    c: [
                        0,
                        0.12,
                        0,
                        0.27,
                        0,
                        0.33
                    ]
                },
                "12": {
                    x: 641.609,
                    y: 359.798,
                    sx: 1.052,
                    sy: 1.052,
                    c: [
                        0,
                        0.11,
                        0,
                        0.38,
                        0,
                        0.45
                    ]
                },
                "13": {
                    x: 641.625,
                    y: 359.797,
                    sx: 1.081,
                    sy: 1.081,
                    c: [
                        0,
                        0.09,
                        0,
                        0.51,
                        0,
                        0.6
                    ]
                },
                "14": {
                    x: 641.688,
                    y: 359.794,
                    sx: 1.108,
                    sy: 1.108,
                    c: [
                        0,
                        0.07,
                        0,
                        0.64,
                        0,
                        0.74
                    ]
                },
                "15": {
                    x: 641.68,
                    y: 359.797,
                    sx: 1.126,
                    sy: 1.126,
                    c: [
                        0,
                        0.06,
                        0,
                        0.72,
                        0,
                        0.84
                    ]
                },
                "16": {
                    x: 641.696,
                    y: 359.765,
                    sx: 1.138,
                    sy: 1.138,
                    c: [
                        0,
                        0.05,
                        0,
                        0.78,
                        0,
                        0.89
                    ]
                },
                "17": {
                    x: 641.67,
                    y: 359.769,
                    sx: 1.143,
                    sy: 1.143,
                    c: [
                        0,
                        0.05,
                        0,
                        0.8,
                        0,
                        0.93
                    ]
                },
                "18": {
                    x: 641.7,
                    y: 359.8,
                    sx: 1.145,
                    sy: 1.145,
                    c: [
                        0,
                        0.05,
                        0,
                        0.81,
                        0,
                        0.93
                    ]
                },
                "19": {
                    x: 641.692,
                    y: 359.772,
                    sx: 1.14,
                    sy: 1.14,
                    c: [
                        0,
                        0.05,
                        0,
                        0.8,
                        0,
                        0.92
                    ]
                },
                "20": {
                    x: 641.664,
                    y: 359.788,
                    sx: 1.124,
                    sy: 1.124,
                    c: [
                        0,
                        0.05,
                        0,
                        0.75,
                        0,
                        0.87
                    ]
                },
                "21": {
                    x: 641.641,
                    y: 359.765,
                    sx: 1.09,
                    sy: 1.09,
                    c: [
                        0,
                        0.07,
                        0,
                        0.66,
                        0,
                        0.76
                    ]
                },
                "22": {
                    x: 641.624,
                    y: 359.805,
                    sx: 1.032,
                    sy: 1.032,
                    c: [
                        0,
                        0.09,
                        0,
                        0.5,
                        0,
                        0.58
                    ]
                },
                "23": {
                    x: 641.582,
                    y: 359.866,
                    sx: 0.967,
                    sy: 0.967,
                    c: [
                        0,
                        0.11,
                        0,
                        0.32,
                        0,
                        0.38
                    ]
                },
                "24": {
                    x: 641.551,
                    y: 359.848,
                    sx: 0.924,
                    sy: 0.924,
                    c: [
                        0,
                        0.13,
                        0,
                        0.21,
                        0,
                        0.25
                    ]
                },
                "25": {
                    x: 641.519,
                    y: 359.879,
                    sx: 0.904,
                    sy: 0.904,
                    c: [
                        0,
                        0.14,
                        0,
                        0.15,
                        0,
                        0.18
                    ]
                },
                "26": {
                    x: 641.55,
                    y: 359.85,
                    sx: 0.899,
                    sy: 0.899,
                    c: [
                        0,
                        0.14,
                        0,
                        0.14,
                        0,
                        0.17
                    ]
                },
                "27": {
                    x: 641.549,
                    y: 359.86,
                    sx: 0.9,
                    sy: 0.9
                },
                "28": {
                    x: 641.512,
                    y: 359.859,
                    sx: 0.905,
                    sy: 0.905
                },
                "29": {
                    x: 641.514,
                    y: 359.832,
                    sx: 0.916,
                    sy: 0.916
                },
                "30": {
                    x: 641.54,
                    y: 359.821,
                    sx: 0.934,
                    sy: 0.934
                },
                "31": {
                    x: 641.547,
                    y: 359.778,
                    sx: 0.958,
                    sy: 0.958
                },
                "32": {
                    x: 641.549,
                    y: 359.783,
                    sx: 0.979,
                    sy: 0.979
                },
                "33": {
                    x: 641.609,
                    y: 359.778,
                    sx: 0.992,
                    sy: 0.992
                },
                "34": {
                    x: 641.565,
                    y: 359.798,
                    sx: 0.998,
                    sy: 0.998
                },
                "35": {
                    x: 641.6,
                    y: 359.8,
                    sx: 1,
                    sy: 1
                }
            })
            .addTimedChild(instance1, 0, 45, {
                "0": {
                    x: 821.6,
                    y: 359.8,
                    sx: 1,
                    sy: 1,
                    c: [
                        0,
                        0.14,
                        0,
                        0.14,
                        0,
                        0.17
                    ]
                },
                "15": {
                    x: 821.578,
                    y: 359.822,
                    sx: 1.001,
                    sy: 1.001,
                    c: [
                        0,
                        0.14,
                        0,
                        0.15,
                        0,
                        0.18
                    ]
                },
                "16": {
                    x: 821.589,
                    y: 359.797,
                    sx: 1.006,
                    sy: 1.006,
                    c: [
                        0,
                        0.13,
                        0,
                        0.16,
                        0,
                        0.2
                    ]
                },
                "17": {
                    x: 821.61,
                    y: 359.803,
                    sx: 1.015,
                    sy: 1.015,
                    c: [
                        0,
                        0.13,
                        0,
                        0.2,
                        0,
                        0.25
                    ]
                },
                "18": {
                    x: 821.621,
                    y: 359.795,
                    sx: 1.03,
                    sy: 1.03,
                    c: [
                        0,
                        0.12,
                        0,
                        0.27,
                        0,
                        0.33
                    ]
                },
                "19": {
                    x: 821.638,
                    y: 359.798,
                    sx: 1.052,
                    sy: 1.052,
                    c: [
                        0,
                        0.11,
                        0,
                        0.38,
                        0,
                        0.45
                    ]
                },
                "20": {
                    x: 821.657,
                    y: 359.797,
                    sx: 1.081,
                    sy: 1.081,
                    c: [
                        0,
                        0.09,
                        0,
                        0.51,
                        0,
                        0.6
                    ]
                },
                "21": {
                    x: 821.671,
                    y: 359.794,
                    sx: 1.108,
                    sy: 1.108,
                    c: [
                        0,
                        0.07,
                        0,
                        0.64,
                        0,
                        0.74
                    ]
                },
                "22": {
                    x: 821.691,
                    y: 359.797,
                    sx: 1.126,
                    sy: 1.126,
                    c: [
                        0,
                        0.06,
                        0,
                        0.72,
                        0,
                        0.84
                    ]
                },
                "23": {
                    x: 821.668,
                    y: 359.765,
                    sx: 1.138,
                    sy: 1.138,
                    c: [
                        0,
                        0.05,
                        0,
                        0.78,
                        0,
                        0.89
                    ]
                },
                "24": {
                    x: 821.694,
                    y: 359.769,
                    sx: 1.143,
                    sy: 1.143,
                    c: [
                        0,
                        0.05,
                        0,
                        0.8,
                        0,
                        0.93
                    ]
                },
                "25": {
                    x: 821.7,
                    y: 359.8,
                    sx: 1.145,
                    sy: 1.145,
                    c: [
                        0,
                        0.05,
                        0,
                        0.81,
                        0,
                        0.93
                    ]
                },
                "26": {
                    x: 821.705,
                    y: 359.779,
                    sx: 1.141,
                    sy: 1.141,
                    c: [
                        0,
                        0.05,
                        0,
                        0.8,
                        0,
                        0.92
                    ]
                },
                "27": {
                    x: 821.741,
                    y: 359.764,
                    sx: 1.129,
                    sy: 1.129,
                    c: [
                        0,
                        0.05,
                        0,
                        0.76,
                        0,
                        0.89
                    ]
                },
                "28": {
                    x: 821.686,
                    y: 359.78,
                    sx: 1.104,
                    sy: 1.104,
                    c: [
                        0,
                        0.06,
                        0,
                        0.7,
                        0,
                        0.81
                    ]
                },
                "29": {
                    x: 821.692,
                    y: 359.793,
                    sx: 1.061,
                    sy: 1.061,
                    c: [
                        0,
                        0.08,
                        0,
                        0.58,
                        0,
                        0.67
                    ]
                },
                "30": {
                    x: 821.638,
                    y: 359.847,
                    sx: 1.002,
                    sy: 1.002,
                    c: [
                        0,
                        0.1,
                        0,
                        0.42,
                        0,
                        0.49
                    ]
                },
                "31": {
                    x: 821.599,
                    sx: 0.95,
                    sy: 0.95,
                    c: [
                        0,
                        0.12,
                        0,
                        0.27,
                        0,
                        0.33
                    ]
                },
                "32": {
                    x: 821.568,
                    y: 359.838,
                    sx: 0.918,
                    sy: 0.918,
                    c: [
                        0,
                        0.13,
                        0,
                        0.19,
                        0,
                        0.23
                    ]
                },
                "33": {
                    x: 821.597,
                    y: 359.85,
                    sx: 0.903,
                    sy: 0.903,
                    c: [
                        0,
                        0.14,
                        0,
                        0.15,
                        0,
                        0.18
                    ]
                },
                "34": {
                    x: 821.55,
                    sx: 0.899,
                    sy: 0.899,
                    c: [
                        0,
                        0.14,
                        0,
                        0.14,
                        0,
                        0.17
                    ]
                },
                "35": {
                    x: 821.529,
                    y: 359.86,
                    sx: 0.9,
                    sy: 0.9
                },
                "36": {
                    x: 821.542,
                    y: 359.859,
                    sx: 0.905,
                    sy: 0.905
                },
                "37": {
                    x: 821.556,
                    y: 359.832,
                    sx: 0.916,
                    sy: 0.916
                },
                "38": {
                    x: 821.553,
                    y: 359.821,
                    sx: 0.934,
                    sy: 0.934
                },
                "39": {
                    x: 821.566,
                    y: 359.778,
                    sx: 0.958,
                    sy: 0.958
                },
                "40": {
                    x: 821.597,
                    y: 359.783,
                    sx: 0.979,
                    sy: 0.979
                },
                "41": {
                    x: 821.606,
                    y: 359.778,
                    sx: 0.992,
                    sy: 0.992
                },
                "42": {
                    x: 821.597,
                    y: 359.798,
                    sx: 0.998,
                    sy: 0.998
                },
                "43": {
                    x: 821.6,
                    y: 359.8,
                    sx: 1,
                    sy: 1
                }
            });
    });

    lib.thinking_dots3.assets = {
        "thinking_dots3_atlas_1": "images/thinking_dots3_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.thinking_dots3,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 45,
        library: lib
    };
}
