// jibo-anim-db-animations/timelines/turnon_01.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;
    var Graphics = PIXI.Graphics;
    var shapes = PIXI.animate.ShapesCache;

    var Graphic1 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 44, loop: false });
        var instance1 = new Graphics()
            .drawCommands(shapes.turnon01[0]);
        this.addTimedChild(instance1);
    });

    var Graphic2 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 44, loop: false });
        var instance1 = new Sprite(fromFrame("ring1"))
            .setTransform(-211.65, -211.65, 0.645, 0.645);
        this.addTimedChild(instance1);
    });

    var Graphic3 = MovieClip.extend(function (mode) {
        MovieClip.call(this, { mode: mode, duration: 44, loop: false });
        var instance1 = new Sprite(fromFrame("glow1"))
            .setTransform(-321.75, -315.65, 0.856, 0.856);
        this.addTimedChild(instance1);
    });

    lib.turnon01 = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 44,
            framerate: 30,
            loop: false
        });
        var instance3 = new Graphic3(MovieClip.SYNCHED);
        var instance2 = new Graphic2(MovieClip.SYNCHED);
        var instance1 = new Graphic1(MovieClip.SYNCHED);
        this.addTimedChild(instance3, 0, 44, {
                "0": {
                    x: 642,
                    y: 356.3,
                    sx: 0.292,
                    sy: 0.292,
                    a: 0
                },
                "1": {
                    x: 641.993,
                    y: 356.264,
                    sx: 0.296,
                    sy: 0.296
                },
                "2": {
                    y: 356.215,
                    sx: 0.308,
                    sy: 0.308,
                    a: 0.02
                },
                "3": {
                    x: 642.028,
                    y: 356.108,
                    sx: 0.329,
                    sy: 0.329,
                    a: 0.04
                },
                "4": {
                    x: 642.032,
                    y: 355.886,
                    sx: 0.362,
                    sy: 0.362,
                    a: 0.08
                },
                "5": {
                    x: 642.027,
                    y: 355.643,
                    sx: 0.407,
                    sy: 0.407,
                    a: 0.13
                },
                "6": {
                    x: 642.05,
                    y: 355.291,
                    sx: 0.468,
                    sy: 0.468,
                    a: 0.2
                },
                "7": {
                    x: 642.069,
                    y: 354.832,
                    sx: 0.545,
                    sy: 0.545,
                    a: 0.29
                },
                "8": {
                    x: 642.122,
                    y: 354.288,
                    sx: 0.639,
                    sy: 0.639,
                    a: 0.39
                },
                "9": {
                    x: 642.163,
                    y: 353.649,
                    sx: 0.746,
                    sy: 0.746,
                    a: 0.52
                },
                "10": {
                    x: 642.174,
                    y: 353.04,
                    sx: 0.856,
                    sy: 0.856,
                    a: 0.64
                },
                "11": {
                    x: 642.194,
                    y: 352.43,
                    sx: 0.956,
                    sy: 0.956,
                    a: 0.76
                },
                "12": {
                    x: 642.239,
                    y: 351.952,
                    sx: 1.039,
                    sy: 1.039,
                    a: 0.85
                },
                "13": {
                    x: 642.242,
                    y: 351.616,
                    sx: 1.1,
                    sy: 1.1,
                    a: 0.92
                },
                "14": {
                    x: 642.289,
                    y: 351.381,
                    sx: 1.14,
                    sy: 1.14,
                    a: 0.97
                },
                "15": {
                    x: 642.265,
                    y: 351.246,
                    sx: 1.161,
                    sy: 1.161,
                    a: 0.99
                },
                "16": {
                    x: 642.25,
                    y: 351.15,
                    sx: 1.168,
                    sy: 1.168,
                    a: 1
                },
                "17": {
                    a: 0.93
                },
                "18": {
                    a: 0.86
                },
                "19": {
                    a: 0.79
                },
                "20": {
                    a: 0.73
                },
                "21": {
                    a: 0.66
                },
                "22": {
                    a: 0.61
                },
                "23": {
                    a: 0.55
                },
                "24": {
                    a: 0.5
                },
                "25": {
                    a: 0.45
                },
                "26": {
                    a: 0.39
                },
                "27": {
                    a: 0.35
                },
                "28": {
                    a: 0.31
                },
                "29": {
                    a: 0.27
                },
                "30": {
                    a: 0.23
                },
                "31": {
                    a: 0.2
                },
                "32": {
                    a: 0.16
                },
                "33": {
                    a: 0.14
                },
                "34": {
                    a: 0.11
                },
                "35": {
                    a: 0.09
                },
                "36": {
                    a: 0.07
                },
                "37": {
                    a: 0.05
                },
                "38": {
                    a: 0.04
                },
                "39": {
                    a: 0.02
                },
                "40": {
                    a: 0.01
                },
                "41": {
                    a: 0
                }
            })
            .addTimedChild(instance2, 0, 44, {
                "0": {
                    x: 638,
                    y: 366,
                    sx: 0.371,
                    sy: 0.371,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "1": {
                    x: 638.006,
                    y: 366.015,
                    sx: 0.374,
                    sy: 0.374
                },
                "2": {
                    x: 638.119,
                    y: 365.878,
                    sx: 0.383,
                    sy: 0.383
                },
                "3": {
                    x: 638.264,
                    y: 365.796,
                    sx: 0.399,
                    sy: 0.399
                },
                "4": {
                    x: 638.422,
                    y: 365.645,
                    sx: 0.424,
                    sy: 0.424
                },
                "5": {
                    x: 638.719,
                    y: 365.383,
                    sx: 0.458,
                    sy: 0.458
                },
                "6": {
                    x: 639.064,
                    y: 365.082,
                    sx: 0.504,
                    sy: 0.504
                },
                "7": {
                    x: 639.569,
                    y: 364.707,
                    sx: 0.561,
                    sy: 0.561
                },
                "8": {
                    x: 640.083,
                    y: 364.216,
                    sx: 0.632,
                    sy: 0.632
                },
                "9": {
                    x: 640.775,
                    y: 363.62,
                    sx: 0.716,
                    sy: 0.716
                },
                "10": {
                    x: 641.532,
                    y: 363.013,
                    sx: 0.81,
                    sy: 0.81
                },
                "11": {
                    x: 642.283,
                    y: 362.281,
                    sx: 0.907,
                    sy: 0.907
                },
                "12": {
                    x: 643.054,
                    y: 361.634,
                    sx: 0.999,
                    sy: 0.999
                },
                "13": {
                    x: 643.729,
                    y: 361.084,
                    sx: 1.079,
                    sy: 1.079
                },
                "14": {
                    x: 644.185,
                    y: 360.71,
                    sx: 1.142,
                    sy: 1.142
                },
                "15": {
                    x: 644.622,
                    y: 360.359,
                    sx: 1.189,
                    sy: 1.189
                },
                "16": {
                    x: 644.778,
                    y: 360.144,
                    sx: 1.219,
                    sy: 1.219
                },
                "17": {
                    x: 644.954,
                    y: 360.03,
                    sx: 1.236,
                    sy: 1.236
                },
                "18": {
                    x: 645,
                    y: 360,
                    sx: 1.241,
                    sy: 1.241
                },
                "19": {
                    x: 644.591,
                    y: 359.977,
                    sx: 1.29,
                    sy: 1.29,
                    c: [
                        0.92,
                        0,
                        0.92,
                        0,
                        0.92,
                        0
                    ]
                },
                "20": {
                    x: 644.258,
                    y: 359.983,
                    sx: 1.337,
                    sy: 1.337,
                    c: [
                        0.85,
                        0,
                        0.85,
                        0,
                        0.85,
                        0
                    ]
                },
                "21": {
                    x: 643.872,
                    y: 360.002,
                    sx: 1.382,
                    sy: 1.382,
                    c: [
                        0.77,
                        0,
                        0.77,
                        0,
                        0.77,
                        0
                    ]
                },
                "22": {
                    x: 643.562,
                    y: 359.998,
                    sx: 1.425,
                    sy: 1.425,
                    c: [
                        0.71,
                        0,
                        0.71,
                        0,
                        0.71,
                        0
                    ]
                },
                "23": {
                    x: 643.189,
                    y: 360.002,
                    sx: 1.466,
                    sy: 1.466,
                    c: [
                        0.64,
                        0,
                        0.64,
                        0,
                        0.64,
                        0
                    ]
                },
                "24": {
                    x: 642.893,
                    y: 359.985,
                    sx: 1.505,
                    sy: 1.505,
                    c: [
                        0.58,
                        0,
                        0.58,
                        0,
                        0.58,
                        0
                    ]
                },
                "25": {
                    x: 642.594,
                    y: 359.979,
                    sx: 1.542,
                    sy: 1.542,
                    c: [
                        0.52,
                        0,
                        0.52,
                        0,
                        0.52,
                        0
                    ]
                },
                "26": {
                    x: 642.321,
                    y: 360.003,
                    sx: 1.577,
                    sy: 1.577,
                    c: [
                        0.46,
                        0,
                        0.46,
                        0,
                        0.46,
                        0
                    ]
                },
                "27": {
                    x: 642.036,
                    y: 359.983,
                    sx: 1.61,
                    sy: 1.61,
                    c: [
                        0.41,
                        0,
                        0.41,
                        0,
                        0.41,
                        0
                    ]
                },
                "28": {
                    x: 641.816,
                    y: 359.986,
                    sx: 1.641,
                    sy: 1.641,
                    c: [
                        0.36,
                        0,
                        0.36,
                        0,
                        0.36,
                        0
                    ]
                },
                "29": {
                    x: 641.564,
                    y: 360.012,
                    sx: 1.671,
                    sy: 1.671,
                    c: [
                        0.31,
                        0,
                        0.31,
                        0,
                        0.31,
                        0
                    ]
                },
                "30": {
                    x: 641.368,
                    y: 360.007,
                    sx: 1.698,
                    sy: 1.698,
                    c: [
                        0.27,
                        0,
                        0.27,
                        0,
                        0.27,
                        0
                    ]
                },
                "31": {
                    x: 641.129,
                    y: 359.969,
                    sx: 1.723,
                    sy: 1.723,
                    c: [
                        0.23,
                        0,
                        0.23,
                        0,
                        0.23,
                        0
                    ]
                },
                "32": {
                    x: 640.947,
                    y: 359.998,
                    sx: 1.746,
                    sy: 1.746,
                    c: [
                        0.2,
                        0,
                        0.2,
                        0,
                        0.2,
                        0
                    ]
                },
                "33": {
                    x: 640.822,
                    y: 359.996,
                    sx: 1.767,
                    sy: 1.767,
                    c: [
                        0.16,
                        0,
                        0.16,
                        0,
                        0.16,
                        0
                    ]
                },
                "34": {
                    x: 640.663,
                    y: 359.966,
                    sx: 1.786,
                    sy: 1.786,
                    c: [
                        0.13,
                        0,
                        0.13,
                        0,
                        0.13,
                        0
                    ]
                },
                "35": {
                    x: 640.51,
                    y: 360.005,
                    sx: 1.803,
                    sy: 1.803,
                    c: [
                        0.1,
                        0,
                        0.1,
                        0,
                        0.1,
                        0
                    ]
                },
                "36": {
                    x: 640.425,
                    y: 360.016,
                    sx: 1.818,
                    sy: 1.818,
                    c: [
                        0.08,
                        0,
                        0.08,
                        0,
                        0.08,
                        0
                    ]
                },
                "37": {
                    x: 640.286,
                    y: 359.99,
                    sx: 1.831,
                    sy: 1.831,
                    c: [
                        0.06,
                        0,
                        0.06,
                        0,
                        0.06,
                        0
                    ]
                },
                "38": {
                    x: 640.214,
                    y: 359.987,
                    sx: 1.842,
                    sy: 1.842,
                    c: [
                        0.04,
                        0,
                        0.04,
                        0,
                        0.04,
                        0
                    ]
                },
                "39": {
                    x: 640.139,
                    y: 359.996,
                    sx: 1.851,
                    sy: 1.851,
                    c: [
                        0.03,
                        0,
                        0.03,
                        0,
                        0.03,
                        0
                    ]
                },
                "40": {
                    x: 640.031,
                    y: 359.979,
                    sx: 1.858,
                    sy: 1.858,
                    c: [
                        0.02,
                        0,
                        0.02,
                        0,
                        0.02,
                        0
                    ]
                },
                "41": {
                    x: 640.039,
                    y: 359.985,
                    sx: 1.863,
                    sy: 1.863,
                    c: [
                        0.01,
                        0,
                        0.01,
                        0,
                        0.01,
                        0
                    ]
                },
                "42": {
                    x: 639.994,
                    y: 360.002,
                    sx: 1.866,
                    sy: 1.866,
                    c: [
                        0,
                        0,
                        0,
                        0,
                        0,
                        0
                    ]
                },
                "43": {
                    x: 640,
                    y: 360,
                    sx: 1.867,
                    sy: 1.867
                }
            })
            .addTimedChild(instance1, 0, 44, {
                "0": {
                    x: 627.8,
                    y: 372.7,
                    sx: 0.412,
                    sy: 0.412,
                    c: [
                        0,
                        0,
                        0,
                        0,
                        0,
                        0
                    ]
                },
                "1": {
                    x: 627.769,
                    y: 372.679,
                    sx: 0.418,
                    sy: 0.418,
                    c: [
                        0.01,
                        0,
                        0.01,
                        0,
                        0.01,
                        0
                    ]
                },
                "2": {
                    x: 627.81,
                    y: 372.695,
                    sx: 0.434,
                    sy: 0.434,
                    c: [
                        0.03,
                        0,
                        0.03,
                        0,
                        0.03,
                        0
                    ]
                },
                "3": {
                    x: 627.772,
                    y: 372.714,
                    sx: 0.462,
                    sy: 0.462,
                    c: [
                        0.06,
                        0,
                        0.06,
                        0,
                        0.06,
                        0
                    ]
                },
                "4": {
                    x: 627.793,
                    y: 372.688,
                    sx: 0.501,
                    sy: 0.501,
                    c: [
                        0.11,
                        0,
                        0.11,
                        0,
                        0.11,
                        0
                    ]
                },
                "5": {
                    x: 627.771,
                    y: 372.7,
                    sx: 0.552,
                    sy: 0.552,
                    c: [
                        0.18,
                        0,
                        0.18,
                        0,
                        0.18,
                        0
                    ]
                },
                "6": {
                    x: 627.802,
                    y: 372.704,
                    sx: 0.612,
                    sy: 0.612,
                    c: [
                        0.25,
                        0,
                        0.25,
                        0,
                        0.25,
                        0
                    ]
                },
                "7": {
                    x: 627.769,
                    y: 372.693,
                    sx: 0.682,
                    sy: 0.682,
                    c: [
                        0.34,
                        0,
                        0.34,
                        0,
                        0.34,
                        0
                    ]
                },
                "8": {
                    x: 627.805,
                    y: 372.68,
                    sx: 0.758,
                    sy: 0.758,
                    c: [
                        0.44,
                        0,
                        0.44,
                        0,
                        0.44,
                        0
                    ]
                },
                "9": {
                    x: 627.806,
                    y: 372.668,
                    sx: 0.838,
                    sy: 0.838,
                    c: [
                        0.54,
                        0,
                        0.54,
                        0,
                        0.54,
                        0
                    ]
                },
                "10": {
                    x: 627.794,
                    y: 372.676,
                    sx: 0.918,
                    sy: 0.918,
                    c: [
                        0.64,
                        0,
                        0.64,
                        0,
                        0.64,
                        0
                    ]
                },
                "11": {
                    x: 627.787,
                    y: 372.694,
                    sx: 0.995,
                    sy: 0.995,
                    c: [
                        0.74,
                        0,
                        0.74,
                        0,
                        0.74,
                        0
                    ]
                },
                "12": {
                    x: 627.813,
                    y: 372.688,
                    sx: 1.063,
                    sy: 1.063,
                    c: [
                        0.83,
                        0,
                        0.83,
                        0,
                        0.83,
                        0
                    ]
                },
                "13": {
                    x: 627.798,
                    y: 372.668,
                    sx: 1.12,
                    sy: 1.12,
                    c: [
                        0.9,
                        0,
                        0.9,
                        0,
                        0.9,
                        0
                    ]
                },
                "14": {
                    x: 627.786,
                    y: 372.698,
                    sx: 1.163,
                    sy: 1.163,
                    c: [
                        0.96,
                        0,
                        0.96,
                        0,
                        0.96,
                        0
                    ]
                },
                "15": {
                    x: 627.797,
                    y: 372.681,
                    sx: 1.189,
                    sy: 1.189,
                    c: [
                        0.99,
                        0,
                        0.99,
                        0,
                        0.99,
                        0
                    ]
                },
                "16": {
                    x: 627.8,
                    y: 372.7,
                    sx: 1.198,
                    sy: 1.198,
                    c: [
                        1,
                        0,
                        1,
                        0,
                        1,
                        0
                    ]
                },
                "17": {
                    x: 627.801,
                    y: 372.719,
                    sx: 1.184,
                    sy: 1.184
                },
                "18": {
                    x: 627.79,
                    y: 372.708,
                    sx: 1.17,
                    sy: 1.17
                },
                "19": {
                    x: 627.807,
                    y: 372.712,
                    sx: 1.157,
                    sy: 1.157
                },
                "20": {
                    x: 627.811,
                    y: 372.687,
                    sx: 1.144,
                    sy: 1.144
                },
                "21": {
                    x: 627.792,
                    y: 372.677,
                    sx: 1.132,
                    sy: 1.132
                },
                "22": {
                    x: 627.821,
                    y: 372.693,
                    sx: 1.12,
                    sy: 1.12
                },
                "23": {
                    x: 627.778,
                    y: 372.674,
                    sx: 1.109,
                    sy: 1.109
                },
                "24": {
                    x: 627.781,
                    y: 372.681,
                    sx: 1.098,
                    sy: 1.098
                },
                "25": {
                    x: 627.812,
                    y: 372.703,
                    sx: 1.088,
                    sy: 1.088
                },
                "26": {
                    x: 627.822,
                    y: 372.69,
                    sx: 1.079,
                    sy: 1.079
                },
                "27": {
                    x: 627.818,
                    y: 372.698,
                    sx: 1.07,
                    sy: 1.07
                },
                "28": {
                    x: 627.802,
                    y: 372.676,
                    sx: 1.061,
                    sy: 1.061
                },
                "29": {
                    x: 627.814,
                    y: 372.72,
                    sx: 1.053,
                    sy: 1.053
                },
                "30": {
                    x: 627.822,
                    y: 372.689,
                    sx: 1.046,
                    sy: 1.046
                },
                "31": {
                    x: 627.818,
                    y: 372.679,
                    sx: 1.039,
                    sy: 1.039
                },
                "32": {
                    x: 627.782,
                    sx: 1.033,
                    sy: 1.033
                },
                "33": {
                    x: 627.794,
                    y: 372.704,
                    sx: 1.027,
                    sy: 1.027
                },
                "34": {
                    x: 627.783,
                    y: 372.695,
                    sx: 1.022,
                    sy: 1.022
                },
                "35": {
                    x: 627.809,
                    y: 372.706,
                    sx: 1.017,
                    sy: 1.017
                },
                "36": {
                    x: 627.782,
                    y: 372.694,
                    sx: 1.013,
                    sy: 1.013
                },
                "37": {
                    x: 627.784,
                    y: 372.696,
                    sx: 1.01,
                    sy: 1.01
                },
                "38": {
                    x: 627.803,
                    y: 372.708,
                    sx: 1.007,
                    sy: 1.007
                },
                "39": {
                    x: 627.779,
                    y: 372.701,
                    sx: 1.004,
                    sy: 1.004
                },
                "40": {
                    x: 627.783,
                    y: 372.71,
                    sx: 1.002,
                    sy: 1.002
                },
                "41": {
                    x: 627.814,
                    y: 372.683,
                    sx: 1.001,
                    sy: 1.001
                },
                "42": {
                    x: 627.793,
                    sx: 1,
                    sy: 1
                },
                "43": {
                    x: 627.8,
                    y: 372.7
                }
            });
    });

    lib.turnon01.assets = {
        "turnon01": "images/turnon01.shapes.json",
        "turnon01_atlas_1": "images/turnon01_atlas_1.json",
        "turnon01_atlas_2": "images/turnon01_atlas_2.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.turnon01,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 30,
        totalFrames: 44,
        library: lib
    };
}
