// @be/tutorial/assets/timelines/pan/pan.js
(function (PIXI, lib) {

    var MovieClip = PIXI.animate.MovieClip;
    var Container = PIXI.Container;
    var Sprite = PIXI.Sprite;
    var fromFrame = PIXI.Texture.fromFrame;

    lib.hand = Container.extend(function () {
        Container.call(this);
        var instance1 = new Sprite(fromFrame("pan-hand"));
        this.addChild(instance1);
    });

    lib.fade = Container.extend(function () {
        Container.call(this);
        var instance1 = new Sprite(fromFrame("pan-fade"));
        this.addChild(instance1);
    });

    lib.pan = MovieClip.extend(function () {
        MovieClip.call(this, {
            duration: 50,
            framerate: 24
        });
        var instance2 = new lib.fade();
        var instance1 = new lib.hand();
        this.addTimedChild(instance2, 0, 43, {
                "0": {
                    x: 1188.8,
                    y: 345.45,
                    sx: 0.022,
                    sy: 0.85,
                    ky: 3.142,
                    a: 1
                },
                "1": {
                    x: 1188.712,
                    y: 345.445
                },
                "2": {
                    x: 1188.495
                },
                "3": {
                    x: 1188.089,
                    sx: 0.023
                },
                "4": {
                    x: 1187.354,
                    y: 345.395,
                    sx: 0.025
                },
                "5": {
                    x: 1186.324,
                    sx: 0.027
                },
                "6": {
                    x: 1184.899,
                    y: 345.345,
                    sx: 0.03
                },
                "7": {
                    x: 1182.916,
                    y: 345.295,
                    sx: 0.034
                },
                "8": {
                    x: 1180.29,
                    y: 345.245,
                    sx: 0.039
                },
                "9": {
                    x: 1176.838,
                    y: 345.145,
                    sx: 0.047
                },
                "10": {
                    x: 1172.301,
                    y: 345.045,
                    sx: 0.056
                },
                "11": {
                    x: 1166.582,
                    y: 344.895,
                    sx: 0.068
                },
                "12": {
                    x: 1159.28,
                    y: 344.695,
                    sx: 0.083
                },
                "13": {
                    x: 1150.063,
                    y: 344.495,
                    sx: 0.102
                },
                "14": {
                    x: 1138.544,
                    y: 344.195,
                    sx: 0.126
                },
                "15": {
                    x: 1124.413,
                    y: 343.845,
                    sx: 0.155
                },
                "16": {
                    x: 1107.461,
                    y: 343.395,
                    sx: 0.191
                },
                "17": {
                    x: 1087.354,
                    y: 342.895,
                    sx: 0.232
                },
                "18": {
                    x: 1064.329,
                    y: 342.345,
                    sx: 0.28
                },
                "19": {
                    x: 1038.518,
                    y: 341.695,
                    sx: 0.334
                },
                "20": {
                    x: 1010.383,
                    y: 340.995,
                    sx: 0.392
                },
                "21": {
                    x: 980.359,
                    y: 340.245,
                    sx: 0.454
                },
                "22": {
                    x: 948.984,
                    y: 339.445,
                    sx: 0.519
                },
                "23": {
                    x: 916.566,
                    y: 338.645,
                    sx: 0.587
                },
                "24": {
                    x: 883.555,
                    y: 337.795,
                    sx: 0.655
                },
                "25": {
                    x: 850.152,
                    y: 336.995,
                    sx: 0.724
                },
                "26": {
                    x: 816.543,
                    y: 336.145,
                    sx: 0.794
                },
                "27": {
                    x: 782.863,
                    y: 335.295,
                    sx: 0.864
                },
                "28": {
                    x: 749.227,
                    y: 334.445,
                    sx: 0.934
                },
                "29": {
                    x: 715.82,
                    y: 333.645,
                    sx: 1.003
                },
                "30": {
                    x: 682.555,
                    y: 332.795,
                    sx: 1.072
                },
                "31": {
                    x: 649.477,
                    y: 331.945,
                    sx: 1.141
                },
                "32": {
                    x: 616.2,
                    y: 331.15,
                    sx: 1.209
                },
                "33": {
                    x: 616.191,
                    y: 331.195,
                    a: 0.83
                },
                "34": {
                    a: 0.67
                },
                "35": {
                    a: 0.5
                },
                "36": {
                    a: 0.33
                },
                "37": {
                    a: 0.17
                },
                "38": {
                    x: 616.2,
                    y: 331.15,
                    a: 0
                }
            })
            .addTimedChild(instance1, 0, 50, {
                "0": {
                    x: 1207.2,
                    y: 341.75,
                    sx: 0.8,
                    sy: 0.8,
                    kx: 0,
                    ky: 3.142
                },
                "1": {
                    x: 1207.131,
                    y: 341.745
                },
                "2": {
                    x: 1206.571,
                    y: 341.732,
                    ky: -3.142
                },
                "3": {
                    x: 1205.545,
                    y: 341.674,
                    ky: -3.141
                },
                "4": {
                    x: 1203.99,
                    y: 341.647
                },
                "5": {
                    x: 1201.838,
                    y: 341.482
                },
                "6": {
                    x: 1198.865,
                    y: 341.39,
                    kx: -0.001
                },
                "7": {
                    x: 1194.858,
                    y: 341.247
                },
                "8": {
                    x: 1190.248,
                    y: 341.476,
                    kx: -0.004,
                    ky: -3.137
                },
                "9": {
                    x: 1183.742,
                    y: 341.229,
                    kx: -0.005
                },
                "10": {
                    x: 1175.923,
                    y: 341.305,
                    kx: -0.009,
                    ky: -3.133
                },
                "11": {
                    x: 1165.535,
                    y: 340.879,
                    kx: -0.01,
                    ky: -3.132
                },
                "12": {
                    x: 1153.12,
                    y: 340.782,
                    kx: -0.014,
                    ky: -3.128
                },
                "13": {
                    x: 1137.557,
                    y: 340.555,
                    kx: -0.018,
                    ky: -3.124
                },
                "14": {
                    x: 1118.375,
                    y: 340.164,
                    kx: -0.023,
                    ky: -3.119
                },
                "15": {
                    x: 1095.351,
                    y: 340.104,
                    kx: -0.031,
                    ky: -3.111
                },
                "16": {
                    x: 1067.675,
                    y: 339.87,
                    kx: -0.039,
                    ky: -3.102
                },
                "17": {
                    x: 1035.08,
                    y: 339.326,
                    kx: -0.048,
                    ky: -3.093
                },
                "18": {
                    x: 997.599,
                    y: 338.709,
                    kx: -0.058,
                    ky: -3.084
                },
                "19": {
                    x: 956.252,
                    y: 338.351,
                    kx: -0.07,
                    ky: -3.071
                },
                "20": {
                    x: 911.146,
                    y: 337.894,
                    kx: -0.083,
                    ky: -3.058
                },
                "21": {
                    x: 863.267,
                    y: 337.258,
                    sx: 0.799,
                    sy: 0.799,
                    kx: -0.097,
                    ky: -3.045
                },
                "22": {
                    x: 813.36,
                    y: 336.678,
                    kx: -0.11,
                    ky: -3.031
                },
                "23": {
                    x: 762.699,
                    y: 336.515,
                    kx: -0.127,
                    ky: -3.015
                },
                "24": {
                    x: 710.873,
                    y: 335.868,
                    kx: -0.14,
                    ky: -3.001
                },
                "25": {
                    x: 659.342,
                    y: 335.709,
                    kx: -0.157,
                    ky: -2.984
                },
                "26": {
                    x: 607.458,
                    y: 335.153,
                    kx: -0.171,
                    ky: -2.971
                },
                "27": {
                    x: 555.856,
                    y: 334.589,
                    kx: -0.184,
                    ky: -2.957
                },
                "28": {
                    x: 505.034,
                    y: 334.604,
                    kx: -0.201,
                    ky: -2.941
                },
                "29": {
                    x: 454.442,
                    y: 334.131,
                    kx: -0.214,
                    ky: -2.927
                },
                "30": {
                    x: 404.291,
                    y: 333.73,
                    kx: -0.228,
                    ky: -2.914
                },
                "31": {
                    x: 354.764,
                    y: 333.329,
                    kx: -0.241,
                    ky: -2.9
                },
                "32": {
                    x: 306.35,
                    y: 333.45,
                    kx: -0.258,
                    ky: -2.884
                },
                "39": {
                    x: 320.893,
                    y: 335.073,
                    sx: 0.808,
                    sy: 0.808,
                    ky: -2.883
                },
                "40": {
                    x: 365.286,
                    y: 340.19,
                    sx: 0.836,
                    sy: 0.836,
                    kx: -0.263,
                    ky: -2.879
                },
                "41": {
                    x: 439.416,
                    y: 348.974,
                    sx: 0.882,
                    sy: 0.882,
                    kx: -0.271,
                    ky: -2.87
                },
                "42": {
                    x: 543.088,
                    y: 361.071,
                    sx: 0.948,
                    sy: 0.948,
                    kx: -0.28,
                    ky: -2.861
                },
                "43": {
                    x: 677.191,
                    y: 378.104,
                    sx: 1.031,
                    sy: 1.031,
                    kx: -0.297,
                    ky: -2.845
                },
                "44": {
                    x: 841.3,
                    y: 398.7,
                    sx: 1.135,
                    sy: 1.135,
                    kx: -0.314,
                    ky: -2.828
                },
                "45": {
                    x: 971.197,
                    y: 371.793,
                    sx: 1.014,
                    sy: 1.014,
                    kx: -0.201,
                    ky: -2.941
                },
                "46": {
                    x: 1073.036,
                    y: 355.442,
                    sx: 0.92,
                    sy: 0.92,
                    kx: -0.11,
                    ky: -3.031
                },
                "47": {
                    x: 1147.164,
                    y: 346.974,
                    sx: 0.853,
                    sy: 0.853,
                    kx: -0.049,
                    ky: -3.093
                },
                "48": {
                    x: 1191.748,
                    y: 342.663,
                    sx: 0.813,
                    sy: 0.813,
                    kx: -0.01,
                    ky: -3.132
                },
                "49": {
                    x: 1207.2,
                    y: 341.75,
                    sx: 0.8,
                    sy: 0.8,
                    kx: 0,
                    ky: 3.142
                }
            });
    });

    lib.pan.assets = {
        "pan_atlas_1": "images/pan_atlas_1.json"
    };
})(PIXI, lib = lib || {});
var lib;
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        stage: lib.pan,
        background: 0x0,
        width: 1280,
        height: 720,
        framerate: 24,
        totalFrames: 50,
        library: lib
    };
}
